<!DOCTYPE html>
<html class="writer-html5" lang="en" >
<head>
    <meta charset="utf-8" />
    <meta http-equiv="X-UA-Compatible" content="IE=edge" />
    <meta name="viewport" content="width=device-width, initial-scale=1.0" />
      <link rel="shortcut icon" href="../../../../img/favicon.ico" />
    <title>query_and_csv_production - Genomic Data Extraction</title>
    <link rel="stylesheet" href="../../../../css/theme.css" />
    <link rel="stylesheet" href="../../../../css/theme_extra.css" />
        <link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/highlight.js/10.5.0/styles/github.min.css" />
        <link href="../../../../assets/_mkdocstrings.css" rel="stylesheet" />
    
      <script>
        // Current page data
        var mkdocs_page_name = "query_and_csv_production";
        var mkdocs_page_input_path = "genomic_data_extraction/rna/rna_download_logic/query_and_csv_production.md";
        var mkdocs_page_url = null;
      </script>
    
    <script src="../../../../js/jquery-3.6.0.min.js" defer></script>
    <!--[if lt IE 9]>
      <script src="../../../../js/html5shiv.min.js"></script>
    <![endif]-->
      <script src="https://cdnjs.cloudflare.com/ajax/libs/highlight.js/10.5.0/highlight.min.js"></script>
      <script>hljs.initHighlightingOnLoad();</script> 
</head>

<body class="wy-body-for-nav" role="document">

  <div class="wy-grid-for-nav">
    <nav data-toggle="wy-nav-shift" class="wy-nav-side stickynav">
    <div class="wy-side-scroll">
      <div class="wy-side-nav-search">
          <a href="../../../.." class="icon icon-home"> Genomic Data Extraction
        </a>
      </div>

      <div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="Navigation menu">
              <ul>
                <li class="toctree-l1"><a class="reference internal" href="../../../..">Home</a>
                </li>
              </ul>
              <p class="caption"><span class="caption-text">genomic_data_extraction</span></p>
              <ul class="current">
                  <li class="toctree-l1"><a class="reference internal" href="#">dna</a>
    <ul>
                <li class="toctree-l2"><a class="reference internal" href="../../../dna/dna_extraction/">dna_extraction</a>
                </li>
                <li class="toctree-l2"><a class="reference internal" href="../../../dna/dna_feature_extraction/">dna_feature_extraction</a>
                </li>
                <li class="toctree-l2"><a class="reference internal" href="../../../dna/ensembl_api/">ensembl_api</a>
                </li>
    </ul>
                  </li>
                  <li class="toctree-l1 current"><a class="reference internal current" href="#">rna</a>
    <ul class="current">
                <li class="toctree-l2"><a class="reference internal" href="#">data_conversion_helper_functions</a>
    <ul>
                <li class="toctree-l3"><a class="reference internal" href="../../data_conversion_helper_functions/convert_quantsf_to_csv/">convert_quantsf_to_csv</a>
                </li>
                <li class="toctree-l3"><a class="reference internal" href="../../data_conversion_helper_functions/create_expression_matrix/">create_expression_matrix</a>
                </li>
                <li class="toctree-l3"><a class="reference internal" href="../../data_conversion_helper_functions/create_samplesheet_csv/">create_samplesheet_csv</a>
                </li>
                <li class="toctree-l3"><a class="reference internal" href="../../data_conversion_helper_functions/process_expression_matrix/">process_expression_matrix</a>
                </li>
    </ul>
                </li>
                <li class="toctree-l2 current"><a class="reference internal current" href="#">rna_download_logic</a>
    <ul class="current">
                <li class="toctree-l3"><a class="reference internal" href="../mRNA_fastq_download/">mRNA_fastq_download</a>
                </li>
                <li class="toctree-l3 current"><a class="reference internal current" href="./">query_and_csv_production</a>
    <ul class="current">
    </ul>
                </li>
    </ul>
                </li>
                <li class="toctree-l2"><a class="reference internal" href="../../rna_extraction/">rna_extraction</a>
                </li>
    </ul>
                  </li>
              </ul>
      </div>
    </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap">
      <nav class="wy-nav-top" role="navigation" aria-label="Mobile navigation menu">
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="../../../..">Genomic Data Extraction</a>
        
      </nav>
      <div class="wy-nav-content">
        <div class="rst-content"><div role="navigation" aria-label="breadcrumbs navigation">
  <ul class="wy-breadcrumbs">
    <li><a href="../../../.." class="icon icon-home" alt="Docs"></a> &raquo;</li>
          <li>genomic_data_extraction &raquo;</li>
          <li>rna &raquo;</li>
          <li>rna_download_logic &raquo;</li>
      <li>query_and_csv_production</li>
    <li class="wy-breadcrumbs-aside">
          <a href="https://github.com/TransferLearningForDNA/genomic_data_extraction/edit/master/docs/genomic_data_extraction/rna/rna_download_logic/query_and_csv_production.md" class="icon icon-github"> Edit on GitHub</a>
    </li>
  </ul>
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
            <div class="section" itemprop="articleBody">
              
                <div class="doc doc-object doc-module">



<a id="rna.rna_download_logic.query_and_csv_production"></a>
  <div class="doc doc-contents first">

  

  <div class="doc doc-children">










<div class="doc doc-object doc-function">



<h2 id="rna.rna_download_logic.query_and_csv_production.SRX_to_SRR_csv" class="doc doc-heading">
          <code class="highlight language-python">SRX_to_SRR_csv(species_srx_map, output_file)</code>

</h2>


  <div class="doc doc-contents ">
  
      <p>Save a CSV file with columns for species, taxonomy_id, srx_id, and corresponding srr_ids.</p>
<p>This function processes each species and its SRX IDs to fetch SRR IDs and taxonomy IDs from the SRA database,
and then compiles this information into a CSV file.
This file can be used for downloading data and for logic handling in later stages of processing and testing.</p>



<table class="field-list">
  <colgroup>
    <col class="field-name" />
    <col class="field-body" />
  </colgroup>
  <tbody valign="top">
    <tr class="field">
      <th class="field-name">Parameters:</th>
      <td class="field-body">
        <ul class="first simple">
            <li>
              <b><code>species_srx_map</code></b>
                  (<code><span title="typing.Dict">Dict</span>[str, <span title="typing.List">List</span>[str]]</code>)
              –
              <div class="doc-md-description">
                <p>A dictionary with species names as keys and lists of SRX (experiment accession) IDs as values.</p>
              </div>
            </li>
            <li>
              <b><code>output_file</code></b>
                  (<code>str</code>)
              –
              <div class="doc-md-description">
                <p>The path to the output CSV file where the data will be saved.</p>
              </div>
            </li>
        </ul>
      </td>
    </tr>
  </tbody>
</table>
          <details class="quote">
            <summary>Source code in <code>rna/rna_download_logic/query_and_csv_production.py</code></summary>
            <pre class="highlight"><code class="language-python">def SRX_to_SRR_csv(species_srx_map: Dict[str, List[str]], output_file: str) -&gt; None:
    """Save a CSV file with columns for species, taxonomy_id, srx_id, and corresponding srr_ids.

    This function processes each species and its SRX IDs to fetch SRR IDs and taxonomy IDs from the SRA database,
    and then compiles this information into a CSV file.
    This file can be used for downloading data and for logic handling in later stages of processing and testing.

    Args:
        species_srx_map (Dict[str, List[str]]): A dictionary with species names as keys and lists of SRX (experiment accession) IDs as values.
        output_file (str): The path to the output CSV file where the data will be saved.
    """
    db = SRAweb()
    data_rows = []

    for species, srx_ids in species_srx_map.items():
        print(f"Processing {species} with {len(srx_ids)} SRX IDs...")

        for srx_id in srx_ids:
            df = db.sra_metadata([srx_id])
            srr_ids = df["run_accession"].unique()

            if not df.empty and "organism_taxid" in df.columns:
                taxonomy_id = df.iloc[0]["organism_taxid"]
            else:
                taxonomy_id = None  # Use None or an appropriate placeholder if taxonomy ID is not available

            # For each SRX ID, store its corresponding SRR IDs along with the taxonomy ID
            for srr_id in srr_ids:
                data_row = {
                    "species": species,
                    "taxonomy_id": taxonomy_id,
                    "srx_id": srx_id,
                    "srr_id": srr_id,
                }
                data_rows.append(data_row)

    # Convert the list of dictionaries to a DataFrame
    df_output = pd.DataFrame(data_rows)

    # Save the df to a CSV file
    df_output.to_csv(output_file, index=False)
    print(f"Data saved to {output_file}")</code></pre>
          </details>
  </div>

</div>


<div class="doc doc-object doc-function">



<h2 id="rna.rna_download_logic.query_and_csv_production.query_and_get_srx_accession_ids" class="doc doc-heading">
          <code class="highlight language-python">query_and_get_srx_accession_ids(species_data, limit=10)</code>

</h2>


  <div class="doc doc-contents ">
  
      <p>Get experiment accession numbers (SRX IDs) for each species.</p>



<table class="field-list">
  <colgroup>
    <col class="field-name" />
    <col class="field-body" />
  </colgroup>
  <tbody valign="top">
    <tr class="field">
      <th class="field-name">Parameters:</th>
      <td class="field-body">
        <ul class="first simple">
            <li>
              <b><code>species_data</code></b>
                  (<code><span title="typing.Dict">Dict</span>[str, int]</code>)
              –
              <div class="doc-md-description">
                <p>A dictionary with species names as keys and taxonomy IDs as values.</p>
              </div>
            </li>
            <li>
              <b><code>limit</code></b>
                  (<code><span title="typing.Optional">Optional</span>[int]</code>, default:
                      <code>10</code>
)
              –
              <div class="doc-md-description">
                <p>Maximum number of experiment accessions IDs to query per species (defaults to 10).</p>
              </div>
            </li>
        </ul>
      </td>
    </tr>
  </tbody>
</table>


<table class="field-list">
  <colgroup>
    <col class="field-name" />
    <col class="field-body" />
  </colgroup>
  <tbody valign="top">
    <tr class="field">
    <th class="field-name">Returns:</th>
    <td class="field-body">
      <ul class="first simple">
          <li>
                <code><span title="typing.Dict">Dict</span>[str, <span title="typing.List">List</span>[str]]</code>
            –
            <div class="doc-md-description">
              <p>Dict[str, List[str]: A dictionary with species names as keys and SRX (experiment accession) ID as values.</p>
            </div>
          </li>
      </ul>
    </td>
    </tr>
  </tbody>
</table>
          <details class="quote">
            <summary>Source code in <code>rna/rna_download_logic/query_and_csv_production.py</code></summary>
            <pre class="highlight"><code class="language-python">def query_and_get_srx_accession_ids(
    species_data: Dict[str, int], limit: Optional[int] = 10
) -&gt; Dict[str, List[str]]:
    """Get experiment accession numbers (SRX IDs) for each species.

    Args:
        species_data (Dict[str, int]): A dictionary with species names as keys and taxonomy IDs as values.
        limit (Optional[int]): Maximum number of experiment accessions IDs to query per species (defaults to 10).

    Returns:
        Dict[str, List[str]: A dictionary with species names as keys and SRX (experiment accession) ID as values.
    """
    # Dictionary to store species names and their SRX IDs
    species_srx_map = {}

    for species, tax_id in species_data.items():
        df = query_sra(species, tax_id, limit=limit)

        if df is not None and not df.empty:
            srx_ids = df["experiment_accession"].unique()
            species_srx_map[species] = srx_ids
        else:
            print(f"No results or empty DataFrame for {species}")

    for species, srx_ids in species_srx_map.items():
        print(f"\n{species}: {', '.join(srx_ids)}")

    return species_srx_map</code></pre>
          </details>
  </div>

</div>


<div class="doc doc-object doc-function">



<h2 id="rna.rna_download_logic.query_and_csv_production.query_sra" class="doc doc-heading">
          <code class="highlight language-python">query_sra(species, taxonomy_id, limit=10)</code>

</h2>


  <div class="doc doc-contents ">
  
      <p>Query the SRA database for RNA-seq species metadata results given a species name and taxonomy ID.</p>



<table class="field-list">
  <colgroup>
    <col class="field-name" />
    <col class="field-body" />
  </colgroup>
  <tbody valign="top">
    <tr class="field">
      <th class="field-name">Parameters:</th>
      <td class="field-body">
        <ul class="first simple">
            <li>
              <b><code>species</code></b>
                  (<code>str</code>)
              –
              <div class="doc-md-description">
                <p>Species scientific name (format e.g. 'Homo sapiens')</p>
              </div>
            </li>
            <li>
              <b><code>taxonomy_id</code></b>
                  (<code>int</code>)
              –
              <div class="doc-md-description">
                <p>Taxonomy ID of the species (e.g. 9606)</p>
              </div>
            </li>
            <li>
              <b><code>limit</code></b>
                  (<code><span title="typing.Optional">Optional</span>[int]</code>, default:
                      <code>10</code>
)
              –
              <div class="doc-md-description">
                <p>Maximum number of experiment accessions IDs to query (defaults to 10).</p>
              </div>
            </li>
        </ul>
      </td>
    </tr>
  </tbody>
</table>


<table class="field-list">
  <colgroup>
    <col class="field-name" />
    <col class="field-body" />
  </colgroup>
  <tbody valign="top">
    <tr class="field">
    <th class="field-name">Returns:</th>
    <td class="field-body">
      <ul class="first simple">
          <li>
<b><code>DataFrame</code></b>(                <code><span title="pandas.DataFrame">DataFrame</span> | None</code>
)            –
            <div class="doc-md-description">
              <p>Species metadata RNA-seq results (contains experiment accession IDs)</p>
            </div>
          </li>
      </ul>
    </td>
    </tr>
  </tbody>
</table>
          <details class="quote">
            <summary>Source code in <code>rna/rna_download_logic/query_and_csv_production.py</code></summary>
            <pre class="highlight"><code class="language-python">def query_sra(species: str, taxonomy_id: int, limit: Optional[int] = 10) -&gt; pd.DataFrame | None:
    """Query the SRA database for RNA-seq species metadata results given a species name and taxonomy ID.

    Args:
        species (str): Species scientific name (format e.g. 'Homo sapiens')
        taxonomy_id (int): Taxonomy ID of the species (e.g. 9606)
        limit (Optional[int]): Maximum number of experiment accessions IDs to query (defaults to 10).

    Returns:
        DataFrame: Species metadata RNA-seq results (contains experiment accession IDs)
    """
    # First, search for 'RNA-Seq' strategy
    try:
        sra_search_rna_seq = SraSearch(
            organism=species, layout="paired", strategy=["RNA-Seq"], return_max=limit
        )
        sra_search_rna_seq.search()
        df_rna_seq = sra_search_rna_seq.get_df()

        # If 'RNA-Seq' results are fewer than 50, search with 'OTHER' strategy
        df_other = pd.DataFrame()
        if len(df_rna_seq) &lt; limit:
            additional_results_needed = limit - len(df_rna_seq)
            sra_search_other = SraSearch(
                organism=species,
                layout="paired",
                strategy=["OTHER"],
                return_max=additional_results_needed,
            )
            sra_search_other.search()
            df_other = sra_search_other.get_df()

        # Combine the two DataFrames
        df_combined = pd.concat([df_rna_seq, df_other], ignore_index=True)
        # print(f"{species} df_combined:", df_combined.columns)

        # Filter these combined dataframes to ensure they match the taxonomy ID
        df_filtered = df_combined[df_combined["sample_taxon_id"] == str(taxonomy_id)]

        return df_filtered if not df_filtered.empty else None

    except (ValueError, KeyError, AttributeError) as e:
        print(f"Error querying {species}: {e}")
        return None
    except Exception as e:
        print(
            f"Error querying {species}, possibly due to external libraries or API (SraSearch): {e}"
        )
        return None</code></pre>
          </details>
  </div>

</div>


<div class="doc doc-object doc-function">



<h2 id="rna.rna_download_logic.query_and_csv_production.view_srx_metadata" class="doc doc-heading">
          <code class="highlight language-python">view_srx_metadata(species_srx_map)</code>

</h2>


  <div class="doc doc-contents ">
  
      <p>Get metadata for all species and experiment accession numbers.</p>



<table class="field-list">
  <colgroup>
    <col class="field-name" />
    <col class="field-body" />
  </colgroup>
  <tbody valign="top">
    <tr class="field">
      <th class="field-name">Parameters:</th>
      <td class="field-body">
        <ul class="first simple">
            <li>
              <b><code>species_srx_map</code></b>
                  (<code><span title="typing.Dict">Dict</span>[str, <span title="typing.List">List</span>[str]]</code>)
              –
              <div class="doc-md-description">
                <p>A dictionary with species names as keys and SRX (experiment accession) IDs as values.</p>
              </div>
            </li>
        </ul>
      </td>
    </tr>
  </tbody>
</table>


<table class="field-list">
  <colgroup>
    <col class="field-name" />
    <col class="field-body" />
  </colgroup>
  <tbody valign="top">
    <tr class="field">
    <th class="field-name">Returns:</th>
    <td class="field-body">
      <ul class="first simple">
          <li>
                <code><span title="typing.Dict">Dict</span>[str, <span title="pandas.DataFrame">DataFrame</span>]</code>
            –
            <div class="doc-md-description">
              <p>Dict[str, pd.DataFrame]: A dictionary with species name as keys and dataframes with experiment metadata as values.</p>
            </div>
          </li>
      </ul>
    </td>
    </tr>
  </tbody>
</table>
          <details class="quote">
            <summary>Source code in <code>rna/rna_download_logic/query_and_csv_production.py</code></summary>
            <pre class="highlight"><code class="language-python">def view_srx_metadata(species_srx_map: Dict[str, List[str]]) -&gt; Dict[str, pd.DataFrame]:
    """Get metadata for all species and experiment accession numbers.

    Args:
        species_srx_map (Dict[str, List[str]]): A dictionary with species names as keys and SRX (experiment accession) IDs as values.

    Returns:
        Dict[str, pd.DataFrame]: A dictionary with species name as keys and dataframes with experiment metadata as values.
    """
    db = SRAweb()

    all_species_metadata = {}

    for species, srx_ids in species_srx_map.items():
        list_ids = list(srx_ids)
        print(f"Processing {species} with {len(list_ids)} SRX IDs...")

        df = db.sra_metadata(list_ids)

        all_species_metadata[species] = df
    print(list(all_species_metadata[0].columns))

    return all_species_metadata</code></pre>
          </details>
  </div>

</div>



  </div>

  </div>

</div>
              
            </div>
          </div><footer>
    <div class="rst-footer-buttons" role="navigation" aria-label="Footer Navigation">
        <a href="../mRNA_fastq_download/" class="btn btn-neutral float-left" title="mRNA_fastq_download"><span class="icon icon-circle-arrow-left"></span> Previous</a>
        <a href="../../rna_extraction/" class="btn btn-neutral float-right" title="rna_extraction">Next <span class="icon icon-circle-arrow-right"></span></a>
    </div>

  <hr/>

  <div role="contentinfo">
    <!-- Copyright etc -->
  </div>

  Built with <a href="https://www.mkdocs.org/">MkDocs</a> using a <a href="https://github.com/readthedocs/sphinx_rtd_theme">theme</a> provided by <a href="https://readthedocs.org">Read the Docs</a>.
</footer>
          
        </div>
      </div>

    </section>

  </div>

  <div class="rst-versions" role="note" aria-label="Versions">
  <span class="rst-current-version" data-toggle="rst-current-version">
    
        <span>
          <a href="https://github.com/TransferLearningForDNA/genomic_data_extraction" class="fa fa-github" style="color: #fcfcfc"> GitHub</a>
        </span>
    
    
      <span><a href="../mRNA_fastq_download/" style="color: #fcfcfc">&laquo; Previous</a></span>
    
    
      <span><a href="../../rna_extraction/" style="color: #fcfcfc">Next &raquo;</a></span>
    
  </span>
</div>
    <script>var base_url = '../../../..';</script>
    <script src="../../../../js/theme_extra.js" defer></script>
    <script src="../../../../js/theme.js" defer></script>
    <script defer>
        window.onload = function () {
            SphinxRtdTheme.Navigation.enable(true);
        };
    </script>

</body>
</html>
