<!DOCTYPE html>
<html class="writer-html5" lang="en" >
<head>
    <meta charset="utf-8" />
    <meta http-equiv="X-UA-Compatible" content="IE=edge" />
    <meta name="viewport" content="width=device-width, initial-scale=1.0" />
      <link rel="shortcut icon" href="../../../../img/favicon.ico" />
    <title>query_and_csv_production - Genomic Data Extraction</title>
    <link rel="stylesheet" href="../../../../css/theme.css" />
    <link rel="stylesheet" href="../../../../css/theme_extra.css" />
        <link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/highlight.js/10.5.0/styles/github.min.css" />
        <link href="../../../../assets/_mkdocstrings.css" rel="stylesheet" />
    
      <script>
        // Current page data
        var mkdocs_page_name = "query_and_csv_production";
        var mkdocs_page_input_path = "genomic_data_extraction\\rna\\rna_download_logic\\query_and_csv_production.md";
        var mkdocs_page_url = null;
      </script>
    
    <script src="../../../../js/jquery-3.6.0.min.js" defer></script>
    <!--[if lt IE 9]>
      <script src="../../../../js/html5shiv.min.js"></script>
    <![endif]-->
      <script src="https://cdnjs.cloudflare.com/ajax/libs/highlight.js/10.5.0/highlight.min.js"></script>
      <script>hljs.initHighlightingOnLoad();</script> 
</head>

<body class="wy-body-for-nav" role="document">

  <div class="wy-grid-for-nav">
    <nav data-toggle="wy-nav-shift" class="wy-nav-side stickynav">
    <div class="wy-side-scroll">
      <div class="wy-side-nav-search">
          <a href="../../../.." class="icon icon-home"> Genomic Data Extraction
        </a>
      </div>

      <div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="Navigation menu">
              <ul>
                <li class="toctree-l1"><a class="reference internal" href="../../../..">Home</a>
                </li>
              </ul>
              <p class="caption"><span class="caption-text">genomic_data_extraction</span></p>
              <ul class="current">
                  <li class="toctree-l1"><a class="reference internal" href="#">dna</a>
    <ul>
                <li class="toctree-l2"><a class="reference internal" href="../../../dna/dna_extraction/">dna_extraction</a>
                </li>
                <li class="toctree-l2"><a class="reference internal" href="../../../dna/dna_feature_extraction/">dna_feature_extraction</a>
                </li>
                <li class="toctree-l2"><a class="reference internal" href="../../../dna/ensembl_api/">ensembl_api</a>
                </li>
    </ul>
                  </li>
                  <li class="toctree-l1 current"><a class="reference internal current" href="#">rna</a>
    <ul class="current">
                <li class="toctree-l2"><a class="reference internal" href="#">data_conversion_helper_functions</a>
    <ul>
                <li class="toctree-l3"><a class="" href="../../conversion_helper_functions/convert_quantsf_to_csv.md">convert_quantsf_to_csv</a>
                </li>
                <li class="toctree-l3"><a class="" href="../../conversion_helper_functions/create_expression_matrix.md">create_expression_matrix</a>
                </li>
                <li class="toctree-l3"><a class="" href="../../conversion_helper_functions/create_samplesheet_csv.md">create_samplesheet_csv</a>
                </li>
                <li class="toctree-l3"><a class="" href="../../conversion_helper_functions/process_expression_matrix.md">process_expression_matrix</a>
                </li>
    </ul>
                </li>
                <li class="toctree-l2 current"><a class="reference internal current" href="#">rna_download_logic</a>
    <ul class="current">
                <li class="toctree-l3"><a class="reference internal" href="../mRNA_fastq_download/">mRNA_fastq_download</a>
                </li>
                <li class="toctree-l3 current"><a class="reference internal current" href="./">query_and_csv_production</a>
    <ul class="current">
    </ul>
                </li>
    </ul>
                </li>
                <li class="toctree-l2"><a class="reference internal" href="../../rna_extraction/">rna_extraction</a>
                </li>
    </ul>
                  </li>
              </ul>
      </div>
    </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap">
      <nav class="wy-nav-top" role="navigation" aria-label="Mobile navigation menu">
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="../../../..">Genomic Data Extraction</a>
        
      </nav>
      <div class="wy-nav-content">
        <div class="rst-content"><div role="navigation" aria-label="breadcrumbs navigation">
  <ul class="wy-breadcrumbs">
    <li><a href="../../../.." class="icon icon-home" alt="Docs"></a> &raquo;</li>
          <li>genomic_data_extraction &raquo;</li>
          <li>rna &raquo;</li>
          <li>rna_download_logic &raquo;</li>
      <li>query_and_csv_production</li>
    <li class="wy-breadcrumbs-aside">
          <a href="https://github.com/TransferLearningForDNA/genomic_data_extraction/edit/master/docs/genomic_data_extraction/rna/rna_download_logic/query_and_csv_production.md" class="icon icon-github"> Edit on GitHub</a>
    </li>
  </ul>
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
            <div class="section" itemprop="articleBody">
              
                <div class="doc doc-object doc-module">



<a id="rna.rna_download_logic.query_and_csv_production"></a>
  <div class="doc doc-contents first">

  

  <div class="doc doc-children">










<div class="doc doc-object doc-function">



<h2 id="rna.rna_download_logic.query_and_csv_production.SRX_to_SRR_csv" class="doc doc-heading">
          <code class="highlight language-python"><span class="n">SRX_to_SRR_csv</span><span class="p">(</span><span class="n">species_srx_map</span><span class="p">,</span> <span class="n">output_file</span><span class="p">)</span></code>

</h2>


  <div class="doc doc-contents ">
  
      <p>Save a CSV file with columns for species, taxonomy_id, srx_id, and corresponding srr_ids.</p>
<p>This function processes each species and its SRX IDs to fetch SRR IDs and taxonomy IDs from the SRA database,
and then compiles this information into a CSV file.
This file can be used for downloading data and for logic handling in later stages of processing and testing.</p>



<table class="field-list">
  <colgroup>
    <col class="field-name" />
    <col class="field-body" />
  </colgroup>
  <tbody valign="top">
    <tr class="field">
      <th class="field-name">Parameters:</th>
      <td class="field-body">
        <ul class="first simple">
            <li>
              <b><code>species_srx_map</code></b>
                  (<code><span title="typing.Dict">Dict</span>[str, <span title="typing.List">List</span>[str]]</code>)
              –
              <div class="doc-md-description">
                <p>A dictionary with species names as keys and lists of SRX (experiment accession) IDs as values.</p>
              </div>
            </li>
            <li>
              <b><code>output_file</code></b>
                  (<code>str</code>)
              –
              <div class="doc-md-description">
                <p>The path to the output CSV file where the data will be saved.</p>
              </div>
            </li>
        </ul>
      </td>
    </tr>
  </tbody>
</table>
          <details class="quote">
            <summary>Source code in <code>rna\rna_download_logic\query_and_csv_production.py</code></summary>
            <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">112</span>
<span class="normal">113</span>
<span class="normal">114</span>
<span class="normal">115</span>
<span class="normal">116</span>
<span class="normal">117</span>
<span class="normal">118</span>
<span class="normal">119</span>
<span class="normal">120</span>
<span class="normal">121</span>
<span class="normal">122</span>
<span class="normal">123</span>
<span class="normal">124</span>
<span class="normal">125</span>
<span class="normal">126</span>
<span class="normal">127</span>
<span class="normal">128</span>
<span class="normal">129</span>
<span class="normal">130</span>
<span class="normal">131</span>
<span class="normal">132</span>
<span class="normal">133</span>
<span class="normal">134</span>
<span class="normal">135</span>
<span class="normal">136</span>
<span class="normal">137</span>
<span class="normal">138</span>
<span class="normal">139</span>
<span class="normal">140</span>
<span class="normal">141</span>
<span class="normal">142</span>
<span class="normal">143</span>
<span class="normal">144</span>
<span class="normal">145</span>
<span class="normal">146</span>
<span class="normal">147</span>
<span class="normal">148</span>
<span class="normal">149</span>
<span class="normal">150</span>
<span class="normal">151</span>
<span class="normal">152</span>
<span class="normal">153</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span> <span class="nf">SRX_to_SRR_csv</span><span class="p">(</span><span class="n">species_srx_map</span><span class="p">:</span> <span class="n">Dict</span><span class="p">[</span><span class="nb">str</span><span class="p">,</span> <span class="n">List</span><span class="p">[</span><span class="nb">str</span><span class="p">]],</span> <span class="n">output_file</span><span class="p">:</span> <span class="nb">str</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="kc">None</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Save a CSV file with columns for species, taxonomy_id, srx_id, and corresponding srr_ids.</span>

<span class="sd">    This function processes each species and its SRX IDs to fetch SRR IDs and taxonomy IDs from the SRA database,</span>
<span class="sd">    and then compiles this information into a CSV file.</span>
<span class="sd">    This file can be used for downloading data and for logic handling in later stages of processing and testing.</span>

<span class="sd">    Args:</span>
<span class="sd">        species_srx_map (Dict[str, List[str]]): A dictionary with species names as keys and lists of SRX (experiment accession) IDs as values.</span>
<span class="sd">        output_file (str): The path to the output CSV file where the data will be saved.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">db</span> <span class="o">=</span> <span class="n">SRAweb</span><span class="p">()</span>
    <span class="n">data_rows</span> <span class="o">=</span> <span class="p">[]</span>

    <span class="k">for</span> <span class="n">species</span><span class="p">,</span> <span class="n">srx_ids</span> <span class="ow">in</span> <span class="n">species_srx_map</span><span class="o">.</span><span class="n">items</span><span class="p">():</span>
        <span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Processing </span><span class="si">{</span><span class="n">species</span><span class="si">}</span><span class="s2"> with </span><span class="si">{</span><span class="nb">len</span><span class="p">(</span><span class="n">srx_ids</span><span class="p">)</span><span class="si">}</span><span class="s2"> SRX IDs...&quot;</span><span class="p">)</span>

        <span class="k">for</span> <span class="n">srx_id</span> <span class="ow">in</span> <span class="n">srx_ids</span><span class="p">:</span>
            <span class="n">df</span> <span class="o">=</span> <span class="n">db</span><span class="o">.</span><span class="n">sra_metadata</span><span class="p">([</span><span class="n">srx_id</span><span class="p">])</span>
            <span class="n">srr_ids</span> <span class="o">=</span> <span class="n">df</span><span class="p">[</span><span class="s2">&quot;run_accession&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">unique</span><span class="p">()</span>

            <span class="k">if</span> <span class="ow">not</span> <span class="n">df</span><span class="o">.</span><span class="n">empty</span> <span class="ow">and</span> <span class="s2">&quot;organism_taxid&quot;</span> <span class="ow">in</span> <span class="n">df</span><span class="o">.</span><span class="n">columns</span><span class="p">:</span>
                <span class="n">taxonomy_id</span> <span class="o">=</span> <span class="n">df</span><span class="o">.</span><span class="n">iloc</span><span class="p">[</span><span class="mi">0</span><span class="p">][</span><span class="s2">&quot;organism_taxid&quot;</span><span class="p">]</span>
            <span class="k">else</span><span class="p">:</span>
                <span class="n">taxonomy_id</span> <span class="o">=</span> <span class="kc">None</span>  <span class="c1"># Use None or an appropriate placeholder if taxonomy ID is not available</span>

            <span class="c1"># For each SRX ID, store its corresponding SRR IDs along with the taxonomy ID</span>
            <span class="k">for</span> <span class="n">srr_id</span> <span class="ow">in</span> <span class="n">srr_ids</span><span class="p">:</span>
                <span class="n">data_row</span> <span class="o">=</span> <span class="p">{</span>
                    <span class="s2">&quot;species&quot;</span><span class="p">:</span> <span class="n">species</span><span class="p">,</span>
                    <span class="s2">&quot;taxonomy_id&quot;</span><span class="p">:</span> <span class="n">taxonomy_id</span><span class="p">,</span>
                    <span class="s2">&quot;srx_id&quot;</span><span class="p">:</span> <span class="n">srx_id</span><span class="p">,</span>
                    <span class="s2">&quot;srr_id&quot;</span><span class="p">:</span> <span class="n">srr_id</span><span class="p">,</span>
                <span class="p">}</span>
                <span class="n">data_rows</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">data_row</span><span class="p">)</span>

    <span class="c1"># Convert the list of dictionaries to a DataFrame</span>
    <span class="n">df_output</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">(</span><span class="n">data_rows</span><span class="p">)</span>

    <span class="c1"># Save the df to a CSV file</span>
    <span class="n">df_output</span><span class="o">.</span><span class="n">to_csv</span><span class="p">(</span><span class="n">output_file</span><span class="p">,</span> <span class="n">index</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>
    <span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Data saved to </span><span class="si">{</span><span class="n">output_file</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
</code></pre></div></td></tr></table></div>
          </details>
  </div>

</div>


<div class="doc doc-object doc-function">



<h2 id="rna.rna_download_logic.query_and_csv_production.query_and_get_srx_accession_ids" class="doc doc-heading">
          <code class="highlight language-python"><span class="n">query_and_get_srx_accession_ids</span><span class="p">(</span><span class="n">species_data</span><span class="p">)</span></code>

</h2>


  <div class="doc doc-contents ">
  
      <p>Get experiment accession numbers (SRX IDs) for each species.</p>



<table class="field-list">
  <colgroup>
    <col class="field-name" />
    <col class="field-body" />
  </colgroup>
  <tbody valign="top">
    <tr class="field">
      <th class="field-name">Parameters:</th>
      <td class="field-body">
        <ul class="first simple">
            <li>
              <b><code>species_data</code></b>
                  (<code><span title="typing.Dict">Dict</span>[str, int]</code>)
              –
              <div class="doc-md-description">
                <p>A dictionary with species names as keys and taxonomy IDs as values.</p>
              </div>
            </li>
        </ul>
      </td>
    </tr>
  </tbody>
</table>


<table class="field-list">
  <colgroup>
    <col class="field-name" />
    <col class="field-body" />
  </colgroup>
  <tbody valign="top">
    <tr class="field">
    <th class="field-name">Returns:</th>
    <td class="field-body">
      <ul class="first simple">
          <li>
                <code><span title="typing.Dict">Dict</span>[str, <span title="typing.List">List</span>[str]]</code>
            –
            <div class="doc-md-description">
              <p>Dict[str, List[str]: A dictionary with species names as keys and SRX (experiment accession) ID as values.</p>
            </div>
          </li>
      </ul>
    </td>
    </tr>
  </tbody>
</table>
          <details class="quote">
            <summary>Source code in <code>rna\rna_download_logic\query_and_csv_production.py</code></summary>
            <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">58</span>
<span class="normal">59</span>
<span class="normal">60</span>
<span class="normal">61</span>
<span class="normal">62</span>
<span class="normal">63</span>
<span class="normal">64</span>
<span class="normal">65</span>
<span class="normal">66</span>
<span class="normal">67</span>
<span class="normal">68</span>
<span class="normal">69</span>
<span class="normal">70</span>
<span class="normal">71</span>
<span class="normal">72</span>
<span class="normal">73</span>
<span class="normal">74</span>
<span class="normal">75</span>
<span class="normal">76</span>
<span class="normal">77</span>
<span class="normal">78</span>
<span class="normal">79</span>
<span class="normal">80</span>
<span class="normal">81</span>
<span class="normal">82</span>
<span class="normal">83</span>
<span class="normal">84</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span> <span class="nf">query_and_get_srx_accession_ids</span><span class="p">(</span>
    <span class="n">species_data</span><span class="p">:</span> <span class="n">Dict</span><span class="p">[</span><span class="nb">str</span><span class="p">,</span> <span class="nb">int</span><span class="p">]</span>
<span class="p">)</span> <span class="o">-&gt;</span> <span class="n">Dict</span><span class="p">[</span><span class="nb">str</span><span class="p">,</span> <span class="n">List</span><span class="p">[</span><span class="nb">str</span><span class="p">]]:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Get experiment accession numbers (SRX IDs) for each species.</span>

<span class="sd">    Args:</span>
<span class="sd">        species_data (Dict[str, int]): A dictionary with species names as keys and taxonomy IDs as values.</span>

<span class="sd">    Returns:</span>
<span class="sd">        Dict[str, List[str]: A dictionary with species names as keys and SRX (experiment accession) ID as values.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="c1"># Dictionary to store species names and their SRX IDs</span>
    <span class="n">species_srx_map</span> <span class="o">=</span> <span class="p">{}</span>

    <span class="k">for</span> <span class="n">species</span><span class="p">,</span> <span class="n">tax_id</span> <span class="ow">in</span> <span class="n">species_data</span><span class="o">.</span><span class="n">items</span><span class="p">():</span>
        <span class="n">df</span> <span class="o">=</span> <span class="n">query_sra</span><span class="p">(</span><span class="n">species</span><span class="p">,</span> <span class="n">tax_id</span><span class="p">)</span>

        <span class="k">if</span> <span class="n">df</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span> <span class="ow">and</span> <span class="ow">not</span> <span class="n">df</span><span class="o">.</span><span class="n">empty</span><span class="p">:</span>
            <span class="n">srx_ids</span> <span class="o">=</span> <span class="n">df</span><span class="p">[</span><span class="s2">&quot;experiment_accession&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">unique</span><span class="p">()</span>
            <span class="n">species_srx_map</span><span class="p">[</span><span class="n">species</span><span class="p">]</span> <span class="o">=</span> <span class="n">srx_ids</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;No results or empty DataFrame for </span><span class="si">{</span><span class="n">species</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>

    <span class="k">for</span> <span class="n">species</span><span class="p">,</span> <span class="n">srx_ids</span> <span class="ow">in</span> <span class="n">species_srx_map</span><span class="o">.</span><span class="n">items</span><span class="p">():</span>
        <span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;</span><span class="se">\n</span><span class="si">{</span><span class="n">species</span><span class="si">}</span><span class="s2">: </span><span class="si">{</span><span class="s1">&#39;, &#39;</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">srx_ids</span><span class="p">)</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>

    <span class="k">return</span> <span class="n">species_srx_map</span>
</code></pre></div></td></tr></table></div>
          </details>
  </div>

</div>


<div class="doc doc-object doc-function">



<h2 id="rna.rna_download_logic.query_and_csv_production.query_sra" class="doc doc-heading">
          <code class="highlight language-python"><span class="n">query_sra</span><span class="p">(</span><span class="n">species</span><span class="p">,</span> <span class="n">taxonomy_id</span><span class="p">)</span></code>

</h2>


  <div class="doc doc-contents ">
  
      <p>Query the SRA database for RNA-seq species metadata results given a species name and taxonomy ID.</p>



<table class="field-list">
  <colgroup>
    <col class="field-name" />
    <col class="field-body" />
  </colgroup>
  <tbody valign="top">
    <tr class="field">
      <th class="field-name">Parameters:</th>
      <td class="field-body">
        <ul class="first simple">
            <li>
              <b><code>species</code></b>
                  (<code>str</code>)
              –
              <div class="doc-md-description">
                <p>Species scientific name (format e.g. 'Homo sapiens')</p>
              </div>
            </li>
            <li>
              <b><code>taxonomy_id</code></b>
                  (<code>int</code>)
              –
              <div class="doc-md-description">
                <p>Taxonomy ID of the species (e.g. 9606)</p>
              </div>
            </li>
        </ul>
      </td>
    </tr>
  </tbody>
</table>


<table class="field-list">
  <colgroup>
    <col class="field-name" />
    <col class="field-body" />
  </colgroup>
  <tbody valign="top">
    <tr class="field">
    <th class="field-name">Returns:</th>
    <td class="field-body">
      <ul class="first simple">
          <li>
<b><code>DataFrame</code></b>(                <code><span title="pandas.DataFrame">DataFrame</span> | None</code>
)            –
            <div class="doc-md-description">
              <p>Species metadata RNA-seq results (contains experiment accession IDs)</p>
            </div>
          </li>
      </ul>
    </td>
    </tr>
  </tbody>
</table>
          <details class="quote">
            <summary>Source code in <code>rna\rna_download_logic\query_and_csv_production.py</code></summary>
            <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal"> 8</span>
<span class="normal"> 9</span>
<span class="normal">10</span>
<span class="normal">11</span>
<span class="normal">12</span>
<span class="normal">13</span>
<span class="normal">14</span>
<span class="normal">15</span>
<span class="normal">16</span>
<span class="normal">17</span>
<span class="normal">18</span>
<span class="normal">19</span>
<span class="normal">20</span>
<span class="normal">21</span>
<span class="normal">22</span>
<span class="normal">23</span>
<span class="normal">24</span>
<span class="normal">25</span>
<span class="normal">26</span>
<span class="normal">27</span>
<span class="normal">28</span>
<span class="normal">29</span>
<span class="normal">30</span>
<span class="normal">31</span>
<span class="normal">32</span>
<span class="normal">33</span>
<span class="normal">34</span>
<span class="normal">35</span>
<span class="normal">36</span>
<span class="normal">37</span>
<span class="normal">38</span>
<span class="normal">39</span>
<span class="normal">40</span>
<span class="normal">41</span>
<span class="normal">42</span>
<span class="normal">43</span>
<span class="normal">44</span>
<span class="normal">45</span>
<span class="normal">46</span>
<span class="normal">47</span>
<span class="normal">48</span>
<span class="normal">49</span>
<span class="normal">50</span>
<span class="normal">51</span>
<span class="normal">52</span>
<span class="normal">53</span>
<span class="normal">54</span>
<span class="normal">55</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span> <span class="nf">query_sra</span><span class="p">(</span><span class="n">species</span><span class="p">:</span> <span class="nb">str</span><span class="p">,</span> <span class="n">taxonomy_id</span><span class="p">:</span> <span class="nb">int</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span> <span class="o">|</span> <span class="kc">None</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Query the SRA database for RNA-seq species metadata results given a species name and taxonomy ID.</span>

<span class="sd">    Args:</span>
<span class="sd">        species (str): Species scientific name (format e.g. &#39;Homo sapiens&#39;)</span>
<span class="sd">        taxonomy_id (int): Taxonomy ID of the species (e.g. 9606)</span>

<span class="sd">    Returns:</span>
<span class="sd">        DataFrame: Species metadata RNA-seq results (contains experiment accession IDs)</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="c1"># First, search for &#39;RNA-Seq&#39; strategy</span>
    <span class="k">try</span><span class="p">:</span>
        <span class="n">sra_search_rna_seq</span> <span class="o">=</span> <span class="n">SraSearch</span><span class="p">(</span>
            <span class="n">organism</span><span class="o">=</span><span class="n">species</span><span class="p">,</span> <span class="n">layout</span><span class="o">=</span><span class="s2">&quot;paired&quot;</span><span class="p">,</span> <span class="n">strategy</span><span class="o">=</span><span class="p">[</span><span class="s2">&quot;RNA-Seq&quot;</span><span class="p">],</span> <span class="n">return_max</span><span class="o">=</span><span class="mi">50</span>
        <span class="p">)</span>
        <span class="n">sra_search_rna_seq</span><span class="o">.</span><span class="n">search</span><span class="p">()</span>
        <span class="n">df_rna_seq</span> <span class="o">=</span> <span class="n">sra_search_rna_seq</span><span class="o">.</span><span class="n">get_df</span><span class="p">()</span>

        <span class="c1"># If &#39;RNA-Seq&#39; results are fewer than 50, search with &#39;OTHER&#39; strategy</span>
        <span class="n">df_other</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">()</span>
        <span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="n">df_rna_seq</span><span class="p">)</span> <span class="o">&lt;</span> <span class="mi">50</span><span class="p">:</span>
            <span class="n">additional_results_needed</span> <span class="o">=</span> <span class="mi">50</span> <span class="o">-</span> <span class="nb">len</span><span class="p">(</span><span class="n">df_rna_seq</span><span class="p">)</span>
            <span class="n">sra_search_other</span> <span class="o">=</span> <span class="n">SraSearch</span><span class="p">(</span>
                <span class="n">organism</span><span class="o">=</span><span class="n">species</span><span class="p">,</span>
                <span class="n">layout</span><span class="o">=</span><span class="s2">&quot;paired&quot;</span><span class="p">,</span>
                <span class="n">strategy</span><span class="o">=</span><span class="p">[</span><span class="s2">&quot;OTHER&quot;</span><span class="p">],</span>
                <span class="n">return_max</span><span class="o">=</span><span class="n">additional_results_needed</span><span class="p">,</span>
            <span class="p">)</span>
            <span class="n">sra_search_other</span><span class="o">.</span><span class="n">search</span><span class="p">()</span>
            <span class="n">df_other</span> <span class="o">=</span> <span class="n">sra_search_other</span><span class="o">.</span><span class="n">get_df</span><span class="p">()</span>

        <span class="c1"># Combine the two DataFrames</span>
        <span class="n">df_combined</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">concat</span><span class="p">([</span><span class="n">df_rna_seq</span><span class="p">,</span> <span class="n">df_other</span><span class="p">],</span> <span class="n">ignore_index</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
        <span class="c1"># print(f&quot;{species} df_combined:&quot;, df_combined.columns)</span>

        <span class="c1"># Filter these combined dataframes to ensure they match the taxonomy ID</span>
        <span class="n">df_filtered</span> <span class="o">=</span> <span class="n">df_combined</span><span class="p">[</span><span class="n">df_combined</span><span class="p">[</span><span class="s2">&quot;sample_taxon_id&quot;</span><span class="p">]</span> <span class="o">==</span> <span class="nb">str</span><span class="p">(</span><span class="n">taxonomy_id</span><span class="p">)]</span>

        <span class="k">return</span> <span class="n">df_filtered</span> <span class="k">if</span> <span class="ow">not</span> <span class="n">df_filtered</span><span class="o">.</span><span class="n">empty</span> <span class="k">else</span> <span class="kc">None</span>

    <span class="k">except</span> <span class="p">(</span><span class="ne">ValueError</span><span class="p">,</span> <span class="ne">KeyError</span><span class="p">,</span> <span class="ne">AttributeError</span><span class="p">)</span> <span class="k">as</span> <span class="n">e</span><span class="p">:</span>
        <span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Error querying </span><span class="si">{</span><span class="n">species</span><span class="si">}</span><span class="s2">: </span><span class="si">{</span><span class="n">e</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
        <span class="k">return</span> <span class="kc">None</span>
    <span class="k">except</span> <span class="ne">Exception</span> <span class="k">as</span> <span class="n">e</span><span class="p">:</span>
        <span class="nb">print</span><span class="p">(</span>
            <span class="sa">f</span><span class="s2">&quot;Error querying </span><span class="si">{</span><span class="n">species</span><span class="si">}</span><span class="s2">, possibly due to external libraries or API (SraSearch): </span><span class="si">{</span><span class="n">e</span><span class="si">}</span><span class="s2">&quot;</span>
        <span class="p">)</span>
        <span class="k">return</span> <span class="kc">None</span>
</code></pre></div></td></tr></table></div>
          </details>
  </div>

</div>


<div class="doc doc-object doc-function">



<h2 id="rna.rna_download_logic.query_and_csv_production.view_srx_metadata" class="doc doc-heading">
          <code class="highlight language-python"><span class="n">view_srx_metadata</span><span class="p">(</span><span class="n">species_srx_map</span><span class="p">)</span></code>

</h2>


  <div class="doc doc-contents ">
  
      <p>Get metadata for all species and experiment accession numbers.</p>



<table class="field-list">
  <colgroup>
    <col class="field-name" />
    <col class="field-body" />
  </colgroup>
  <tbody valign="top">
    <tr class="field">
      <th class="field-name">Parameters:</th>
      <td class="field-body">
        <ul class="first simple">
            <li>
              <b><code>species_srx_map</code></b>
                  (<code><span title="typing.Dict">Dict</span>[str, <span title="typing.List">List</span>[str]]</code>)
              –
              <div class="doc-md-description">
                <p>A dictionary with species names as keys and SRX (experiment accession) IDs as values.</p>
              </div>
            </li>
        </ul>
      </td>
    </tr>
  </tbody>
</table>


<table class="field-list">
  <colgroup>
    <col class="field-name" />
    <col class="field-body" />
  </colgroup>
  <tbody valign="top">
    <tr class="field">
    <th class="field-name">Returns:</th>
    <td class="field-body">
      <ul class="first simple">
          <li>
                <code><span title="typing.Dict">Dict</span>[str, <span title="pandas.DataFrame">DataFrame</span>]</code>
            –
            <div class="doc-md-description">
              <p>Dict[str, pd.DataFrame]: A dictionary with species name as keys and dataframes with experiment metadata as values.</p>
            </div>
          </li>
      </ul>
    </td>
    </tr>
  </tbody>
</table>
          <details class="quote">
            <summary>Source code in <code>rna\rna_download_logic\query_and_csv_production.py</code></summary>
            <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal"> 87</span>
<span class="normal"> 88</span>
<span class="normal"> 89</span>
<span class="normal"> 90</span>
<span class="normal"> 91</span>
<span class="normal"> 92</span>
<span class="normal"> 93</span>
<span class="normal"> 94</span>
<span class="normal"> 95</span>
<span class="normal"> 96</span>
<span class="normal"> 97</span>
<span class="normal"> 98</span>
<span class="normal"> 99</span>
<span class="normal">100</span>
<span class="normal">101</span>
<span class="normal">102</span>
<span class="normal">103</span>
<span class="normal">104</span>
<span class="normal">105</span>
<span class="normal">106</span>
<span class="normal">107</span>
<span class="normal">108</span>
<span class="normal">109</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span> <span class="nf">view_srx_metadata</span><span class="p">(</span><span class="n">species_srx_map</span><span class="p">:</span> <span class="n">Dict</span><span class="p">[</span><span class="nb">str</span><span class="p">,</span> <span class="n">List</span><span class="p">[</span><span class="nb">str</span><span class="p">]])</span> <span class="o">-&gt;</span> <span class="n">Dict</span><span class="p">[</span><span class="nb">str</span><span class="p">,</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">]:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Get metadata for all species and experiment accession numbers.</span>

<span class="sd">    Args:</span>
<span class="sd">        species_srx_map (Dict[str, List[str]]): A dictionary with species names as keys and SRX (experiment accession) IDs as values.</span>

<span class="sd">    Returns:</span>
<span class="sd">        Dict[str, pd.DataFrame]: A dictionary with species name as keys and dataframes with experiment metadata as values.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">db</span> <span class="o">=</span> <span class="n">SRAweb</span><span class="p">()</span>

    <span class="n">all_species_metadata</span> <span class="o">=</span> <span class="p">{}</span>

    <span class="k">for</span> <span class="n">species</span><span class="p">,</span> <span class="n">srx_ids</span> <span class="ow">in</span> <span class="n">species_srx_map</span><span class="o">.</span><span class="n">items</span><span class="p">():</span>
        <span class="n">list_ids</span> <span class="o">=</span> <span class="nb">list</span><span class="p">(</span><span class="n">srx_ids</span><span class="p">)</span>
        <span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Processing </span><span class="si">{</span><span class="n">species</span><span class="si">}</span><span class="s2"> with </span><span class="si">{</span><span class="nb">len</span><span class="p">(</span><span class="n">list_ids</span><span class="p">)</span><span class="si">}</span><span class="s2"> SRX IDs...&quot;</span><span class="p">)</span>

        <span class="n">df</span> <span class="o">=</span> <span class="n">db</span><span class="o">.</span><span class="n">sra_metadata</span><span class="p">(</span><span class="n">list_ids</span><span class="p">)</span>

        <span class="n">all_species_metadata</span><span class="p">[</span><span class="n">species</span><span class="p">]</span> <span class="o">=</span> <span class="n">df</span>
    <span class="nb">print</span><span class="p">(</span><span class="nb">list</span><span class="p">(</span><span class="n">all_species_metadata</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="o">.</span><span class="n">columns</span><span class="p">))</span>

    <span class="k">return</span> <span class="n">all_species_metadata</span>
</code></pre></div></td></tr></table></div>
          </details>
  </div>

</div>



  </div>

  </div>

</div>
              
            </div>
          </div><footer>
    <div class="rst-footer-buttons" role="navigation" aria-label="Footer Navigation">
        <a href="../mRNA_fastq_download/" class="btn btn-neutral float-left" title="mRNA_fastq_download"><span class="icon icon-circle-arrow-left"></span> Previous</a>
        <a href="../../rna_extraction/" class="btn btn-neutral float-right" title="rna_extraction">Next <span class="icon icon-circle-arrow-right"></span></a>
    </div>

  <hr/>

  <div role="contentinfo">
    <!-- Copyright etc -->
  </div>

  Built with <a href="https://www.mkdocs.org/">MkDocs</a> using a <a href="https://github.com/readthedocs/sphinx_rtd_theme">theme</a> provided by <a href="https://readthedocs.org">Read the Docs</a>.
</footer>
          
        </div>
      </div>

    </section>

  </div>

  <div class="rst-versions" role="note" aria-label="Versions">
  <span class="rst-current-version" data-toggle="rst-current-version">
    
        <span>
          <a href="https://github.com/TransferLearningForDNA/genomic_data_extraction" class="fa fa-github" style="color: #fcfcfc"> GitHub</a>
        </span>
    
    
      <span><a href="../mRNA_fastq_download/" style="color: #fcfcfc">&laquo; Previous</a></span>
    
    
      <span><a href="../../rna_extraction/" style="color: #fcfcfc">Next &raquo;</a></span>
    
  </span>
</div>
    <script>var base_url = '../../../..';</script>
    <script src="../../../../js/theme_extra.js" defer></script>
    <script src="../../../../js/theme.js" defer></script>
    <script defer>
        window.onload = function () {
            SphinxRtdTheme.Navigation.enable(true);
        };
    </script>

</body>
</html>
