{
  "cells": [
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "CgHb92kfVryX"
      },
      "source": [
        "# Using Emsembl-REST"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "Vt5KaGiiYbqY"
      },
      "source": [
        "## Import statements"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 8,
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "2B0_kEtcVm4A",
        "outputId": "c58f1b25-0590-4aa8-a073-03af45df908b"
      },
      "outputs": [
        {
          "name": "stdout",
          "output_type": "stream",
          "text": [
            "Requirement already satisfied: ensembl_rest in c:\\users\\aayus\\appdata\\local\\packages\\pythonsoftwarefoundation.python.3.10_qbz5n2kfra8p0\\localcache\\local-packages\\python310\\site-packages (0.3.4)\n",
            "Requirement already satisfied: importlib-resources in c:\\users\\aayus\\appdata\\local\\packages\\pythonsoftwarefoundation.python.3.10_qbz5n2kfra8p0\\localcache\\local-packages\\python310\\site-packages (from ensembl_rest) (6.1.1)\n",
            "Requirement already satisfied: requests in c:\\users\\aayus\\appdata\\local\\packages\\pythonsoftwarefoundation.python.3.10_qbz5n2kfra8p0\\localcache\\local-packages\\python310\\site-packages (from ensembl_rest) (2.31.0)\n",
            "Requirement already satisfied: simplejson in c:\\users\\aayus\\appdata\\local\\packages\\pythonsoftwarefoundation.python.3.10_qbz5n2kfra8p0\\localcache\\local-packages\\python310\\site-packages (from ensembl_rest) (3.19.2)\n",
            "Requirement already satisfied: intervaltree in c:\\users\\aayus\\appdata\\local\\packages\\pythonsoftwarefoundation.python.3.10_qbz5n2kfra8p0\\localcache\\local-packages\\python310\\site-packages (from ensembl_rest) (3.1.0)\n",
            "Requirement already satisfied: sortedcontainers<3.0,>=2.0 in c:\\users\\aayus\\appdata\\local\\packages\\pythonsoftwarefoundation.python.3.10_qbz5n2kfra8p0\\localcache\\local-packages\\python310\\site-packages (from intervaltree->ensembl_rest) (2.4.0)\n",
            "Requirement already satisfied: charset-normalizer<4,>=2 in c:\\users\\aayus\\appdata\\local\\packages\\pythonsoftwarefoundation.python.3.10_qbz5n2kfra8p0\\localcache\\local-packages\\python310\\site-packages (from requests->ensembl_rest) (3.3.0)\n",
            "Requirement already satisfied: idna<4,>=2.5 in c:\\users\\aayus\\appdata\\local\\packages\\pythonsoftwarefoundation.python.3.10_qbz5n2kfra8p0\\localcache\\local-packages\\python310\\site-packages (from requests->ensembl_rest) (3.4)\n",
            "Requirement already satisfied: urllib3<3,>=1.21.1 in c:\\users\\aayus\\appdata\\local\\packages\\pythonsoftwarefoundation.python.3.10_qbz5n2kfra8p0\\localcache\\local-packages\\python310\\site-packages (from requests->ensembl_rest) (2.0.6)\n",
            "Requirement already satisfied: certifi>=2017.4.17 in c:\\users\\aayus\\appdata\\local\\packages\\pythonsoftwarefoundation.python.3.10_qbz5n2kfra8p0\\localcache\\local-packages\\python310\\site-packages (from requests->ensembl_rest) (2023.7.22)\n",
            "Note: you may need to restart the kernel to use updated packages.\n"
          ]
        }
      ],
      "source": [
        "pip install ensembl_rest"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 9,
      "metadata": {
        "id": "vz8EnuT1V0Dg"
      },
      "outputs": [],
      "source": [
        "import ensembl_rest\n",
        "import pandas as pd\n",
        "import requests\n",
        "import csv\n",
        "import re\n",
        "import sys\n"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 10,
      "metadata": {
        "id": "HN2PsiHeWSal"
      },
      "outputs": [
        {
          "name": "stdout",
          "output_type": "stream",
          "text": [
            "Help on package ensembl_rest:\n",
            "\n",
            "NAME\n",
            "    ensembl_rest\n",
            "\n",
            "PACKAGE CONTENTS\n",
            "    assemblymapper\n",
            "    core (package)\n",
            "    data (package)\n",
            "    ensemblclient\n",
            "    utils\n",
            "\n",
            "FUNCTIONS\n",
            "    VariantAnnotationSet(*args, **kwargs)\n",
            "            Variation GA4GH ``POST ga4gh/variantannotationsets/search``\n",
            "        \n",
            "        Return a list of annotation sets in GA4GH format\n",
            "        \n",
            "        \n",
            "        **Parameters**\n",
            "        \n",
            "        - Required:\n",
            "        \n",
            "          + *Name*: \n",
            "        \n",
            "            * *Type*: String\n",
            "            * *Description*: \n",
            "            * *Default*: -\n",
            "            * *Example Values*: 1\n",
            "        \n",
            "        \n",
            "        - Optional:\n",
            "        \n",
            "          + *Name*: callback\n",
            "        \n",
            "            * *Type*: String\n",
            "            * *Description*: Name of the callback subroutine to be returned by the requested JSONP response. Required ONLY when using JSONP as the serialisation method. Please see , .\n",
            "            * *Default*: -\n",
            "            * *Example Values*: randomlygeneratedname\n",
            "        \n",
            "        \n",
            "          + *Name*: pageSize\n",
            "        \n",
            "            * *Type*: Int\n",
            "            * *Description*: \n",
            "            * *Default*: 10\n",
            "            * *Example Values*: -\n",
            "        \n",
            "        \n",
            "          + *Name*: pageToken\n",
            "        \n",
            "            * *Type*: Int\n",
            "            * *Description*: Identifier showing which page of data to retrieve next\n",
            "            * *Default*: null\n",
            "            * *Example Values*: -\n",
            "        \n",
            "        \n",
            "        **Resource info**\n",
            "        \n",
            "        - **Methods**:  POST\n",
            "        - **Response formats**: json, jsonp\n",
            "        \n",
            "        \n",
            "        **More info**\n",
            "        \n",
            "        https://rest.ensembl.org/documentation/info/VariantAnnotationSet\n",
            "    \n",
            "    VariantAnnotationSet_id(*args, **kwargs)\n",
            "            Variation GA4GH ``GET ga4gh/variantannotationsets/:id``\n",
            "        \n",
            "        Return meta data for a specific annotation set in GA4GH format\n",
            "        \n",
            "        \n",
            "        **Parameters**\n",
            "        \n",
            "        - Required:\n",
            "        \n",
            "          + *Name*: id\n",
            "        \n",
            "            * *Type*: String\n",
            "            * *Description*: VariantAnnotation set id\n",
            "            * *Default*: -\n",
            "            * *Example Values*: Ensembl\n",
            "        \n",
            "        \n",
            "        - Optional:\n",
            "        \n",
            "          + *Name*: callback\n",
            "        \n",
            "            * *Type*: String\n",
            "            * *Description*: Name of the callback subroutine to be returned by the requested JSONP response. Required ONLY when using JSONP as the serialisation method. Please see , .\n",
            "            * *Default*: -\n",
            "            * *Example Values*: randomlygeneratedname\n",
            "        \n",
            "        \n",
            "        **Resource info**\n",
            "        \n",
            "        - **Methods**:  GET\n",
            "        - **Response formats**: json, jsonp\n",
            "        \n",
            "        \n",
            "        **More info**\n",
            "        \n",
            "        https://rest.ensembl.org/documentation/info/VariantAnnotationSet_id\n",
            "    \n",
            "    analysis(*args, **kwargs)\n",
            "            Information ``GET info/analysis/:species``\n",
            "        \n",
            "        List the names of analyses involved in generating Ensembl data.\n",
            "        \n",
            "        \n",
            "        **Parameters**\n",
            "        \n",
            "        - Required:\n",
            "        \n",
            "          + *Name*: species\n",
            "        \n",
            "            * *Type*: String\n",
            "            * *Description*: Species name/alias\n",
            "            * *Default*: -\n",
            "            * *Example Values*: homo_sapiens\n",
            "        \n",
            "        \n",
            "        - Optional:\n",
            "        \n",
            "          + *Name*: callback\n",
            "        \n",
            "            * *Type*: String\n",
            "            * *Description*: Name of the callback subroutine to be returned by the requested JSONP response. Required ONLY when using JSONP as the serialisation method. Please see , .\n",
            "            * *Default*: -\n",
            "            * *Example Values*: randomlygeneratedname\n",
            "        \n",
            "        \n",
            "        **Resource info**\n",
            "        \n",
            "        - **Methods**:  GET\n",
            "        - **Response formats**: json, xml, jsonp\n",
            "        \n",
            "        \n",
            "        **More info**\n",
            "        \n",
            "        https://rest.ensembl.org/documentation/info/analysis\n",
            "    \n",
            "    archive_id_get(*args, **kwargs)\n",
            "            Archive ``GET archive/id/:id``\n",
            "        \n",
            "        Uses the given identifier to return its latest version\n",
            "        \n",
            "        \n",
            "        **Parameters**\n",
            "        \n",
            "        - Required:\n",
            "        \n",
            "          + *Name*: id\n",
            "        \n",
            "            * *Type*: String\n",
            "            * *Description*: An Ensembl stable ID\n",
            "            * *Default*: -\n",
            "            * *Example Values*: ENSG00000157764\n",
            "        \n",
            "        \n",
            "        - Optional:\n",
            "        \n",
            "          + *Name*: callback\n",
            "        \n",
            "            * *Type*: String\n",
            "            * *Description*: Name of the callback subroutine to be returned by the requested JSONP response. Required ONLY when using JSONP as the serialisation method. Please see , .\n",
            "            * *Default*: -\n",
            "            * *Example Values*: randomlygeneratedname\n",
            "        \n",
            "        \n",
            "        **Resource info**\n",
            "        \n",
            "        - **Methods**:  GET\n",
            "        - **Response formats**: json, xml, jsonp\n",
            "        \n",
            "        \n",
            "        **More info**\n",
            "        \n",
            "        https://rest.ensembl.org/documentation/info/archive_id_get\n",
            "    \n",
            "    archive_id_post(*args, **kwargs)\n",
            "            Archive ``POST archive/id``\n",
            "        \n",
            "        Retrieve the latest version for a set of identifiers\n",
            "        \n",
            "        \n",
            "        **Parameters**\n",
            "        \n",
            "        - Required:\n",
            "        \n",
            "          + *Name*: callback\n",
            "        \n",
            "            * *Type*: String\n",
            "            * *Description*: Name of the callback subroutine to be returned by the requested JSONP response. Required ONLY when using JSONP as the serialisation method. Please see , .\n",
            "            * *Default*: -\n",
            "            * *Example Values*: randomlygeneratedname\n",
            "        \n",
            "        \n",
            "        - Optional:\n",
            "        \n",
            "          + *Name*: callback\n",
            "        \n",
            "            * *Type*: String\n",
            "            * *Description*: Name of the callback subroutine to be returned by the requested JSONP response. Required ONLY when using JSONP as the serialisation method. Please see , .\n",
            "            * *Default*: -\n",
            "            * *Example Values*: randomlygeneratedname\n",
            "        \n",
            "        \n",
            "        **Resource info**\n",
            "        \n",
            "        - **Methods**:  POST\n",
            "        - **Response formats**: json, jsonp\n",
            "        - **Maximum POST size**:        1000\n",
            "        \n",
            "        \n",
            "        **More info**\n",
            "        \n",
            "        https://rest.ensembl.org/documentation/info/archive_id_post\n",
            "    \n",
            "    array(*args, **kwargs)\n",
            "            Regulation ``GET regulatory/species/:species/microarray/:microarray/vendor/:vendor``\n",
            "        \n",
            "        Returns information about a specific microarray\n",
            "        \n",
            "        \n",
            "        **Parameters**\n",
            "        \n",
            "        - Required:\n",
            "        \n",
            "          + *Name*: microarray\n",
            "        \n",
            "            * *Type*: String\n",
            "            * *Description*: Microarray name\n",
            "            * *Default*: -\n",
            "            * *Example Values*: HumanWG_6_V2\n",
            "        \n",
            "        \n",
            "          + *Name*: species\n",
            "        \n",
            "            * *Type*: String\n",
            "            * *Description*: Species name/alias\n",
            "            * *Default*: -\n",
            "            * *Example Values*: homo_sapiens\n",
            "        \n",
            "        \n",
            "          + *Name*: vendor\n",
            "        \n",
            "            * *Type*: String\n",
            "            * *Description*: Probe name\n",
            "            * *Default*: -\n",
            "            * *Example Values*: ILMN_1763508\n",
            "        \n",
            "        \n",
            "        - Optional:\n",
            "        \n",
            "          + *Name*: callback\n",
            "        \n",
            "            * *Type*: String\n",
            "            * *Description*: Name of the callback subroutine to be returned by the requested JSONP response. Required ONLY when using JSONP as the serialisation method. Please see , .\n",
            "            * *Default*: -\n",
            "            * *Example Values*: randomlygeneratedname\n",
            "        \n",
            "        \n",
            "        **Resource info**\n",
            "        \n",
            "        - **Methods**:  GET\n",
            "        - **Response formats**: json, xml, jsonp\n",
            "        \n",
            "        \n",
            "        **More info**\n",
            "        \n",
            "        https://rest.ensembl.org/documentation/info/array\n",
            "    \n",
            "    assembly_cdna(*args, **kwargs)\n",
            "            Mapping ``GET map/cdna/:id/:region``\n",
            "        \n",
            "        Convert from cDNA coordinates to genomic coordinates. Output reflects forward orientation coordinates as returned from the Ensembl API.\n",
            "        \n",
            "        \n",
            "        **Parameters**\n",
            "        \n",
            "        - Required:\n",
            "        \n",
            "          + *Name*: id\n",
            "        \n",
            "            * *Type*: String\n",
            "            * *Description*: An Ensembl stable ID\n",
            "            * *Default*: -\n",
            "            * *Example Values*: ENST00000288602\n",
            "        \n",
            "        \n",
            "          + *Name*: region\n",
            "        \n",
            "            * *Type*: String\n",
            "            * *Description*: Query region\n",
            "            * *Default*: -\n",
            "            * *Example Values*: 100..300\n",
            "        \n",
            "        \n",
            "        - Optional:\n",
            "        \n",
            "          + *Name*: callback\n",
            "        \n",
            "            * *Type*: String\n",
            "            * *Description*: Name of the callback subroutine to be returned by the requested JSONP response. Required ONLY when using JSONP as the serialisation method. Please see , .\n",
            "            * *Default*: -\n",
            "            * *Example Values*: randomlygeneratedname\n",
            "        \n",
            "        \n",
            "          + *Name*: include_original_region\n",
            "        \n",
            "            * *Type*: Boolean(0,1)\n",
            "            * *Description*: Include original input region (cDNA coordinates) along with the target region (genomic coordinates) mappings.\n",
            "            * *Default*: 0\n",
            "            * *Example Values*: -\n",
            "        \n",
            "        \n",
            "          + *Name*: species\n",
            "        \n",
            "            * *Type*: String\n",
            "            * *Description*: Species name/alias\n",
            "            * *Default*: -\n",
            "            * *Example Values*: homo_sapiens, human\n",
            "        \n",
            "        \n",
            "        **Resource info**\n",
            "        \n",
            "        - **Methods**:  GET\n",
            "        - **Response formats**: json, xml, jsonp\n",
            "        \n",
            "        \n",
            "        **More info**\n",
            "        \n",
            "        https://rest.ensembl.org/documentation/info/assembly_cdna\n",
            "    \n",
            "    assembly_cds(*args, **kwargs)\n",
            "            Mapping ``GET map/cds/:id/:region``\n",
            "        \n",
            "        Convert from CDS coordinates to genomic coordinates. Output reflects forward orientation coordinates as returned from the Ensembl API.\n",
            "        \n",
            "        \n",
            "        **Parameters**\n",
            "        \n",
            "        - Required:\n",
            "        \n",
            "          + *Name*: id\n",
            "        \n",
            "            * *Type*: String\n",
            "            * *Description*: An Ensembl stable ID\n",
            "            * *Default*: -\n",
            "            * *Example Values*: ENST00000288602\n",
            "        \n",
            "        \n",
            "          + *Name*: region\n",
            "        \n",
            "            * *Type*: String\n",
            "            * *Description*: Query region\n",
            "            * *Default*: -\n",
            "            * *Example Values*: 1..1000\n",
            "        \n",
            "        \n",
            "        - Optional:\n",
            "        \n",
            "          + *Name*: callback\n",
            "        \n",
            "            * *Type*: String\n",
            "            * *Description*: Name of the callback subroutine to be returned by the requested JSONP response. Required ONLY when using JSONP as the serialisation method. Please see , .\n",
            "            * *Default*: -\n",
            "            * *Example Values*: randomlygeneratedname\n",
            "        \n",
            "        \n",
            "          + *Name*: include_original_region\n",
            "        \n",
            "            * *Type*: Boolean(0,1)\n",
            "            * *Description*: Include original input region (cds coordinates) along with the target region (genomic coordinates) mappings.\n",
            "            * *Default*: 0\n",
            "            * *Example Values*: -\n",
            "        \n",
            "        \n",
            "          + *Name*: species\n",
            "        \n",
            "            * *Type*: String\n",
            "            * *Description*: Species name/alias\n",
            "            * *Default*: -\n",
            "            * *Example Values*: homo_sapiens, human\n",
            "        \n",
            "        \n",
            "        **Resource info**\n",
            "        \n",
            "        - **Methods**:  GET\n",
            "        - **Response formats**: json, xml, jsonp\n",
            "        \n",
            "        \n",
            "        **More info**\n",
            "        \n",
            "        https://rest.ensembl.org/documentation/info/assembly_cds\n",
            "    \n",
            "    assembly_info(*args, **kwargs)\n",
            "            Information ``GET info/assembly/:species``\n",
            "        \n",
            "        List the currently available assemblies for a species, along with toplevel sequences, chromosomes and cytogenetic bands.\n",
            "        \n",
            "        \n",
            "        **Parameters**\n",
            "        \n",
            "        - Required:\n",
            "        \n",
            "          + *Name*: species\n",
            "        \n",
            "            * *Type*: String\n",
            "            * *Description*: Species name/alias\n",
            "            * *Default*: -\n",
            "            * *Example Values*: homo_sapiens, human\n",
            "        \n",
            "        \n",
            "        - Optional:\n",
            "        \n",
            "          + *Name*: bands\n",
            "        \n",
            "            * *Type*: Boolean(0,1)\n",
            "            * *Description*: If set to 1, include karyotype band information. Only display if band information is available\n",
            "            * *Default*: 0\n",
            "            * *Example Values*: -\n",
            "        \n",
            "        \n",
            "          + *Name*: callback\n",
            "        \n",
            "            * *Type*: String\n",
            "            * *Description*: Name of the callback subroutine to be returned by the requested JSONP response. Required ONLY when using JSONP as the serialisation method. Please see , .\n",
            "            * *Default*: -\n",
            "            * *Example Values*: randomlygeneratedname\n",
            "        \n",
            "        \n",
            "          + *Name*: synonyms\n",
            "        \n",
            "            * *Type*: Boolean(0,1)\n",
            "            * *Description*: If set to 1, include information about known synonyms.\n",
            "            * *Default*: 0\n",
            "            * *Example Values*: -\n",
            "        \n",
            "        \n",
            "        **Resource info**\n",
            "        \n",
            "        - **Methods**:  GET\n",
            "        - **Response formats**: json, xml, jsonp\n",
            "        \n",
            "        \n",
            "        **More info**\n",
            "        \n",
            "        https://rest.ensembl.org/documentation/info/assembly_info\n",
            "    \n",
            "    assembly_map(*args, **kwargs)\n",
            "            Mapping ``GET map/:species/:asm_one/:region/:asm_two``\n",
            "        \n",
            "        Convert the co-ordinates of one assembly to another\n",
            "        \n",
            "        \n",
            "        **Parameters**\n",
            "        \n",
            "        - Required:\n",
            "        \n",
            "          + *Name*: asm_one\n",
            "        \n",
            "            * *Type*: String\n",
            "            * *Description*: Version of the input assembly\n",
            "            * *Default*: -\n",
            "            * *Example Values*: GRCh37\n",
            "        \n",
            "        \n",
            "          + *Name*: asm_two\n",
            "        \n",
            "            * *Type*: String\n",
            "            * *Description*: Version of the output assembly\n",
            "            * *Default*: -\n",
            "            * *Example Values*: GRCh38\n",
            "        \n",
            "        \n",
            "          + *Name*: region\n",
            "        \n",
            "            * *Type*: String\n",
            "            * *Description*: Query region\n",
            "            * *Default*: -\n",
            "            * *Example Values*: X:1000000..1000100:1, X:1000000..1000100:-1, X:1000000..1000100\n",
            "        \n",
            "        \n",
            "          + *Name*: species\n",
            "        \n",
            "            * *Type*: String\n",
            "            * *Description*: Species name/alias\n",
            "            * *Default*: -\n",
            "            * *Example Values*: homo_sapiens, human\n",
            "        \n",
            "        \n",
            "        - Optional:\n",
            "        \n",
            "          + *Name*: callback\n",
            "        \n",
            "            * *Type*: String\n",
            "            * *Description*: Name of the callback subroutine to be returned by the requested JSONP response. Required ONLY when using JSONP as the serialisation method. Please see , .\n",
            "            * *Default*: -\n",
            "            * *Example Values*: randomlygeneratedname\n",
            "        \n",
            "        \n",
            "          + *Name*: coord_system\n",
            "        \n",
            "            * *Type*: String\n",
            "            * *Description*: Name of the input coordinate system\n",
            "            * *Default*: chromosome\n",
            "            * *Example Values*: chromosome\n",
            "        \n",
            "        \n",
            "          + *Name*: target_coord_system\n",
            "        \n",
            "            * *Type*: String\n",
            "            * *Description*: Name of the output coordinate system\n",
            "            * *Default*: chromosome\n",
            "            * *Example Values*: chromosome\n",
            "        \n",
            "        \n",
            "        **Resource info**\n",
            "        \n",
            "        - **Methods**:  GET\n",
            "        - **Response formats**: json, xml, jsonp\n",
            "        \n",
            "        \n",
            "        **More info**\n",
            "        \n",
            "        https://rest.ensembl.org/documentation/info/assembly_map\n",
            "    \n",
            "    assembly_stats(*args, **kwargs)\n",
            "            Information ``GET info/assembly/:species/:region_name``\n",
            "        \n",
            "        Returns information about the specified toplevel sequence region for the given species.\n",
            "        \n",
            "        \n",
            "        **Parameters**\n",
            "        \n",
            "        - Required:\n",
            "        \n",
            "          + *Name*: region_name\n",
            "        \n",
            "            * *Type*: String\n",
            "            * *Description*: The (top level) sequence region name.\n",
            "            * *Default*: -\n",
            "            * *Example Values*: X\n",
            "        \n",
            "        \n",
            "          + *Name*: species\n",
            "        \n",
            "            * *Type*: String\n",
            "            * *Description*: Species name/alias\n",
            "            * *Default*: -\n",
            "            * *Example Values*: homo_sapiens, human\n",
            "        \n",
            "        \n",
            "        - Optional:\n",
            "        \n",
            "          + *Name*: bands\n",
            "        \n",
            "            * *Type*: Boolean(0,1)\n",
            "            * *Description*: If set to 1, include karyotype band information. Only display if band information is available\n",
            "            * *Default*: 0\n",
            "            * *Example Values*: -\n",
            "        \n",
            "        \n",
            "          + *Name*: callback\n",
            "        \n",
            "            * *Type*: String\n",
            "            * *Description*: Name of the callback subroutine to be returned by the requested JSONP response. Required ONLY when using JSONP as the serialisation method. Please see , .\n",
            "            * *Default*: -\n",
            "            * *Example Values*: randomlygeneratedname\n",
            "        \n",
            "        \n",
            "          + *Name*: synonyms\n",
            "        \n",
            "            * *Type*: Boolean(0,1)\n",
            "            * *Description*: If set to 1, include information about known synonyms\n",
            "            * *Default*: 0\n",
            "            * *Example Values*: -\n",
            "        \n",
            "        \n",
            "        **Resource info**\n",
            "        \n",
            "        - **Methods**:  GET\n",
            "        - **Response formats**: json, xml, jsonp\n",
            "        \n",
            "        \n",
            "        **More info**\n",
            "        \n",
            "        https://rest.ensembl.org/documentation/info/assembly_stats\n",
            "    \n",
            "    assembly_translation(*args, **kwargs)\n",
            "            Mapping ``GET map/translation/:id/:region``\n",
            "        \n",
            "        Convert from protein (translation) coordinates to genomic coordinates. Output reflects forward orientation coordinates as returned from the Ensembl API.\n",
            "        \n",
            "        \n",
            "        **Parameters**\n",
            "        \n",
            "        - Required:\n",
            "        \n",
            "          + *Name*: id\n",
            "        \n",
            "            * *Type*: String\n",
            "            * *Description*: An Ensembl stable ID\n",
            "            * *Default*: -\n",
            "            * *Example Values*: ENSP00000288602\n",
            "        \n",
            "        \n",
            "          + *Name*: region\n",
            "        \n",
            "            * *Type*: String\n",
            "            * *Description*: Query region\n",
            "            * *Default*: -\n",
            "            * *Example Values*: 100..300\n",
            "        \n",
            "        \n",
            "        - Optional:\n",
            "        \n",
            "          + *Name*: callback\n",
            "        \n",
            "            * *Type*: String\n",
            "            * *Description*: Name of the callback subroutine to be returned by the requested JSONP response. Required ONLY when using JSONP as the serialisation method. Please see , .\n",
            "            * *Default*: -\n",
            "            * *Example Values*: randomlygeneratedname\n",
            "        \n",
            "        \n",
            "          + *Name*: species\n",
            "        \n",
            "            * *Type*: String\n",
            "            * *Description*: Species name/alias\n",
            "            * *Default*: -\n",
            "            * *Example Values*: homo_sapiens, human\n",
            "        \n",
            "        \n",
            "        **Resource info**\n",
            "        \n",
            "        - **Methods**:  GET\n",
            "        - **Response formats**: json, xml, jsonp\n",
            "        \n",
            "        \n",
            "        **More info**\n",
            "        \n",
            "        https://rest.ensembl.org/documentation/info/assembly_translation\n",
            "    \n",
            "    beacon_get(*args, **kwargs)\n",
            "            Variation GA4GH ``GET ga4gh/beacon``\n",
            "        \n",
            "        Return Beacon information\n",
            "        \n",
            "        \n",
            "        **Parameters**\n",
            "        \n",
            "        - Required:\n",
            "        \n",
            "          + *Name*: callback\n",
            "        \n",
            "            * *Type*: String\n",
            "            * *Description*: Name of the callback subroutine to be returned by the requested JSONP response. Required ONLY when using JSONP as the serialisation method. Please see , .\n",
            "            * *Default*: -\n",
            "            * *Example Values*: randomlygeneratedname\n",
            "        \n",
            "        \n",
            "        - Optional:\n",
            "        \n",
            "          + *Name*: callback\n",
            "        \n",
            "            * *Type*: String\n",
            "            * *Description*: Name of the callback subroutine to be returned by the requested JSONP response. Required ONLY when using JSONP as the serialisation method. Please see , .\n",
            "            * *Default*: -\n",
            "            * *Example Values*: randomlygeneratedname\n",
            "        \n",
            "        \n",
            "        **Resource info**\n",
            "        \n",
            "        - **Methods**:  GET\n",
            "        - **Response formats**: json, xml, jsonp\n",
            "        \n",
            "        \n",
            "        **More info**\n",
            "        \n",
            "        https://rest.ensembl.org/documentation/info/beacon_get\n",
            "    \n",
            "    beacon_query_get(*args, **kwargs)\n",
            "            Variation GA4GH ``GET ga4gh/beacon/query``\n",
            "        \n",
            "        Return the Beacon response for allele information\n",
            "        \n",
            "        \n",
            "        **Parameters**\n",
            "        \n",
            "        - Required:\n",
            "        \n",
            "          + *Name*: alternateBases\n",
            "        \n",
            "            * *Type*: String\n",
            "            * *Description*: The bases that appear instead of the reference bases. Accepted values: see the ALT field in VCF 4.2 specification (, ).\n",
            "            * *Default*: -\n",
            "            * *Example Values*: C\n",
            "        \n",
            "        \n",
            "          + *Name*: assemblyId\n",
            "        \n",
            "            * *Type*: String\n",
            "            * *Description*: Assembly identifier (GRC notation, e.g. GRCh38).\n",
            "            * *Default*: -\n",
            "            * *Example Values*: GRCh38\n",
            "        \n",
            "        \n",
            "          + *Name*: referenceBases\n",
            "        \n",
            "            * *Type*: String\n",
            "            * *Description*: \n",
            "            * *Default*: -\n",
            "            * *Example Values*: G\n",
            "        \n",
            "        \n",
            "          + *Name*: referenceName\n",
            "        \n",
            "            * *Type*: String\n",
            "            * *Description*: Reference name (chromosome). Accepted values: 1-22, X, Y.\n",
            "            * *Default*: -\n",
            "            * *Example Values*: 9\n",
            "        \n",
            "        \n",
            "          + *Name*: start\n",
            "        \n",
            "            * *Type*: Int\n",
            "            * *Description*: Position, allele locus (0-based). Accepted values: non-negative integers smaller than reference length.\n",
            "            * *Default*: -\n",
            "            * *Example Values*: 22125503\n",
            "        \n",
            "        \n",
            "        - Optional:\n",
            "        \n",
            "          + *Name*: callback\n",
            "        \n",
            "            * *Type*: String\n",
            "            * *Description*: Name of the callback subroutine to be returned by the requested JSONP response. Required ONLY when using JSONP as the serialisation method. Please see , .\n",
            "            * *Default*: -\n",
            "            * *Example Values*: randomlygeneratedname\n",
            "        \n",
            "        \n",
            "          + *Name*: datasetIds\n",
            "        \n",
            "            * *Type*: array of strings\n",
            "            * *Description*: Identifiers of datasets. Option not used currently as single dataset.\n",
            "            * *Default*: -\n",
            "            * *Example Values*: -\n",
            "        \n",
            "        \n",
            "          + *Name*: includeDatasetResponses\n",
            "        \n",
            "            * *Type*: boolean\n",
            "            * *Description*: Indicator of whether responses for individual datasets should be included.\n",
            "            * *Default*: false\n",
            "            * *Example Values*: -\n",
            "        \n",
            "        \n",
            "        **Resource info**\n",
            "        \n",
            "        - **Methods**:  GET\n",
            "        - **Response formats**: json, jsonp\n",
            "        \n",
            "        \n",
            "        **More info**\n",
            "        \n",
            "        https://rest.ensembl.org/documentation/info/beacon_query_get\n",
            "    \n",
            "    beacon_query_post(*args, **kwargs)\n",
            "            Variation GA4GH ``POST ga4gh/beacon/query``\n",
            "        \n",
            "        Return the Beacon response for allele information\n",
            "        \n",
            "        \n",
            "        **Parameters**\n",
            "        \n",
            "        - Required:\n",
            "        \n",
            "          + *Name*: alternateBases\n",
            "        \n",
            "            * *Type*: String\n",
            "            * *Description*: The bases that appear instead of the reference bases. Accepted values: see the ALT field in VCF 4.2 specification (, ).\n",
            "            * *Default*: -\n",
            "            * *Example Values*: C\n",
            "        \n",
            "        \n",
            "          + *Name*: assemblyId\n",
            "        \n",
            "            * *Type*: String\n",
            "            * *Description*: Assembly identifier (GRC notation, e.g. GRCh38).\n",
            "            * *Default*: -\n",
            "            * *Example Values*: GRCh38\n",
            "        \n",
            "        \n",
            "          + *Name*: referenceBases\n",
            "        \n",
            "            * *Type*: String\n",
            "            * *Description*: \n",
            "            * *Default*: -\n",
            "            * *Example Values*: G\n",
            "        \n",
            "        \n",
            "          + *Name*: referenceName\n",
            "        \n",
            "            * *Type*: String\n",
            "            * *Description*: Reference name (chromosome). Accepted values: 1-22, X, Y.\n",
            "            * *Default*: -\n",
            "            * *Example Values*: 9\n",
            "        \n",
            "        \n",
            "          + *Name*: start\n",
            "        \n",
            "            * *Type*: Int\n",
            "            * *Description*: Position, allele locus (0-based). Accepted values: non-negative integers smaller than reference length.\n",
            "            * *Default*: -\n",
            "            * *Example Values*: 22125503\n",
            "        \n",
            "        \n",
            "        - Optional:\n",
            "        \n",
            "          + *Name*: callback\n",
            "        \n",
            "            * *Type*: String\n",
            "            * *Description*: Name of the callback subroutine to be returned by the requested JSONP response. Required ONLY when using JSONP as the serialisation method. Please see , .\n",
            "            * *Default*: -\n",
            "            * *Example Values*: randomlygeneratedname\n",
            "        \n",
            "        \n",
            "          + *Name*: datasetIds\n",
            "        \n",
            "            * *Type*: array of strings\n",
            "            * *Description*: Identifiers of datasets. Option not used currently as single dataset.\n",
            "            * *Default*: -\n",
            "            * *Example Values*: -\n",
            "        \n",
            "        \n",
            "          + *Name*: includeDatasetResponses\n",
            "        \n",
            "            * *Type*: boolean\n",
            "            * *Description*: Indicator of whether responses for individual datasets should be included.\n",
            "            * *Default*: false\n",
            "            * *Example Values*: -\n",
            "        \n",
            "        \n",
            "        **Resource info**\n",
            "        \n",
            "        - **Methods**:  POST\n",
            "        - **Response formats**: json, jsonp\n",
            "        \n",
            "        \n",
            "        **More info**\n",
            "        \n",
            "        https://rest.ensembl.org/documentation/info/beacon_query_post\n",
            "    \n",
            "    biotypes(*args, **kwargs)\n",
            "            Information ``GET info/biotypes/:species``\n",
            "        \n",
            "        List the functional classifications of gene models that Ensembl associates with a particular species. Useful for restricting the type of genes/transcripts retrieved by other endpoints.\n",
            "        \n",
            "        \n",
            "        **Parameters**\n",
            "        \n",
            "        - Required:\n",
            "        \n",
            "          + *Name*: species\n",
            "        \n",
            "            * *Type*: String\n",
            "            * *Description*: Species name/alias\n",
            "            * *Default*: -\n",
            "            * *Example Values*: homo_sapiens\n",
            "        \n",
            "        \n",
            "        - Optional:\n",
            "        \n",
            "          + *Name*: callback\n",
            "        \n",
            "            * *Type*: String\n",
            "            * *Description*: Name of the callback subroutine to be returned by the requested JSONP response. Required ONLY when using JSONP as the serialisation method. Please see , .\n",
            "            * *Default*: -\n",
            "            * *Example Values*: randomlygeneratedname\n",
            "        \n",
            "        \n",
            "        **Resource info**\n",
            "        \n",
            "        - **Methods**:  GET\n",
            "        - **Response formats**: json, xml, jsonp\n",
            "        \n",
            "        \n",
            "        **More info**\n",
            "        \n",
            "        https://rest.ensembl.org/documentation/info/biotypes\n",
            "    \n",
            "    biotypes_groups(*args, **kwargs)\n",
            "            Information ``GET info/biotypes/groups/:group/:object_type``\n",
            "        \n",
            "        Without argument the list of available biotype groups is returned. With :group argument provided, list the properties of biotypes within that group. Object type (gene or transcript) can be provided for filtering.\n",
            "        \n",
            "        \n",
            "        **Parameters**\n",
            "        \n",
            "        - Required:\n",
            "        \n",
            "          + *Name*: callback\n",
            "        \n",
            "            * *Type*: String\n",
            "            * *Description*: Name of the callback subroutine to be returned by the requested JSONP response. Required ONLY when using JSONP as the serialisation method. Please see , .\n",
            "            * *Default*: -\n",
            "            * *Example Values*: randomlygeneratedname\n",
            "        \n",
            "        \n",
            "          + *Name*: group\n",
            "        \n",
            "            * *Type*: String\n",
            "            * *Description*: Biotype group\n",
            "            * *Default*: -\n",
            "            * *Example Values*: coding\n",
            "        \n",
            "        \n",
            "          + *Name*: object_type\n",
            "        \n",
            "            * *Type*: String\n",
            "            * *Description*: Object type (gene or transcript)\n",
            "            * *Default*: -\n",
            "            * *Example Values*: gene\n",
            "        \n",
            "        \n",
            "        - Optional:\n",
            "        \n",
            "          + *Name*: callback\n",
            "        \n",
            "            * *Type*: String\n",
            "            * *Description*: Name of the callback subroutine to be returned by the requested JSONP response. Required ONLY when using JSONP as the serialisation method. Please see , .\n",
            "            * *Default*: -\n",
            "            * *Example Values*: randomlygeneratedname\n",
            "        \n",
            "        \n",
            "          + *Name*: group\n",
            "        \n",
            "            * *Type*: String\n",
            "            * *Description*: Biotype group\n",
            "            * *Default*: -\n",
            "            * *Example Values*: coding\n",
            "        \n",
            "        \n",
            "          + *Name*: object_type\n",
            "        \n",
            "            * *Type*: String\n",
            "            * *Description*: Object type (gene or transcript)\n",
            "            * *Default*: -\n",
            "            * *Example Values*: gene\n",
            "        \n",
            "        \n",
            "        **Resource info**\n",
            "        \n",
            "        - **Methods**:  GET\n",
            "        - **Response formats**: json, yaml, jsonp\n",
            "        \n",
            "        \n",
            "        **More info**\n",
            "        \n",
            "        https://rest.ensembl.org/documentation/info/biotypes_groups\n",
            "    \n",
            "    biotypes_name(*args, **kwargs)\n",
            "            Information ``GET info/biotypes/name/:name/:object_type``\n",
            "        \n",
            "        List the properties of biotypes with a given name. Object type (gene or transcript) can be provided for filtering.\n",
            "        \n",
            "        \n",
            "        **Parameters**\n",
            "        \n",
            "        - Required:\n",
            "        \n",
            "          + *Name*: name\n",
            "        \n",
            "            * *Type*: String\n",
            "            * *Description*: Biotype name\n",
            "            * *Default*: -\n",
            "            * *Example Values*: protein_coding\n",
            "        \n",
            "        \n",
            "        - Optional:\n",
            "        \n",
            "          + *Name*: callback\n",
            "        \n",
            "            * *Type*: String\n",
            "            * *Description*: Name of the callback subroutine to be returned by the requested JSONP response. Required ONLY when using JSONP as the serialisation method. Please see , .\n",
            "            * *Default*: -\n",
            "            * *Example Values*: randomlygeneratedname\n",
            "        \n",
            "        \n",
            "          + *Name*: object_type\n",
            "        \n",
            "            * *Type*: String\n",
            "            * *Description*: Object type (gene or transcript)\n",
            "            * *Default*: -\n",
            "            * *Example Values*: gene\n",
            "        \n",
            "        \n",
            "        **Resource info**\n",
            "        \n",
            "        - **Methods**:  GET\n",
            "        - **Response formats**: json, yaml, jsonp\n",
            "        \n",
            "        \n",
            "        **More info**\n",
            "        \n",
            "        https://rest.ensembl.org/documentation/info/biotypes_name\n",
            "    \n",
            "    cafe_tree(*args, **kwargs)\n",
            "            Comparative Genomics ``GET cafe/genetree/id/:id``\n",
            "        \n",
            "        Retrieves a cafe tree of the gene tree using the gene tree stable identifier\n",
            "        \n",
            "        \n",
            "        **Parameters**\n",
            "        \n",
            "        - Required:\n",
            "        \n",
            "          + *Name*: id\n",
            "        \n",
            "            * *Type*: String\n",
            "            * *Description*: An Ensembl genetree ID\n",
            "            * *Default*: -\n",
            "            * *Example Values*: ENSGT00390000003602\n",
            "        \n",
            "        \n",
            "        - Optional:\n",
            "        \n",
            "          + *Name*: callback\n",
            "        \n",
            "            * *Type*: String\n",
            "            * *Description*: Name of the callback subroutine to be returned by the requested JSONP response. Required ONLY when using JSONP as the serialisation method. Please see , .\n",
            "            * *Default*: -\n",
            "            * *Example Values*: randomlygeneratedname\n",
            "        \n",
            "        \n",
            "          + *Name*: compara\n",
            "        \n",
            "            * *Type*: String\n",
            "            * *Description*: Name of the compara database to use. Multiple comparas exist on a server for separate species divisions\n",
            "            * *Default*: vertebrates\n",
            "            * *Example Values*: vertebrates\n",
            "        \n",
            "        \n",
            "          + *Name*: nh_format\n",
            "        \n",
            "            * *Type*: String\n",
            "            * *Description*: The format of a NH (New Hampshire) request. Available only with the default setting to allow us to return the cafe tree with Taxa names appended with number of members and the p_value. example : homo_sapiens_3_0.123 where 3 is the number of members and 0.123 is the p value\n",
            "            * *Default*: Preset with default cafe parameters\n",
            "            * *Example Values*: -\n",
            "        \n",
            "        \n",
            "        **Resource info**\n",
            "        \n",
            "        - **Methods**:  GET\n",
            "        - **Response formats**: nh, json, jsonp\n",
            "        \n",
            "        \n",
            "        **More info**\n",
            "        \n",
            "        https://rest.ensembl.org/documentation/info/cafe_tree\n",
            "    \n",
            "    cafe_tree_member_id(*args, **kwargs)\n",
            "            Comparative Genomics ``GET cafe/genetree/member/id/:id``\n",
            "        \n",
            "        Retrieves the cafe tree of the gene tree that contains the gene / transcript / translation stable identifier\n",
            "        \n",
            "        \n",
            "        **Parameters**\n",
            "        \n",
            "        - Required:\n",
            "        \n",
            "          + *Name*: id\n",
            "        \n",
            "            * *Type*: String\n",
            "            * *Description*: An Ensembl stable ID\n",
            "            * *Default*: -\n",
            "            * *Example Values*: ENSG00000167664\n",
            "        \n",
            "        \n",
            "        - Optional:\n",
            "        \n",
            "          + *Name*: callback\n",
            "        \n",
            "            * *Type*: String\n",
            "            * *Description*: Name of the callback subroutine to be returned by the requested JSONP response. Required ONLY when using JSONP as the serialisation method. Please see , .\n",
            "            * *Default*: -\n",
            "            * *Example Values*: randomlygeneratedname\n",
            "        \n",
            "        \n",
            "          + *Name*: compara\n",
            "        \n",
            "            * *Type*: String\n",
            "            * *Description*: Name of the compara database to use. Multiple comparas exist on a server for separate species divisions\n",
            "            * *Default*: vertebrates\n",
            "            * *Example Values*: vertebrates\n",
            "        \n",
            "        \n",
            "          + *Name*: db_type\n",
            "        \n",
            "            * *Type*: String\n",
            "            * *Description*: Restrict the search to a database other than the default. Useful if you need to use a DB other than core\n",
            "            * *Default*: -\n",
            "            * *Example Values*: core\n",
            "        \n",
            "        \n",
            "          + *Name*: nh_format\n",
            "        \n",
            "            * *Type*: \n",
            "            * *Description*: The format of a NH (New Hampshire) request. Available only with the default setting to allow us to return the cafe tree with Taxa names appended with number of members and the p_value. example : homo_sapiens_3_0.123 where 3 is the number of members and 0.123 is the p value\n",
            "            * *Default*: Preset with default cafe parameters\n",
            "            * *Example Values*: -\n",
            "        \n",
            "        \n",
            "          + *Name*: object_type\n",
            "        \n",
            "            * *Type*: String\n",
            "            * *Description*: Filter by feature type\n",
            "            * *Default*: -\n",
            "            * *Example Values*: gene, transcript\n",
            "        \n",
            "        \n",
            "          + *Name*: species\n",
            "        \n",
            "            * *Type*: String\n",
            "            * *Description*: Species name/alias\n",
            "            * *Default*: -\n",
            "            * *Example Values*: homo_sapiens, human\n",
            "        \n",
            "        \n",
            "        **Resource info**\n",
            "        \n",
            "        - **Methods**:  GET\n",
            "        - **Response formats**: nh, json, jsonp\n",
            "        \n",
            "        \n",
            "        **More info**\n",
            "        \n",
            "        https://rest.ensembl.org/documentation/info/cafe_tree_member_id\n",
            "    \n",
            "    cafe_tree_member_symbol(*args, **kwargs)\n",
            "            Comparative Genomics ``GET cafe/genetree/member/symbol/:species/:symbol``\n",
            "        \n",
            "        Retrieves the cafe tree of the gene tree that contains the gene identified by a symbol\n",
            "        \n",
            "        \n",
            "        **Parameters**\n",
            "        \n",
            "        - Required:\n",
            "        \n",
            "          + *Name*: species\n",
            "        \n",
            "            * *Type*: String\n",
            "            * *Description*: Species name/alias\n",
            "            * *Default*: -\n",
            "            * *Example Values*: homo_sapiens, human\n",
            "        \n",
            "        \n",
            "          + *Name*: symbol\n",
            "        \n",
            "            * *Type*: String\n",
            "            * *Description*: Symbol or display name of a gene\n",
            "            * *Default*: -\n",
            "            * *Example Values*: BRCA2\n",
            "        \n",
            "        \n",
            "        - Optional:\n",
            "        \n",
            "          + *Name*: callback\n",
            "        \n",
            "            * *Type*: String\n",
            "            * *Description*: Name of the callback subroutine to be returned by the requested JSONP response. Required ONLY when using JSONP as the serialisation method. Please see , .\n",
            "            * *Default*: -\n",
            "            * *Example Values*: randomlygeneratedname\n",
            "        \n",
            "        \n",
            "          + *Name*: compara\n",
            "        \n",
            "            * *Type*: String\n",
            "            * *Description*: Name of the compara database to use. Multiple comparas exist on a server for separate species divisions\n",
            "            * *Default*: vertebrates\n",
            "            * *Example Values*: vertebrates\n",
            "        \n",
            "        \n",
            "          + *Name*: db_type\n",
            "        \n",
            "            * *Type*: String\n",
            "            * *Description*: Restrict the search to a database other than the default. Useful if you need to use a DB other than core\n",
            "            * *Default*: core\n",
            "            * *Example Values*: core, otherfeatures\n",
            "        \n",
            "        \n",
            "          + *Name*: external_db\n",
            "        \n",
            "            * *Type*: String\n",
            "            * *Description*: Filter by external database\n",
            "            * *Default*: -\n",
            "            * *Example Values*: HGNC\n",
            "        \n",
            "        \n",
            "          + *Name*: nh_format\n",
            "        \n",
            "            * *Type*: \n",
            "            * *Description*: The format of a NH (New Hampshire) request. Available only with the default setting to allow us to return the cafe tree with Taxa names appended with number of members and the p_value. example : homo_sapiens_3_0.123 where 3 is the number of members and 0.123 is the p value\n",
            "            * *Default*: Preset with default cafe parameters\n",
            "            * *Example Values*: -\n",
            "        \n",
            "        \n",
            "          + *Name*: object_type\n",
            "        \n",
            "            * *Type*: String\n",
            "            * *Description*: Filter by feature type\n",
            "            * *Default*: -\n",
            "            * *Example Values*: gene, transcript\n",
            "        \n",
            "        \n",
            "        **Resource info**\n",
            "        \n",
            "        - **Methods**:  GET\n",
            "        - **Response formats**: nh, json, jsonp\n",
            "        \n",
            "        \n",
            "        **More info**\n",
            "        \n",
            "        https://rest.ensembl.org/documentation/info/cafe_tree_member_symbol\n",
            "    \n",
            "    compara_methods(*args, **kwargs)\n",
            "            Information ``GET info/compara/methods``\n",
            "        \n",
            "        List all compara analyses available (an analysis defines the type of comparative data).\n",
            "        \n",
            "        \n",
            "        **Parameters**\n",
            "        \n",
            "        - Required:\n",
            "        \n",
            "          + *Name*: callback\n",
            "        \n",
            "            * *Type*: String\n",
            "            * *Description*: Name of the callback subroutine to be returned by the requested JSONP response. Required ONLY when using JSONP as the serialisation method. Please see , .\n",
            "            * *Default*: -\n",
            "            * *Example Values*: randomlygeneratedname\n",
            "        \n",
            "        \n",
            "          + *Name*: class\n",
            "        \n",
            "            * *Type*: String\n",
            "            * *Description*: The class of the method to query for. Regular expression patterns are supported.\n",
            "            * *Default*: -\n",
            "            * *Example Values*: GenomicAlign\n",
            "        \n",
            "        \n",
            "          + *Name*: compara\n",
            "        \n",
            "            * *Type*: String\n",
            "            * *Description*: Name of the compara database to use. Multiple comparas exist on a server for separate species divisions\n",
            "            * *Default*: vertebrates\n",
            "            * *Example Values*: vertebrates\n",
            "        \n",
            "        \n",
            "        - Optional:\n",
            "        \n",
            "          + *Name*: callback\n",
            "        \n",
            "            * *Type*: String\n",
            "            * *Description*: Name of the callback subroutine to be returned by the requested JSONP response. Required ONLY when using JSONP as the serialisation method. Please see , .\n",
            "            * *Default*: -\n",
            "            * *Example Values*: randomlygeneratedname\n",
            "        \n",
            "        \n",
            "          + *Name*: class\n",
            "        \n",
            "            * *Type*: String\n",
            "            * *Description*: The class of the method to query for. Regular expression patterns are supported.\n",
            "            * *Default*: -\n",
            "            * *Example Values*: GenomicAlign\n",
            "        \n",
            "        \n",
            "          + *Name*: compara\n",
            "        \n",
            "            * *Type*: String\n",
            "            * *Description*: Name of the compara database to use. Multiple comparas exist on a server for separate species divisions\n",
            "            * *Default*: vertebrates\n",
            "            * *Example Values*: vertebrates\n",
            "        \n",
            "        \n",
            "        **Resource info**\n",
            "        \n",
            "        - **Methods**:  GET\n",
            "        - **Response formats**: json, yaml, xml, jsonp\n",
            "        \n",
            "        \n",
            "        **More info**\n",
            "        \n",
            "        https://rest.ensembl.org/documentation/info/compara_methods\n",
            "    \n",
            "    compara_species_sets(*args, **kwargs)\n",
            "            Information ``GET info/compara/species_sets/:method``\n",
            "        \n",
            "        List all collections of species analysed with the specified compara method.\n",
            "        \n",
            "        \n",
            "        **Parameters**\n",
            "        \n",
            "        - Required:\n",
            "        \n",
            "          + *Name*: method\n",
            "        \n",
            "            * *Type*: String\n",
            "            * *Description*: Filter by compara method. Use one the methods returned by ,  endpoint.\n",
            "            * *Default*: -\n",
            "            * *Example Values*: EPO\n",
            "        \n",
            "        \n",
            "        - Optional:\n",
            "        \n",
            "          + *Name*: callback\n",
            "        \n",
            "            * *Type*: String\n",
            "            * *Description*: Name of the callback subroutine to be returned by the requested JSONP response. Required ONLY when using JSONP as the serialisation method. Please see , .\n",
            "            * *Default*: -\n",
            "            * *Example Values*: randomlygeneratedname\n",
            "        \n",
            "        \n",
            "          + *Name*: compara\n",
            "        \n",
            "            * *Type*: String\n",
            "            * *Description*: Name of the compara database to use. Multiple comparas exist on a server for separate species divisions\n",
            "            * *Default*: vertebrates\n",
            "            * *Example Values*: vertebrates\n",
            "        \n",
            "        \n",
            "        **Resource info**\n",
            "        \n",
            "        - **Methods**:  GET\n",
            "        - **Response formats**: json, yaml, xml, jsonp\n",
            "        \n",
            "        \n",
            "        **More info**\n",
            "        \n",
            "        https://rest.ensembl.org/documentation/info/compara_species_sets\n",
            "    \n",
            "    comparas(*args, **kwargs)\n",
            "            Information ``GET info/comparas``\n",
            "        \n",
            "        Lists all available comparative genomics databases and their data release. DEPRECATED: use info/genomes/division instead.\n",
            "        \n",
            "        \n",
            "        **Parameters**\n",
            "        \n",
            "        - Required:\n",
            "        \n",
            "          + *Name*: callback\n",
            "        \n",
            "            * *Type*: String\n",
            "            * *Description*: Name of the callback subroutine to be returned by the requested JSONP response. Required ONLY when using JSONP as the serialisation method. Please see , .\n",
            "            * *Default*: -\n",
            "            * *Example Values*: randomlygeneratedname\n",
            "        \n",
            "        \n",
            "        - Optional:\n",
            "        \n",
            "          + *Name*: callback\n",
            "        \n",
            "            * *Type*: String\n",
            "            * *Description*: Name of the callback subroutine to be returned by the requested JSONP response. Required ONLY when using JSONP as the serialisation method. Please see , .\n",
            "            * *Default*: -\n",
            "            * *Example Values*: randomlygeneratedname\n",
            "        \n",
            "        \n",
            "        **Resource info**\n",
            "        \n",
            "        - **Methods**:  GET\n",
            "        - **Response formats**: json, xml, jsonp\n",
            "        \n",
            "        \n",
            "        **More info**\n",
            "        \n",
            "        https://rest.ensembl.org/documentation/info/comparas\n",
            "    \n",
            "    data(*args, **kwargs)\n",
            "            Information ``GET info/data``\n",
            "        \n",
            "        Shows the data releases available on this REST server. May return more than one release (unfrequent non-standard Ensembl configuration).\n",
            "        \n",
            "        \n",
            "        **Parameters**\n",
            "        \n",
            "        - Required:\n",
            "        \n",
            "          + *Name*: callback\n",
            "        \n",
            "            * *Type*: String\n",
            "            * *Description*: Name of the callback subroutine to be returned by the requested JSONP response. Required ONLY when using JSONP as the serialisation method. Please see , .\n",
            "            * *Default*: -\n",
            "            * *Example Values*: randomlygeneratedname\n",
            "        \n",
            "        \n",
            "        - Optional:\n",
            "        \n",
            "          + *Name*: callback\n",
            "        \n",
            "            * *Type*: String\n",
            "            * *Description*: Name of the callback subroutine to be returned by the requested JSONP response. Required ONLY when using JSONP as the serialisation method. Please see , .\n",
            "            * *Default*: -\n",
            "            * *Example Values*: randomlygeneratedname\n",
            "        \n",
            "        \n",
            "        **Resource info**\n",
            "        \n",
            "        - **Methods**:  GET\n",
            "        - **Response formats**: json, xml, jsonp\n",
            "        \n",
            "        \n",
            "        **More info**\n",
            "        \n",
            "        https://rest.ensembl.org/documentation/info/data\n",
            "    \n",
            "    eg_version(*args, **kwargs)\n",
            "            Information ``GET info/eg_version``\n",
            "        \n",
            "        Returns the Ensembl Genomes version of the databases backing this service\n",
            "        \n",
            "        \n",
            "        **Parameters**\n",
            "        \n",
            "        - Required:\n",
            "        \n",
            "          + *Name*: callback\n",
            "        \n",
            "            * *Type*: String\n",
            "            * *Description*: Name of the callback subroutine to be returned by the requested JSONP response. Required ONLY when using JSONP as the serialisation method. Please see , .\n",
            "            * *Default*: -\n",
            "            * *Example Values*: randomlygeneratedname\n",
            "        \n",
            "        \n",
            "        - Optional:\n",
            "        \n",
            "          + *Name*: callback\n",
            "        \n",
            "            * *Type*: String\n",
            "            * *Description*: Name of the callback subroutine to be returned by the requested JSONP response. Required ONLY when using JSONP as the serialisation method. Please see , .\n",
            "            * *Default*: -\n",
            "            * *Example Values*: randomlygeneratedname\n",
            "        \n",
            "        \n",
            "        **Resource info**\n",
            "        \n",
            "        - **Methods**:  GET\n",
            "        - **Response formats**: json, xml, jsonp\n",
            "        \n",
            "        \n",
            "        **More info**\n",
            "        \n",
            "        https://rest.ensembl.org/documentation/info/eg_version\n",
            "    \n",
            "    external_dbs(*args, **kwargs)\n",
            "            Information ``GET info/external_dbs/:species``\n",
            "        \n",
            "        Lists all available external sources for a species.\n",
            "        \n",
            "        \n",
            "        **Parameters**\n",
            "        \n",
            "        - Required:\n",
            "        \n",
            "          + *Name*: species\n",
            "        \n",
            "            * *Type*: String\n",
            "            * *Description*: Species name/alias\n",
            "            * *Default*: -\n",
            "            * *Example Values*: homo_sapiens\n",
            "        \n",
            "        \n",
            "        - Optional:\n",
            "        \n",
            "          + *Name*: callback\n",
            "        \n",
            "            * *Type*: String\n",
            "            * *Description*: Name of the callback subroutine to be returned by the requested JSONP response. Required ONLY when using JSONP as the serialisation method. Please see , .\n",
            "            * *Default*: -\n",
            "            * *Example Values*: randomlygeneratedname\n",
            "        \n",
            "        \n",
            "          + *Name*: feature\n",
            "        \n",
            "            * *Type*: Enum(dna_align_feature protein_align_feature unmapped_object xref seq_region_synonym)\n",
            "            * *Description*: Only return external DB entries for a given feature.\n",
            "            * *Default*: -\n",
            "            * *Example Values*: xref, dna_align_feature\n",
            "        \n",
            "        \n",
            "          + *Name*: filter\n",
            "        \n",
            "            * *Type*: String\n",
            "            * *Description*: Restrict external DB searches to a single source or pattern. SQL-LIKE patterns are supported.\n",
            "            * *Default*: -\n",
            "            * *Example Values*: HGNC, GO%\n",
            "        \n",
            "        \n",
            "        **Resource info**\n",
            "        \n",
            "        - **Methods**:  GET\n",
            "        - **Response formats**: json, xml, jsonp\n",
            "        \n",
            "        \n",
            "        **More info**\n",
            "        \n",
            "        https://rest.ensembl.org/documentation/info/external_dbs\n",
            "    \n",
            "    family(*args, **kwargs)\n",
            "            Comparative Genomics ``GET family/id/:id``\n",
            "        \n",
            "        Retrieves a family information using the family stable identifier\n",
            "        \n",
            "        \n",
            "        **Parameters**\n",
            "        \n",
            "        - Required:\n",
            "        \n",
            "          + *Name*: id\n",
            "        \n",
            "            * *Type*: String\n",
            "            * *Description*: An Ensembl family ID\n",
            "            * *Default*: -\n",
            "            * *Example Values*: PTHR15573\n",
            "        \n",
            "        \n",
            "        - Optional:\n",
            "        \n",
            "          + *Name*: aligned\n",
            "        \n",
            "            * *Type*: Boolean\n",
            "            * *Description*: Return the aligned string if true. Otherwise, return the original sequence (no insertions). Note: If aligned=1, this will override sequence=none\n",
            "            * *Default*: 1\n",
            "            * *Example Values*: -\n",
            "        \n",
            "        \n",
            "          + *Name*: callback\n",
            "        \n",
            "            * *Type*: String\n",
            "            * *Description*: Name of the callback subroutine to be returned by the requested JSONP response. Required ONLY when using JSONP as the serialisation method. Please see , .\n",
            "            * *Default*: -\n",
            "            * *Example Values*: randomlygeneratedname\n",
            "        \n",
            "        \n",
            "          + *Name*: compara\n",
            "        \n",
            "            * *Type*: String\n",
            "            * *Description*: Name of the compara database to use. Multiple comparas exist on a server for separate species divisions\n",
            "            * *Default*: vertebrates\n",
            "            * *Example Values*: vertebrates\n",
            "        \n",
            "        \n",
            "          + *Name*: member_source\n",
            "        \n",
            "            * *Type*: Enum(all, ensembl, uniprot)\n",
            "            * *Description*: The source of the family members that you want returned\n",
            "            * *Default*: all\n",
            "            * *Example Values*: -\n",
            "        \n",
            "        \n",
            "          + *Name*: sequence\n",
            "        \n",
            "            * *Type*: Enum(none, cdna, protein)\n",
            "            * *Description*: The type of sequence to bring back. Setting it to none results in no sequence being returned. Note: dependant on the setting for \"aligned\", If aligned=1, this will override sequence=none\n",
            "            * *Default*: protein\n",
            "            * *Example Values*: -\n",
            "        \n",
            "        \n",
            "        **Resource info**\n",
            "        \n",
            "        - **Methods**:  GET\n",
            "        - **Response formats**: json, jsonp\n",
            "        \n",
            "        \n",
            "        **More info**\n",
            "        \n",
            "        https://rest.ensembl.org/documentation/info/family\n",
            "    \n",
            "    family_member_id(*args, **kwargs)\n",
            "            Comparative Genomics ``GET family/member/id/:id``\n",
            "        \n",
            "        Retrieves the information for all the families that contains the gene / transcript / translation stable identifier\n",
            "        \n",
            "        \n",
            "        **Parameters**\n",
            "        \n",
            "        - Required:\n",
            "        \n",
            "          + *Name*: id\n",
            "        \n",
            "            * *Type*: String\n",
            "            * *Description*: An Ensembl stable ID\n",
            "            * *Default*: -\n",
            "            * *Example Values*: ENSG00000167664\n",
            "        \n",
            "        \n",
            "        - Optional:\n",
            "        \n",
            "          + *Name*: aligned\n",
            "        \n",
            "            * *Type*: Boolean\n",
            "            * *Description*: Return the aligned string if true. Otherwise, return the original sequence (no insertions). Note: If aligned=1, this will override sequence=none\n",
            "            * *Default*: 1\n",
            "            * *Example Values*: -\n",
            "        \n",
            "        \n",
            "          + *Name*: callback\n",
            "        \n",
            "            * *Type*: String\n",
            "            * *Description*: Name of the callback subroutine to be returned by the requested JSONP response. Required ONLY when using JSONP as the serialisation method. Please see , .\n",
            "            * *Default*: -\n",
            "            * *Example Values*: randomlygeneratedname\n",
            "        \n",
            "        \n",
            "          + *Name*: compara\n",
            "        \n",
            "            * *Type*: String\n",
            "            * *Description*: Name of the compara database to use. Multiple comparas exist on a server for separate species divisions\n",
            "            * *Default*: vertebrates\n",
            "            * *Example Values*: vertebrates\n",
            "        \n",
            "        \n",
            "          + *Name*: member_source\n",
            "        \n",
            "            * *Type*: Enum(all, ensembl, uniprot)\n",
            "            * *Description*: The source of the family members that you want returned\n",
            "            * *Default*: all\n",
            "            * *Example Values*: -\n",
            "        \n",
            "        \n",
            "          + *Name*: sequence\n",
            "        \n",
            "            * *Type*: Enum(none, cdna, protein)\n",
            "            * *Description*: The type of sequence to bring back. Setting it to none results in no sequence being returned. Note: dependant on the setting for \"aligned\", If aligned=1, this will override sequence=none\n",
            "            * *Default*: protein\n",
            "            * *Example Values*: -\n",
            "        \n",
            "        \n",
            "        **Resource info**\n",
            "        \n",
            "        - **Methods**:  GET\n",
            "        - **Response formats**: json, jsonp\n",
            "        \n",
            "        \n",
            "        **More info**\n",
            "        \n",
            "        https://rest.ensembl.org/documentation/info/family_member_id\n",
            "    \n",
            "    family_member_symbol(*args, **kwargs)\n",
            "            Comparative Genomics ``GET family/member/symbol/:species/:symbol``\n",
            "        \n",
            "        Retrieves the information for all the families that contains the gene identified by a symbol\n",
            "        \n",
            "        \n",
            "        **Parameters**\n",
            "        \n",
            "        - Required:\n",
            "        \n",
            "          + *Name*: species\n",
            "        \n",
            "            * *Type*: String\n",
            "            * *Description*: Species name/alias\n",
            "            * *Default*: -\n",
            "            * *Example Values*: homo_sapiens, human\n",
            "        \n",
            "        \n",
            "          + *Name*: symbol\n",
            "        \n",
            "            * *Type*: String\n",
            "            * *Description*: Symbol or display name of a gene\n",
            "            * *Default*: -\n",
            "            * *Example Values*: BRCA2\n",
            "        \n",
            "        \n",
            "        - Optional:\n",
            "        \n",
            "          + *Name*: aligned\n",
            "        \n",
            "            * *Type*: Boolean\n",
            "            * *Description*: Return the aligned string if true. Otherwise, return the original sequence (no insertions). Note: If aligned=1, this will override sequence=none\n",
            "            * *Default*: 1\n",
            "            * *Example Values*: -\n",
            "        \n",
            "        \n",
            "          + *Name*: callback\n",
            "        \n",
            "            * *Type*: String\n",
            "            * *Description*: Name of the callback subroutine to be returned by the requested JSONP response. Required ONLY when using JSONP as the serialisation method. Please see , .\n",
            "            * *Default*: -\n",
            "            * *Example Values*: randomlygeneratedname\n",
            "        \n",
            "        \n",
            "          + *Name*: compara\n",
            "        \n",
            "            * *Type*: String\n",
            "            * *Description*: Name of the compara database to use. Multiple comparas exist on a server for separate species divisions\n",
            "            * *Default*: vertebrates\n",
            "            * *Example Values*: vertebrates\n",
            "        \n",
            "        \n",
            "          + *Name*: db_type\n",
            "        \n",
            "            * *Type*: String\n",
            "            * *Description*: Restrict the search to a database other than the default. Useful if you need to use a DB other than core\n",
            "            * *Default*: core\n",
            "            * *Example Values*: core, otherfeatures\n",
            "        \n",
            "        \n",
            "          + *Name*: external_db\n",
            "        \n",
            "            * *Type*: String\n",
            "            * *Description*: Filter by external database\n",
            "            * *Default*: -\n",
            "            * *Example Values*: HGNC\n",
            "        \n",
            "        \n",
            "          + *Name*: member_source\n",
            "        \n",
            "            * *Type*: Enum(all, ensembl, uniprot)\n",
            "            * *Description*: The source of the family members that you want returned\n",
            "            * *Default*: all\n",
            "            * *Example Values*: -\n",
            "        \n",
            "        \n",
            "          + *Name*: object_type\n",
            "        \n",
            "            * *Type*: String\n",
            "            * *Description*: Filter by feature type\n",
            "            * *Default*: -\n",
            "            * *Example Values*: gene, transcript\n",
            "        \n",
            "        \n",
            "          + *Name*: sequence\n",
            "        \n",
            "            * *Type*: Enum(none, cdna, protein)\n",
            "            * *Description*: The type of sequence to bring back. Setting it to none results in no sequence being returned. Note: dependant on the setting for \"aligned\", If aligned=1, this will override sequence=none\n",
            "            * *Default*: protein\n",
            "            * *Example Values*: -\n",
            "        \n",
            "        \n",
            "        **Resource info**\n",
            "        \n",
            "        - **Methods**:  GET\n",
            "        - **Response formats**: json, jsonp\n",
            "        \n",
            "        \n",
            "        **More info**\n",
            "        \n",
            "        https://rest.ensembl.org/documentation/info/family_member_symbol\n",
            "    \n",
            "    features_id(*args, **kwargs)\n",
            "            Variation GA4GH ``GET ga4gh/features/:id``\n",
            "        \n",
            "        Return the GA4GH record for a specific sequence feature given its identifier\n",
            "        \n",
            "        \n",
            "        **Parameters**\n",
            "        \n",
            "        - Required:\n",
            "        \n",
            "          + *Name*: id\n",
            "        \n",
            "            * *Type*: String\n",
            "            * *Description*: Feature id\n",
            "            * *Default*: -\n",
            "            * *Example Values*: ENST00000408937.7\n",
            "        \n",
            "        \n",
            "        - Optional:\n",
            "        \n",
            "          + *Name*: callback\n",
            "        \n",
            "            * *Type*: String\n",
            "            * *Description*: Name of the callback subroutine to be returned by the requested JSONP response. Required ONLY when using JSONP as the serialisation method. Please see , .\n",
            "            * *Default*: -\n",
            "            * *Example Values*: randomlygeneratedname\n",
            "        \n",
            "        \n",
            "        **Resource info**\n",
            "        \n",
            "        - **Methods**:  GET\n",
            "        - **Response formats**: json, xml, jsonp\n",
            "        \n",
            "        \n",
            "        **More info**\n",
            "        \n",
            "        https://rest.ensembl.org/documentation/info/features_id\n",
            "    \n",
            "    features_post(*args, **kwargs)\n",
            "            Variation GA4GH ``POST ga4gh/features/search``\n",
            "        \n",
            "        Return a list of sequence annotation features in GA4GH format\n",
            "        \n",
            "        \n",
            "        **Parameters**\n",
            "        \n",
            "        - Required:\n",
            "        \n",
            "          + *Name*: end\n",
            "        \n",
            "            * *Type*: Int\n",
            "            * *Description*: Return features within a window with this end location\n",
            "            * *Default*: -\n",
            "            * *Example Values*: 1109000\n",
            "        \n",
            "        \n",
            "          + *Name*: referenceName\n",
            "        \n",
            "            * *Type*: string\n",
            "            * *Description*: Return features on this reference\n",
            "            * *Default*: -\n",
            "            * *Example Values*: 6\n",
            "        \n",
            "        \n",
            "          + *Name*: start\n",
            "        \n",
            "            * *Type*: Int\n",
            "            * *Description*: Return features within a window with this start location\n",
            "            * *Default*: -\n",
            "            * *Example Values*: 1108000\n",
            "        \n",
            "        \n",
            "        - Optional:\n",
            "        \n",
            "          + *Name*: callback\n",
            "        \n",
            "            * *Type*: String\n",
            "            * *Description*: Name of the callback subroutine to be returned by the requested JSONP response. Required ONLY when using JSONP as the serialisation method. Please see , .\n",
            "            * *Default*: -\n",
            "            * *Example Values*: randomlygeneratedname\n",
            "        \n",
            "        \n",
            "          + *Name*: featureTypes\n",
            "        \n",
            "            * *Type*: array of strings\n",
            "            * *Description*: Return features of this type (requires SO terms)\n",
            "            * *Default*: -\n",
            "            * *Example Values*: [transcript]\n",
            "        \n",
            "        \n",
            "          + *Name*: featuresetId\n",
            "        \n",
            "            * *Type*: String\n",
            "            * *Description*: Return features in this set\n",
            "            * *Default*: -\n",
            "            * *Example Values*: Ensembl\n",
            "        \n",
            "        \n",
            "          + *Name*: pageSize\n",
            "        \n",
            "            * *Type*: Int\n",
            "            * *Description*: Number of features to return per request\n",
            "            * *Default*: 10\n",
            "            * *Example Values*: -\n",
            "        \n",
            "        \n",
            "          + *Name*: pageToken\n",
            "        \n",
            "            * *Type*: Int\n",
            "            * *Description*: Identifier showing which page of data to retrieve next\n",
            "            * *Default*: null\n",
            "            * *Example Values*: -\n",
            "        \n",
            "        \n",
            "          + *Name*: parentId\n",
            "        \n",
            "            * *Type*: String\n",
            "            * *Description*: Return the child features of this feature\n",
            "            * *Default*: -\n",
            "            * *Example Values*: ENSG00000176515.1\n",
            "        \n",
            "        \n",
            "        **Resource info**\n",
            "        \n",
            "        - **Methods**:  POST\n",
            "        - **Response formats**: json, jsonp\n",
            "        \n",
            "        \n",
            "        **More info**\n",
            "        \n",
            "        https://rest.ensembl.org/documentation/info/features_post\n",
            "    \n",
            "    fetch_all_epigenomes(*args, **kwargs)\n",
            "            Regulation ``GET regulatory/species/:species/epigenome``\n",
            "        \n",
            "        Returns information about all epigenomes available for the given species\n",
            "        \n",
            "        \n",
            "        **Parameters**\n",
            "        \n",
            "        - Required:\n",
            "        \n",
            "          + *Name*: species\n",
            "        \n",
            "            * *Type*: String\n",
            "            * *Description*: Species name/alias\n",
            "            * *Default*: -\n",
            "            * *Example Values*: homo_sapiens\n",
            "        \n",
            "        \n",
            "        - Optional:\n",
            "        \n",
            "          + *Name*: callback\n",
            "        \n",
            "            * *Type*: String\n",
            "            * *Description*: Name of the callback subroutine to be returned by the requested JSONP response. Required ONLY when using JSONP as the serialisation method. Please see , .\n",
            "            * *Default*: -\n",
            "            * *Example Values*: randomlygeneratedname\n",
            "        \n",
            "        \n",
            "        **Resource info**\n",
            "        \n",
            "        - **Methods**:  GET\n",
            "        - **Response formats**: json, xml, jsonp\n",
            "        \n",
            "        \n",
            "        **More info**\n",
            "        \n",
            "        https://rest.ensembl.org/documentation/info/fetch_all_epigenomes\n",
            "    \n",
            "    gacallSet(*args, **kwargs)\n",
            "            Variation GA4GH ``POST ga4gh/callsets/search``\n",
            "        \n",
            "        Return a list of sets of genotype calls for specific samples in GA4GH format\n",
            "        \n",
            "        \n",
            "        **Parameters**\n",
            "        \n",
            "        - Required:\n",
            "        \n",
            "          + *Name*: \n",
            "        \n",
            "            * *Type*: String\n",
            "            * *Description*: \n",
            "            * *Default*: -\n",
            "            * *Example Values*: 1\n",
            "        \n",
            "        \n",
            "        - Optional:\n",
            "        \n",
            "          + *Name*: callback\n",
            "        \n",
            "            * *Type*: String\n",
            "            * *Description*: Name of the callback subroutine to be returned by the requested JSONP response. Required ONLY when using JSONP as the serialisation method. Please see , .\n",
            "            * *Default*: -\n",
            "            * *Example Values*: randomlygeneratedname\n",
            "        \n",
            "        \n",
            "          + *Name*: name\n",
            "        \n",
            "            * *Type*: String\n",
            "            * *Description*: Return callSets by name\n",
            "            * *Default*: -\n",
            "            * *Example Values*: NA19777\n",
            "        \n",
            "        \n",
            "          + *Name*: pageSize\n",
            "        \n",
            "            * *Type*: Int\n",
            "            * *Description*: Number of callSets to return per request\n",
            "            * *Default*: 10\n",
            "            * *Example Values*: -\n",
            "        \n",
            "        \n",
            "          + *Name*: pageToken\n",
            "        \n",
            "            * *Type*: Int\n",
            "            * *Description*: Identifier showing which page of data to retrieve next\n",
            "            * *Default*: null\n",
            "            * *Example Values*: -\n",
            "        \n",
            "        \n",
            "        **Resource info**\n",
            "        \n",
            "        - **Methods**:  POST\n",
            "        - **Response formats**: json, jsonp\n",
            "        \n",
            "        \n",
            "        **More info**\n",
            "        \n",
            "        https://rest.ensembl.org/documentation/info/gacallSet\n",
            "    \n",
            "    gacallset_id(*args, **kwargs)\n",
            "            Variation GA4GH ``GET ga4gh/callsets/:id``\n",
            "        \n",
            "        Return the GA4GH record for a specific CallSet given its identifier\n",
            "        \n",
            "        \n",
            "        **Parameters**\n",
            "        \n",
            "        - Required:\n",
            "        \n",
            "          + *Name*: id\n",
            "        \n",
            "            * *Type*: String\n",
            "            * *Description*: CallSet id\n",
            "            * *Default*: -\n",
            "            * *Example Values*: 1\n",
            "        \n",
            "        \n",
            "        - Optional:\n",
            "        \n",
            "          + *Name*: callback\n",
            "        \n",
            "            * *Type*: String\n",
            "            * *Description*: Name of the callback subroutine to be returned by the requested JSONP response. Required ONLY when using JSONP as the serialisation method. Please see , .\n",
            "            * *Default*: -\n",
            "            * *Example Values*: randomlygeneratedname\n",
            "        \n",
            "        \n",
            "        **Resource info**\n",
            "        \n",
            "        - **Methods**:  GET\n",
            "        - **Response formats**: json, xml, jsonp\n",
            "        \n",
            "        \n",
            "        **More info**\n",
            "        \n",
            "        https://rest.ensembl.org/documentation/info/gacallset_id\n",
            "    \n",
            "    gadataset(*args, **kwargs)\n",
            "            Variation GA4GH ``POST ga4gh/datasets/search``\n",
            "        \n",
            "        Return a list of datasets in GA4GH format\n",
            "        \n",
            "        \n",
            "        **Parameters**\n",
            "        \n",
            "        - Required:\n",
            "        \n",
            "          + *Name*: callback\n",
            "        \n",
            "            * *Type*: String\n",
            "            * *Description*: Name of the callback subroutine to be returned by the requested JSONP response. Required ONLY when using JSONP as the serialisation method. Please see , .\n",
            "            * *Default*: -\n",
            "            * *Example Values*: randomlygeneratedname\n",
            "        \n",
            "        \n",
            "          + *Name*: pageSize\n",
            "        \n",
            "            * *Type*: Int\n",
            "            * *Description*: Number of dataSets to return per request\n",
            "            * *Default*: 10\n",
            "            * *Example Values*: -\n",
            "        \n",
            "        \n",
            "          + *Name*: pageToken\n",
            "        \n",
            "            * *Type*: Int\n",
            "            * *Description*: Identifier showing which page of data to retrieve next\n",
            "            * *Default*: null\n",
            "            * *Example Values*: -\n",
            "        \n",
            "        \n",
            "        - Optional:\n",
            "        \n",
            "          + *Name*: callback\n",
            "        \n",
            "            * *Type*: String\n",
            "            * *Description*: Name of the callback subroutine to be returned by the requested JSONP response. Required ONLY when using JSONP as the serialisation method. Please see , .\n",
            "            * *Default*: -\n",
            "            * *Example Values*: randomlygeneratedname\n",
            "        \n",
            "        \n",
            "          + *Name*: pageSize\n",
            "        \n",
            "            * *Type*: Int\n",
            "            * *Description*: Number of dataSets to return per request\n",
            "            * *Default*: 10\n",
            "            * *Example Values*: -\n",
            "        \n",
            "        \n",
            "          + *Name*: pageToken\n",
            "        \n",
            "            * *Type*: Int\n",
            "            * *Description*: Identifier showing which page of data to retrieve next\n",
            "            * *Default*: null\n",
            "            * *Example Values*: -\n",
            "        \n",
            "        \n",
            "        **Resource info**\n",
            "        \n",
            "        - **Methods**:  POST\n",
            "        - **Response formats**: json, jsonp\n",
            "        \n",
            "        \n",
            "        **More info**\n",
            "        \n",
            "        https://rest.ensembl.org/documentation/info/gadataset\n",
            "    \n",
            "    gadataset_id(*args, **kwargs)\n",
            "            Variation GA4GH ``GET ga4gh/datasets/:id``\n",
            "        \n",
            "        Return the GA4GH record for a specific dataset given its identifier\n",
            "        \n",
            "        \n",
            "        **Parameters**\n",
            "        \n",
            "        - Required:\n",
            "        \n",
            "          + *Name*: id\n",
            "        \n",
            "            * *Type*: String\n",
            "            * *Description*: Dataset id\n",
            "            * *Default*: -\n",
            "            * *Example Values*: 6e340c4d1e333c7a676b1710d2e3953c\n",
            "        \n",
            "        \n",
            "        - Optional:\n",
            "        \n",
            "          + *Name*: callback\n",
            "        \n",
            "            * *Type*: String\n",
            "            * *Description*: Name of the callback subroutine to be returned by the requested JSONP response. Required ONLY when using JSONP as the serialisation method. Please see , .\n",
            "            * *Default*: -\n",
            "            * *Example Values*: randomlygeneratedname\n",
            "        \n",
            "        \n",
            "        **Resource info**\n",
            "        \n",
            "        - **Methods**:  GET\n",
            "        - **Response formats**: json, xml, jsonp\n",
            "        \n",
            "        \n",
            "        **More info**\n",
            "        \n",
            "        https://rest.ensembl.org/documentation/info/gadataset_id\n",
            "    \n",
            "    gafeatureset(*args, **kwargs)\n",
            "            Variation GA4GH ``POST ga4gh/featuresets/search``\n",
            "        \n",
            "        Return a list of feature sets in GA4GH format\n",
            "        \n",
            "        \n",
            "        **Parameters**\n",
            "        \n",
            "        - Required:\n",
            "        \n",
            "          + *Name*: datasetId\n",
            "        \n",
            "            * *Type*: String\n",
            "            * *Description*: Return featureSets by dataSet Identifier\n",
            "            * *Default*: -\n",
            "            * *Example Values*: Ensembl\n",
            "        \n",
            "        \n",
            "        - Optional:\n",
            "        \n",
            "          + *Name*: callback\n",
            "        \n",
            "            * *Type*: String\n",
            "            * *Description*: Name of the callback subroutine to be returned by the requested JSONP response. Required ONLY when using JSONP as the serialisation method. Please see , .\n",
            "            * *Default*: -\n",
            "            * *Example Values*: randomlygeneratedname\n",
            "        \n",
            "        \n",
            "          + *Name*: pageSize\n",
            "        \n",
            "            * *Type*: Int\n",
            "            * *Description*: Number of featureSets to return per request\n",
            "            * *Default*: 10\n",
            "            * *Example Values*: -\n",
            "        \n",
            "        \n",
            "          + *Name*: pageToken\n",
            "        \n",
            "            * *Type*: Int\n",
            "            * *Description*: Identifier showing which page of data to retrieve next\n",
            "            * *Default*: null\n",
            "            * *Example Values*: -\n",
            "        \n",
            "        \n",
            "        **Resource info**\n",
            "        \n",
            "        - **Methods**:  POST\n",
            "        - **Response formats**: json, jsonp\n",
            "        \n",
            "        \n",
            "        **More info**\n",
            "        \n",
            "        https://rest.ensembl.org/documentation/info/gafeatureset\n",
            "    \n",
            "    gafeatureset_id(*args, **kwargs)\n",
            "            Variation GA4GH ``GET ga4gh/featuresets/:id``\n",
            "        \n",
            "        Return the GA4GH record for a specific featureSet given its identifier\n",
            "        \n",
            "        \n",
            "        **Parameters**\n",
            "        \n",
            "        - Required:\n",
            "        \n",
            "          + *Name*: id\n",
            "        \n",
            "            * *Type*: String\n",
            "            * *Description*: featureSet id\n",
            "            * *Default*: -\n",
            "            * *Example Values*: Ensembl\n",
            "        \n",
            "        \n",
            "        - Optional:\n",
            "        \n",
            "          + *Name*: callback\n",
            "        \n",
            "            * *Type*: String\n",
            "            * *Description*: Name of the callback subroutine to be returned by the requested JSONP response. Required ONLY when using JSONP as the serialisation method. Please see , .\n",
            "            * *Default*: -\n",
            "            * *Example Values*: randomlygeneratedname\n",
            "        \n",
            "        \n",
            "        **Resource info**\n",
            "        \n",
            "        - **Methods**:  GET\n",
            "        - **Response formats**: json, xml, jsonp\n",
            "        \n",
            "        \n",
            "        **More info**\n",
            "        \n",
            "        https://rest.ensembl.org/documentation/info/gafeatureset_id\n",
            "    \n",
            "    gavariant_id(*args, **kwargs)\n",
            "            Variation GA4GH ``GET ga4gh/variants/:id``\n",
            "        \n",
            "        Return the GA4GH record for a specific variant given its identifier.\n",
            "        \n",
            "        \n",
            "        **Parameters**\n",
            "        \n",
            "        - Required:\n",
            "        \n",
            "          + *Name*: id\n",
            "        \n",
            "            * *Type*: String\n",
            "            * *Description*: Variation id\n",
            "            * *Default*: -\n",
            "            * *Example Values*: 1:rs1333049\n",
            "        \n",
            "        \n",
            "        - Optional:\n",
            "        \n",
            "          + *Name*: callback\n",
            "        \n",
            "            * *Type*: String\n",
            "            * *Description*: Name of the callback subroutine to be returned by the requested JSONP response. Required ONLY when using JSONP as the serialisation method. Please see , .\n",
            "            * *Default*: -\n",
            "            * *Example Values*: randomlygeneratedname\n",
            "        \n",
            "        \n",
            "        **Resource info**\n",
            "        \n",
            "        - **Methods**:  GET\n",
            "        - **Response formats**: json, xml, jsonp\n",
            "        \n",
            "        \n",
            "        **More info**\n",
            "        \n",
            "        https://rest.ensembl.org/documentation/info/gavariant_id\n",
            "    \n",
            "    gavariantannotations(*args, **kwargs)\n",
            "            Variation GA4GH ``POST ga4gh/variantannotations/search``\n",
            "        \n",
            "        Return variant annotation information in GA4GH format for a region on a reference sequence\n",
            "        \n",
            "        \n",
            "        **Parameters**\n",
            "        \n",
            "        - Required:\n",
            "        \n",
            "          + *Name*: \n",
            "        \n",
            "            * *Type*: String\n",
            "            * *Description*: \n",
            "            * *Default*: -\n",
            "            * *Example Values*: Ensembl\n",
            "        \n",
            "        \n",
            "        - Optional:\n",
            "        \n",
            "          + *Name*: callback\n",
            "        \n",
            "            * *Type*: String\n",
            "            * *Description*: Name of the callback subroutine to be returned by the requested JSONP response. Required ONLY when using JSONP as the serialisation method. Please see , .\n",
            "            * *Default*: -\n",
            "            * *Example Values*: randomlygeneratedname\n",
            "        \n",
            "        \n",
            "          + *Name*: effects\n",
            "        \n",
            "            * *Type*: [OntologyTerm]\n",
            "            * *Description*: \n",
            "            * *Default*: -\n",
            "            * *Example Values*: [ {\"sourceName\":\"SO\", \"term\":\"missense_variant\", \"id\":\"SO:0001583\", \"sourceVersion\": \"\"}]\n",
            "        \n",
            "        \n",
            "          + *Name*: end\n",
            "        \n",
            "            * *Type*: Int\n",
            "            * *Description*: End position of region (zero-based, exclusive)\n",
            "            * *Default*: -\n",
            "            * *Example Values*: 25221500\n",
            "        \n",
            "        \n",
            "          + *Name*: pageSize\n",
            "        \n",
            "            * *Type*: Int\n",
            "            * *Description*: \n",
            "            * *Default*: 10\n",
            "            * *Example Values*: -\n",
            "        \n",
            "        \n",
            "          + *Name*: pageToken\n",
            "        \n",
            "            * *Type*: Int\n",
            "            * *Description*: Identifier showing which page of data to retrieve next\n",
            "            * *Default*: null\n",
            "            * *Example Values*: -\n",
            "        \n",
            "        \n",
            "          + *Name*: referenceId\n",
            "        \n",
            "            * *Type*: String\n",
            "            * *Description*: Reference sequence id\n",
            "            * *Default*: -\n",
            "            * *Example Values*: 9489ae7581e14efcad134f02afafe26c\n",
            "        \n",
            "        \n",
            "          + *Name*: referenceName\n",
            "        \n",
            "            * *Type*: String\n",
            "            * *Description*: Reference sequence name\n",
            "            * *Default*: -\n",
            "            * *Example Values*: 22\n",
            "        \n",
            "        \n",
            "          + *Name*: start\n",
            "        \n",
            "            * *Type*: Int\n",
            "            * *Description*: Start position of region (zero-based, inclusive)\n",
            "            * *Default*: -\n",
            "            * *Example Values*: 25221400\n",
            "        \n",
            "        \n",
            "        **Resource info**\n",
            "        \n",
            "        - **Methods**:  POST\n",
            "        - **Response formats**: json, jsonp\n",
            "        \n",
            "        \n",
            "        **More info**\n",
            "        \n",
            "        https://rest.ensembl.org/documentation/info/gavariantannotations\n",
            "    \n",
            "    gavariants(*args, **kwargs)\n",
            "            Variation GA4GH ``POST ga4gh/variants/search``\n",
            "        \n",
            "        Return variant call information in GA4GH format for a region on a reference sequence\n",
            "        \n",
            "        \n",
            "        **Parameters**\n",
            "        \n",
            "        - Required:\n",
            "        \n",
            "          + *Name*: end\n",
            "        \n",
            "            * *Type*: Int\n",
            "            * *Description*: End position of region (zero-based, exclusive)\n",
            "            * *Default*: -\n",
            "            * *Example Values*: 25455087\n",
            "        \n",
            "        \n",
            "          + *Name*: referenceName\n",
            "        \n",
            "            * *Type*: String\n",
            "            * *Description*: Reference sequence name\n",
            "            * *Default*: -\n",
            "            * *Example Values*: 22\n",
            "        \n",
            "        \n",
            "          + *Name*: start\n",
            "        \n",
            "            * *Type*: Int\n",
            "            * *Description*: Start position of region (zero-based, inclusive)\n",
            "            * *Default*: -\n",
            "            * *Example Values*: 25455086\n",
            "        \n",
            "        \n",
            "          + *Name*: \n",
            "        \n",
            "            * *Type*: String\n",
            "            * *Description*: \n",
            "            * *Default*: -\n",
            "            * *Example Values*: 1\n",
            "        \n",
            "        \n",
            "        - Optional:\n",
            "        \n",
            "          + *Name*: callSetIds\n",
            "        \n",
            "            * *Type*: String\n",
            "            * *Description*: \n",
            "            * *Default*: -\n",
            "            * *Example Values*: [ 1:NA19777 ]\n",
            "        \n",
            "        \n",
            "          + *Name*: callback\n",
            "        \n",
            "            * *Type*: String\n",
            "            * *Description*: Name of the callback subroutine to be returned by the requested JSONP response. Required ONLY when using JSONP as the serialisation method. Please see , .\n",
            "            * *Default*: -\n",
            "            * *Example Values*: randomlygeneratedname\n",
            "        \n",
            "        \n",
            "          + *Name*: pageSize\n",
            "        \n",
            "            * *Type*: Int\n",
            "            * *Description*: \n",
            "            * *Default*: 10\n",
            "            * *Example Values*: -\n",
            "        \n",
            "        \n",
            "          + *Name*: pageToken\n",
            "        \n",
            "            * *Type*: Int\n",
            "            * *Description*: Identifier showing which page of data to retrieve next\n",
            "            * *Default*: null\n",
            "            * *Example Values*: -\n",
            "        \n",
            "        \n",
            "        **Resource info**\n",
            "        \n",
            "        - **Methods**:  POST\n",
            "        - **Response formats**: json, jsonp\n",
            "        \n",
            "        \n",
            "        **More info**\n",
            "        \n",
            "        https://rest.ensembl.org/documentation/info/gavariants\n",
            "    \n",
            "    gavariantset(*args, **kwargs)\n",
            "            Variation GA4GH ``POST ga4gh/variantsets/search``\n",
            "        \n",
            "        Return a list of variant sets in GA4GH format\n",
            "        \n",
            "        \n",
            "        **Parameters**\n",
            "        \n",
            "        - Required:\n",
            "        \n",
            "          + *Name*: datasetId\n",
            "        \n",
            "            * *Type*: String\n",
            "            * *Description*: \n",
            "            * *Default*: -\n",
            "            * *Example Values*: 6e340c4d1e333c7a676b1710d2e3953c\n",
            "        \n",
            "        \n",
            "        - Optional:\n",
            "        \n",
            "          + *Name*: callback\n",
            "        \n",
            "            * *Type*: String\n",
            "            * *Description*: Name of the callback subroutine to be returned by the requested JSONP response. Required ONLY when using JSONP as the serialisation method. Please see , .\n",
            "            * *Default*: -\n",
            "            * *Example Values*: randomlygeneratedname\n",
            "        \n",
            "        \n",
            "          + *Name*: pageSize\n",
            "        \n",
            "            * *Type*: Int\n",
            "            * *Description*: \n",
            "            * *Default*: 10\n",
            "            * *Example Values*: -\n",
            "        \n",
            "        \n",
            "          + *Name*: pageToken\n",
            "        \n",
            "            * *Type*: Int\n",
            "            * *Description*: Identifier showing which page of data to retrieve next\n",
            "            * *Default*: null\n",
            "            * *Example Values*: -\n",
            "        \n",
            "        \n",
            "        **Resource info**\n",
            "        \n",
            "        - **Methods**:  POST\n",
            "        - **Response formats**: json, jsonp\n",
            "        \n",
            "        \n",
            "        **More info**\n",
            "        \n",
            "        https://rest.ensembl.org/documentation/info/gavariantset\n",
            "    \n",
            "    gavariantset_id(*args, **kwargs)\n",
            "            Variation GA4GH ``GET ga4gh/variantsets/:id``\n",
            "        \n",
            "        Return the GA4GH record for a specific VariantSet given its identifier\n",
            "        \n",
            "        \n",
            "        **Parameters**\n",
            "        \n",
            "        - Required:\n",
            "        \n",
            "          + *Name*: id\n",
            "        \n",
            "            * *Type*: String\n",
            "            * *Description*: VariantSet id\n",
            "            * *Default*: -\n",
            "            * *Example Values*: 1\n",
            "        \n",
            "        \n",
            "        - Optional:\n",
            "        \n",
            "          + *Name*: callback\n",
            "        \n",
            "            * *Type*: String\n",
            "            * *Description*: Name of the callback subroutine to be returned by the requested JSONP response. Required ONLY when using JSONP as the serialisation method. Please see , .\n",
            "            * *Default*: -\n",
            "            * *Example Values*: randomlygeneratedname\n",
            "        \n",
            "        \n",
            "        **Resource info**\n",
            "        \n",
            "        - **Methods**:  GET\n",
            "        - **Response formats**: json, xml, jsonp\n",
            "        \n",
            "        \n",
            "        **More info**\n",
            "        \n",
            "        https://rest.ensembl.org/documentation/info/gavariantset_id\n",
            "    \n",
            "    genetree(*args, **kwargs)\n",
            "            Comparative Genomics ``GET genetree/id/:id``\n",
            "        \n",
            "        Retrieves a gene tree for a gene tree stable identifier\n",
            "        \n",
            "        \n",
            "        **Parameters**\n",
            "        \n",
            "        - Required:\n",
            "        \n",
            "          + *Name*: id\n",
            "        \n",
            "            * *Type*: String\n",
            "            * *Description*: An Ensembl genetree ID\n",
            "            * *Default*: -\n",
            "            * *Example Values*: ENSGT00390000003602\n",
            "        \n",
            "        \n",
            "        - Optional:\n",
            "        \n",
            "          + *Name*: aligned\n",
            "        \n",
            "            * *Type*: Boolean\n",
            "            * *Description*: Return the aligned string if true. Otherwise, return the original sequence (no insertions)\n",
            "            * *Default*: 0\n",
            "            * *Example Values*: -\n",
            "        \n",
            "        \n",
            "          + *Name*: callback\n",
            "        \n",
            "            * *Type*: String\n",
            "            * *Description*: Name of the callback subroutine to be returned by the requested JSONP response. Required ONLY when using JSONP as the serialisation method. Please see , .\n",
            "            * *Default*: -\n",
            "            * *Example Values*: randomlygeneratedname\n",
            "        \n",
            "        \n",
            "          + *Name*: cigar_line\n",
            "        \n",
            "            * *Type*: Boolean\n",
            "            * *Description*: Return the aligned sequence encoded in CIGAR format\n",
            "            * *Default*: 0\n",
            "            * *Example Values*: -\n",
            "        \n",
            "        \n",
            "          + *Name*: clusterset_id\n",
            "        \n",
            "            * *Type*: String\n",
            "            * *Description*: Name of the gene-tree resource being queried. Common values are \"default\" for the standard multi-clade trees (which exclude all non-reference strains) and \"murinae\" for the trees spanning all mouse strains. By default, the most inclusive analysis will be selected\n",
            "            * *Default*: -\n",
            "            * *Example Values*: default, murinae\n",
            "        \n",
            "        \n",
            "          + *Name*: compara\n",
            "        \n",
            "            * *Type*: String\n",
            "            * *Description*: Name of the compara database to use. Multiple comparas exist on a server for separate species divisions\n",
            "            * *Default*: vertebrates\n",
            "            * *Example Values*: vertebrates\n",
            "        \n",
            "        \n",
            "          + *Name*: nh_format\n",
            "        \n",
            "            * *Type*: Enum(full, display_label_composite, simple, species, species_short_name, ncbi_taxon, ncbi_name, njtree, phylip)\n",
            "            * *Description*: The format of a NH (New Hampshire) request.\n",
            "            * *Default*: simple\n",
            "            * *Example Values*: -\n",
            "        \n",
            "        \n",
            "          + *Name*: prune_species\n",
            "        \n",
            "            * *Type*: String\n",
            "            * *Description*: Prune the tree by species. Supports all species aliases. Will return a tree with only the species given\n",
            "            * *Default*: -\n",
            "            * *Example Values*: human, cow\n",
            "        \n",
            "        \n",
            "          + *Name*: prune_taxon\n",
            "        \n",
            "            * *Type*: Integer\n",
            "            * *Description*: Prune the tree by taxon. Will return a tree with only the taxons given\n",
            "            * *Default*: -\n",
            "            * *Example Values*: 9606, 10090\n",
            "        \n",
            "        \n",
            "          + *Name*: sequence\n",
            "        \n",
            "            * *Type*: Enum(none, cdna, protein)\n",
            "            * *Description*: The type of sequence to bring back. Setting it to none results in no sequence being returned\n",
            "            * *Default*: protein\n",
            "            * *Example Values*: -\n",
            "        \n",
            "        \n",
            "        **Resource info**\n",
            "        \n",
            "        - **Methods**:  GET\n",
            "        - **Response formats**: phyloxml, orthoxml, nh, json, jsonp\n",
            "        \n",
            "        \n",
            "        **More info**\n",
            "        \n",
            "        https://rest.ensembl.org/documentation/info/genetree\n",
            "    \n",
            "    genetree_member_id(*args, **kwargs)\n",
            "            Comparative Genomics ``GET genetree/member/id/:id``\n",
            "        \n",
            "        Retrieves the gene tree that contains the gene / transcript / translation stable identifier\n",
            "        \n",
            "        \n",
            "        **Parameters**\n",
            "        \n",
            "        - Required:\n",
            "        \n",
            "          + *Name*: id\n",
            "        \n",
            "            * *Type*: String\n",
            "            * *Description*: An Ensembl stable ID\n",
            "            * *Default*: -\n",
            "            * *Example Values*: ENSG00000167664\n",
            "        \n",
            "        \n",
            "        - Optional:\n",
            "        \n",
            "          + *Name*: aligned\n",
            "        \n",
            "            * *Type*: Boolean\n",
            "            * *Description*: Return the aligned string if true. Otherwise, return the original sequence (no insertions)\n",
            "            * *Default*: 0\n",
            "            * *Example Values*: -\n",
            "        \n",
            "        \n",
            "          + *Name*: callback\n",
            "        \n",
            "            * *Type*: String\n",
            "            * *Description*: Name of the callback subroutine to be returned by the requested JSONP response. Required ONLY when using JSONP as the serialisation method. Please see , .\n",
            "            * *Default*: -\n",
            "            * *Example Values*: randomlygeneratedname\n",
            "        \n",
            "        \n",
            "          + *Name*: cigar_line\n",
            "        \n",
            "            * *Type*: Boolean\n",
            "            * *Description*: Return the aligned sequence encoded in CIGAR format\n",
            "            * *Default*: 0\n",
            "            * *Example Values*: -\n",
            "        \n",
            "        \n",
            "          + *Name*: clusterset_id\n",
            "        \n",
            "            * *Type*: String\n",
            "            * *Description*: Name of the gene-tree resource being queried. Common values are \"default\" for the standard multi-clade trees (which exclude all non-reference strains) and \"murinae\" for the trees spanning all mouse strains. By default, the most inclusive analysis will be selected\n",
            "            * *Default*: -\n",
            "            * *Example Values*: default, murinae\n",
            "        \n",
            "        \n",
            "          + *Name*: compara\n",
            "        \n",
            "            * *Type*: String\n",
            "            * *Description*: Name of the compara database to use. Multiple comparas exist on a server for separate species divisions\n",
            "            * *Default*: vertebrates\n",
            "            * *Example Values*: vertebrates\n",
            "        \n",
            "        \n",
            "          + *Name*: db_type\n",
            "        \n",
            "            * *Type*: String\n",
            "            * *Description*: Restrict the search to a database other than the default. Useful if you need to use a DB other than core\n",
            "            * *Default*: -\n",
            "            * *Example Values*: core\n",
            "        \n",
            "        \n",
            "          + *Name*: nh_format\n",
            "        \n",
            "            * *Type*: Enum(full, display_label_composite, simple, species, species_short_name, ncbi_taxon, ncbi_name, njtree, phylip)\n",
            "            * *Description*: The format of a NH (New Hampshire) request.\n",
            "            * *Default*: simple\n",
            "            * *Example Values*: -\n",
            "        \n",
            "        \n",
            "          + *Name*: object_type\n",
            "        \n",
            "            * *Type*: String\n",
            "            * *Description*: Filter by feature type\n",
            "            * *Default*: -\n",
            "            * *Example Values*: gene, transcript\n",
            "        \n",
            "        \n",
            "          + *Name*: prune_species\n",
            "        \n",
            "            * *Type*: String\n",
            "            * *Description*: Prune the tree by species. Supports all species aliases. Will return a tree with only the species given\n",
            "            * *Default*: -\n",
            "            * *Example Values*: human, cow\n",
            "        \n",
            "        \n",
            "          + *Name*: prune_taxon\n",
            "        \n",
            "            * *Type*: Integer\n",
            "            * *Description*: Prune the tree by taxon. Will return a tree with only the taxons given\n",
            "            * *Default*: -\n",
            "            * *Example Values*: 9606, 10090\n",
            "        \n",
            "        \n",
            "          + *Name*: sequence\n",
            "        \n",
            "            * *Type*: Enum(none, cdna, protein)\n",
            "            * *Description*: The type of sequence to bring back. Setting it to none results in no sequence being returned\n",
            "            * *Default*: protein\n",
            "            * *Example Values*: -\n",
            "        \n",
            "        \n",
            "          + *Name*: species\n",
            "        \n",
            "            * *Type*: String\n",
            "            * *Description*: Species name/alias\n",
            "            * *Default*: -\n",
            "            * *Example Values*: homo_sapiens, human\n",
            "        \n",
            "        \n",
            "        **Resource info**\n",
            "        \n",
            "        - **Methods**:  GET\n",
            "        - **Response formats**: phyloxml, orthoxml, nh, json, jsonp\n",
            "        \n",
            "        \n",
            "        **More info**\n",
            "        \n",
            "        https://rest.ensembl.org/documentation/info/genetree_member_id\n",
            "    \n",
            "    genetree_member_symbol(*args, **kwargs)\n",
            "            Comparative Genomics ``GET genetree/member/symbol/:species/:symbol``\n",
            "        \n",
            "        Retrieves the gene tree that contains the gene identified by a symbol\n",
            "        \n",
            "        \n",
            "        **Parameters**\n",
            "        \n",
            "        - Required:\n",
            "        \n",
            "          + *Name*: species\n",
            "        \n",
            "            * *Type*: String\n",
            "            * *Description*: Species name/alias\n",
            "            * *Default*: -\n",
            "            * *Example Values*: homo_sapiens, human\n",
            "        \n",
            "        \n",
            "          + *Name*: symbol\n",
            "        \n",
            "            * *Type*: String\n",
            "            * *Description*: Symbol or display name of a gene\n",
            "            * *Default*: -\n",
            "            * *Example Values*: BRCA2\n",
            "        \n",
            "        \n",
            "        - Optional:\n",
            "        \n",
            "          + *Name*: aligned\n",
            "        \n",
            "            * *Type*: Boolean\n",
            "            * *Description*: Return the aligned string if true. Otherwise, return the original sequence (no insertions)\n",
            "            * *Default*: 0\n",
            "            * *Example Values*: -\n",
            "        \n",
            "        \n",
            "          + *Name*: callback\n",
            "        \n",
            "            * *Type*: String\n",
            "            * *Description*: Name of the callback subroutine to be returned by the requested JSONP response. Required ONLY when using JSONP as the serialisation method. Please see , .\n",
            "            * *Default*: -\n",
            "            * *Example Values*: randomlygeneratedname\n",
            "        \n",
            "        \n",
            "          + *Name*: cigar_line\n",
            "        \n",
            "            * *Type*: Boolean\n",
            "            * *Description*: Return the aligned sequence encoded in CIGAR format\n",
            "            * *Default*: 0\n",
            "            * *Example Values*: -\n",
            "        \n",
            "        \n",
            "          + *Name*: clusterset_id\n",
            "        \n",
            "            * *Type*: String\n",
            "            * *Description*: Name of the gene-tree resource being queried. Common values are \"default\" for the standard multi-clade trees (which exclude all non-reference strains) and \"murinae\" for the trees spanning all mouse strains. By default, the most inclusive analysis will be selected\n",
            "            * *Default*: -\n",
            "            * *Example Values*: default, murinae\n",
            "        \n",
            "        \n",
            "          + *Name*: compara\n",
            "        \n",
            "            * *Type*: String\n",
            "            * *Description*: Name of the compara database to use. Multiple comparas exist on a server for separate species divisions\n",
            "            * *Default*: vertebrates\n",
            "            * *Example Values*: vertebrates\n",
            "        \n",
            "        \n",
            "          + *Name*: db_type\n",
            "        \n",
            "            * *Type*: String\n",
            "            * *Description*: Restrict the search to a database other than the default. Useful if you need to use a DB other than core\n",
            "            * *Default*: core\n",
            "            * *Example Values*: core, otherfeatures\n",
            "        \n",
            "        \n",
            "          + *Name*: external_db\n",
            "        \n",
            "            * *Type*: String\n",
            "            * *Description*: Filter by external database\n",
            "            * *Default*: -\n",
            "            * *Example Values*: HGNC\n",
            "        \n",
            "        \n",
            "          + *Name*: nh_format\n",
            "        \n",
            "            * *Type*: Enum(full, display_label_composite, simple, species, species_short_name, ncbi_taxon, ncbi_name, njtree, phylip)\n",
            "            * *Description*: The format of a NH (New Hampshire) request.\n",
            "            * *Default*: simple\n",
            "            * *Example Values*: -\n",
            "        \n",
            "        \n",
            "          + *Name*: object_type\n",
            "        \n",
            "            * *Type*: String\n",
            "            * *Description*: Filter by feature type\n",
            "            * *Default*: -\n",
            "            * *Example Values*: gene, transcript\n",
            "        \n",
            "        \n",
            "          + *Name*: prune_species\n",
            "        \n",
            "            * *Type*: String\n",
            "            * *Description*: Prune the tree by species. Supports all species aliases. Will return a tree with only the species given\n",
            "            * *Default*: -\n",
            "            * *Example Values*: human, cow\n",
            "        \n",
            "        \n",
            "          + *Name*: prune_taxon\n",
            "        \n",
            "            * *Type*: Integer\n",
            "            * *Description*: Prune the tree by taxon. Will return a tree with only the taxons given\n",
            "            * *Default*: -\n",
            "            * *Example Values*: 9606, 10090\n",
            "        \n",
            "        \n",
            "          + *Name*: sequence\n",
            "        \n",
            "            * *Type*: Enum(none, cdna, protein)\n",
            "            * *Description*: The type of sequence to bring back. Setting it to none results in no sequence being returned\n",
            "            * *Default*: protein\n",
            "            * *Example Values*: -\n",
            "        \n",
            "        \n",
            "        **Resource info**\n",
            "        \n",
            "        - **Methods**:  GET\n",
            "        - **Response formats**: phyloxml, orthoxml, nh, json, jsonp\n",
            "        \n",
            "        \n",
            "        **More info**\n",
            "        \n",
            "        https://rest.ensembl.org/documentation/info/genetree_member_symbol\n",
            "    \n",
            "    genomic_alignment_region(*args, **kwargs)\n",
            "            Comparative Genomics ``GET alignment/region/:species/:region``\n",
            "        \n",
            "        Retrieves genomic alignments as separate blocks based on a region and species\n",
            "        \n",
            "        \n",
            "        **Parameters**\n",
            "        \n",
            "        - Required:\n",
            "        \n",
            "          + *Name*: region\n",
            "        \n",
            "            * *Type*: String\n",
            "            * *Description*: Query region. A maximum of 10Mb is allowed to be requested at any one time\n",
            "            * *Default*: -\n",
            "            * *Example Values*: X:1000000..1000100:1, X:1000000..1000100:-1, X:1000000..1000100\n",
            "        \n",
            "        \n",
            "          + *Name*: species\n",
            "        \n",
            "            * *Type*: String\n",
            "            * *Description*: Species name/alias\n",
            "            * *Default*: -\n",
            "            * *Example Values*: homo_sapiens, human\n",
            "        \n",
            "        \n",
            "        - Optional:\n",
            "        \n",
            "          + *Name*: aligned\n",
            "        \n",
            "            * *Type*: Boolean\n",
            "            * *Description*: Return the aligned string if true. Otherwise, return the original sequence (no insertions)\n",
            "            * *Default*: 1\n",
            "            * *Example Values*: -\n",
            "        \n",
            "        \n",
            "          + *Name*: callback\n",
            "        \n",
            "            * *Type*: String\n",
            "            * *Description*: Name of the callback subroutine to be returned by the requested JSONP response. Required ONLY when using JSONP as the serialisation method. Please see , .\n",
            "            * *Default*: -\n",
            "            * *Example Values*: randomlygeneratedname\n",
            "        \n",
            "        \n",
            "          + *Name*: compact\n",
            "        \n",
            "            * *Type*: Boolean\n",
            "            * *Description*: Applicable to EPO_LOW_COVERAGE alignments. If true, concatenate the low coverage species sequences together to create a single sequence. Otherwise, separates out all sequences.\n",
            "            * *Default*: 1\n",
            "            * *Example Values*: -\n",
            "        \n",
            "        \n",
            "          + *Name*: compara\n",
            "        \n",
            "            * *Type*: String\n",
            "            * *Description*: Name of the compara database to use. Multiple comparas exist on a server for separate species divisions\n",
            "            * *Default*: vertebrates\n",
            "            * *Example Values*: vertebrates\n",
            "        \n",
            "        \n",
            "          + *Name*: display_species_set\n",
            "        \n",
            "            * *Type*: String\n",
            "            * *Description*: Subset of species in the alignment to be displayed (multiple values). All the species in the alignment will be displayed if this is not set. Any valid alias may be used.\n",
            "            * *Default*: -\n",
            "            * *Example Values*: human, chimp, gorilla\n",
            "        \n",
            "        \n",
            "          + *Name*: mask\n",
            "        \n",
            "            * *Type*: Enum(hard,soft)\n",
            "            * *Description*: Request the sequence masked for repeat sequences. Hard will mask all repeats as N's and soft will mask repeats as lowercased characters.\n",
            "            * *Default*: -\n",
            "            * *Example Values*: hard\n",
            "        \n",
            "        \n",
            "          + *Name*: method\n",
            "        \n",
            "            * *Type*: Enum(EPO, EPO_LOW_COVERAGE, PECAN, LASTZ_NET, BLASTZ_NET, TRANSLATED_BLAT_NET, CACTUS_HAL, CACTUS_HAL_PW)\n",
            "            * *Description*: The alignment method\n",
            "            * *Default*: EPO\n",
            "            * *Example Values*: PECAN\n",
            "        \n",
            "        \n",
            "          + *Name*: species_set\n",
            "        \n",
            "            * *Type*: String\n",
            "            * *Description*: The set of species used to define the pairwise alignment (multiple values). Should not be used with the species_set_group parameter. Use ,  with one of the methods listed above to obtain a valid list of species sets. Any valid alias may be used.\n",
            "            * *Default*: -\n",
            "            * *Example Values*: homo_sapiens, mus_musculus\n",
            "        \n",
            "        \n",
            "          + *Name*: species_set_group\n",
            "        \n",
            "            * *Type*: String\n",
            "            * *Description*: The species set group name of the multiple alignment. Should not be used with the species_set parameter. Use ,  with one of the methods listed above to obtain a valid list of group names.\n",
            "            * *Default*: mammals\n",
            "            * *Example Values*: mammals, amniotes, fish, sauropsids, murinae\n",
            "        \n",
            "        \n",
            "        **Resource info**\n",
            "        \n",
            "        - **Methods**:  GET\n",
            "        - **Response formats**: json, xml, phyloxml, jsonp\n",
            "        \n",
            "        \n",
            "        **More info**\n",
            "        \n",
            "        https://rest.ensembl.org/documentation/info/genomic_alignment_region\n",
            "    \n",
            "    get_binding_matrix(*args, **kwargs)\n",
            "            Regulation ``GET species/:species/binding_matrix/:binding_matrix_stable_id/``\n",
            "        \n",
            "        Return the specified binding matrix\n",
            "        \n",
            "        \n",
            "        **Parameters**\n",
            "        \n",
            "        - Required:\n",
            "        \n",
            "          + *Name*: binding_matrix\n",
            "        \n",
            "            * *Type*: String\n",
            "            * *Description*: Stable ID of binding matrix\n",
            "            * *Default*: -\n",
            "            * *Example Values*: ENSPFM0001\n",
            "        \n",
            "        \n",
            "          + *Name*: species\n",
            "        \n",
            "            * *Type*: String\n",
            "            * *Description*: Species name/alias\n",
            "            * *Default*: -\n",
            "            * *Example Values*: homo_sapiens\n",
            "        \n",
            "        \n",
            "        - Optional:\n",
            "        \n",
            "          + *Name*: callback\n",
            "        \n",
            "            * *Type*: String\n",
            "            * *Description*: Name of the callback subroutine to be returned by the requested JSONP response. Required ONLY when using JSONP as the serialisation method. Please see , .\n",
            "            * *Default*: -\n",
            "            * *Example Values*: randomlygeneratedname\n",
            "        \n",
            "        \n",
            "          + *Name*: unit\n",
            "        \n",
            "            * *Type*: String\n",
            "            * *Description*: Unit of the matrix elements\n",
            "            * *Default*: frequencies\n",
            "            * *Example Values*: frequencies, probabilities, bits\n",
            "        \n",
            "        \n",
            "        **Resource info**\n",
            "        \n",
            "        - **Methods**:  GET\n",
            "        - **Response formats**: json, xml, jsonp\n",
            "        \n",
            "        \n",
            "        **More info**\n",
            "        \n",
            "        https://rest.ensembl.org/documentation/info/get_binding_matrix\n",
            "    \n",
            "    homology_ensemblgene(*args, **kwargs)\n",
            "            Comparative Genomics ``GET homology/id/:id``\n",
            "        \n",
            "        Retrieves homology information (orthologs) by Ensembl gene id\n",
            "        \n",
            "        \n",
            "        **Parameters**\n",
            "        \n",
            "        - Required:\n",
            "        \n",
            "          + *Name*: id\n",
            "        \n",
            "            * *Type*: String\n",
            "            * *Description*: An Ensembl stable ID\n",
            "            * *Default*: -\n",
            "            * *Example Values*: ENSG00000157764\n",
            "        \n",
            "        \n",
            "        - Optional:\n",
            "        \n",
            "          + *Name*: aligned\n",
            "        \n",
            "            * *Type*: Boolean\n",
            "            * *Description*: Return the aligned string if true. Otherwise, return the original sequence (no insertions)\n",
            "            * *Default*: 1\n",
            "            * *Example Values*: -\n",
            "        \n",
            "        \n",
            "          + *Name*: callback\n",
            "        \n",
            "            * *Type*: String\n",
            "            * *Description*: Name of the callback subroutine to be returned by the requested JSONP response. Required ONLY when using JSONP as the serialisation method. Please see , .\n",
            "            * *Default*: -\n",
            "            * *Example Values*: randomlygeneratedname\n",
            "        \n",
            "        \n",
            "          + *Name*: cigar_line\n",
            "        \n",
            "            * *Type*: Boolean\n",
            "            * *Description*: Return the aligned sequence encoded in CIGAR format\n",
            "            * *Default*: 1\n",
            "            * *Example Values*: -\n",
            "        \n",
            "        \n",
            "          + *Name*: compara\n",
            "        \n",
            "            * *Type*: String\n",
            "            * *Description*: Name of the compara database to use. Multiple comparas exist on a server for separate species divisions\n",
            "            * *Default*: vertebrates\n",
            "            * *Example Values*: vertebrates\n",
            "        \n",
            "        \n",
            "          + *Name*: format\n",
            "        \n",
            "            * *Type*: Enum(full, condensed)\n",
            "            * *Description*: Layout of the response\n",
            "            * *Default*: full\n",
            "            * *Example Values*: -\n",
            "        \n",
            "        \n",
            "          + *Name*: sequence\n",
            "        \n",
            "            * *Type*: Enum(none, cdna, protein)\n",
            "            * *Description*: The type of sequence to bring back. Setting it to none results in no sequence being returned\n",
            "            * *Default*: protein\n",
            "            * *Example Values*: -\n",
            "        \n",
            "        \n",
            "          + *Name*: target_species\n",
            "        \n",
            "            * *Type*: String\n",
            "            * *Description*: Filter by species. Supports all species aliases\n",
            "            * *Default*: -\n",
            "            * *Example Values*: human, cow\n",
            "        \n",
            "        \n",
            "          + *Name*: target_taxon\n",
            "        \n",
            "            * *Type*: Integer\n",
            "            * *Description*: Filter by taxon\n",
            "            * *Default*: -\n",
            "            * *Example Values*: 9606, 10090\n",
            "        \n",
            "        \n",
            "          + *Name*: type\n",
            "        \n",
            "            * *Type*: Enum(orthologues, paralogues, projections, all)\n",
            "            * *Description*: The type of homology to return from this call. Projections are orthology calls defined between alternative assemblies and the genes shared between them. Useful if you need only one type of homology back from the service\n",
            "            * *Default*: all\n",
            "            * *Example Values*: -\n",
            "        \n",
            "        \n",
            "        **Resource info**\n",
            "        \n",
            "        - **Methods**:  GET\n",
            "        - **Response formats**: json, xml, orthoxml, jsonp\n",
            "        \n",
            "        \n",
            "        **More info**\n",
            "        \n",
            "        https://rest.ensembl.org/documentation/info/homology_ensemblgene\n",
            "    \n",
            "    homology_symbol(*args, **kwargs)\n",
            "            Comparative Genomics ``GET homology/symbol/:species/:symbol``\n",
            "        \n",
            "        Retrieves homology information (orthologs) by symbol\n",
            "        \n",
            "        \n",
            "        **Parameters**\n",
            "        \n",
            "        - Required:\n",
            "        \n",
            "          + *Name*: species\n",
            "        \n",
            "            * *Type*: String\n",
            "            * *Description*: Species name/alias\n",
            "            * *Default*: -\n",
            "            * *Example Values*: homo_sapiens, human\n",
            "        \n",
            "        \n",
            "          + *Name*: symbol\n",
            "        \n",
            "            * *Type*: String\n",
            "            * *Description*: Symbol or display name of a gene\n",
            "            * *Default*: -\n",
            "            * *Example Values*: BRCA2\n",
            "        \n",
            "        \n",
            "        - Optional:\n",
            "        \n",
            "          + *Name*: aligned\n",
            "        \n",
            "            * *Type*: Boolean\n",
            "            * *Description*: Return the aligned string if true. Otherwise, return the original sequence (no insertions)\n",
            "            * *Default*: 1\n",
            "            * *Example Values*: -\n",
            "        \n",
            "        \n",
            "          + *Name*: callback\n",
            "        \n",
            "            * *Type*: String\n",
            "            * *Description*: Name of the callback subroutine to be returned by the requested JSONP response. Required ONLY when using JSONP as the serialisation method. Please see , .\n",
            "            * *Default*: -\n",
            "            * *Example Values*: randomlygeneratedname\n",
            "        \n",
            "        \n",
            "          + *Name*: cigar_line\n",
            "        \n",
            "            * *Type*: Boolean\n",
            "            * *Description*: Return the aligned sequence encoded in CIGAR format\n",
            "            * *Default*: 1\n",
            "            * *Example Values*: -\n",
            "        \n",
            "        \n",
            "          + *Name*: compara\n",
            "        \n",
            "            * *Type*: String\n",
            "            * *Description*: Name of the compara database to use. Multiple comparas exist on a server for separate species divisions\n",
            "            * *Default*: vertebrates\n",
            "            * *Example Values*: vertebrates\n",
            "        \n",
            "        \n",
            "          + *Name*: external_db\n",
            "        \n",
            "            * *Type*: String\n",
            "            * *Description*: Filter by external database\n",
            "            * *Default*: -\n",
            "            * *Example Values*: HGNC\n",
            "        \n",
            "        \n",
            "          + *Name*: format\n",
            "        \n",
            "            * *Type*: Enum(full,condensed)\n",
            "            * *Description*: Layout of the response\n",
            "            * *Default*: full\n",
            "            * *Example Values*: -\n",
            "        \n",
            "        \n",
            "          + *Name*: sequence\n",
            "        \n",
            "            * *Type*: Enum(none, cdna, protein)\n",
            "            * *Description*: The type of sequence to bring back. Setting it to none results in no sequence being returned\n",
            "            * *Default*: protein\n",
            "            * *Example Values*: -\n",
            "        \n",
            "        \n",
            "          + *Name*: target_species\n",
            "        \n",
            "            * *Type*: String\n",
            "            * *Description*: Filter by species. Supports all species aliases\n",
            "            * *Default*: -\n",
            "            * *Example Values*: homo_sapiens, human\n",
            "        \n",
            "        \n",
            "          + *Name*: target_taxon\n",
            "        \n",
            "            * *Type*: Integer\n",
            "            * *Description*: Filter by taxon\n",
            "            * *Default*: -\n",
            "            * *Example Values*: 9606, 10090\n",
            "        \n",
            "        \n",
            "          + *Name*: type\n",
            "        \n",
            "            * *Type*: Enum(orthologues, paralogues, projections, all)\n",
            "            * *Description*: The type of homology to return from this call. Projections are orthology calls defined between alternative assemblies and the genes shared between them. Useful if you need only one type of homology back from the service\n",
            "            * *Default*: all\n",
            "            * *Example Values*: -\n",
            "        \n",
            "        \n",
            "        **Resource info**\n",
            "        \n",
            "        - **Methods**:  GET\n",
            "        - **Response formats**: json, xml, orthoxml, jsonp\n",
            "        \n",
            "        \n",
            "        **More info**\n",
            "        \n",
            "        https://rest.ensembl.org/documentation/info/homology_symbol\n",
            "    \n",
            "    info_divisions(*args, **kwargs)\n",
            "            Information ``GET info/divisions``\n",
            "        \n",
            "        Get list of all Ensembl divisions for which information is available\n",
            "        \n",
            "        \n",
            "        **Parameters**\n",
            "        \n",
            "        - Required:\n",
            "        \n",
            "          + *Name*: callback\n",
            "        \n",
            "            * *Type*: String\n",
            "            * *Description*: Name of the callback subroutine to be returned by the requested JSONP response. Required ONLY when using JSONP as the serialisation method. Please see , .\n",
            "            * *Default*: -\n",
            "            * *Example Values*: randomlygeneratedname\n",
            "        \n",
            "        \n",
            "        - Optional:\n",
            "        \n",
            "          + *Name*: callback\n",
            "        \n",
            "            * *Type*: String\n",
            "            * *Description*: Name of the callback subroutine to be returned by the requested JSONP response. Required ONLY when using JSONP as the serialisation method. Please see , .\n",
            "            * *Default*: -\n",
            "            * *Example Values*: randomlygeneratedname\n",
            "        \n",
            "        \n",
            "        **Resource info**\n",
            "        \n",
            "        - **Methods**:  GET\n",
            "        - **Response formats**: json, xml, jsonp\n",
            "        \n",
            "        \n",
            "        **More info**\n",
            "        \n",
            "        https://rest.ensembl.org/documentation/info/info_divisions\n",
            "    \n",
            "    info_genome(*args, **kwargs)\n",
            "            Information ``GET info/genomes/:genome_name``\n",
            "        \n",
            "        Find information about a given genome\n",
            "        \n",
            "        \n",
            "        **Parameters**\n",
            "        \n",
            "        - Required:\n",
            "        \n",
            "          + *Name*: name\n",
            "        \n",
            "            * *Type*: String\n",
            "            * *Description*: The production name of the genome.\n",
            "            * *Default*: -\n",
            "            * *Example Values*: nanoarchaeum_equitans_kin4_m\n",
            "        \n",
            "        \n",
            "        - Optional:\n",
            "        \n",
            "          + *Name*: callback\n",
            "        \n",
            "            * *Type*: String\n",
            "            * *Description*: Name of the callback subroutine to be returned by the requested JSONP response. Required ONLY when using JSONP as the serialisation method. Please see , .\n",
            "            * *Default*: -\n",
            "            * *Example Values*: randomlygeneratedname\n",
            "        \n",
            "        \n",
            "          + *Name*: expand\n",
            "        \n",
            "            * *Type*: Boolean(0,1)\n",
            "            * *Description*: Expands the information to include details of sequences. Can be very large.\n",
            "            * *Default*: NULL\n",
            "            * *Example Values*: -\n",
            "        \n",
            "        \n",
            "        **Resource info**\n",
            "        \n",
            "        - **Methods**:  GET\n",
            "        - **Response formats**: json, xml, jsonp\n",
            "        \n",
            "        \n",
            "        **More info**\n",
            "        \n",
            "        https://rest.ensembl.org/documentation/info/info_genome\n",
            "    \n",
            "    info_genomes_accession(*args, **kwargs)\n",
            "            Information ``GET info/genomes/accession/:accession``\n",
            "        \n",
            "        Find information about genomes containing a specified INSDC accession\n",
            "        \n",
            "        \n",
            "        **Parameters**\n",
            "        \n",
            "        - Required:\n",
            "        \n",
            "          + *Name*: accession\n",
            "        \n",
            "            * *Type*: String\n",
            "            * *Description*: INSDC sequence accession (optionally versioned)\n",
            "            * *Default*: -\n",
            "            * *Example Values*: U00096\n",
            "        \n",
            "        \n",
            "        - Optional:\n",
            "        \n",
            "          + *Name*: callback\n",
            "        \n",
            "            * *Type*: String\n",
            "            * *Description*: Name of the callback subroutine to be returned by the requested JSONP response. Required ONLY when using JSONP as the serialisation method. Please see , .\n",
            "            * *Default*: -\n",
            "            * *Example Values*: randomlygeneratedname\n",
            "        \n",
            "        \n",
            "          + *Name*: expand\n",
            "        \n",
            "            * *Type*: Boolean(0,1)\n",
            "            * *Description*: Expands the information to include details of sequences. Can be very large.\n",
            "            * *Default*: NULL\n",
            "            * *Example Values*: -\n",
            "        \n",
            "        \n",
            "        **Resource info**\n",
            "        \n",
            "        - **Methods**:  GET\n",
            "        - **Response formats**: json, xml, jsonp\n",
            "        \n",
            "        \n",
            "        **More info**\n",
            "        \n",
            "        https://rest.ensembl.org/documentation/info/info_genomes_accession\n",
            "    \n",
            "    info_genomes_assembly(*args, **kwargs)\n",
            "            Information ``GET info/genomes/assembly/:assembly_id``\n",
            "        \n",
            "        Find information about a genome with a specified assembly\n",
            "        \n",
            "        \n",
            "        **Parameters**\n",
            "        \n",
            "        - Required:\n",
            "        \n",
            "          + *Name*: assembly_id\n",
            "        \n",
            "            * *Type*: String\n",
            "            * *Description*: INSDC assembly ID (optionally versioned)\n",
            "            * *Default*: -\n",
            "            * *Example Values*: GCA_000005005.6\n",
            "        \n",
            "        \n",
            "        - Optional:\n",
            "        \n",
            "          + *Name*: callback\n",
            "        \n",
            "            * *Type*: String\n",
            "            * *Description*: Name of the callback subroutine to be returned by the requested JSONP response. Required ONLY when using JSONP as the serialisation method. Please see , .\n",
            "            * *Default*: -\n",
            "            * *Example Values*: randomlygeneratedname\n",
            "        \n",
            "        \n",
            "          + *Name*: expand\n",
            "        \n",
            "            * *Type*: Boolean(0,1)\n",
            "            * *Description*: Expands the information to include details of sequences. Can be very large.\n",
            "            * *Default*: NULL\n",
            "            * *Example Values*: -\n",
            "        \n",
            "        \n",
            "        **Resource info**\n",
            "        \n",
            "        - **Methods**:  GET\n",
            "        - **Response formats**: json, xml, jsonp\n",
            "        \n",
            "        \n",
            "        **More info**\n",
            "        \n",
            "        https://rest.ensembl.org/documentation/info/info_genomes_assembly\n",
            "    \n",
            "    info_genomes_division(*args, **kwargs)\n",
            "            Information ``GET info/genomes/division/:division_name``\n",
            "        \n",
            "        Find information about all genomes in a given division. May be large for Ensembl Bacteria.\n",
            "        \n",
            "        \n",
            "        **Parameters**\n",
            "        \n",
            "        - Required:\n",
            "        \n",
            "          + *Name*: division\n",
            "        \n",
            "            * *Type*: String\n",
            "            * *Description*: The name of the division.\n",
            "            * *Default*: -\n",
            "            * *Example Values*: EnsemblPlants\n",
            "        \n",
            "        \n",
            "        - Optional:\n",
            "        \n",
            "          + *Name*: callback\n",
            "        \n",
            "            * *Type*: String\n",
            "            * *Description*: Name of the callback subroutine to be returned by the requested JSONP response. Required ONLY when using JSONP as the serialisation method. Please see , .\n",
            "            * *Default*: -\n",
            "            * *Example Values*: randomlygeneratedname\n",
            "        \n",
            "        \n",
            "          + *Name*: expand\n",
            "        \n",
            "            * *Type*: Boolean(0,1)\n",
            "            * *Description*: Expands the information to include details of sequences. Can be very large.\n",
            "            * *Default*: NULL\n",
            "            * *Example Values*: -\n",
            "        \n",
            "        \n",
            "        **Resource info**\n",
            "        \n",
            "        - **Methods**:  GET\n",
            "        - **Response formats**: json, xml, jsonp\n",
            "        \n",
            "        \n",
            "        **More info**\n",
            "        \n",
            "        https://rest.ensembl.org/documentation/info/info_genomes_division\n",
            "    \n",
            "    info_genomes_taxonomy(*args, **kwargs)\n",
            "            Information ``GET info/genomes/taxonomy/:taxon_name``\n",
            "        \n",
            "        Find information about all genomes beneath a given node of the taxonomy\n",
            "        \n",
            "        \n",
            "        **Parameters**\n",
            "        \n",
            "        - Required:\n",
            "        \n",
            "          + *Name*: taxon_name\n",
            "        \n",
            "            * *Type*: String\n",
            "            * *Description*: Taxon name or NCBI taxonomy ID\n",
            "            * *Default*: -\n",
            "            * *Example Values*: Homo sapiens\n",
            "        \n",
            "        \n",
            "        - Optional:\n",
            "        \n",
            "          + *Name*: callback\n",
            "        \n",
            "            * *Type*: String\n",
            "            * *Description*: Name of the callback subroutine to be returned by the requested JSONP response. Required ONLY when using JSONP as the serialisation method. Please see , .\n",
            "            * *Default*: -\n",
            "            * *Example Values*: randomlygeneratedname\n",
            "        \n",
            "        \n",
            "          + *Name*: expand\n",
            "        \n",
            "            * *Type*: Boolean(0,1)\n",
            "            * *Description*: Expands the information to include details of sequences. Can be very large.\n",
            "            * *Default*: NULL\n",
            "            * *Example Values*: -\n",
            "        \n",
            "        \n",
            "        **Resource info**\n",
            "        \n",
            "        - **Methods**:  GET\n",
            "        - **Response formats**: json, xml, jsonp\n",
            "        \n",
            "        \n",
            "        **More info**\n",
            "        \n",
            "        https://rest.ensembl.org/documentation/info/info_genomes_taxonomy\n",
            "    \n",
            "    ld_id_get(*args, **kwargs)\n",
            "            Linkage Disequilibrium ``GET ld/:species/:id/:population_name``\n",
            "        \n",
            "        Computes and returns LD values between the given variant and all other variants in a window centered around the given variant. The window size is set to 500 kb.\n",
            "        \n",
            "        \n",
            "        **Parameters**\n",
            "        \n",
            "        - Required:\n",
            "        \n",
            "          + *Name*: id\n",
            "        \n",
            "            * *Type*: String\n",
            "            * *Description*: Variant id\n",
            "            * *Default*: -\n",
            "            * *Example Values*: rs56116432\n",
            "        \n",
            "        \n",
            "          + *Name*: population_name\n",
            "        \n",
            "            * *Type*: String\n",
            "            * *Description*: \n",
            "            * *Default*: -\n",
            "            * *Example Values*: 1000GENOMES:phase_3:KHV\n",
            "        \n",
            "        \n",
            "          + *Name*: species\n",
            "        \n",
            "            * *Type*: String\n",
            "            * *Description*: Species name/alias\n",
            "            * *Default*: -\n",
            "            * *Example Values*: homo_sapiens, human\n",
            "        \n",
            "        \n",
            "        - Optional:\n",
            "        \n",
            "          + *Name*: attribs\n",
            "        \n",
            "            * *Type*: Boolean\n",
            "            * *Description*: \n",
            "            * *Default*: 0\n",
            "            * *Example Values*: -\n",
            "        \n",
            "        \n",
            "          + *Name*: callback\n",
            "        \n",
            "            * *Type*: String\n",
            "            * *Description*: Name of the callback subroutine to be returned by the requested JSONP response. Required ONLY when using JSONP as the serialisation method. Please see , .\n",
            "            * *Default*: -\n",
            "            * *Example Values*: randomlygeneratedname\n",
            "        \n",
            "        \n",
            "          + *Name*: d_prime\n",
            "        \n",
            "            * *Type*: Float\n",
            "            * *Description*: \n",
            "            * *Default*: 0\n",
            "            * *Example Values*: 1.0\n",
            "        \n",
            "        \n",
            "          + *Name*: r2\n",
            "        \n",
            "            * *Type*: Float\n",
            "            * *Description*: \n",
            "            * *Default*: 0\n",
            "            * *Example Values*: 0.85\n",
            "        \n",
            "        \n",
            "          + *Name*: window_size\n",
            "        \n",
            "            * *Type*: Integer\n",
            "            * *Description*: \n",
            "            * *Default*: 500\n",
            "            * *Example Values*: 500\n",
            "        \n",
            "        \n",
            "        **Resource info**\n",
            "        \n",
            "        - **Methods**:  GET\n",
            "        - **Response formats**: json, xml, jsonp\n",
            "        \n",
            "        \n",
            "        **More info**\n",
            "        \n",
            "        https://rest.ensembl.org/documentation/info/ld_id_get\n",
            "    \n",
            "    ld_pairwise_get(*args, **kwargs)\n",
            "            Linkage Disequilibrium ``GET ld/:species/pairwise/:id1/:id2``\n",
            "        \n",
            "        Computes and returns LD values between the given variants.\n",
            "        \n",
            "        \n",
            "        **Parameters**\n",
            "        \n",
            "        - Required:\n",
            "        \n",
            "          + *Name*: id1\n",
            "        \n",
            "            * *Type*: String\n",
            "            * *Description*: Variant id1\n",
            "            * *Default*: -\n",
            "            * *Example Values*: rs6792369\n",
            "        \n",
            "        \n",
            "          + *Name*: id2\n",
            "        \n",
            "            * *Type*: String\n",
            "            * *Description*: Variant id2\n",
            "            * *Default*: -\n",
            "            * *Example Values*: rs1042779\n",
            "        \n",
            "        \n",
            "          + *Name*: species\n",
            "        \n",
            "            * *Type*: String\n",
            "            * *Description*: Species name/alias\n",
            "            * *Default*: -\n",
            "            * *Example Values*: homo_sapiens, human\n",
            "        \n",
            "        \n",
            "        - Optional:\n",
            "        \n",
            "          + *Name*: callback\n",
            "        \n",
            "            * *Type*: String\n",
            "            * *Description*: Name of the callback subroutine to be returned by the requested JSONP response. Required ONLY when using JSONP as the serialisation method. Please see , .\n",
            "            * *Default*: -\n",
            "            * *Example Values*: randomlygeneratedname\n",
            "        \n",
            "        \n",
            "          + *Name*: d_prime\n",
            "        \n",
            "            * *Type*: Float\n",
            "            * *Description*: \n",
            "            * *Default*: 0\n",
            "            * *Example Values*: 1.0\n",
            "        \n",
            "        \n",
            "          + *Name*: population_name\n",
            "        \n",
            "            * *Type*: String\n",
            "            * *Description*: \n",
            "            * *Default*: 0\n",
            "            * *Example Values*: 1000GENOMES:phase_3:KHV\n",
            "        \n",
            "        \n",
            "          + *Name*: r2\n",
            "        \n",
            "            * *Type*: Float\n",
            "            * *Description*: \n",
            "            * *Default*: 0\n",
            "            * *Example Values*: 0.85\n",
            "        \n",
            "        \n",
            "        **Resource info**\n",
            "        \n",
            "        - **Methods**:  GET\n",
            "        - **Response formats**: json, xml, jsonp\n",
            "        \n",
            "        \n",
            "        **More info**\n",
            "        \n",
            "        https://rest.ensembl.org/documentation/info/ld_pairwise_get\n",
            "    \n",
            "    ld_region_get(*args, **kwargs)\n",
            "            Linkage Disequilibrium ``GET ld/:species/region/:region/:population_name``\n",
            "        \n",
            "        Computes and returns LD values between all pairs of variants in the defined region.\n",
            "        \n",
            "        \n",
            "        **Parameters**\n",
            "        \n",
            "        - Required:\n",
            "        \n",
            "          + *Name*: population_name\n",
            "        \n",
            "            * *Type*: String\n",
            "            * *Description*: \n",
            "            * *Default*: -\n",
            "            * *Example Values*: 1000GENOMES:phase_3:KHV\n",
            "        \n",
            "        \n",
            "          + *Name*: region\n",
            "        \n",
            "            * *Type*: String\n",
            "            * *Description*: Query region. A maximum of 1Mb is allowed.\n",
            "            * *Default*: -\n",
            "            * *Example Values*: 6:25837556..25843455\n",
            "        \n",
            "        \n",
            "          + *Name*: species\n",
            "        \n",
            "            * *Type*: String\n",
            "            * *Description*: Species name/alias\n",
            "            * *Default*: -\n",
            "            * *Example Values*: homo_sapiens, human\n",
            "        \n",
            "        \n",
            "        - Optional:\n",
            "        \n",
            "          + *Name*: callback\n",
            "        \n",
            "            * *Type*: String\n",
            "            * *Description*: Name of the callback subroutine to be returned by the requested JSONP response. Required ONLY when using JSONP as the serialisation method. Please see , .\n",
            "            * *Default*: -\n",
            "            * *Example Values*: randomlygeneratedname\n",
            "        \n",
            "        \n",
            "          + *Name*: d_prime\n",
            "        \n",
            "            * *Type*: Float\n",
            "            * *Description*: \n",
            "            * *Default*: 0\n",
            "            * *Example Values*: 1.0\n",
            "        \n",
            "        \n",
            "          + *Name*: r2\n",
            "        \n",
            "            * *Type*: Float\n",
            "            * *Description*: \n",
            "            * *Default*: 0\n",
            "            * *Example Values*: 0.85\n",
            "        \n",
            "        \n",
            "        **Resource info**\n",
            "        \n",
            "        - **Methods**:  GET\n",
            "        - **Response formats**: json, xml, jsonp\n",
            "        \n",
            "        \n",
            "        **More info**\n",
            "        \n",
            "        https://rest.ensembl.org/documentation/info/ld_region_get\n",
            "    \n",
            "    list_all_microarrays(*args, **kwargs)\n",
            "            Regulation ``GET regulatory/species/:species/microarray``\n",
            "        \n",
            "        Returns information about all microarrays available for the given species\n",
            "        \n",
            "        \n",
            "        **Parameters**\n",
            "        \n",
            "        - Required:\n",
            "        \n",
            "          + *Name*: species\n",
            "        \n",
            "            * *Type*: String\n",
            "            * *Description*: Species name/alias\n",
            "            * *Default*: -\n",
            "            * *Example Values*: homo_sapiens\n",
            "        \n",
            "        \n",
            "        - Optional:\n",
            "        \n",
            "          + *Name*: callback\n",
            "        \n",
            "            * *Type*: String\n",
            "            * *Description*: Name of the callback subroutine to be returned by the requested JSONP response. Required ONLY when using JSONP as the serialisation method. Please see , .\n",
            "            * *Default*: -\n",
            "            * *Example Values*: randomlygeneratedname\n",
            "        \n",
            "        \n",
            "        **Resource info**\n",
            "        \n",
            "        - **Methods**:  GET\n",
            "        - **Response formats**: json, xml, jsonp\n",
            "        \n",
            "        \n",
            "        **More info**\n",
            "        \n",
            "        https://rest.ensembl.org/documentation/info/list_all_microarrays\n",
            "    \n",
            "    lookup(*args, **kwargs)\n",
            "            Lookup ``GET lookup/id/:id``\n",
            "        \n",
            "        Find the species and database for a single identifier e.g. gene, transcript, protein\n",
            "        \n",
            "        \n",
            "        **Parameters**\n",
            "        \n",
            "        - Required:\n",
            "        \n",
            "          + *Name*: id\n",
            "        \n",
            "            * *Type*: String\n",
            "            * *Description*: An Ensembl stable ID\n",
            "            * *Default*: -\n",
            "            * *Example Values*: ENSG00000157764\n",
            "        \n",
            "        \n",
            "        - Optional:\n",
            "        \n",
            "          + *Name*: callback\n",
            "        \n",
            "            * *Type*: String\n",
            "            * *Description*: Name of the callback subroutine to be returned by the requested JSONP response. Required ONLY when using JSONP as the serialisation method. Please see , .\n",
            "            * *Default*: -\n",
            "            * *Example Values*: randomlygeneratedname\n",
            "        \n",
            "        \n",
            "          + *Name*: db_type\n",
            "        \n",
            "            * *Type*: String\n",
            "            * *Description*: Restrict the search to a database other than the default. Useful if you need to use a DB other than core\n",
            "            * *Default*: -\n",
            "            * *Example Values*: core, otherfeatures\n",
            "        \n",
            "        \n",
            "          + *Name*: expand\n",
            "        \n",
            "            * *Type*: Boolean(0,1)\n",
            "            * *Description*: Expands the search to include any connected features. e.g. If the object is a gene, its transcripts, translations and exons will be returned as well.\n",
            "            * *Default*: 0\n",
            "            * *Example Values*: -\n",
            "        \n",
            "        \n",
            "          + *Name*: format\n",
            "        \n",
            "            * *Type*: Enum(full,condensed)\n",
            "            * *Description*: Specify the formats to emit from this endpoint\n",
            "            * *Default*: full\n",
            "            * *Example Values*: -\n",
            "        \n",
            "        \n",
            "          + *Name*: phenotypes\n",
            "        \n",
            "            * *Type*: Boolean(0,1)\n",
            "            * *Description*: Include phenotypes. Only available for gene objects.\n",
            "            * *Default*: 0\n",
            "            * *Example Values*: -\n",
            "        \n",
            "        \n",
            "          + *Name*: species\n",
            "        \n",
            "            * *Type*: String\n",
            "            * *Description*: Species name/alias\n",
            "            * *Default*: -\n",
            "            * *Example Values*: homo_sapiens, human\n",
            "        \n",
            "        \n",
            "          + *Name*: utr\n",
            "        \n",
            "            * *Type*: Boolean(0,1)\n",
            "            * *Description*: Include 5' and 3' UTR features. Only available if the expand option is used.\n",
            "            * *Default*: 0\n",
            "            * *Example Values*: -\n",
            "        \n",
            "        \n",
            "        **Resource info**\n",
            "        \n",
            "        - **Methods**:  GET\n",
            "        - **Response formats**: json, xml, jsonp\n",
            "        \n",
            "        \n",
            "        **More info**\n",
            "        \n",
            "        https://rest.ensembl.org/documentation/info/lookup\n",
            "    \n",
            "    lookup_post(*args, **kwargs)\n",
            "            Lookup ``POST lookup/id``\n",
            "        \n",
            "        Find the species and database for several identifiers. IDs that are not found are returned with no data.\n",
            "        \n",
            "        \n",
            "        **Parameters**\n",
            "        \n",
            "        - Required:\n",
            "        \n",
            "          + *Name*: callback\n",
            "        \n",
            "            * *Type*: String\n",
            "            * *Description*: Name of the callback subroutine to be returned by the requested JSONP response. Required ONLY when using JSONP as the serialisation method. Please see , .\n",
            "            * *Default*: -\n",
            "            * *Example Values*: randomlygeneratedname\n",
            "        \n",
            "        \n",
            "          + *Name*: db_type\n",
            "        \n",
            "            * *Type*: String\n",
            "            * *Description*: Restrict the search to a database other than the default. Useful if you need to use a DB other than core\n",
            "            * *Default*: -\n",
            "            * *Example Values*: core, otherfeatures\n",
            "        \n",
            "        \n",
            "          + *Name*: expand\n",
            "        \n",
            "            * *Type*: Boolean(0,1)\n",
            "            * *Description*: Expands the search to include any connected features. e.g. If the object is a gene, its transcripts, translations and exons will be returned as well.\n",
            "            * *Default*: 0\n",
            "            * *Example Values*: -\n",
            "        \n",
            "        \n",
            "          + *Name*: format\n",
            "        \n",
            "            * *Type*: Enum(full,condensed)\n",
            "            * *Description*: Specify the formats to emit from this endpoint\n",
            "            * *Default*: full\n",
            "            * *Example Values*: -\n",
            "        \n",
            "        \n",
            "          + *Name*: object_type\n",
            "        \n",
            "            * *Type*: String\n",
            "            * *Description*: Filter by feature type\n",
            "            * *Default*: -\n",
            "            * *Example Values*: gene, transcript\n",
            "        \n",
            "        \n",
            "          + *Name*: species\n",
            "        \n",
            "            * *Type*: String\n",
            "            * *Description*: Species name/alias. Causes problems if the species doesn't match the identifiers in the POST body.\n",
            "            * *Default*: -\n",
            "            * *Example Values*: homo_sapiens, human\n",
            "        \n",
            "        \n",
            "        - Optional:\n",
            "        \n",
            "          + *Name*: callback\n",
            "        \n",
            "            * *Type*: String\n",
            "            * *Description*: Name of the callback subroutine to be returned by the requested JSONP response. Required ONLY when using JSONP as the serialisation method. Please see , .\n",
            "            * *Default*: -\n",
            "            * *Example Values*: randomlygeneratedname\n",
            "        \n",
            "        \n",
            "          + *Name*: db_type\n",
            "        \n",
            "            * *Type*: String\n",
            "            * *Description*: Restrict the search to a database other than the default. Useful if you need to use a DB other than core\n",
            "            * *Default*: -\n",
            "            * *Example Values*: core, otherfeatures\n",
            "        \n",
            "        \n",
            "          + *Name*: expand\n",
            "        \n",
            "            * *Type*: Boolean(0,1)\n",
            "            * *Description*: Expands the search to include any connected features. e.g. If the object is a gene, its transcripts, translations and exons will be returned as well.\n",
            "            * *Default*: 0\n",
            "            * *Example Values*: -\n",
            "        \n",
            "        \n",
            "          + *Name*: format\n",
            "        \n",
            "            * *Type*: Enum(full,condensed)\n",
            "            * *Description*: Specify the formats to emit from this endpoint\n",
            "            * *Default*: full\n",
            "            * *Example Values*: -\n",
            "        \n",
            "        \n",
            "          + *Name*: object_type\n",
            "        \n",
            "            * *Type*: String\n",
            "            * *Description*: Filter by feature type\n",
            "            * *Default*: -\n",
            "            * *Example Values*: gene, transcript\n",
            "        \n",
            "        \n",
            "          + *Name*: species\n",
            "        \n",
            "            * *Type*: String\n",
            "            * *Description*: Species name/alias. Causes problems if the species doesn't match the identifiers in the POST body.\n",
            "            * *Default*: -\n",
            "            * *Example Values*: homo_sapiens, human\n",
            "        \n",
            "        \n",
            "        **Resource info**\n",
            "        \n",
            "        - **Methods**:  POST\n",
            "        - **Response formats**: json, jsonp\n",
            "        - **Maximum POST size**:        1000\n",
            "        \n",
            "        \n",
            "        **More info**\n",
            "        \n",
            "        https://rest.ensembl.org/documentation/info/lookup_post\n",
            "    \n",
            "    ontology_ancestors(*args, **kwargs)\n",
            "            Ontologies and Taxonomy ``GET ontology/ancestors/:id``\n",
            "        \n",
            "        Reconstruct the entire ancestry of a term from is_a and part_of relationships\n",
            "        \n",
            "        \n",
            "        **Parameters**\n",
            "        \n",
            "        - Required:\n",
            "        \n",
            "          + *Name*: id\n",
            "        \n",
            "            * *Type*: String\n",
            "            * *Description*: An ontology term identifier\n",
            "            * *Default*: -\n",
            "            * *Example Values*: GO:0005667\n",
            "        \n",
            "        \n",
            "        - Optional:\n",
            "        \n",
            "          + *Name*: callback\n",
            "        \n",
            "            * *Type*: String\n",
            "            * *Description*: Name of the callback subroutine to be returned by the requested JSONP response. Required ONLY when using JSONP as the serialisation method. Please see , .\n",
            "            * *Default*: -\n",
            "            * *Example Values*: randomlygeneratedname\n",
            "        \n",
            "        \n",
            "          + *Name*: ontology\n",
            "        \n",
            "            * *Type*: String\n",
            "            * *Description*: Filter by ontology. Used to disambiguate terms which are shared between ontologies such as GO and EFO\n",
            "            * *Default*: -\n",
            "            * *Example Values*: GO\n",
            "        \n",
            "        \n",
            "        **Resource info**\n",
            "        \n",
            "        - **Methods**:  GET\n",
            "        - **Response formats**: json, xml, yaml, jsonp\n",
            "        \n",
            "        \n",
            "        **More info**\n",
            "        \n",
            "        https://rest.ensembl.org/documentation/info/ontology_ancestors\n",
            "    \n",
            "    ontology_ancestors_chart(*args, **kwargs)\n",
            "            Ontologies and Taxonomy ``GET ontology/ancestors/chart/:id``\n",
            "        \n",
            "        Reconstruct the entire ancestry of a term from is_a and part_of relationships.\n",
            "        \n",
            "        \n",
            "        **Parameters**\n",
            "        \n",
            "        - Required:\n",
            "        \n",
            "          + *Name*: id\n",
            "        \n",
            "            * *Type*: String\n",
            "            * *Description*: An ontology term identifier\n",
            "            * *Default*: -\n",
            "            * *Example Values*: GO:0005667\n",
            "        \n",
            "        \n",
            "        - Optional:\n",
            "        \n",
            "          + *Name*: callback\n",
            "        \n",
            "            * *Type*: String\n",
            "            * *Description*: Name of the callback subroutine to be returned by the requested JSONP response. Required ONLY when using JSONP as the serialisation method. Please see , .\n",
            "            * *Default*: -\n",
            "            * *Example Values*: randomlygeneratedname\n",
            "        \n",
            "        \n",
            "          + *Name*: ontology\n",
            "        \n",
            "            * *Type*: String\n",
            "            * *Description*: Filter by ontology. Used to disambiguate terms which are shared between ontologies such as GO and EFO\n",
            "            * *Default*: -\n",
            "            * *Example Values*: GO\n",
            "        \n",
            "        \n",
            "        **Resource info**\n",
            "        \n",
            "        - **Methods**:  GET\n",
            "        - **Response formats**: json, xml, jsonp\n",
            "        \n",
            "        \n",
            "        **More info**\n",
            "        \n",
            "        https://rest.ensembl.org/documentation/info/ontology_ancestors_chart\n",
            "    \n",
            "    ontology_descendants(*args, **kwargs)\n",
            "            Ontologies and Taxonomy ``GET ontology/descendants/:id``\n",
            "        \n",
            "        Find all the terms descended from a given term. By default searches are conducted within the namespace of the given identifier\n",
            "        \n",
            "        \n",
            "        **Parameters**\n",
            "        \n",
            "        - Required:\n",
            "        \n",
            "          + *Name*: id\n",
            "        \n",
            "            * *Type*: String\n",
            "            * *Description*: An ontology term identifier\n",
            "            * *Default*: -\n",
            "            * *Example Values*: GO:0005667\n",
            "        \n",
            "        \n",
            "        - Optional:\n",
            "        \n",
            "          + *Name*: callback\n",
            "        \n",
            "            * *Type*: String\n",
            "            * *Description*: Name of the callback subroutine to be returned by the requested JSONP response. Required ONLY when using JSONP as the serialisation method. Please see , .\n",
            "            * *Default*: -\n",
            "            * *Example Values*: randomlygeneratedname\n",
            "        \n",
            "        \n",
            "          + *Name*: closest_term\n",
            "        \n",
            "            * *Type*: Boolean\n",
            "            * *Description*: If true return only the closest terms to the specified term\n",
            "            * *Default*: -\n",
            "            * *Example Values*: -\n",
            "        \n",
            "        \n",
            "          + *Name*: ontology\n",
            "        \n",
            "            * *Type*: String\n",
            "            * *Description*: Filter by ontology. Used to disambiguate terms which are shared between ontologies such as GO and EFO\n",
            "            * *Default*: -\n",
            "            * *Example Values*: GO\n",
            "        \n",
            "        \n",
            "          + *Name*: subset\n",
            "        \n",
            "            * *Type*: String\n",
            "            * *Description*: Filter terms by the specified subset\n",
            "            * *Default*: -\n",
            "            * *Example Values*: goslim_generic, goslim_metagenomics\n",
            "        \n",
            "        \n",
            "          + *Name*: zero_distance\n",
            "        \n",
            "            * *Type*: Boolean\n",
            "            * *Description*: Return terms with a distance of 0\n",
            "            * *Default*: -\n",
            "            * *Example Values*: -\n",
            "        \n",
            "        \n",
            "        **Resource info**\n",
            "        \n",
            "        - **Methods**:  GET\n",
            "        - **Response formats**: json, xml, jsonp\n",
            "        \n",
            "        \n",
            "        **More info**\n",
            "        \n",
            "        https://rest.ensembl.org/documentation/info/ontology_descendants\n",
            "    \n",
            "    ontology_id(*args, **kwargs)\n",
            "            Ontologies and Taxonomy ``GET ontology/id/:id``\n",
            "        \n",
            "        Search for an ontological term by its namespaced identifier\n",
            "        \n",
            "        \n",
            "        **Parameters**\n",
            "        \n",
            "        - Required:\n",
            "        \n",
            "          + *Name*: id\n",
            "        \n",
            "            * *Type*: String\n",
            "            * *Description*: An ontology term identifier\n",
            "            * *Default*: -\n",
            "            * *Example Values*: GO:0005667\n",
            "        \n",
            "        \n",
            "        - Optional:\n",
            "        \n",
            "          + *Name*: callback\n",
            "        \n",
            "            * *Type*: String\n",
            "            * *Description*: Name of the callback subroutine to be returned by the requested JSONP response. Required ONLY when using JSONP as the serialisation method. Please see , .\n",
            "            * *Default*: -\n",
            "            * *Example Values*: randomlygeneratedname\n",
            "        \n",
            "        \n",
            "          + *Name*: relation\n",
            "        \n",
            "            * *Type*: String\n",
            "            * *Description*: The types of relationships to include in the output. Fetches all relations by default\n",
            "            * *Default*: -\n",
            "            * *Example Values*: is_a, part_of\n",
            "        \n",
            "        \n",
            "          + *Name*: simple\n",
            "        \n",
            "            * *Type*: Boolean\n",
            "            * *Description*: If set the API will avoid the fetching of parent and child terms\n",
            "            * *Default*: 0\n",
            "            * *Example Values*: -\n",
            "        \n",
            "        \n",
            "        **Resource info**\n",
            "        \n",
            "        - **Methods**:  GET\n",
            "        - **Response formats**: json, xml, yaml, jsonp\n",
            "        \n",
            "        \n",
            "        **More info**\n",
            "        \n",
            "        https://rest.ensembl.org/documentation/info/ontology_id\n",
            "    \n",
            "    ontology_name(*args, **kwargs)\n",
            "            Ontologies and Taxonomy ``GET ontology/name/:name``\n",
            "        \n",
            "        Search for a list of ontological terms by their name\n",
            "        \n",
            "        \n",
            "        **Parameters**\n",
            "        \n",
            "        - Required:\n",
            "        \n",
            "          + *Name*: name\n",
            "        \n",
            "            * *Type*: String\n",
            "            * *Description*: An ontology name. SQL wildcards are supported\n",
            "            * *Default*: -\n",
            "            * *Example Values*: transcription factor complex\n",
            "        \n",
            "        \n",
            "        - Optional:\n",
            "        \n",
            "          + *Name*: callback\n",
            "        \n",
            "            * *Type*: String\n",
            "            * *Description*: Name of the callback subroutine to be returned by the requested JSONP response. Required ONLY when using JSONP as the serialisation method. Please see , .\n",
            "            * *Default*: -\n",
            "            * *Example Values*: randomlygeneratedname\n",
            "        \n",
            "        \n",
            "          + *Name*: ontology\n",
            "        \n",
            "            * *Type*: String\n",
            "            * *Description*: Filter by ontology. Used to disambiguate terms which are shared between ontologies such as GO and EFO\n",
            "            * *Default*: -\n",
            "            * *Example Values*: GO\n",
            "        \n",
            "        \n",
            "          + *Name*: relation\n",
            "        \n",
            "            * *Type*: String\n",
            "            * *Description*: The types of relationships to include in the output. Fetches all relations by default\n",
            "            * *Default*: -\n",
            "            * *Example Values*: is_a, part_of\n",
            "        \n",
            "        \n",
            "          + *Name*: simple\n",
            "        \n",
            "            * *Type*: Boolean\n",
            "            * *Description*: If set the API will avoid the fetching of parent and child terms\n",
            "            * *Default*: 0\n",
            "            * *Example Values*: -\n",
            "        \n",
            "        \n",
            "        **Resource info**\n",
            "        \n",
            "        - **Methods**:  GET\n",
            "        - **Response formats**: json, xml, yaml, jsonp\n",
            "        \n",
            "        \n",
            "        **More info**\n",
            "        \n",
            "        https://rest.ensembl.org/documentation/info/ontology_name\n",
            "    \n",
            "    overlap_id(*args, **kwargs)\n",
            "            Overlap ``GET overlap/id/:id``\n",
            "        \n",
            "        Retrieves features (e.g. genes, transcripts, variants and more) that overlap a region defined by the given identifier.\n",
            "        \n",
            "        \n",
            "        **Parameters**\n",
            "        \n",
            "        - Required:\n",
            "        \n",
            "          + *Name*: feature\n",
            "        \n",
            "            * *Type*: Enum(band, gene, transcript, cds, exon, repeat, simple, misc, variation, somatic_variation, structural_variation, somatic_structural_variation, constrained, regulatory,  motif, chipseq, array_probe)\n",
            "            * *Description*: The type of feature to retrieve. Multiple values are accepted.\n",
            "            * *Default*: none\n",
            "            * *Example Values*: -\n",
            "        \n",
            "        \n",
            "          + *Name*: id\n",
            "        \n",
            "            * *Type*: String\n",
            "            * *Description*: An Ensembl stable ID\n",
            "            * *Default*: -\n",
            "            * *Example Values*: ENSG00000157764\n",
            "        \n",
            "        \n",
            "        - Optional:\n",
            "        \n",
            "          + *Name*: biotype\n",
            "        \n",
            "            * *Type*: String\n",
            "            * *Description*: The functional classification of the gene or transcript to fetch. Cannot be used in conjunction with logic_name when querying transcripts.\n",
            "            * *Default*: -\n",
            "            * *Example Values*: protein_coding\n",
            "        \n",
            "        \n",
            "          + *Name*: callback\n",
            "        \n",
            "            * *Type*: String\n",
            "            * *Description*: Name of the callback subroutine to be returned by the requested JSONP response. Required ONLY when using JSONP as the serialisation method. Please see , .\n",
            "            * *Default*: -\n",
            "            * *Example Values*: randomlygeneratedname\n",
            "        \n",
            "        \n",
            "          + *Name*: db_type\n",
            "        \n",
            "            * *Type*: String\n",
            "            * *Description*: Restrict the search to a database other than the default. Useful if you need to use a DB other than core\n",
            "            * *Default*: -\n",
            "            * *Example Values*: core\n",
            "        \n",
            "        \n",
            "          + *Name*: logic_name\n",
            "        \n",
            "            * *Type*: String\n",
            "            * *Description*: Limit retrieval of genes, transcripts and exons by a given name of an analysis.\n",
            "            * *Default*: -\n",
            "            * *Example Values*: -\n",
            "        \n",
            "        \n",
            "          + *Name*: misc_set\n",
            "        \n",
            "            * *Type*: String\n",
            "            * *Description*: Miscellaneous set which groups together feature entries. Consult the DB or returned data sets to discover what is available.\n",
            "            * *Default*: -\n",
            "            * *Example Values*: cloneset_30k\n",
            "        \n",
            "        \n",
            "          + *Name*: object_type\n",
            "        \n",
            "            * *Type*: String\n",
            "            * *Description*: Filter by feature type\n",
            "            * *Default*: -\n",
            "            * *Example Values*: gene\n",
            "        \n",
            "        \n",
            "          + *Name*: so_term\n",
            "        \n",
            "            * *Type*: String\n",
            "            * *Description*: \n",
            "            * *Default*: -\n",
            "            * *Example Values*: SO:0001650\n",
            "        \n",
            "        \n",
            "          + *Name*: species\n",
            "        \n",
            "            * *Type*: String\n",
            "            * *Description*: Species name/alias.\n",
            "            * *Default*: -\n",
            "            * *Example Values*: homo_sapiens\n",
            "        \n",
            "        \n",
            "          + *Name*: species_set\n",
            "        \n",
            "            * *Type*: String\n",
            "            * *Description*: Filter by species set for retrieving constrained elements.\n",
            "            * *Default*: mammals\n",
            "            * *Example Values*: -\n",
            "        \n",
            "        \n",
            "          + *Name*: \n",
            "        \n",
            "            * *Type*: String\n",
            "            * *Description*: \n",
            "            * *Default*: -\n",
            "            * *Example Values*: ClinVar\n",
            "        \n",
            "        \n",
            "        **Resource info**\n",
            "        \n",
            "        - **Methods**:  GET\n",
            "        - **Response formats**: json, xml, gff3, bed, jsonp\n",
            "        - **Slice length**:     5e6\n",
            "        \n",
            "        \n",
            "        **More info**\n",
            "        \n",
            "        https://rest.ensembl.org/documentation/info/overlap_id\n",
            "    \n",
            "    overlap_region(*args, **kwargs)\n",
            "            Overlap ``GET overlap/region/:species/:region``\n",
            "        \n",
            "        Retrieves features (e.g. genes, transcripts, variants and more) that overlap a given region.\n",
            "        \n",
            "        \n",
            "        **Parameters**\n",
            "        \n",
            "        - Required:\n",
            "        \n",
            "          + *Name*: feature\n",
            "        \n",
            "            * *Type*: Enum(band, gene, transcript, cds, exon, repeat, simple, misc, variation, somatic_variation, structural_variation, somatic_structural_variation, constrained, regulatory, motif, peak, other_regulatory, array_probe)\n",
            "            * *Description*: The type of feature to retrieve. Multiple values are accepted.\n",
            "            * *Default*: none\n",
            "            * *Example Values*: -\n",
            "        \n",
            "        \n",
            "          + *Name*: region\n",
            "        \n",
            "            * *Type*: String\n",
            "            * *Description*: Query region. A maximum of 5Mb is allowed to be requested at any one time\n",
            "            * *Default*: -\n",
            "            * *Example Values*: X:1..1000:1, X:1..1000:-1, X:1..1000\n",
            "        \n",
            "        \n",
            "          + *Name*: species\n",
            "        \n",
            "            * *Type*: String\n",
            "            * *Description*: Species name/alias.\n",
            "            * *Default*: -\n",
            "            * *Example Values*: homo_sapiens\n",
            "        \n",
            "        \n",
            "        - Optional:\n",
            "        \n",
            "          + *Name*: biotype\n",
            "        \n",
            "            * *Type*: String\n",
            "            * *Description*: Functional classification of the gene or transcript to fetch. Cannot be used in conjunction with logic_name when querying transcripts.\n",
            "            * *Default*: -\n",
            "            * *Example Values*: protein_coding\n",
            "        \n",
            "        \n",
            "          + *Name*: callback\n",
            "        \n",
            "            * *Type*: String\n",
            "            * *Description*: Name of the callback subroutine to be returned by the requested JSONP response. Required ONLY when using JSONP as the serialisation method. Please see , .\n",
            "            * *Default*: -\n",
            "            * *Example Values*: randomlygeneratedname\n",
            "        \n",
            "        \n",
            "          + *Name*: db_type\n",
            "        \n",
            "            * *Type*: String\n",
            "            * *Description*: \n",
            "            * *Default*: core\n",
            "            * *Example Values*: core, otherfeatures\n",
            "        \n",
            "        \n",
            "          + *Name*: logic_name\n",
            "        \n",
            "            * *Type*: String\n",
            "            * *Description*: Limit retrieval of genes, transcripts and exons by the name of analysis.\n",
            "            * *Default*: -\n",
            "            * *Example Values*: -\n",
            "        \n",
            "        \n",
            "          + *Name*: misc_set\n",
            "        \n",
            "            * *Type*: String\n",
            "            * *Description*: Miscellaneous set which groups together feature entries. Consult the DB or returned data sets to discover what is available.\n",
            "            * *Default*: -\n",
            "            * *Example Values*: cloneset_30k\n",
            "        \n",
            "        \n",
            "          + *Name*: so_term\n",
            "        \n",
            "            * *Type*: String\n",
            "            * *Description*: \n",
            "            * *Default*: -\n",
            "            * *Example Values*: SO:0001650\n",
            "        \n",
            "        \n",
            "          + *Name*: species_set\n",
            "        \n",
            "            * *Type*: String\n",
            "            * *Description*: The species set name for retrieving constrained elements.\n",
            "            * *Default*: mammals\n",
            "            * *Example Values*: -\n",
            "        \n",
            "        \n",
            "          + *Name*: trim_downstream\n",
            "        \n",
            "            * *Type*: Boolean\n",
            "            * *Description*: Do not return features which overlap the downstream end of the region.\n",
            "            * *Default*: 0\n",
            "            * *Example Values*: -\n",
            "        \n",
            "        \n",
            "          + *Name*: trim_upstream\n",
            "        \n",
            "            * *Type*: Boolean\n",
            "            * *Description*: Do not return features which overlap upstream end of the region.\n",
            "            * *Default*: 0\n",
            "            * *Example Values*: -\n",
            "        \n",
            "        \n",
            "          + *Name*: \n",
            "        \n",
            "            * *Type*: String\n",
            "            * *Description*: \n",
            "            * *Default*: -\n",
            "            * *Example Values*: ClinVar\n",
            "        \n",
            "        \n",
            "        **Resource info**\n",
            "        \n",
            "        - **Methods**:  GET\n",
            "        - **Response formats**: json, xml, gff3, bed, jsonp\n",
            "        - **Slice length**:     5e6\n",
            "        \n",
            "        \n",
            "        **More info**\n",
            "        \n",
            "        https://rest.ensembl.org/documentation/info/overlap_region\n",
            "    \n",
            "    overlap_translation(*args, **kwargs)\n",
            "            Overlap ``GET overlap/translation/:id``\n",
            "        \n",
            "        Retrieve features related to a specific Translation as described by its stable ID (e.g. domains, variants).\n",
            "        \n",
            "        \n",
            "        **Parameters**\n",
            "        \n",
            "        - Required:\n",
            "        \n",
            "          + *Name*: id\n",
            "        \n",
            "            * *Type*: String\n",
            "            * *Description*: An Ensembl stable ID\n",
            "            * *Default*: -\n",
            "            * *Example Values*: ENSP00000288602\n",
            "        \n",
            "        \n",
            "        - Optional:\n",
            "        \n",
            "          + *Name*: callback\n",
            "        \n",
            "            * *Type*: String\n",
            "            * *Description*: Name of the callback subroutine to be returned by the requested JSONP response. Required ONLY when using JSONP as the serialisation method. Please see , .\n",
            "            * *Default*: -\n",
            "            * *Example Values*: randomlygeneratedname\n",
            "        \n",
            "        \n",
            "          + *Name*: db_type\n",
            "        \n",
            "            * *Type*: String\n",
            "            * *Description*: Restrict the search to a database other than the default. Useful if you need to use a DB other than core\n",
            "            * *Default*: -\n",
            "            * *Example Values*: core\n",
            "        \n",
            "        \n",
            "          + *Name*: feature\n",
            "        \n",
            "            * *Type*: Enum(transcript_variation, protein_feature, residue_overlap, translation_exon, somatic_transcript_variation)\n",
            "            * *Description*: Specify the type of features requested for the translation.\n",
            "            * *Default*: protein_feature\n",
            "            * *Example Values*: -\n",
            "        \n",
            "        \n",
            "          + *Name*: so_term\n",
            "        \n",
            "            * *Type*: String\n",
            "            * *Description*: \n",
            "            * *Default*: -\n",
            "            * *Example Values*: SO:0001650\n",
            "        \n",
            "        \n",
            "          + *Name*: species\n",
            "        \n",
            "            * *Type*: String\n",
            "            * *Description*: Species name/alias.\n",
            "            * *Default*: -\n",
            "            * *Example Values*: homo_sapiens\n",
            "        \n",
            "        \n",
            "          + *Name*: type\n",
            "        \n",
            "            * *Type*: String\n",
            "            * *Description*: Type of data to filter by. By default, all features are returned. Can specify a domain or consequence type.\n",
            "            * *Default*: none\n",
            "            * *Example Values*: low_complexity\n",
            "        \n",
            "        \n",
            "        **Resource info**\n",
            "        \n",
            "        - **Methods**:  GET\n",
            "        - **Response formats**: json, xml, jsonp\n",
            "        - **Slice length**:     5e6\n",
            "        \n",
            "        \n",
            "        **More info**\n",
            "        \n",
            "        https://rest.ensembl.org/documentation/info/overlap_translation\n",
            "    \n",
            "    phenotype_accession(*args, **kwargs)\n",
            "            Phenotype annotations ``GET /phenotype/accession/:species/:accession``\n",
            "        \n",
            "        Return phenotype annotations for genomic features given a phenotype ontology accession\n",
            "        \n",
            "        \n",
            "        **Parameters**\n",
            "        \n",
            "        - Required:\n",
            "        \n",
            "          + *Name*: accession\n",
            "        \n",
            "            * *Type*: String\n",
            "            * *Description*: phenotype ontology accession\n",
            "            * *Default*: -\n",
            "            * *Example Values*: EFO:0003900\n",
            "        \n",
            "        \n",
            "          + *Name*: species\n",
            "        \n",
            "            * *Type*: String\n",
            "            * *Description*: Species name/alias\n",
            "            * *Default*: -\n",
            "            * *Example Values*: homo_sapiens, human\n",
            "        \n",
            "        \n",
            "        - Optional:\n",
            "        \n",
            "          + *Name*: callback\n",
            "        \n",
            "            * *Type*: String\n",
            "            * *Description*: Name of the callback subroutine to be returned by the requested JSONP response. Required ONLY when using JSONP as the serialisation method. Please see , .\n",
            "            * *Default*: -\n",
            "            * *Example Values*: randomlygeneratedname\n",
            "        \n",
            "        \n",
            "          + *Name*: include_children\n",
            "        \n",
            "            * *Type*: Boolean(0,1)\n",
            "            * *Description*: Include annotations attached to child terms\n",
            "            * *Default*: 0\n",
            "            * *Example Values*: -\n",
            "        \n",
            "        \n",
            "          + *Name*: include_pubmed_id\n",
            "        \n",
            "            * *Type*: Boolean(0,1)\n",
            "            * *Description*: Include the pubmed_ids\n",
            "            * *Default*: 0\n",
            "            * *Example Values*: -\n",
            "        \n",
            "        \n",
            "          + *Name*: include_review_status\n",
            "        \n",
            "            * *Type*: Boolean(0,1)\n",
            "            * *Description*: Include the review_status information\n",
            "            * *Default*: 0\n",
            "            * *Example Values*: -\n",
            "        \n",
            "        \n",
            "          + *Name*: source\n",
            "        \n",
            "            * *Type*: String\n",
            "            * *Description*: Restrict to annotations from a specific source.\n",
            "            * *Default*: undef\n",
            "            * *Example Values*: -\n",
            "        \n",
            "        \n",
            "        **Resource info**\n",
            "        \n",
            "        - **Methods**:  GET\n",
            "        - **Response formats**: json, xml, jsonp\n",
            "        \n",
            "        \n",
            "        **More info**\n",
            "        \n",
            "        https://rest.ensembl.org/documentation/info/phenotype_accession\n",
            "    \n",
            "    phenotype_gene(*args, **kwargs)\n",
            "            Phenotype annotations ``GET /phenotype/gene/:species/:gene``\n",
            "        \n",
            "        Return phenotype annotations for a given gene.\n",
            "        \n",
            "        \n",
            "        **Parameters**\n",
            "        \n",
            "        - Required:\n",
            "        \n",
            "          + *Name*: gene\n",
            "        \n",
            "            * *Type*: String\n",
            "            * *Description*: Query gene name or Ensembl stable ID.\n",
            "            * *Default*: -\n",
            "            * *Example Values*: ENSG00000157764\n",
            "        \n",
            "        \n",
            "          + *Name*: species\n",
            "        \n",
            "            * *Type*: String\n",
            "            * *Description*: Species name/alias\n",
            "            * *Default*: -\n",
            "            * *Example Values*: homo_sapiens, human\n",
            "        \n",
            "        \n",
            "        - Optional:\n",
            "        \n",
            "          + *Name*: callback\n",
            "        \n",
            "            * *Type*: String\n",
            "            * *Description*: Name of the callback subroutine to be returned by the requested JSONP response. Required ONLY when using JSONP as the serialisation method. Please see , .\n",
            "            * *Default*: -\n",
            "            * *Example Values*: randomlygeneratedname\n",
            "        \n",
            "        \n",
            "          + *Name*: include_associated\n",
            "        \n",
            "            * *Type*: Boolean(0,1)\n",
            "            * *Description*: \n",
            "            * *Default*: 0\n",
            "            * *Example Values*: -\n",
            "        \n",
            "        \n",
            "          + *Name*: include_overlap\n",
            "        \n",
            "            * *Type*: Boolean(0,1)\n",
            "            * *Description*: Include phenotypes of features overlapping the gene.\n",
            "            * *Default*: 0\n",
            "            * *Example Values*: -\n",
            "        \n",
            "        \n",
            "          + *Name*: include_pubmed_id\n",
            "        \n",
            "            * *Type*: Boolean(0,1)\n",
            "            * *Description*: Include the pubmed_ids\n",
            "            * *Default*: 0\n",
            "            * *Example Values*: -\n",
            "        \n",
            "        \n",
            "          + *Name*: include_review_status\n",
            "        \n",
            "            * *Type*: Boolean(0,1)\n",
            "            * *Description*: Include the review_status information\n",
            "            * *Default*: 0\n",
            "            * *Example Values*: -\n",
            "        \n",
            "        \n",
            "          + *Name*: include_submitter\n",
            "        \n",
            "            * *Type*: Boolean(0,1)\n",
            "            * *Description*: Include the submitter names\n",
            "            * *Default*: 0\n",
            "            * *Example Values*: -\n",
            "        \n",
            "        \n",
            "        **Resource info**\n",
            "        \n",
            "        - **Methods**:  GET\n",
            "        - **Response formats**: json, xml, jsonp\n",
            "        \n",
            "        \n",
            "        **More info**\n",
            "        \n",
            "        https://rest.ensembl.org/documentation/info/phenotype_gene\n",
            "    \n",
            "    phenotype_region(*args, **kwargs)\n",
            "            Phenotype annotations ``GET /phenotype/region/:species/:region``\n",
            "        \n",
            "        Return phenotype annotations that overlap a given genomic region.\n",
            "        \n",
            "        \n",
            "        **Parameters**\n",
            "        \n",
            "        - Required:\n",
            "        \n",
            "          + *Name*: region\n",
            "        \n",
            "            * *Type*: String\n",
            "            * *Description*: Query region. A maximum of 5Mb is allowed to be requested at any one time\n",
            "            * *Default*: -\n",
            "            * *Example Values*: 9:22125500-22136000:1, 9:22125500-22136000:-1, 9:22125500-22136000\n",
            "        \n",
            "        \n",
            "          + *Name*: species\n",
            "        \n",
            "            * *Type*: String\n",
            "            * *Description*: Species name/alias\n",
            "            * *Default*: -\n",
            "            * *Example Values*: homo_sapiens, human\n",
            "        \n",
            "        \n",
            "        - Optional:\n",
            "        \n",
            "          + *Name*: callback\n",
            "        \n",
            "            * *Type*: String\n",
            "            * *Description*: Name of the callback subroutine to be returned by the requested JSONP response. Required ONLY when using JSONP as the serialisation method. Please see , .\n",
            "            * *Default*: -\n",
            "            * *Example Values*: randomlygeneratedname\n",
            "        \n",
            "        \n",
            "          + *Name*: feature_type\n",
            "        \n",
            "            * *Type*: String\n",
            "            * *Description*: Restrict to phenotype annotations from a specific feature type.\n",
            "            * *Default*: -\n",
            "            * *Example Values*: Variation, StructuralVariation, Gene, QTL\n",
            "        \n",
            "        \n",
            "          + *Name*: include_pubmed_id\n",
            "        \n",
            "            * *Type*: Boolean(0,1)\n",
            "            * *Description*: Include the pubmed_ids\n",
            "            * *Default*: 0\n",
            "            * *Example Values*: -\n",
            "        \n",
            "        \n",
            "          + *Name*: include_review_status\n",
            "        \n",
            "            * *Type*: Boolean(0,1)\n",
            "            * *Description*: Include the review_status information\n",
            "            * *Default*: 0\n",
            "            * *Example Values*: -\n",
            "        \n",
            "        \n",
            "          + *Name*: include_submitter\n",
            "        \n",
            "            * *Type*: Boolean(0,1)\n",
            "            * *Description*: Include the submitter names\n",
            "            * *Default*: 0\n",
            "            * *Example Values*: -\n",
            "        \n",
            "        \n",
            "          + *Name*: only_phenotypes\n",
            "        \n",
            "            * *Type*: Boolean(0,1)\n",
            "            * *Description*: Only returns associated phenotype description and mapped ontology accessions for a lighter output.\n",
            "            * *Default*: 0\n",
            "            * *Example Values*: -\n",
            "        \n",
            "        \n",
            "        **Resource info**\n",
            "        \n",
            "        - **Methods**:  GET\n",
            "        - **Response formats**: json, xml, jsonp\n",
            "        \n",
            "        \n",
            "        **More info**\n",
            "        \n",
            "        https://rest.ensembl.org/documentation/info/phenotype_region\n",
            "    \n",
            "    phenotype_term(*args, **kwargs)\n",
            "            Phenotype annotations ``GET /phenotype/term/:species/:term``\n",
            "        \n",
            "        Return phenotype annotations for genomic features given a phenotype ontology term\n",
            "        \n",
            "        \n",
            "        **Parameters**\n",
            "        \n",
            "        - Required:\n",
            "        \n",
            "          + *Name*: species\n",
            "        \n",
            "            * *Type*: String\n",
            "            * *Description*: Species name/alias\n",
            "            * *Default*: -\n",
            "            * *Example Values*: homo_sapiens, human\n",
            "        \n",
            "        \n",
            "          + *Name*: term\n",
            "        \n",
            "            * *Type*: String\n",
            "            * *Description*: phenotype ontology term\n",
            "            * *Default*: -\n",
            "            * *Example Values*: coffee consumption\n",
            "        \n",
            "        \n",
            "        - Optional:\n",
            "        \n",
            "          + *Name*: callback\n",
            "        \n",
            "            * *Type*: String\n",
            "            * *Description*: Name of the callback subroutine to be returned by the requested JSONP response. Required ONLY when using JSONP as the serialisation method. Please see , .\n",
            "            * *Default*: -\n",
            "            * *Example Values*: randomlygeneratedname\n",
            "        \n",
            "        \n",
            "          + *Name*: include_children\n",
            "        \n",
            "            * *Type*: Boolean(0,1)\n",
            "            * *Description*: Include annotations attached to child terms\n",
            "            * *Default*: 0\n",
            "            * *Example Values*: -\n",
            "        \n",
            "        \n",
            "          + *Name*: include_pubmed_id\n",
            "        \n",
            "            * *Type*: Boolean(0,1)\n",
            "            * *Description*: Include the pubmed_ids\n",
            "            * *Default*: 0\n",
            "            * *Example Values*: -\n",
            "        \n",
            "        \n",
            "          + *Name*: include_review_status\n",
            "        \n",
            "            * *Type*: Boolean(0,1)\n",
            "            * *Description*: Include the review_status information\n",
            "            * *Default*: 0\n",
            "            * *Example Values*: -\n",
            "        \n",
            "        \n",
            "          + *Name*: source\n",
            "        \n",
            "            * *Type*: String\n",
            "            * *Description*: Restrict to annotations from a specific source.\n",
            "            * *Default*: undef\n",
            "            * *Example Values*: -\n",
            "        \n",
            "        \n",
            "        **Resource info**\n",
            "        \n",
            "        - **Methods**:  GET\n",
            "        - **Response formats**: json, xml, jsonp\n",
            "        \n",
            "        \n",
            "        **More info**\n",
            "        \n",
            "        https://rest.ensembl.org/documentation/info/phenotype_term\n",
            "    \n",
            "    ping(*args, **kwargs)\n",
            "            Information ``GET info/ping``\n",
            "        \n",
            "        Checks if the service is alive.\n",
            "        \n",
            "        \n",
            "        **Parameters**\n",
            "        \n",
            "        - Required:\n",
            "        \n",
            "          + *Name*: callback\n",
            "        \n",
            "            * *Type*: String\n",
            "            * *Description*: Name of the callback subroutine to be returned by the requested JSONP response. Required ONLY when using JSONP as the serialisation method. Please see , .\n",
            "            * *Default*: -\n",
            "            * *Example Values*: randomlygeneratedname\n",
            "        \n",
            "        \n",
            "        - Optional:\n",
            "        \n",
            "          + *Name*: callback\n",
            "        \n",
            "            * *Type*: String\n",
            "            * *Description*: Name of the callback subroutine to be returned by the requested JSONP response. Required ONLY when using JSONP as the serialisation method. Please see , .\n",
            "            * *Default*: -\n",
            "            * *Example Values*: randomlygeneratedname\n",
            "        \n",
            "        \n",
            "        **Resource info**\n",
            "        \n",
            "        - **Methods**:  GET\n",
            "        - **Response formats**: json, xml, jsonp\n",
            "        \n",
            "        \n",
            "        **More info**\n",
            "        \n",
            "        https://rest.ensembl.org/documentation/info/ping\n",
            "    \n",
            "    probe(*args, **kwargs)\n",
            "            Regulation ``GET regulatory/species/:species/microarray/:microarray/probe/:probe``\n",
            "        \n",
            "        Returns information about a specific probe from a microarray\n",
            "        \n",
            "        \n",
            "        **Parameters**\n",
            "        \n",
            "        - Required:\n",
            "        \n",
            "          + *Name*: microarray\n",
            "        \n",
            "            * *Type*: String\n",
            "            * *Description*: Microarray name\n",
            "            * *Default*: -\n",
            "            * *Example Values*: HumanWG_6_V2\n",
            "        \n",
            "        \n",
            "          + *Name*: probe\n",
            "        \n",
            "            * *Type*: String\n",
            "            * *Description*: Probe name\n",
            "            * *Default*: -\n",
            "            * *Example Values*: ILMN_1763508\n",
            "        \n",
            "        \n",
            "          + *Name*: species\n",
            "        \n",
            "            * *Type*: String\n",
            "            * *Description*: Species name/alias\n",
            "            * *Default*: -\n",
            "            * *Example Values*: homo_sapiens\n",
            "        \n",
            "        \n",
            "        - Optional:\n",
            "        \n",
            "          + *Name*: callback\n",
            "        \n",
            "            * *Type*: String\n",
            "            * *Description*: Name of the callback subroutine to be returned by the requested JSONP response. Required ONLY when using JSONP as the serialisation method. Please see , .\n",
            "            * *Default*: -\n",
            "            * *Example Values*: randomlygeneratedname\n",
            "        \n",
            "        \n",
            "          + *Name*: gene\n",
            "        \n",
            "            * *Type*: Boolean(0,1)\n",
            "            * *Description*: Has to be used in conjunction with transcript. Displays the associated gene\n",
            "            * *Default*: 0\n",
            "            * *Example Values*: -\n",
            "        \n",
            "        \n",
            "          + *Name*: transcripts\n",
            "        \n",
            "            * *Type*: Boolean(0,1)\n",
            "            * *Description*: Displays the transcripts linked to this probe\n",
            "            * *Default*: 0\n",
            "            * *Example Values*: -\n",
            "        \n",
            "        \n",
            "        **Resource info**\n",
            "        \n",
            "        - **Methods**:  GET\n",
            "        - **Response formats**: json, xml, jsonp\n",
            "        \n",
            "        \n",
            "        **More info**\n",
            "        \n",
            "        https://rest.ensembl.org/documentation/info/probe\n",
            "    \n",
            "    probe_set(*args, **kwargs)\n",
            "            Regulation ``GET regulatory/species/:species/microarray/:microarray/probe_set/:probe_set``\n",
            "        \n",
            "        Returns information about a specific probe_set from a microarray\n",
            "        \n",
            "        \n",
            "        **Parameters**\n",
            "        \n",
            "        - Required:\n",
            "        \n",
            "          + *Name*: microarray\n",
            "        \n",
            "            * *Type*: String\n",
            "            * *Description*: Microarray name\n",
            "            * *Default*: -\n",
            "            * *Example Values*: HG-U133_Plus_2\n",
            "        \n",
            "        \n",
            "          + *Name*: probe_set\n",
            "        \n",
            "            * *Type*: String\n",
            "            * *Description*: ProbeSet name\n",
            "            * *Default*: -\n",
            "            * *Example Values*: 202820_at\n",
            "        \n",
            "        \n",
            "          + *Name*: species\n",
            "        \n",
            "            * *Type*: String\n",
            "            * *Description*: Species name/alias\n",
            "            * *Default*: -\n",
            "            * *Example Values*: homo_sapiens\n",
            "        \n",
            "        \n",
            "        - Optional:\n",
            "        \n",
            "          + *Name*: callback\n",
            "        \n",
            "            * *Type*: String\n",
            "            * *Description*: Name of the callback subroutine to be returned by the requested JSONP response. Required ONLY when using JSONP as the serialisation method. Please see , .\n",
            "            * *Default*: -\n",
            "            * *Example Values*: randomlygeneratedname\n",
            "        \n",
            "        \n",
            "          + *Name*: gene\n",
            "        \n",
            "            * *Type*: Boolean(0,1)\n",
            "            * *Description*: Has to be used in conjunction with transcript. Displays the associated gene\n",
            "            * *Default*: 0\n",
            "            * *Example Values*: -\n",
            "        \n",
            "        \n",
            "          + *Name*: transcripts\n",
            "        \n",
            "            * *Type*: Boolean(0,1)\n",
            "            * *Description*: Displays the transcripts linked to this probe\n",
            "            * *Default*: 0\n",
            "            * *Example Values*: -\n",
            "        \n",
            "        \n",
            "        **Resource info**\n",
            "        \n",
            "        - **Methods**:  GET\n",
            "        - **Response formats**: json, xml, jsonp\n",
            "        \n",
            "        \n",
            "        **More info**\n",
            "        \n",
            "        https://rest.ensembl.org/documentation/info/probe_set\n",
            "    \n",
            "    referenceSets(*args, **kwargs)\n",
            "            Variation GA4GH ``POST ga4gh/referencesets/search``\n",
            "        \n",
            "        Return a list of reference sets in GA4GH format\n",
            "        \n",
            "        \n",
            "        **Parameters**\n",
            "        \n",
            "        - Required:\n",
            "        \n",
            "          + *Name*: accession\n",
            "        \n",
            "            * *Type*: String\n",
            "            * *Description*: Return referenceSet information for a specific accession\n",
            "            * *Default*: -\n",
            "            * *Example Values*: 1\n",
            "        \n",
            "        \n",
            "          + *Name*: callback\n",
            "        \n",
            "            * *Type*: String\n",
            "            * *Description*: Name of the callback subroutine to be returned by the requested JSONP response. Required ONLY when using JSONP as the serialisation method. Please see , .\n",
            "            * *Default*: -\n",
            "            * *Example Values*: randomlygeneratedname\n",
            "        \n",
            "        \n",
            "          + *Name*: pageSize\n",
            "        \n",
            "            * *Type*: Int\n",
            "            * *Description*: Number of referenceSets to return per request\n",
            "            * *Default*: 10\n",
            "            * *Example Values*: -\n",
            "        \n",
            "        \n",
            "          + *Name*: pageToken\n",
            "        \n",
            "            * *Type*: Int\n",
            "            * *Description*: Identifier showing which page of data to retrieve next\n",
            "            * *Default*: null\n",
            "            * *Example Values*: -\n",
            "        \n",
            "        \n",
            "        - Optional:\n",
            "        \n",
            "          + *Name*: accession\n",
            "        \n",
            "            * *Type*: String\n",
            "            * *Description*: Return referenceSet information for a specific accession\n",
            "            * *Default*: -\n",
            "            * *Example Values*: 1\n",
            "        \n",
            "        \n",
            "          + *Name*: callback\n",
            "        \n",
            "            * *Type*: String\n",
            "            * *Description*: Name of the callback subroutine to be returned by the requested JSONP response. Required ONLY when using JSONP as the serialisation method. Please see , .\n",
            "            * *Default*: -\n",
            "            * *Example Values*: randomlygeneratedname\n",
            "        \n",
            "        \n",
            "          + *Name*: pageSize\n",
            "        \n",
            "            * *Type*: Int\n",
            "            * *Description*: Number of referenceSets to return per request\n",
            "            * *Default*: 10\n",
            "            * *Example Values*: -\n",
            "        \n",
            "        \n",
            "          + *Name*: pageToken\n",
            "        \n",
            "            * *Type*: Int\n",
            "            * *Description*: Identifier showing which page of data to retrieve next\n",
            "            * *Default*: null\n",
            "            * *Example Values*: -\n",
            "        \n",
            "        \n",
            "        **Resource info**\n",
            "        \n",
            "        - **Methods**:  POST\n",
            "        - **Response formats**: json, jsonp\n",
            "        \n",
            "        \n",
            "        **More info**\n",
            "        \n",
            "        https://rest.ensembl.org/documentation/info/referenceSets\n",
            "    \n",
            "    referenceSets_id(*args, **kwargs)\n",
            "            Variation GA4GH ``GET ga4gh/referencesets/:id``\n",
            "        \n",
            "        Return data for a specific reference set in GA4GH format\n",
            "        \n",
            "        \n",
            "        **Parameters**\n",
            "        \n",
            "        - Required:\n",
            "        \n",
            "          + *Name*: id\n",
            "        \n",
            "            * *Type*: String\n",
            "            * *Description*: Reference set id\n",
            "            * *Default*: -\n",
            "            * *Example Values*: -\n",
            "        \n",
            "        \n",
            "        - Optional:\n",
            "        \n",
            "          + *Name*: callback\n",
            "        \n",
            "            * *Type*: String\n",
            "            * *Description*: Name of the callback subroutine to be returned by the requested JSONP response. Required ONLY when using JSONP as the serialisation method. Please see , .\n",
            "            * *Default*: -\n",
            "            * *Example Values*: randomlygeneratedname\n",
            "        \n",
            "        \n",
            "        **Resource info**\n",
            "        \n",
            "        - **Methods**:  GET\n",
            "        - **Response formats**: json, jsonp\n",
            "        \n",
            "        \n",
            "        **More info**\n",
            "        \n",
            "        https://rest.ensembl.org/documentation/info/referenceSets_id\n",
            "    \n",
            "    references(*args, **kwargs)\n",
            "            Variation GA4GH ``POST ga4gh/references/search``\n",
            "        \n",
            "        Return a list of reference sequences in GA4GH format\n",
            "        \n",
            "        \n",
            "        **Parameters**\n",
            "        \n",
            "        - Required:\n",
            "        \n",
            "          + *Name*: referenceSetId\n",
            "        \n",
            "            * *Type*: string\n",
            "            * *Description*: Return references for a referenceSet\n",
            "            * *Default*: -\n",
            "            * *Example Values*: GRCh38\n",
            "        \n",
            "        \n",
            "        - Optional:\n",
            "        \n",
            "          + *Name*: accession\n",
            "        \n",
            "            * *Type*: string\n",
            "            * *Description*: Return reference information for a specific accession\n",
            "            * *Default*: -\n",
            "            * *Example Values*: NC_000021.9\n",
            "        \n",
            "        \n",
            "          + *Name*: callback\n",
            "        \n",
            "            * *Type*: String\n",
            "            * *Description*: Name of the callback subroutine to be returned by the requested JSONP response. Required ONLY when using JSONP as the serialisation method. Please see , .\n",
            "            * *Default*: -\n",
            "            * *Example Values*: randomlygeneratedname\n",
            "        \n",
            "        \n",
            "          + *Name*: md5checksum\n",
            "        \n",
            "            * *Type*: string\n",
            "            * *Description*: Return reference information for the md5checksum of the sequence\n",
            "            * *Default*: -\n",
            "            * *Example Values*: 9489ae7581e14efcad134f02afafe26c\n",
            "        \n",
            "        \n",
            "          + *Name*: pageSize\n",
            "        \n",
            "            * *Type*: Int\n",
            "            * *Description*: Number of references to return per request\n",
            "            * *Default*: 10\n",
            "            * *Example Values*: -\n",
            "        \n",
            "        \n",
            "          + *Name*: pageToken\n",
            "        \n",
            "            * *Type*: Int\n",
            "            * *Description*: Identifier showing which page of data to retrieve next\n",
            "            * *Default*: null\n",
            "            * *Example Values*: -\n",
            "        \n",
            "        \n",
            "        **Resource info**\n",
            "        \n",
            "        - **Methods**:  POST\n",
            "        - **Response formats**: json, jsonp\n",
            "        \n",
            "        \n",
            "        **More info**\n",
            "        \n",
            "        https://rest.ensembl.org/documentation/info/references\n",
            "    \n",
            "    references_id(*args, **kwargs)\n",
            "            Variation GA4GH ``GET ga4gh/references/:id``\n",
            "        \n",
            "        Return data for a specific reference in GA4GH format by id\n",
            "        \n",
            "        \n",
            "        **Parameters**\n",
            "        \n",
            "        - Required:\n",
            "        \n",
            "          + *Name*: id\n",
            "        \n",
            "            * *Type*: String\n",
            "            * *Description*: Reference id\n",
            "            * *Default*: -\n",
            "            * *Example Values*: 9489ae7581e14efcad134f02afafe26c\n",
            "        \n",
            "        \n",
            "        - Optional:\n",
            "        \n",
            "          + *Name*: callback\n",
            "        \n",
            "            * *Type*: String\n",
            "            * *Description*: Name of the callback subroutine to be returned by the requested JSONP response. Required ONLY when using JSONP as the serialisation method. Please see , .\n",
            "            * *Default*: -\n",
            "            * *Example Values*: randomlygeneratedname\n",
            "        \n",
            "        \n",
            "        **Resource info**\n",
            "        \n",
            "        - **Methods**:  GET\n",
            "        - **Response formats**: json, jsonp\n",
            "        \n",
            "        \n",
            "        **More info**\n",
            "        \n",
            "        https://rest.ensembl.org/documentation/info/references_id\n",
            "    \n",
            "    regulatory_id(*args, **kwargs)\n",
            "            Regulation ``GET regulatory/species/:species/id/:id``\n",
            "        \n",
            "        Returns a RegulatoryFeature given its stable ID (e.g. ENSR00000006733)\n",
            "        \n",
            "        \n",
            "        **Parameters**\n",
            "        \n",
            "        - Required:\n",
            "        \n",
            "          + *Name*: id\n",
            "        \n",
            "            * *Type*: String\n",
            "            * *Description*: RegulatoryFeature stable ID\n",
            "            * *Default*: -\n",
            "            * *Example Values*: ENSR00000006733\n",
            "        \n",
            "        \n",
            "          + *Name*: species\n",
            "        \n",
            "            * *Type*: String\n",
            "            * *Description*: Species name/alias\n",
            "            * *Default*: -\n",
            "            * *Example Values*: homo_sapiens\n",
            "        \n",
            "        \n",
            "        - Optional:\n",
            "        \n",
            "          + *Name*: activity\n",
            "        \n",
            "            * *Type*: Boolean(0,1)\n",
            "            * *Description*: Returns the activity of the Regulatory Feature in each Epigenome for the given species\n",
            "            * *Default*: 0\n",
            "            * *Example Values*: -\n",
            "        \n",
            "        \n",
            "          + *Name*: callback\n",
            "        \n",
            "            * *Type*: String\n",
            "            * *Description*: Name of the callback subroutine to be returned by the requested JSONP response. Required ONLY when using JSONP as the serialisation method. Please see , .\n",
            "            * *Default*: -\n",
            "            * *Example Values*: randomlygeneratedname\n",
            "        \n",
            "        \n",
            "        **Resource info**\n",
            "        \n",
            "        - **Methods**:  GET\n",
            "        - **Response formats**: json, xml, jsonp\n",
            "        \n",
            "        \n",
            "        **More info**\n",
            "        \n",
            "        https://rest.ensembl.org/documentation/info/regulatory_id\n",
            "    \n",
            "    rest(*args, **kwargs)\n",
            "            Information ``GET info/rest``\n",
            "        \n",
            "        Shows the current version of the Ensembl REST API.\n",
            "        \n",
            "        \n",
            "        **Parameters**\n",
            "        \n",
            "        - Required:\n",
            "        \n",
            "          + *Name*: callback\n",
            "        \n",
            "            * *Type*: String\n",
            "            * *Description*: Name of the callback subroutine to be returned by the requested JSONP response. Required ONLY when using JSONP as the serialisation method. Please see , .\n",
            "            * *Default*: -\n",
            "            * *Example Values*: randomlygeneratedname\n",
            "        \n",
            "        \n",
            "        - Optional:\n",
            "        \n",
            "          + *Name*: callback\n",
            "        \n",
            "            * *Type*: String\n",
            "            * *Description*: Name of the callback subroutine to be returned by the requested JSONP response. Required ONLY when using JSONP as the serialisation method. Please see , .\n",
            "            * *Default*: -\n",
            "            * *Example Values*: randomlygeneratedname\n",
            "        \n",
            "        \n",
            "        **Resource info**\n",
            "        \n",
            "        - **Methods**:  GET\n",
            "        - **Response formats**: json, xml, jsonp\n",
            "        \n",
            "        \n",
            "        **More info**\n",
            "        \n",
            "        https://rest.ensembl.org/documentation/info/rest\n",
            "    \n",
            "    sequence_id(*args, **kwargs)\n",
            "            Sequence ``GET sequence/id/:id``\n",
            "        \n",
            "        Request multiple types of sequence by stable identifier. Supports feature masking and expand options.\n",
            "        \n",
            "        \n",
            "        **Parameters**\n",
            "        \n",
            "        - Required:\n",
            "        \n",
            "          + *Name*: id\n",
            "        \n",
            "            * *Type*: String\n",
            "            * *Description*: An Ensembl stable ID\n",
            "            * *Default*: -\n",
            "            * *Example Values*: ENSG00000157764, ENSG00000157764.fasta (supported on some deployments)\n",
            "        \n",
            "        \n",
            "        - Optional:\n",
            "        \n",
            "          + *Name*: callback\n",
            "        \n",
            "            * *Type*: String\n",
            "            * *Description*: Name of the callback subroutine to be returned by the requested JSONP response. Required ONLY when using JSONP as the serialisation method. Please see , .\n",
            "            * *Default*: -\n",
            "            * *Example Values*: randomlygeneratedname\n",
            "        \n",
            "        \n",
            "          + *Name*: db_type\n",
            "        \n",
            "            * *Type*: String\n",
            "            * *Description*: Restrict the search to a database other than the default. Useful if you need to use a DB other than core\n",
            "            * *Default*: -\n",
            "            * *Example Values*: core\n",
            "        \n",
            "        \n",
            "          + *Name*: end\n",
            "        \n",
            "            * *Type*: Int\n",
            "            * *Description*: Trim the end of the sequence by this many basepairs. Trimming is relative to reading direction and in the coordinate system of the stable identifier. Parameter can not be used in conjunction with expand_5prime or expand_3prime.\n",
            "            * *Default*: -\n",
            "            * *Example Values*: 1000\n",
            "        \n",
            "        \n",
            "          + *Name*: expand_3prime\n",
            "        \n",
            "            * *Type*: Int\n",
            "            * *Description*: Expand the sequence downstream of the sequence by this many basepairs. Only available when using genomic sequence type.\n",
            "            * *Default*: -\n",
            "            * *Example Values*: 1000\n",
            "        \n",
            "        \n",
            "          + *Name*: expand_5prime\n",
            "        \n",
            "            * *Type*: Int\n",
            "            * *Description*: Expand the sequence upstream of the sequence by this many basepairs. Only available when using genomic sequence type.\n",
            "            * *Default*: -\n",
            "            * *Example Values*: 1000\n",
            "        \n",
            "        \n",
            "          + *Name*: format\n",
            "        \n",
            "            * *Type*: Enum(fasta)\n",
            "            * *Description*: Format of the data\n",
            "            * *Default*: -\n",
            "            * *Example Values*: fasta\n",
            "        \n",
            "        \n",
            "          + *Name*: mask\n",
            "        \n",
            "            * *Type*: Enum(hard,soft)\n",
            "            * *Description*: Request the sequence masked for repeat sequences. Hard will mask all repeats as N's and soft will mask repeats as lowercased characters. Only available when using genomic sequence type.\n",
            "            * *Default*: -\n",
            "            * *Example Values*: hard\n",
            "        \n",
            "        \n",
            "          + *Name*: mask_feature\n",
            "        \n",
            "            * *Type*: Boolean\n",
            "            * *Description*: Mask features on the sequence. If sequence is genomic, mask introns. If sequence is cDNA, mask UTRs. Incompatible with the 'mask' option\n",
            "            * *Default*: 0\n",
            "            * *Example Values*: -\n",
            "        \n",
            "        \n",
            "          + *Name*: multiple_sequences\n",
            "        \n",
            "            * *Type*: Boolean\n",
            "            * *Description*: Allow the service to return more than 1 sequence per identifier. This is useful when querying for a gene but using a type such as protein.\n",
            "            * *Default*: 0\n",
            "            * *Example Values*: -\n",
            "        \n",
            "        \n",
            "          + *Name*: object_type\n",
            "        \n",
            "            * *Type*: String\n",
            "            * *Description*: Filter by feature type\n",
            "            * *Default*: -\n",
            "            * *Example Values*: gene\n",
            "        \n",
            "        \n",
            "          + *Name*: species\n",
            "        \n",
            "            * *Type*: String\n",
            "            * *Description*: Species name/alias\n",
            "            * *Default*: -\n",
            "            * *Example Values*: homo_sapiens\n",
            "        \n",
            "        \n",
            "          + *Name*: start\n",
            "        \n",
            "            * *Type*: Int\n",
            "            * *Description*: Trim the start of the sequence by this many basepairs. Trimming is relative to reading direction and in the coordinate system of the stable identifier. Parameter can not be used in conjunction with expand_5prime or expand_3prime.\n",
            "            * *Default*: -\n",
            "            * *Example Values*: 1000\n",
            "        \n",
            "        \n",
            "          + *Name*: type\n",
            "        \n",
            "            * *Type*: Enum(genomic,cds,cdna,protein)\n",
            "            * *Description*: Type of sequence. Defaults to genomic where applicable, i.e. not translations. cdna refers to the spliced transcript sequence with UTR; cds refers to the spliced transcript sequence without UTR.\n",
            "            * *Default*: genomic\n",
            "            * *Example Values*: cds\n",
            "        \n",
            "        \n",
            "        **Resource info**\n",
            "        \n",
            "        - **Methods**:  GET\n",
            "        - **Response formats**: fasta, json, seqxml, text, yaml, jsonp\n",
            "        - **Slice length**:     1e7\n",
            "        \n",
            "        \n",
            "        **More info**\n",
            "        \n",
            "        https://rest.ensembl.org/documentation/info/sequence_id\n",
            "    \n",
            "    sequence_id_post(*args, **kwargs)\n",
            "            Sequence ``POST sequence/id``\n",
            "        \n",
            "        Request multiple types of sequence by a stable identifier list.\n",
            "        \n",
            "        \n",
            "        **Parameters**\n",
            "        \n",
            "        - Required:\n",
            "        \n",
            "          + *Name*: callback\n",
            "        \n",
            "            * *Type*: String\n",
            "            * *Description*: Name of the callback subroutine to be returned by the requested JSONP response. Required ONLY when using JSONP as the serialisation method. Please see , .\n",
            "            * *Default*: -\n",
            "            * *Example Values*: randomlygeneratedname\n",
            "        \n",
            "        \n",
            "          + *Name*: db_type\n",
            "        \n",
            "            * *Type*: String\n",
            "            * *Description*: Restrict the search to a database other than the default. Useful if you need to use a DB other than core\n",
            "            * *Default*: -\n",
            "            * *Example Values*: core\n",
            "        \n",
            "        \n",
            "          + *Name*: end\n",
            "        \n",
            "            * *Type*: Int\n",
            "            * *Description*: Trim the end of the sequence by this many basepairs. Trimming is relative to reading direction and in the coordinate system of the stable identifier. Parameter can not be used in conjunction with expand_5prime or expand_3prime.\n",
            "            * *Default*: -\n",
            "            * *Example Values*: 1000\n",
            "        \n",
            "        \n",
            "          + *Name*: expand_3prime\n",
            "        \n",
            "            * *Type*: Int\n",
            "            * *Description*: Expand the sequence downstream of the sequence by this many basepairs. Only available when using genomic sequence type.\n",
            "            * *Default*: -\n",
            "            * *Example Values*: 1000\n",
            "        \n",
            "        \n",
            "          + *Name*: expand_5prime\n",
            "        \n",
            "            * *Type*: Int\n",
            "            * *Description*: Expand the sequence upstream of the sequence by this many basepairs. Only available when using genomic sequence type.\n",
            "            * *Default*: -\n",
            "            * *Example Values*: 1000\n",
            "        \n",
            "        \n",
            "          + *Name*: format\n",
            "        \n",
            "            * *Type*: Enum(fasta)\n",
            "            * *Description*: Format of the data\n",
            "            * *Default*: -\n",
            "            * *Example Values*: fasta\n",
            "        \n",
            "        \n",
            "          + *Name*: mask\n",
            "        \n",
            "            * *Type*: Enum(hard,soft)\n",
            "            * *Description*: Request the sequence masked for repeat sequences. Hard will mask all repeats as N's and soft will mask repeats as lowercased characters. Only available when using genomic sequence type.\n",
            "            * *Default*: -\n",
            "            * *Example Values*: hard\n",
            "        \n",
            "        \n",
            "          + *Name*: mask_feature\n",
            "        \n",
            "            * *Type*: Boolean\n",
            "            * *Description*: Mask features on the sequence. If sequence is genomic, mask introns. If sequence is cDNA, mask UTRs. Incompatible with the 'mask' option\n",
            "            * *Default*: 0\n",
            "            * *Example Values*: -\n",
            "        \n",
            "        \n",
            "          + *Name*: object_type\n",
            "        \n",
            "            * *Type*: String\n",
            "            * *Description*: Filter by feature type\n",
            "            * *Default*: -\n",
            "            * *Example Values*: gene\n",
            "        \n",
            "        \n",
            "          + *Name*: species\n",
            "        \n",
            "            * *Type*: String\n",
            "            * *Description*: Species name/alias\n",
            "            * *Default*: -\n",
            "            * *Example Values*: homo_sapiens\n",
            "        \n",
            "        \n",
            "          + *Name*: start\n",
            "        \n",
            "            * *Type*: Int\n",
            "            * *Description*: Trim the start of the sequence by this many basepairs. Trimming is relative to reading direction and in the coordinate system of the stable identifier. Parameter can not be used in conjunction with expand_5prime or expand_3prime.\n",
            "            * *Default*: -\n",
            "            * *Example Values*: 1000\n",
            "        \n",
            "        \n",
            "          + *Name*: type\n",
            "        \n",
            "            * *Type*: Enum(genomic,cds,cdna,protein)\n",
            "            * *Description*: Type of sequence. Defaults to genomic where applicable, i.e. not translations. cdna refers to the spliced transcript sequence with UTR; cds refers to the spliced transcript sequence without UTR.\n",
            "            * *Default*: genomic\n",
            "            * *Example Values*: cds\n",
            "        \n",
            "        \n",
            "        - Optional:\n",
            "        \n",
            "          + *Name*: callback\n",
            "        \n",
            "            * *Type*: String\n",
            "            * *Description*: Name of the callback subroutine to be returned by the requested JSONP response. Required ONLY when using JSONP as the serialisation method. Please see , .\n",
            "            * *Default*: -\n",
            "            * *Example Values*: randomlygeneratedname\n",
            "        \n",
            "        \n",
            "          + *Name*: db_type\n",
            "        \n",
            "            * *Type*: String\n",
            "            * *Description*: Restrict the search to a database other than the default. Useful if you need to use a DB other than core\n",
            "            * *Default*: -\n",
            "            * *Example Values*: core\n",
            "        \n",
            "        \n",
            "          + *Name*: end\n",
            "        \n",
            "            * *Type*: Int\n",
            "            * *Description*: Trim the end of the sequence by this many basepairs. Trimming is relative to reading direction and in the coordinate system of the stable identifier. Parameter can not be used in conjunction with expand_5prime or expand_3prime.\n",
            "            * *Default*: -\n",
            "            * *Example Values*: 1000\n",
            "        \n",
            "        \n",
            "          + *Name*: expand_3prime\n",
            "        \n",
            "            * *Type*: Int\n",
            "            * *Description*: Expand the sequence downstream of the sequence by this many basepairs. Only available when using genomic sequence type.\n",
            "            * *Default*: -\n",
            "            * *Example Values*: 1000\n",
            "        \n",
            "        \n",
            "          + *Name*: expand_5prime\n",
            "        \n",
            "            * *Type*: Int\n",
            "            * *Description*: Expand the sequence upstream of the sequence by this many basepairs. Only available when using genomic sequence type.\n",
            "            * *Default*: -\n",
            "            * *Example Values*: 1000\n",
            "        \n",
            "        \n",
            "          + *Name*: format\n",
            "        \n",
            "            * *Type*: Enum(fasta)\n",
            "            * *Description*: Format of the data\n",
            "            * *Default*: -\n",
            "            * *Example Values*: fasta\n",
            "        \n",
            "        \n",
            "          + *Name*: mask\n",
            "        \n",
            "            * *Type*: Enum(hard,soft)\n",
            "            * *Description*: Request the sequence masked for repeat sequences. Hard will mask all repeats as N's and soft will mask repeats as lowercased characters. Only available when using genomic sequence type.\n",
            "            * *Default*: -\n",
            "            * *Example Values*: hard\n",
            "        \n",
            "        \n",
            "          + *Name*: mask_feature\n",
            "        \n",
            "            * *Type*: Boolean\n",
            "            * *Description*: Mask features on the sequence. If sequence is genomic, mask introns. If sequence is cDNA, mask UTRs. Incompatible with the 'mask' option\n",
            "            * *Default*: 0\n",
            "            * *Example Values*: -\n",
            "        \n",
            "        \n",
            "          + *Name*: object_type\n",
            "        \n",
            "            * *Type*: String\n",
            "            * *Description*: Filter by feature type\n",
            "            * *Default*: -\n",
            "            * *Example Values*: gene\n",
            "        \n",
            "        \n",
            "          + *Name*: species\n",
            "        \n",
            "            * *Type*: String\n",
            "            * *Description*: Species name/alias\n",
            "            * *Default*: -\n",
            "            * *Example Values*: homo_sapiens\n",
            "        \n",
            "        \n",
            "          + *Name*: start\n",
            "        \n",
            "            * *Type*: Int\n",
            "            * *Description*: Trim the start of the sequence by this many basepairs. Trimming is relative to reading direction and in the coordinate system of the stable identifier. Parameter can not be used in conjunction with expand_5prime or expand_3prime.\n",
            "            * *Default*: -\n",
            "            * *Example Values*: 1000\n",
            "        \n",
            "        \n",
            "          + *Name*: type\n",
            "        \n",
            "            * *Type*: Enum(genomic,cds,cdna,protein)\n",
            "            * *Description*: Type of sequence. Defaults to genomic where applicable, i.e. not translations. cdna refers to the spliced transcript sequence with UTR; cds refers to the spliced transcript sequence without UTR.\n",
            "            * *Default*: genomic\n",
            "            * *Example Values*: cds\n",
            "        \n",
            "        \n",
            "        **Resource info**\n",
            "        \n",
            "        - **Methods**:  POST\n",
            "        - **Response formats**: json, jsonp\n",
            "        - **Maximum POST size**:        50\n",
            "        - **Slice length**:     1e7\n",
            "        \n",
            "        \n",
            "        **More info**\n",
            "        \n",
            "        https://rest.ensembl.org/documentation/info/sequence_id_post\n",
            "    \n",
            "    sequence_region(*args, **kwargs)\n",
            "            Sequence ``GET sequence/region/:species/:region``\n",
            "        \n",
            "        Returns the genomic sequence of the specified region of the given species. Supports feature masking and expand options.\n",
            "        \n",
            "        \n",
            "        **Parameters**\n",
            "        \n",
            "        - Required:\n",
            "        \n",
            "          + *Name*: region\n",
            "        \n",
            "            * *Type*: String\n",
            "            * *Description*: Query region. A maximum of 10Mb is allowed to be requested at any one time\n",
            "            * *Default*: -\n",
            "            * *Example Values*: X:1000000..1000100:1, X:1000000..1000100:-1, X:1000000..1000100\n",
            "        \n",
            "        \n",
            "          + *Name*: species\n",
            "        \n",
            "            * *Type*: String\n",
            "            * *Description*: Species name/alias\n",
            "            * *Default*: -\n",
            "            * *Example Values*: homo_sapiens, human\n",
            "        \n",
            "        \n",
            "        - Optional:\n",
            "        \n",
            "          + *Name*: callback\n",
            "        \n",
            "            * *Type*: String\n",
            "            * *Description*: Name of the callback subroutine to be returned by the requested JSONP response. Required ONLY when using JSONP as the serialisation method. Please see , .\n",
            "            * *Default*: -\n",
            "            * *Example Values*: randomlygeneratedname\n",
            "        \n",
            "        \n",
            "          + *Name*: coord_system\n",
            "        \n",
            "            * *Type*: String\n",
            "            * *Description*: Filter by coordinate system name\n",
            "            * *Default*: -\n",
            "            * *Example Values*: contig, seqlevel\n",
            "        \n",
            "        \n",
            "          + *Name*: coord_system_version\n",
            "        \n",
            "            * *Type*: String\n",
            "            * *Description*: Filter by coordinate system version\n",
            "            * *Default*: -\n",
            "            * *Example Values*: GRCh37\n",
            "        \n",
            "        \n",
            "          + *Name*: expand_3prime\n",
            "        \n",
            "            * *Type*: Int\n",
            "            * *Description*: Expand the sequence downstream of the sequence by this many basepairs. Only available when using genomic sequence type.\n",
            "            * *Default*: -\n",
            "            * *Example Values*: 1000\n",
            "        \n",
            "        \n",
            "          + *Name*: expand_5prime\n",
            "        \n",
            "            * *Type*: Int\n",
            "            * *Description*: Expand the sequence upstream of the sequence by this many basepairs. Only available when using genomic sequence type.\n",
            "            * *Default*: -\n",
            "            * *Example Values*: 1000\n",
            "        \n",
            "        \n",
            "          + *Name*: format\n",
            "        \n",
            "            * *Type*: Enum(fasta)\n",
            "            * *Description*: Format of the data.\n",
            "            * *Default*: -\n",
            "            * *Example Values*: fasta\n",
            "        \n",
            "        \n",
            "          + *Name*: mask\n",
            "        \n",
            "            * *Type*: Enum(hard,soft)\n",
            "            * *Description*: Request the sequence masked for repeat sequences. Hard will mask all repeats as N's and soft will mask repeats as lower cased characters. Only available when using genomic sequence type.\n",
            "            * *Default*: -\n",
            "            * *Example Values*: hard\n",
            "        \n",
            "        \n",
            "          + *Name*: mask_feature\n",
            "        \n",
            "            * *Type*: Boolean\n",
            "            * *Description*: Mask features on the sequence. If sequence is genomic, mask introns. If sequence is cDNA, mask UTRs. Incompatible with the 'mask' option\n",
            "            * *Default*: 0\n",
            "            * *Example Values*: 1\n",
            "        \n",
            "        \n",
            "        **Resource info**\n",
            "        \n",
            "        - **Methods**:  GET\n",
            "        - **Response formats**: fasta, json, seqxml, text, yaml, jsonp\n",
            "        - **Slice length**:     1e7\n",
            "        \n",
            "        \n",
            "        **More info**\n",
            "        \n",
            "        https://rest.ensembl.org/documentation/info/sequence_region\n",
            "    \n",
            "    sequence_region_post(*args, **kwargs)\n",
            "            Sequence ``POST sequence/region/:species``\n",
            "        \n",
            "        Request multiple types of sequence by a list of regions.\n",
            "        \n",
            "        \n",
            "        **Parameters**\n",
            "        \n",
            "        - Required:\n",
            "        \n",
            "          + *Name*: species\n",
            "        \n",
            "            * *Type*: String\n",
            "            * *Description*: Species name/alias\n",
            "            * *Default*: -\n",
            "            * *Example Values*: homo_sapiens, human\n",
            "        \n",
            "        \n",
            "        - Optional:\n",
            "        \n",
            "          + *Name*: callback\n",
            "        \n",
            "            * *Type*: String\n",
            "            * *Description*: Name of the callback subroutine to be returned by the requested JSONP response. Required ONLY when using JSONP as the serialisation method. Please see , .\n",
            "            * *Default*: -\n",
            "            * *Example Values*: randomlygeneratedname\n",
            "        \n",
            "        \n",
            "          + *Name*: coord_system\n",
            "        \n",
            "            * *Type*: String\n",
            "            * *Description*: Filter by coordinate system name\n",
            "            * *Default*: -\n",
            "            * *Example Values*: contig, seqlevel\n",
            "        \n",
            "        \n",
            "          + *Name*: coord_system_version\n",
            "        \n",
            "            * *Type*: String\n",
            "            * *Description*: Filter by coordinate system version\n",
            "            * *Default*: -\n",
            "            * *Example Values*: GRCh37\n",
            "        \n",
            "        \n",
            "          + *Name*: expand_3prime\n",
            "        \n",
            "            * *Type*: Int\n",
            "            * *Description*: Expand the sequence downstream of the sequence by this many basepairs. Only available when using genomic sequence type.\n",
            "            * *Default*: -\n",
            "            * *Example Values*: 1000\n",
            "        \n",
            "        \n",
            "          + *Name*: expand_5prime\n",
            "        \n",
            "            * *Type*: Int\n",
            "            * *Description*: Expand the sequence upstream of the sequence by this many basepairs. Only available when using genomic sequence type.\n",
            "            * *Default*: -\n",
            "            * *Example Values*: 1000\n",
            "        \n",
            "        \n",
            "          + *Name*: format\n",
            "        \n",
            "            * *Type*: Enum(fasta)\n",
            "            * *Description*: Format of the data.\n",
            "            * *Default*: -\n",
            "            * *Example Values*: fasta\n",
            "        \n",
            "        \n",
            "          + *Name*: mask\n",
            "        \n",
            "            * *Type*: Enum(hard,soft)\n",
            "            * *Description*: Request the sequence masked for repeat sequences. Hard will mask all repeats as N's and soft will mask repeats as lower cased characters. Only available when using genomic sequence type.\n",
            "            * *Default*: -\n",
            "            * *Example Values*: hard\n",
            "        \n",
            "        \n",
            "          + *Name*: mask_feature\n",
            "        \n",
            "            * *Type*: Boolean\n",
            "            * *Description*: Mask features on the sequence. If sequence is genomic, mask introns. If sequence is cDNA, mask UTRs. Incompatible with the 'mask' option\n",
            "            * *Default*: 0\n",
            "            * *Example Values*: 1\n",
            "        \n",
            "        \n",
            "        **Resource info**\n",
            "        \n",
            "        - **Methods**:  POST\n",
            "        - **Response formats**: json, jsonp\n",
            "        - **Maximum POST size**:        50\n",
            "        - **Slice length**:     1e7\n",
            "        \n",
            "        \n",
            "        **More info**\n",
            "        \n",
            "        https://rest.ensembl.org/documentation/info/sequence_region_post\n",
            "    \n",
            "    software(*args, **kwargs)\n",
            "            Information ``GET info/software``\n",
            "        \n",
            "        Shows the current version of the Ensembl API used by the REST server.\n",
            "        \n",
            "        \n",
            "        **Parameters**\n",
            "        \n",
            "        - Required:\n",
            "        \n",
            "          + *Name*: callback\n",
            "        \n",
            "            * *Type*: String\n",
            "            * *Description*: Name of the callback subroutine to be returned by the requested JSONP response. Required ONLY when using JSONP as the serialisation method. Please see , .\n",
            "            * *Default*: -\n",
            "            * *Example Values*: randomlygeneratedname\n",
            "        \n",
            "        \n",
            "        - Optional:\n",
            "        \n",
            "          + *Name*: callback\n",
            "        \n",
            "            * *Type*: String\n",
            "            * *Description*: Name of the callback subroutine to be returned by the requested JSONP response. Required ONLY when using JSONP as the serialisation method. Please see , .\n",
            "            * *Default*: -\n",
            "            * *Example Values*: randomlygeneratedname\n",
            "        \n",
            "        \n",
            "        **Resource info**\n",
            "        \n",
            "        - **Methods**:  GET\n",
            "        - **Response formats**: json, xml, jsonp\n",
            "        \n",
            "        \n",
            "        **More info**\n",
            "        \n",
            "        https://rest.ensembl.org/documentation/info/software\n",
            "    \n",
            "    species(*args, **kwargs)\n",
            "            Information ``GET info/species``\n",
            "        \n",
            "        Lists all available species, their aliases, available adaptor groups and data release.\n",
            "        \n",
            "        \n",
            "        **Parameters**\n",
            "        \n",
            "        - Required:\n",
            "        \n",
            "          + *Name*: callback\n",
            "        \n",
            "            * *Type*: String\n",
            "            * *Description*: Name of the callback subroutine to be returned by the requested JSONP response. Required ONLY when using JSONP as the serialisation method. Please see , .\n",
            "            * *Default*: -\n",
            "            * *Example Values*: randomlygeneratedname\n",
            "        \n",
            "        \n",
            "          + *Name*: division\n",
            "        \n",
            "            * *Type*: String\n",
            "            * *Description*: Filter by Ensembl or Ensembl Genomes division.\n",
            "            * *Default*: EnsemblVertebrates\n",
            "            * *Example Values*: EnsemblVertebrates\n",
            "        \n",
            "        \n",
            "          + *Name*: hide_strain_info\n",
            "        \n",
            "            * *Type*: Boolean(0,1)\n",
            "            * *Description*: Show/hide strain and strain_collection info in the output\n",
            "            * *Default*: 0\n",
            "            * *Example Values*: -\n",
            "        \n",
            "        \n",
            "          + *Name*: strain_collection\n",
            "        \n",
            "            * *Type*: String\n",
            "            * *Description*: Filter by strain_collection.\n",
            "            * *Default*: -\n",
            "            * *Example Values*: mouse\n",
            "        \n",
            "        \n",
            "        - Optional:\n",
            "        \n",
            "          + *Name*: callback\n",
            "        \n",
            "            * *Type*: String\n",
            "            * *Description*: Name of the callback subroutine to be returned by the requested JSONP response. Required ONLY when using JSONP as the serialisation method. Please see , .\n",
            "            * *Default*: -\n",
            "            * *Example Values*: randomlygeneratedname\n",
            "        \n",
            "        \n",
            "          + *Name*: division\n",
            "        \n",
            "            * *Type*: String\n",
            "            * *Description*: Filter by Ensembl or Ensembl Genomes division.\n",
            "            * *Default*: EnsemblVertebrates\n",
            "            * *Example Values*: EnsemblVertebrates\n",
            "        \n",
            "        \n",
            "          + *Name*: hide_strain_info\n",
            "        \n",
            "            * *Type*: Boolean(0,1)\n",
            "            * *Description*: Show/hide strain and strain_collection info in the output\n",
            "            * *Default*: 0\n",
            "            * *Example Values*: -\n",
            "        \n",
            "        \n",
            "          + *Name*: strain_collection\n",
            "        \n",
            "            * *Type*: String\n",
            "            * *Description*: Filter by strain_collection.\n",
            "            * *Default*: -\n",
            "            * *Example Values*: mouse\n",
            "        \n",
            "        \n",
            "        **Resource info**\n",
            "        \n",
            "        - **Methods**:  GET\n",
            "        - **Response formats**: json, xml, jsonp\n",
            "        \n",
            "        \n",
            "        **More info**\n",
            "        \n",
            "        https://rest.ensembl.org/documentation/info/species\n",
            "    \n",
            "    species_id(*args, **kwargs)\n",
            "            EQTL ``GET eqtl/stable_id/:species/:stable_id``\n",
            "        \n",
            "        Returns the p-value for each SNP in a given gene (e.g. ENSG00000227232)\n",
            "        \n",
            "        \n",
            "        **Parameters**\n",
            "        \n",
            "        - Required:\n",
            "        \n",
            "          + *Name*: species\n",
            "        \n",
            "            * *Type*: string\n",
            "            * *Description*: Species name/alias\n",
            "            * *Default*: -\n",
            "            * *Example Values*: homo_sapiens\n",
            "        \n",
            "        \n",
            "          + *Name*: stable_id\n",
            "        \n",
            "            * *Type*: String\n",
            "            * *Description*: Ensembl stable ID\n",
            "            * *Default*: -\n",
            "            * *Example Values*: ENSG00000122435\n",
            "        \n",
            "        \n",
            "        - Optional:\n",
            "        \n",
            "          + *Name*: callback\n",
            "        \n",
            "            * *Type*: String\n",
            "            * *Description*: Name of the callback subroutine to be returned by the requested JSONP response. Required ONLY when using JSONP as the serialisation method. Please see , .\n",
            "            * *Default*: -\n",
            "            * *Example Values*: randomlygeneratedname\n",
            "        \n",
            "        \n",
            "          + *Name*: statistic\n",
            "        \n",
            "            * *Type*: String\n",
            "            * *Description*: Filter by statistic\n",
            "            * *Default*: -\n",
            "            * *Example Values*: p-value, beta\n",
            "        \n",
            "        \n",
            "          + *Name*: tissue\n",
            "        \n",
            "            * *Type*: String\n",
            "            * *Description*: Tissue of interest [Stomach, Thyroid, Whole_Blood]\n",
            "            * *Default*: -\n",
            "            * *Example Values*: Whole_Blood\n",
            "        \n",
            "        \n",
            "          + *Name*: \n",
            "        \n",
            "            * *Type*: String\n",
            "            * *Description*: rsID (Reference SNP cluster ID)\n",
            "            * *Default*: -\n",
            "            * *Example Values*: rs123\n",
            "        \n",
            "        \n",
            "        **Resource info**\n",
            "        \n",
            "        - **Methods**:  GET\n",
            "        - **Response formats**: json, xml, jsonp\n",
            "        \n",
            "        \n",
            "        **More info**\n",
            "        \n",
            "        https://rest.ensembl.org/documentation/info/species_id\n",
            "    \n",
            "    species_variant(*args, **kwargs)\n",
            "            EQTL ``GET eqtl/variant_name/:species/:variant_name``\n",
            "        \n",
            "        Returns the p-values for a SNP (e.g. rs123)\n",
            "        \n",
            "        \n",
            "        **Parameters**\n",
            "        \n",
            "        - Required:\n",
            "        \n",
            "          + *Name*: species\n",
            "        \n",
            "            * *Type*: string\n",
            "            * *Description*: Species name/alias\n",
            "            * *Default*: -\n",
            "            * *Example Values*: homo_sapiens\n",
            "        \n",
            "        \n",
            "          + *Name*: \n",
            "        \n",
            "            * *Type*: String\n",
            "            * *Description*: rsID (Reference SNP cluster ID)\n",
            "            * *Default*: -\n",
            "            * *Example Values*: rs123\n",
            "        \n",
            "        \n",
            "        - Optional:\n",
            "        \n",
            "          + *Name*: callback\n",
            "        \n",
            "            * *Type*: String\n",
            "            * *Description*: Name of the callback subroutine to be returned by the requested JSONP response. Required ONLY when using JSONP as the serialisation method. Please see , .\n",
            "            * *Default*: -\n",
            "            * *Example Values*: randomlygeneratedname\n",
            "        \n",
            "        \n",
            "          + *Name*: id\n",
            "        \n",
            "            * *Type*: String\n",
            "            * *Description*: Ensembl stable ID\n",
            "            * *Default*: -\n",
            "            * *Example Values*: ENSG00000122435\n",
            "        \n",
            "        \n",
            "          + *Name*: statistic\n",
            "        \n",
            "            * *Type*: String\n",
            "            * *Description*: Filter by statistic\n",
            "            * *Default*: -\n",
            "            * *Example Values*: p-value, beta\n",
            "        \n",
            "        \n",
            "          + *Name*: tissue\n",
            "        \n",
            "            * *Type*: String\n",
            "            * *Description*: Tissue of interest [Stomach, Thyroid, Whole_Blood]\n",
            "            * *Default*: -\n",
            "            * *Example Values*: Whole_Blood\n",
            "        \n",
            "        \n",
            "        **Resource info**\n",
            "        \n",
            "        - **Methods**:  GET\n",
            "        - **Response formats**: json, xml, jsonp\n",
            "        \n",
            "        \n",
            "        **More info**\n",
            "        \n",
            "        https://rest.ensembl.org/documentation/info/species_variant\n",
            "    \n",
            "    symbol_lookup(*args, **kwargs)\n",
            "            Lookup ``GET lookup/symbol/:species/:symbol``\n",
            "        \n",
            "        Find the species and database for a symbol in a linked external database\n",
            "        \n",
            "        \n",
            "        **Parameters**\n",
            "        \n",
            "        - Required:\n",
            "        \n",
            "          + *Name*: species\n",
            "        \n",
            "            * *Type*: String\n",
            "            * *Description*: Species name/alias\n",
            "            * *Default*: -\n",
            "            * *Example Values*: homo_sapiens, human\n",
            "        \n",
            "        \n",
            "          + *Name*: symbol\n",
            "        \n",
            "            * *Type*: String\n",
            "            * *Description*: A name or symbol from an annotation source has been linked to a genetic feature\n",
            "            * *Default*: -\n",
            "            * *Example Values*: BRCA2\n",
            "        \n",
            "        \n",
            "        - Optional:\n",
            "        \n",
            "          + *Name*: callback\n",
            "        \n",
            "            * *Type*: String\n",
            "            * *Description*: Name of the callback subroutine to be returned by the requested JSONP response. Required ONLY when using JSONP as the serialisation method. Please see , .\n",
            "            * *Default*: -\n",
            "            * *Example Values*: randomlygeneratedname\n",
            "        \n",
            "        \n",
            "          + *Name*: expand\n",
            "        \n",
            "            * *Type*: Boolean(0,1)\n",
            "            * *Description*: Expands the search to include any connected features. e.g. If the object is a gene, its transcripts, translations and exons will be returned as well.\n",
            "            * *Default*: NULL\n",
            "            * *Example Values*: -\n",
            "        \n",
            "        \n",
            "          + *Name*: format\n",
            "        \n",
            "            * *Type*: Enum(full,condensed)\n",
            "            * *Description*: Specify the layout of the response\n",
            "            * *Default*: full\n",
            "            * *Example Values*: -\n",
            "        \n",
            "        \n",
            "        **Resource info**\n",
            "        \n",
            "        - **Methods**:  GET\n",
            "        - **Response formats**: json, xml, jsonp\n",
            "        \n",
            "        \n",
            "        **More info**\n",
            "        \n",
            "        https://rest.ensembl.org/documentation/info/symbol_lookup\n",
            "    \n",
            "    symbol_post(*args, **kwargs)\n",
            "            Lookup ``POST lookup/symbol/:species/:symbol``\n",
            "        \n",
            "        Find the species and database for a set of symbols in a linked external database. Unknown symbols are omitted from the response.\n",
            "        \n",
            "        \n",
            "        **Parameters**\n",
            "        \n",
            "        - Required:\n",
            "        \n",
            "          + *Name*: species\n",
            "        \n",
            "            * *Type*: String\n",
            "            * *Description*: Species name/alias for the whole batch of symbols\n",
            "            * *Default*: -\n",
            "            * *Example Values*: homo_sapiens, human\n",
            "        \n",
            "        \n",
            "        - Optional:\n",
            "        \n",
            "          + *Name*: callback\n",
            "        \n",
            "            * *Type*: String\n",
            "            * *Description*: Name of the callback subroutine to be returned by the requested JSONP response. Required ONLY when using JSONP as the serialisation method. Please see , .\n",
            "            * *Default*: -\n",
            "            * *Example Values*: randomlygeneratedname\n",
            "        \n",
            "        \n",
            "          + *Name*: expand\n",
            "        \n",
            "            * *Type*: Boolean(0,1)\n",
            "            * *Description*: Expands the search to include any connected features. e.g. If the object is a gene, its transcripts, translations and exons will be returned as well.\n",
            "            * *Default*: NULL\n",
            "            * *Example Values*: -\n",
            "        \n",
            "        \n",
            "          + *Name*: format\n",
            "        \n",
            "            * *Type*: Enum(full,condensed)\n",
            "            * *Description*: Specify the layout of the response\n",
            "            * *Default*: full\n",
            "            * *Example Values*: -\n",
            "        \n",
            "        \n",
            "        **Resource info**\n",
            "        \n",
            "        - **Methods**:  POST\n",
            "        - **Response formats**: json, xml, jsonp\n",
            "        - **Maximum POST size**:        1000\n",
            "        \n",
            "        \n",
            "        **More info**\n",
            "        \n",
            "        https://rest.ensembl.org/documentation/info/symbol_post\n",
            "    \n",
            "    taxonomy_classification(*args, **kwargs)\n",
            "            Ontologies and Taxonomy ``GET taxonomy/classification/:id``\n",
            "        \n",
            "        Return the taxonomic classification of a taxon node\n",
            "        \n",
            "        \n",
            "        **Parameters**\n",
            "        \n",
            "        - Required:\n",
            "        \n",
            "          + *Name*: id\n",
            "        \n",
            "            * *Type*: String\n",
            "            * *Description*: A taxon identifier. Can be a NCBI taxon id or a name\n",
            "            * *Default*: -\n",
            "            * *Example Values*: 9606, Homo sapiens\n",
            "        \n",
            "        \n",
            "        - Optional:\n",
            "        \n",
            "          + *Name*: callback\n",
            "        \n",
            "            * *Type*: String\n",
            "            * *Description*: Name of the callback subroutine to be returned by the requested JSONP response. Required ONLY when using JSONP as the serialisation method. Please see , .\n",
            "            * *Default*: -\n",
            "            * *Example Values*: randomlygeneratedname\n",
            "        \n",
            "        \n",
            "        **Resource info**\n",
            "        \n",
            "        - **Methods**:  GET\n",
            "        - **Response formats**: json, xml, yaml, jsonp\n",
            "        \n",
            "        \n",
            "        **More info**\n",
            "        \n",
            "        https://rest.ensembl.org/documentation/info/taxonomy_classification\n",
            "    \n",
            "    taxonomy_id(*args, **kwargs)\n",
            "            Ontologies and Taxonomy ``GET taxonomy/id/:id``\n",
            "        \n",
            "        Search for a taxonomic term by its identifier or name\n",
            "        \n",
            "        \n",
            "        **Parameters**\n",
            "        \n",
            "        - Required:\n",
            "        \n",
            "          + *Name*: id\n",
            "        \n",
            "            * *Type*: String\n",
            "            * *Description*: A taxon identifier. Can be a NCBI taxon id or a name\n",
            "            * *Default*: -\n",
            "            * *Example Values*: 9606, Homo sapiens\n",
            "        \n",
            "        \n",
            "        - Optional:\n",
            "        \n",
            "          + *Name*: callback\n",
            "        \n",
            "            * *Type*: String\n",
            "            * *Description*: Name of the callback subroutine to be returned by the requested JSONP response. Required ONLY when using JSONP as the serialisation method. Please see , .\n",
            "            * *Default*: -\n",
            "            * *Example Values*: randomlygeneratedname\n",
            "        \n",
            "        \n",
            "          + *Name*: simple\n",
            "        \n",
            "            * *Type*: Boolean\n",
            "            * *Description*: If set the API will avoid the fetching of parent and child terms\n",
            "            * *Default*: 0\n",
            "            * *Example Values*: -\n",
            "        \n",
            "        \n",
            "        **Resource info**\n",
            "        \n",
            "        - **Methods**:  GET\n",
            "        - **Response formats**: json, xml, yaml, jsonp\n",
            "        \n",
            "        \n",
            "        **More info**\n",
            "        \n",
            "        https://rest.ensembl.org/documentation/info/taxonomy_id\n",
            "    \n",
            "    taxonomy_name(*args, **kwargs)\n",
            "            Ontologies and Taxonomy ``GET taxonomy/name/:name``\n",
            "        \n",
            "        Search for a taxonomic id by a non-scientific name\n",
            "        \n",
            "        \n",
            "        **Parameters**\n",
            "        \n",
            "        - Required:\n",
            "        \n",
            "          + *Name*: name\n",
            "        \n",
            "            * *Type*: String\n",
            "            * *Description*: A non-scientific species name. Can include SQL wildcards\n",
            "            * *Default*: -\n",
            "            * *Example Values*: Homo%25\n",
            "        \n",
            "        \n",
            "        - Optional:\n",
            "        \n",
            "          + *Name*: callback\n",
            "        \n",
            "            * *Type*: String\n",
            "            * *Description*: Name of the callback subroutine to be returned by the requested JSONP response. Required ONLY when using JSONP as the serialisation method. Please see , .\n",
            "            * *Default*: -\n",
            "            * *Example Values*: randomlygeneratedname\n",
            "        \n",
            "        \n",
            "        **Resource info**\n",
            "        \n",
            "        - **Methods**:  GET\n",
            "        - **Response formats**: json, xml, yaml, jsonp\n",
            "        \n",
            "        \n",
            "        **More info**\n",
            "        \n",
            "        https://rest.ensembl.org/documentation/info/taxonomy_name\n",
            "    \n",
            "    tissues(*args, **kwargs)\n",
            "            EQTL ``GET eqtl/tissue/:species/``\n",
            "        \n",
            "        Returns all tissues currently available in the DB\n",
            "        \n",
            "        \n",
            "        **Parameters**\n",
            "        \n",
            "        - Required:\n",
            "        \n",
            "          + *Name*: callback\n",
            "        \n",
            "            * *Type*: String\n",
            "            * *Description*: Name of the callback subroutine to be returned by the requested JSONP response. Required ONLY when using JSONP as the serialisation method. Please see , .\n",
            "            * *Default*: -\n",
            "            * *Example Values*: randomlygeneratedname\n",
            "        \n",
            "        \n",
            "        - Optional:\n",
            "        \n",
            "          + *Name*: callback\n",
            "        \n",
            "            * *Type*: String\n",
            "            * *Description*: Name of the callback subroutine to be returned by the requested JSONP response. Required ONLY when using JSONP as the serialisation method. Please see , .\n",
            "            * *Default*: -\n",
            "            * *Example Values*: randomlygeneratedname\n",
            "        \n",
            "        \n",
            "        **Resource info**\n",
            "        \n",
            "        - **Methods**:  GET\n",
            "        - **Response formats**: json, xml, jsonp\n",
            "        \n",
            "        \n",
            "        **More info**\n",
            "        \n",
            "        https://rest.ensembl.org/documentation/info/tissues\n",
            "    \n",
            "    transcript_haplotypes_get(*args, **kwargs)\n",
            "            Transcript Haplotypes ``GET transcript_haplotypes/:species/:id``\n",
            "        \n",
            "        Computes observed transcript haplotype sequences based on phased genotype data\n",
            "        \n",
            "        \n",
            "        **Parameters**\n",
            "        \n",
            "        - Required:\n",
            "        \n",
            "          + *Name*: id\n",
            "        \n",
            "            * *Type*: String\n",
            "            * *Description*: Transcript stable id\n",
            "            * *Default*: -\n",
            "            * *Example Values*: ENST00000288602\n",
            "        \n",
            "        \n",
            "          + *Name*: species\n",
            "        \n",
            "            * *Type*: String\n",
            "            * *Description*: Species name/alias\n",
            "            * *Default*: -\n",
            "            * *Example Values*: homo_sapiens\n",
            "        \n",
            "        \n",
            "        - Optional:\n",
            "        \n",
            "          + *Name*: aligned_sequences\n",
            "        \n",
            "            * *Type*: Boolean(0,1)\n",
            "            * *Description*: Include aligned sequences used to generate differences\n",
            "            * *Default*: 0\n",
            "            * *Example Values*: 1\n",
            "        \n",
            "        \n",
            "          + *Name*: callback\n",
            "        \n",
            "            * *Type*: String\n",
            "            * *Description*: Name of the callback subroutine to be returned by the requested JSONP response. Required ONLY when using JSONP as the serialisation method. Please see , .\n",
            "            * *Default*: -\n",
            "            * *Example Values*: randomlygeneratedname\n",
            "        \n",
            "        \n",
            "          + *Name*: samples\n",
            "        \n",
            "            * *Type*: Boolean(0,1)\n",
            "            * *Description*: Include sample-haplotype assignments\n",
            "            * *Default*: 0\n",
            "            * *Example Values*: 1\n",
            "        \n",
            "        \n",
            "          + *Name*: sequence\n",
            "        \n",
            "            * *Type*: Boolean(0,1)\n",
            "            * *Description*: Include raw sequences\n",
            "            * *Default*: 0\n",
            "            * *Example Values*: 1\n",
            "        \n",
            "        \n",
            "        **Resource info**\n",
            "        \n",
            "        - **Methods**:  GET\n",
            "        - **Response formats**: json, jsonp\n",
            "        \n",
            "        \n",
            "        **More info**\n",
            "        \n",
            "        https://rest.ensembl.org/documentation/info/transcript_haplotypes_get\n",
            "    \n",
            "    variant_recoder(*args, **kwargs)\n",
            "            Variation ``GET variant_recoder/:species/:id``\n",
            "        \n",
            "        Translate a variant identifier, HGVS notation or genomic SPDI notation to all possible variant IDs, HGVS and genomic SPDI\n",
            "        \n",
            "        \n",
            "        **Parameters**\n",
            "        \n",
            "        - Required:\n",
            "        \n",
            "          + *Name*: id\n",
            "        \n",
            "            * *Type*: String\n",
            "            * *Description*: Variant ID, HGVS notation or genomic SPDI notation\n",
            "            * *Default*: -\n",
            "            * *Example Values*: rs56116432, AGT:c.803T>C, NC_000023.11:284252:C:G\n",
            "        \n",
            "        \n",
            "          + *Name*: species\n",
            "        \n",
            "            * *Type*: String\n",
            "            * *Description*: Species name/alias\n",
            "            * *Default*: -\n",
            "            * *Example Values*: homo_sapiens, human\n",
            "        \n",
            "        \n",
            "        - Optional:\n",
            "        \n",
            "          + *Name*: callback\n",
            "        \n",
            "            * *Type*: String\n",
            "            * *Description*: Name of the callback subroutine to be returned by the requested JSONP response. Required ONLY when using JSONP as the serialisation method. Please see , .\n",
            "            * *Default*: -\n",
            "            * *Example Values*: randomlygeneratedname\n",
            "        \n",
            "        \n",
            "          + *Name*: fields\n",
            "        \n",
            "            * *Type*: String\n",
            "            * *Description*: \n",
            "            * *Default*: id,hgvsg,hgvsc,hgvsp,spdi\n",
            "            * *Example Values*: -\n",
            "        \n",
            "        \n",
            "        **Resource info**\n",
            "        \n",
            "        - **Methods**:  GET\n",
            "        - **Response formats**: json, xml, jsonp\n",
            "        \n",
            "        \n",
            "        **More info**\n",
            "        \n",
            "        https://rest.ensembl.org/documentation/info/variant_recoder\n",
            "    \n",
            "    variant_recoder_post(*args, **kwargs)\n",
            "            Variation ``POST variant_recoder/:species``\n",
            "        \n",
            "        Translate a list of variant identifiers, HGVS notations or genomic SPDI notations to all possible variant IDs, HGVS and genomic SPDI\n",
            "        \n",
            "        \n",
            "        **Parameters**\n",
            "        \n",
            "        - Required:\n",
            "        \n",
            "          + *Name*: species\n",
            "        \n",
            "            * *Type*: String\n",
            "            * *Description*: Species name/alias\n",
            "            * *Default*: -\n",
            "            * *Example Values*: homo_sapiens, human\n",
            "        \n",
            "        \n",
            "        - Optional:\n",
            "        \n",
            "          + *Name*: callback\n",
            "        \n",
            "            * *Type*: String\n",
            "            * *Description*: Name of the callback subroutine to be returned by the requested JSONP response. Required ONLY when using JSONP as the serialisation method. Please see , .\n",
            "            * *Default*: -\n",
            "            * *Example Values*: randomlygeneratedname\n",
            "        \n",
            "        \n",
            "          + *Name*: fields\n",
            "        \n",
            "            * *Type*: String\n",
            "            * *Description*: \n",
            "            * *Default*: id,hgvsg,hgvsc,hgvsp,spdi\n",
            "            * *Example Values*: -\n",
            "        \n",
            "        \n",
            "        **Resource info**\n",
            "        \n",
            "        - **Methods**:  POST\n",
            "        - **Response formats**: json, xml, jsonp\n",
            "        - **Maximum POST size**:        200\n",
            "        \n",
            "        \n",
            "        **More info**\n",
            "        \n",
            "        https://rest.ensembl.org/documentation/info/variant_recoder_post\n",
            "    \n",
            "    variation(*args, **kwargs)\n",
            "            Information ``GET info/variation/:species``\n",
            "        \n",
            "        List the variation sources used in Ensembl for a species.\n",
            "        \n",
            "        \n",
            "        **Parameters**\n",
            "        \n",
            "        - Required:\n",
            "        \n",
            "          + *Name*: species\n",
            "        \n",
            "            * *Type*: String\n",
            "            * *Description*: Species name/alias\n",
            "            * *Default*: -\n",
            "            * *Example Values*: homo_sapiens\n",
            "        \n",
            "        \n",
            "        - Optional:\n",
            "        \n",
            "          + *Name*: callback\n",
            "        \n",
            "            * *Type*: String\n",
            "            * *Description*: Name of the callback subroutine to be returned by the requested JSONP response. Required ONLY when using JSONP as the serialisation method. Please see , .\n",
            "            * *Default*: -\n",
            "            * *Example Values*: randomlygeneratedname\n",
            "        \n",
            "        \n",
            "          + *Name*: filter\n",
            "        \n",
            "            * *Type*: String\n",
            "            * *Description*: \n",
            "            * *Default*: -\n",
            "            * *Example Values*: dbSNP, ClinVar, OMIM, UniProt, HGMD\n",
            "        \n",
            "        \n",
            "        **Resource info**\n",
            "        \n",
            "        - **Methods**:  GET\n",
            "        - **Response formats**: json, xml, jsonp\n",
            "        \n",
            "        \n",
            "        **More info**\n",
            "        \n",
            "        https://rest.ensembl.org/documentation/info/variation\n",
            "    \n",
            "    variation_consequence_types(*args, **kwargs)\n",
            "            Information ``GET info/variation/consequence_types``\n",
            "        \n",
            "        Lists all variant consequence types.\n",
            "        \n",
            "        \n",
            "        **Parameters**\n",
            "        \n",
            "        - Required:\n",
            "        \n",
            "          + *Name*: callback\n",
            "        \n",
            "            * *Type*: String\n",
            "            * *Description*: Name of the callback subroutine to be returned by the requested JSONP response. Required ONLY when using JSONP as the serialisation method. Please see , .\n",
            "            * *Default*: -\n",
            "            * *Example Values*: randomlygeneratedname\n",
            "        \n",
            "        \n",
            "        - Optional:\n",
            "        \n",
            "          + *Name*: callback\n",
            "        \n",
            "            * *Type*: String\n",
            "            * *Description*: Name of the callback subroutine to be returned by the requested JSONP response. Required ONLY when using JSONP as the serialisation method. Please see , .\n",
            "            * *Default*: -\n",
            "            * *Example Values*: randomlygeneratedname\n",
            "        \n",
            "        \n",
            "        **Resource info**\n",
            "        \n",
            "        - **Methods**:  GET\n",
            "        - **Response formats**: json, xml, jsonp\n",
            "        \n",
            "        \n",
            "        **More info**\n",
            "        \n",
            "        https://rest.ensembl.org/documentation/info/variation_consequence_types\n",
            "    \n",
            "    variation_id(*args, **kwargs)\n",
            "            Variation ``GET variation/:species/:id``\n",
            "        \n",
            "        Uses a variant identifier (e.g. rsID) to return the variation features including optional genotype, phenotype and population data\n",
            "        \n",
            "        \n",
            "        **Parameters**\n",
            "        \n",
            "        - Required:\n",
            "        \n",
            "          + *Name*: id\n",
            "        \n",
            "            * *Type*: String\n",
            "            * *Description*: Variant id\n",
            "            * *Default*: -\n",
            "            * *Example Values*: rs56116432\n",
            "        \n",
            "        \n",
            "          + *Name*: species\n",
            "        \n",
            "            * *Type*: String\n",
            "            * *Description*: Species name/alias\n",
            "            * *Default*: -\n",
            "            * *Example Values*: homo_sapiens, human\n",
            "        \n",
            "        \n",
            "        - Optional:\n",
            "        \n",
            "          + *Name*: callback\n",
            "        \n",
            "            * *Type*: String\n",
            "            * *Description*: Name of the callback subroutine to be returned by the requested JSONP response. Required ONLY when using JSONP as the serialisation method. Please see , .\n",
            "            * *Default*: -\n",
            "            * *Example Values*: randomlygeneratedname\n",
            "        \n",
            "        \n",
            "          + *Name*: genotypes\n",
            "        \n",
            "            * *Type*: Boolean(0,1)\n",
            "            * *Description*: Include individual genotypes\n",
            "            * *Default*: 0\n",
            "            * *Example Values*: -\n",
            "        \n",
            "        \n",
            "          + *Name*: genotyping_chips\n",
            "        \n",
            "            * *Type*: Boolean(0,1)\n",
            "            * *Description*: Include genotyping chips information\n",
            "            * *Default*: 0\n",
            "            * *Example Values*: -\n",
            "        \n",
            "        \n",
            "          + *Name*: phenotypes\n",
            "        \n",
            "            * *Type*: Boolean(0,1)\n",
            "            * *Description*: Include phenotypes\n",
            "            * *Default*: 0\n",
            "            * *Example Values*: -\n",
            "        \n",
            "        \n",
            "          + *Name*: pops\n",
            "        \n",
            "            * *Type*: Boolean(0,1)\n",
            "            * *Description*: Include population allele frequencies\n",
            "            * *Default*: 0\n",
            "            * *Example Values*: -\n",
            "        \n",
            "        \n",
            "          + *Name*: population_genotypes\n",
            "        \n",
            "            * *Type*: Boolean(0,1)\n",
            "            * *Description*: Include population genotype frequencies\n",
            "            * *Default*: 0\n",
            "            * *Example Values*: -\n",
            "        \n",
            "        \n",
            "        **Resource info**\n",
            "        \n",
            "        - **Methods**:  GET\n",
            "        - **Response formats**: json, xml, jsonp\n",
            "        \n",
            "        \n",
            "        **More info**\n",
            "        \n",
            "        https://rest.ensembl.org/documentation/info/variation_id\n",
            "    \n",
            "    variation_pmcid_get(*args, **kwargs)\n",
            "            Variation ``GET variation/:species/pmcid/:pmcid``\n",
            "        \n",
            "        Fetch variants by publication using PubMed Central reference number (PMCID)\n",
            "        \n",
            "        \n",
            "        **Parameters**\n",
            "        \n",
            "        - Required:\n",
            "        \n",
            "          + *Name*: pmcid\n",
            "        \n",
            "            * *Type*: String\n",
            "            * *Description*: PubMed Central reference number (PMCID)\n",
            "            * *Default*: -\n",
            "            * *Example Values*: PMC5002951\n",
            "        \n",
            "        \n",
            "          + *Name*: species\n",
            "        \n",
            "            * *Type*: String\n",
            "            * *Description*: Species name/alias\n",
            "            * *Default*: -\n",
            "            * *Example Values*: homo_sapiens, human\n",
            "        \n",
            "        \n",
            "        - Optional:\n",
            "        \n",
            "          + *Name*: callback\n",
            "        \n",
            "            * *Type*: String\n",
            "            * *Description*: Name of the callback subroutine to be returned by the requested JSONP response. Required ONLY when using JSONP as the serialisation method. Please see , .\n",
            "            * *Default*: -\n",
            "            * *Example Values*: randomlygeneratedname\n",
            "        \n",
            "        \n",
            "        **Resource info**\n",
            "        \n",
            "        - **Methods**:  GET\n",
            "        - **Response formats**: json, xml, jsonp\n",
            "        \n",
            "        \n",
            "        **More info**\n",
            "        \n",
            "        https://rest.ensembl.org/documentation/info/variation_pmcid_get\n",
            "    \n",
            "    variation_pmid_get(*args, **kwargs)\n",
            "            Variation ``GET variation/:species/pmid/:pmid``\n",
            "        \n",
            "        Fetch variants by publication using PubMed reference number (PMID)\n",
            "        \n",
            "        \n",
            "        **Parameters**\n",
            "        \n",
            "        - Required:\n",
            "        \n",
            "          + *Name*: pmid\n",
            "        \n",
            "            * *Type*: String\n",
            "            * *Description*: PubMed reference number (PMID)\n",
            "            * *Default*: -\n",
            "            * *Example Values*: 26318936\n",
            "        \n",
            "        \n",
            "          + *Name*: species\n",
            "        \n",
            "            * *Type*: String\n",
            "            * *Description*: Species name/alias\n",
            "            * *Default*: -\n",
            "            * *Example Values*: homo_sapiens, human\n",
            "        \n",
            "        \n",
            "        - Optional:\n",
            "        \n",
            "          + *Name*: callback\n",
            "        \n",
            "            * *Type*: String\n",
            "            * *Description*: Name of the callback subroutine to be returned by the requested JSONP response. Required ONLY when using JSONP as the serialisation method. Please see , .\n",
            "            * *Default*: -\n",
            "            * *Example Values*: randomlygeneratedname\n",
            "        \n",
            "        \n",
            "        **Resource info**\n",
            "        \n",
            "        - **Methods**:  GET\n",
            "        - **Response formats**: json, xml, jsonp\n",
            "        \n",
            "        \n",
            "        **More info**\n",
            "        \n",
            "        https://rest.ensembl.org/documentation/info/variation_pmid_get\n",
            "    \n",
            "    variation_population_name(*args, **kwargs)\n",
            "            Information ``GET info/variation/populations/:species:/:population_name``\n",
            "        \n",
            "        List all individuals for a population from a species\n",
            "        \n",
            "        \n",
            "        **Parameters**\n",
            "        \n",
            "        - Required:\n",
            "        \n",
            "          + *Name*: population_name\n",
            "        \n",
            "            * *Type*: String\n",
            "            * *Description*: Population name\n",
            "            * *Default*: -\n",
            "            * *Example Values*: 1000GENOMES:phase_3:ASW\n",
            "        \n",
            "        \n",
            "          + *Name*: species\n",
            "        \n",
            "            * *Type*: String\n",
            "            * *Description*: Species name/alias\n",
            "            * *Default*: -\n",
            "            * *Example Values*: human\n",
            "        \n",
            "        \n",
            "        - Optional:\n",
            "        \n",
            "          + *Name*: callback\n",
            "        \n",
            "            * *Type*: String\n",
            "            * *Description*: Name of the callback subroutine to be returned by the requested JSONP response. Required ONLY when using JSONP as the serialisation method. Please see , .\n",
            "            * *Default*: -\n",
            "            * *Example Values*: randomlygeneratedname\n",
            "        \n",
            "        \n",
            "        **Resource info**\n",
            "        \n",
            "        - **Methods**:  GET\n",
            "        - **Response formats**: json, xml, jsonp\n",
            "        \n",
            "        \n",
            "        **More info**\n",
            "        \n",
            "        https://rest.ensembl.org/documentation/info/variation_population_name\n",
            "    \n",
            "    variation_populations(*args, **kwargs)\n",
            "            Information ``GET info/variation/populations/:species``\n",
            "        \n",
            "        List all populations for a species\n",
            "        \n",
            "        \n",
            "        **Parameters**\n",
            "        \n",
            "        - Required:\n",
            "        \n",
            "          + *Name*: species\n",
            "        \n",
            "            * *Type*: String\n",
            "            * *Description*: Species name/alias\n",
            "            * *Default*: -\n",
            "            * *Example Values*: homo_sapiens\n",
            "        \n",
            "        \n",
            "        - Optional:\n",
            "        \n",
            "          + *Name*: callback\n",
            "        \n",
            "            * *Type*: String\n",
            "            * *Description*: Name of the callback subroutine to be returned by the requested JSONP response. Required ONLY when using JSONP as the serialisation method. Please see , .\n",
            "            * *Default*: -\n",
            "            * *Example Values*: randomlygeneratedname\n",
            "        \n",
            "        \n",
            "          + *Name*: filter\n",
            "        \n",
            "            * *Type*: String\n",
            "            * *Description*: Restrict populations returned to e.g. only populations with LD data. It is highly recommended to set a filter and to avoid loading the complete list of populations.\n",
            "            * *Default*: -\n",
            "            * *Example Values*: LD\n",
            "        \n",
            "        \n",
            "        **Resource info**\n",
            "        \n",
            "        - **Methods**:  GET\n",
            "        - **Response formats**: json, xml, jsonp\n",
            "        \n",
            "        \n",
            "        **More info**\n",
            "        \n",
            "        https://rest.ensembl.org/documentation/info/variation_populations\n",
            "    \n",
            "    variation_post(*args, **kwargs)\n",
            "            Variation ``POST variation/:species/``\n",
            "        \n",
            "        Uses a list of variant identifiers (e.g. rsID) to return the variation features including optional genotype, phenotype and population data\n",
            "        \n",
            "        \n",
            "        **Parameters**\n",
            "        \n",
            "        - Required:\n",
            "        \n",
            "          + *Name*: species\n",
            "        \n",
            "            * *Type*: String\n",
            "            * *Description*: Species name/alias for the whole batch of symbols\n",
            "            * *Default*: -\n",
            "            * *Example Values*: homo_sapiens, human\n",
            "        \n",
            "        \n",
            "        - Optional:\n",
            "        \n",
            "          + *Name*: callback\n",
            "        \n",
            "            * *Type*: String\n",
            "            * *Description*: Name of the callback subroutine to be returned by the requested JSONP response. Required ONLY when using JSONP as the serialisation method. Please see , .\n",
            "            * *Default*: -\n",
            "            * *Example Values*: randomlygeneratedname\n",
            "        \n",
            "        \n",
            "          + *Name*: genotypes\n",
            "        \n",
            "            * *Type*: Boolean(0,1)\n",
            "            * *Description*: Include individual genotypes\n",
            "            * *Default*: 0\n",
            "            * *Example Values*: -\n",
            "        \n",
            "        \n",
            "          + *Name*: phenotypes\n",
            "        \n",
            "            * *Type*: Boolean(0,1)\n",
            "            * *Description*: Include phenotypes\n",
            "            * *Default*: 0\n",
            "            * *Example Values*: -\n",
            "        \n",
            "        \n",
            "          + *Name*: pops\n",
            "        \n",
            "            * *Type*: Boolean(0,1)\n",
            "            * *Description*: Include population allele frequencies\n",
            "            * *Default*: 0\n",
            "            * *Example Values*: -\n",
            "        \n",
            "        \n",
            "          + *Name*: population_genotypes\n",
            "        \n",
            "            * *Type*: Boolean(0,1)\n",
            "            * *Description*: Include population genotype frequencies\n",
            "            * *Default*: 0\n",
            "            * *Example Values*: -\n",
            "        \n",
            "        \n",
            "        **Resource info**\n",
            "        \n",
            "        - **Methods**:  POST\n",
            "        - **Response formats**: json, xml, jsonp\n",
            "        - **Maximum POST size**:        200\n",
            "        \n",
            "        \n",
            "        **More info**\n",
            "        \n",
            "        https://rest.ensembl.org/documentation/info/variation_post\n",
            "    \n",
            "    vep_hgvs_get(*args, **kwargs)\n",
            "            VEP ``GET vep/:species/hgvs/:hgvs_notation``\n",
            "        \n",
            "        Fetch variant consequences based on a HGVS notation\n",
            "        \n",
            "        \n",
            "        **Parameters**\n",
            "        \n",
            "        - Required:\n",
            "        \n",
            "          + *Name*: hgvs_notation\n",
            "        \n",
            "            * *Type*: String\n",
            "            * *Description*: HGVS notation. May be genomic (g), coding (c) or protein (p), with reference to chromosome name, gene name, transcript ID or protein ID.\n",
            "            * *Default*: -\n",
            "            * *Example Values*: AGT:c.803T>C, 9:g.22125504G>C, ENST00000003084:c.1431_1433delTTC\n",
            "        \n",
            "        \n",
            "          + *Name*: species\n",
            "        \n",
            "            * *Type*: String\n",
            "            * *Description*: Species name/alias\n",
            "            * *Default*: -\n",
            "            * *Example Values*: homo_sapiens, human\n",
            "        \n",
            "        \n",
            "        - Optional:\n",
            "        \n",
            "          + *Name*: Blosum62\n",
            "        \n",
            "            * *Type*: Boolean\n",
            "            * *Description*: \n",
            "            * *Default*: 0\n",
            "            * *Example Values*: -\n",
            "        \n",
            "        \n",
            "          + *Name*: Conservation\n",
            "        \n",
            "            * *Type*: Boolean\n",
            "            * *Description*: \n",
            "            * *Default*: 0\n",
            "            * *Example Values*: -\n",
            "        \n",
            "        \n",
            "          + *Name*: GeneSplicer\n",
            "        \n",
            "            * *Type*: Boolean\n",
            "            * *Description*: \n",
            "            * *Default*: 0\n",
            "            * *Example Values*: -\n",
            "        \n",
            "        \n",
            "          + *Name*: MaxEntScan\n",
            "        \n",
            "            * *Type*: Boolean\n",
            "            * *Description*: \n",
            "            * *Default*: 0\n",
            "            * *Example Values*: -\n",
            "        \n",
            "        \n",
            "          + *Name*: Phenotypes\n",
            "        \n",
            "            * *Type*: Boolean\n",
            "            * *Description*: \n",
            "            * *Default*: 0\n",
            "            * *Example Values*: -\n",
            "        \n",
            "        \n",
            "          + *Name*: appris\n",
            "        \n",
            "            * *Type*: Boolean\n",
            "            * *Description*: Include APPRIS isoform annotation\n",
            "            * *Default*: 0\n",
            "            * *Example Values*: -\n",
            "        \n",
            "        \n",
            "          + *Name*: callback\n",
            "        \n",
            "            * *Type*: String\n",
            "            * *Description*: Name of the callback subroutine to be returned by the requested JSONP response. Required ONLY when using JSONP as the serialisation method. Please see , .\n",
            "            * *Default*: -\n",
            "            * *Example Values*: randomlygeneratedname\n",
            "        \n",
            "        \n",
            "          + *Name*: canonical\n",
            "        \n",
            "            * *Type*: Boolean\n",
            "            * *Description*: Include a flag indicating the canonical transcript for a gene\n",
            "            * *Default*: 0\n",
            "            * *Example Values*: -\n",
            "        \n",
            "        \n",
            "          + *Name*: ccds\n",
            "        \n",
            "            * *Type*: Boolean\n",
            "            * *Description*: Include CCDS transcript identifiers\n",
            "            * *Default*: 0\n",
            "            * *Example Values*: -\n",
            "        \n",
            "        \n",
            "          + *Name*: dbNSFP\n",
            "        \n",
            "            * *Type*: String\n",
            "            * *Description*: \n",
            "            * *Default*: Not used\n",
            "            * *Example Values*: LRT_pred,MutationTaster_pred\n",
            "        \n",
            "        \n",
            "          + *Name*: dbscSNV\n",
            "        \n",
            "            * *Type*: Boolean\n",
            "            * *Description*: \n",
            "            * *Default*: 0\n",
            "            * *Example Values*: -\n",
            "        \n",
            "        \n",
            "          + *Name*: distance\n",
            "        \n",
            "            * *Type*: Integer\n",
            "            * *Description*: Change the distance to transcript for which VEP assigns upstream and downstream consequences\n",
            "            * *Default*: 5000\n",
            "            * *Example Values*: -\n",
            "        \n",
            "        \n",
            "          + *Name*: domains\n",
            "        \n",
            "            * *Type*: Boolean\n",
            "            * *Description*: Include names of overlapping protein domains\n",
            "            * *Default*: 0\n",
            "            * *Example Values*: -\n",
            "        \n",
            "        \n",
            "          + *Name*: failed\n",
            "        \n",
            "            * *Type*: Boolean\n",
            "            * *Description*: \n",
            "            * *Default*: 0\n",
            "            * *Example Values*: -\n",
            "        \n",
            "        \n",
            "          + *Name*: hgvs\n",
            "        \n",
            "            * *Type*: Boolean\n",
            "            * *Description*: Include HGVS nomenclature based on Ensembl stable identifiers\n",
            "            * *Default*: 0\n",
            "            * *Example Values*: -\n",
            "        \n",
            "        \n",
            "          + *Name*: merged\n",
            "        \n",
            "            * *Type*: Boolean\n",
            "            * *Description*: Use merged Ensembl and RefSeq transcript set to report consequences (human only)\n",
            "            * *Default*: 0\n",
            "            * *Example Values*: -\n",
            "        \n",
            "        \n",
            "          + *Name*: miRNA\n",
            "        \n",
            "            * *Type*: Boolean\n",
            "            * *Description*: \n",
            "            * *Default*: 0\n",
            "            * *Example Values*: -\n",
            "        \n",
            "        \n",
            "          + *Name*: minimal\n",
            "        \n",
            "            * *Type*: Boolean\n",
            "            * *Description*: Convert alleles to their most minimal representation before consequence calculation i.e. sequence that is identical between each pair of reference and alternate alleles is trimmed off from both ends, with coordinates adjusted accordingly. Note this may lead to discrepancies between input coordinates and coordinates reported by VEP relative to transcript sequences\n",
            "            * *Default*: 0\n",
            "            * *Example Values*: -\n",
            "        \n",
            "        \n",
            "          + *Name*: numbers\n",
            "        \n",
            "            * *Type*: Boolean\n",
            "            * *Description*: Include affected exon and intron positions within the transcript\n",
            "            * *Default*: 0\n",
            "            * *Example Values*: -\n",
            "        \n",
            "        \n",
            "          + *Name*: protein\n",
            "        \n",
            "            * *Type*: Boolean\n",
            "            * *Description*: Include Ensembl protein identifiers\n",
            "            * *Default*: 0\n",
            "            * *Example Values*: -\n",
            "        \n",
            "        \n",
            "          + *Name*: refseq\n",
            "        \n",
            "            * *Type*: Boolean\n",
            "            * *Description*: Use RefSeq transcript set to report consequences (human only)\n",
            "            * *Default*: 0\n",
            "            * *Example Values*: -\n",
            "        \n",
            "        \n",
            "          + *Name*: transcript_id\n",
            "        \n",
            "            * *Type*: String\n",
            "            * *Description*: Filter results by Transcript ID\n",
            "            * *Default*: Not Used\n",
            "            * *Example Values*: -\n",
            "        \n",
            "        \n",
            "          + *Name*: tsl\n",
            "        \n",
            "            * *Type*: Boolean\n",
            "            * *Description*: Include transcript support level (TSL) annotation\n",
            "            * *Default*: 0\n",
            "            * *Example Values*: -\n",
            "        \n",
            "        \n",
            "          + *Name*: uniprot\n",
            "        \n",
            "            * *Type*: Boolean\n",
            "            * *Description*: Include best match accessions for translated protein products from three UniProt-related databases (SWISSPROT, TREMBL and UniParc)\n",
            "            * *Default*: 0\n",
            "            * *Example Values*: -\n",
            "        \n",
            "        \n",
            "          + *Name*: \n",
            "        \n",
            "            * *Type*: Boolean\n",
            "            * *Description*: \n",
            "            * *Default*: 0\n",
            "            * *Example Values*: -\n",
            "        \n",
            "        \n",
            "          + *Name*: xref_refseq\n",
            "        \n",
            "            * *Type*: Boolean\n",
            "            * *Description*: Include aligned RefSeq mRNA identifiers for transcript. NB: theRefSeq and Ensembl transcripts aligned in this way MAY NOT, AND FREQUENTLY WILL NOT, match exactly in sequence, exon structure and protein product\n",
            "            * *Default*: 0\n",
            "            * *Example Values*: -\n",
            "        \n",
            "        \n",
            "        **Resource info**\n",
            "        \n",
            "        - **Methods**:  GET\n",
            "        - **Response formats**: json, xml, jsonp\n",
            "        \n",
            "        \n",
            "        **More info**\n",
            "        \n",
            "        https://rest.ensembl.org/documentation/info/vep_hgvs_get\n",
            "    \n",
            "    vep_hgvs_post(*args, **kwargs)\n",
            "            VEP ``POST vep/:species/hgvs``\n",
            "        \n",
            "        Fetch variant consequences for multiple HGVS notations\n",
            "        \n",
            "        \n",
            "        **Parameters**\n",
            "        \n",
            "        - Required:\n",
            "        \n",
            "          + *Name*: species\n",
            "        \n",
            "            * *Type*: String\n",
            "            * *Description*: Species name/alias\n",
            "            * *Default*: -\n",
            "            * *Example Values*: homo_sapiens, human\n",
            "        \n",
            "        \n",
            "        - Optional:\n",
            "        \n",
            "          + *Name*: Blosum62\n",
            "        \n",
            "            * *Type*: Boolean\n",
            "            * *Description*: \n",
            "            * *Default*: 0\n",
            "            * *Example Values*: -\n",
            "        \n",
            "        \n",
            "          + *Name*: GeneSplicer\n",
            "        \n",
            "            * *Type*: Boolean\n",
            "            * *Description*: \n",
            "            * *Default*: 0\n",
            "            * *Example Values*: -\n",
            "        \n",
            "        \n",
            "          + *Name*: MaxEntScan\n",
            "        \n",
            "            * *Type*: Boolean\n",
            "            * *Description*: \n",
            "            * *Default*: 0\n",
            "            * *Example Values*: -\n",
            "        \n",
            "        \n",
            "          + *Name*: Phenotypes\n",
            "        \n",
            "            * *Type*: Boolean\n",
            "            * *Description*: \n",
            "            * *Default*: 0\n",
            "            * *Example Values*: -\n",
            "        \n",
            "        \n",
            "          + *Name*: appris\n",
            "        \n",
            "            * *Type*: Boolean\n",
            "            * *Description*: Include APPRIS isoform annotation\n",
            "            * *Default*: 0\n",
            "            * *Example Values*: -\n",
            "        \n",
            "        \n",
            "          + *Name*: callback\n",
            "        \n",
            "            * *Type*: String\n",
            "            * *Description*: Name of the callback subroutine to be returned by the requested JSONP response. Required ONLY when using JSONP as the serialisation method. Please see , .\n",
            "            * *Default*: -\n",
            "            * *Example Values*: randomlygeneratedname\n",
            "        \n",
            "        \n",
            "          + *Name*: canonical\n",
            "        \n",
            "            * *Type*: Boolean\n",
            "            * *Description*: Include a flag indicating the canonical transcript for a gene\n",
            "            * *Default*: 0\n",
            "            * *Example Values*: -\n",
            "        \n",
            "        \n",
            "          + *Name*: ccds\n",
            "        \n",
            "            * *Type*: Boolean\n",
            "            * *Description*: Include CCDS transcript identifiers\n",
            "            * *Default*: 0\n",
            "            * *Example Values*: -\n",
            "        \n",
            "        \n",
            "          + *Name*: dbNSFP\n",
            "        \n",
            "            * *Type*: String\n",
            "            * *Description*: \n",
            "            * *Default*: Not used\n",
            "            * *Example Values*: LRT_pred,MutationTaster_pred\n",
            "        \n",
            "        \n",
            "          + *Name*: dbscSNV\n",
            "        \n",
            "            * *Type*: Boolean\n",
            "            * *Description*: \n",
            "            * *Default*: 0\n",
            "            * *Example Values*: -\n",
            "        \n",
            "        \n",
            "          + *Name*: distance\n",
            "        \n",
            "            * *Type*: Integer\n",
            "            * *Description*: Change the distance to transcript for which VEP assigns upstream and downstream consequences\n",
            "            * *Default*: 5000\n",
            "            * *Example Values*: -\n",
            "        \n",
            "        \n",
            "          + *Name*: domains\n",
            "        \n",
            "            * *Type*: Boolean\n",
            "            * *Description*: Include names of overlapping protein domains\n",
            "            * *Default*: 0\n",
            "            * *Example Values*: -\n",
            "        \n",
            "        \n",
            "          + *Name*: failed\n",
            "        \n",
            "            * *Type*: Boolean\n",
            "            * *Description*: \n",
            "            * *Default*: 0\n",
            "            * *Example Values*: -\n",
            "        \n",
            "        \n",
            "          + *Name*: hgvs\n",
            "        \n",
            "            * *Type*: Boolean\n",
            "            * *Description*: Include HGVS nomenclature based on Ensembl stable identifiers\n",
            "            * *Default*: 0\n",
            "            * *Example Values*: -\n",
            "        \n",
            "        \n",
            "          + *Name*: merged\n",
            "        \n",
            "            * *Type*: Boolean\n",
            "            * *Description*: Use merged Ensembl and RefSeq transcript set to report consequences (human only)\n",
            "            * *Default*: 0\n",
            "            * *Example Values*: -\n",
            "        \n",
            "        \n",
            "          + *Name*: miRNA\n",
            "        \n",
            "            * *Type*: Boolean\n",
            "            * *Description*: \n",
            "            * *Default*: 0\n",
            "            * *Example Values*: -\n",
            "        \n",
            "        \n",
            "          + *Name*: minimal\n",
            "        \n",
            "            * *Type*: Boolean\n",
            "            * *Description*: Convert alleles to their most minimal representation before consequence calculation i.e. sequence that is identical between each pair of reference and alternate alleles is trimmed off from both ends, with coordinates adjusted accordingly. Note this may lead to discrepancies between input coordinates and coordinates reported by VEP relative to transcript sequences\n",
            "            * *Default*: 0\n",
            "            * *Example Values*: -\n",
            "        \n",
            "        \n",
            "          + *Name*: numbers\n",
            "        \n",
            "            * *Type*: Boolean\n",
            "            * *Description*: Include affected exon and intron positions within the transcript\n",
            "            * *Default*: 0\n",
            "            * *Example Values*: -\n",
            "        \n",
            "        \n",
            "          + *Name*: protein\n",
            "        \n",
            "            * *Type*: Boolean\n",
            "            * *Description*: Include Ensembl protein identifiers\n",
            "            * *Default*: 0\n",
            "            * *Example Values*: -\n",
            "        \n",
            "        \n",
            "          + *Name*: refseq\n",
            "        \n",
            "            * *Type*: Boolean\n",
            "            * *Description*: Use RefSeq transcript set to report consequences (human only)\n",
            "            * *Default*: 0\n",
            "            * *Example Values*: -\n",
            "        \n",
            "        \n",
            "          + *Name*: transcript_id\n",
            "        \n",
            "            * *Type*: String\n",
            "            * *Description*: Filter results by Transcript ID\n",
            "            * *Default*: Not Used\n",
            "            * *Example Values*: -\n",
            "        \n",
            "        \n",
            "          + *Name*: tsl\n",
            "        \n",
            "            * *Type*: Boolean\n",
            "            * *Description*: Include transcript support level (TSL) annotation\n",
            "            * *Default*: 0\n",
            "            * *Example Values*: -\n",
            "        \n",
            "        \n",
            "          + *Name*: uniprot\n",
            "        \n",
            "            * *Type*: Boolean\n",
            "            * *Description*: Include best match accessions for translated protein products from three UniProt-related databases (SWISSPROT, TREMBL and UniParc)\n",
            "            * *Default*: 0\n",
            "            * *Example Values*: -\n",
            "        \n",
            "        \n",
            "          + *Name*: \n",
            "        \n",
            "            * *Type*: Boolean\n",
            "            * *Description*: \n",
            "            * *Default*: 0\n",
            "            * *Example Values*: -\n",
            "        \n",
            "        \n",
            "          + *Name*: xref_refseq\n",
            "        \n",
            "            * *Type*: Boolean\n",
            "            * *Description*: Include aligned RefSeq mRNA identifiers for transcript. NB: theRefSeq and Ensembl transcripts aligned in this way MAY NOT, AND FREQUENTLY WILL NOT, match exactly in sequence, exon structure and protein product\n",
            "            * *Default*: 0\n",
            "            * *Example Values*: -\n",
            "        \n",
            "        \n",
            "        **Resource info**\n",
            "        \n",
            "        - **Methods**:  POST\n",
            "        - **Response formats**: json, xml, jsonp\n",
            "        - **Maximum POST size**:        200\n",
            "        \n",
            "        \n",
            "        **More info**\n",
            "        \n",
            "        https://rest.ensembl.org/documentation/info/vep_hgvs_post\n",
            "    \n",
            "    vep_id_get(*args, **kwargs)\n",
            "            VEP ``GET vep/:species/id/:id``\n",
            "        \n",
            "        Fetch variant consequences based on a variant identifier\n",
            "        \n",
            "        \n",
            "        **Parameters**\n",
            "        \n",
            "        - Required:\n",
            "        \n",
            "          + *Name*: id\n",
            "        \n",
            "            * *Type*: String\n",
            "            * *Description*: Query ID. Supports dbSNP, COSMIC and HGMD identifiers\n",
            "            * *Default*: -\n",
            "            * *Example Values*: rs56116432, COSM476\n",
            "        \n",
            "        \n",
            "          + *Name*: species\n",
            "        \n",
            "            * *Type*: String\n",
            "            * *Description*: Species name/alias\n",
            "            * *Default*: -\n",
            "            * *Example Values*: homo_sapiens, human\n",
            "        \n",
            "        \n",
            "        - Optional:\n",
            "        \n",
            "          + *Name*: Blosum62\n",
            "        \n",
            "            * *Type*: Boolean\n",
            "            * *Description*: \n",
            "            * *Default*: 0\n",
            "            * *Example Values*: -\n",
            "        \n",
            "        \n",
            "          + *Name*: Conservation\n",
            "        \n",
            "            * *Type*: Boolean\n",
            "            * *Description*: \n",
            "            * *Default*: 0\n",
            "            * *Example Values*: -\n",
            "        \n",
            "        \n",
            "          + *Name*: GeneSplicer\n",
            "        \n",
            "            * *Type*: Boolean\n",
            "            * *Description*: \n",
            "            * *Default*: 0\n",
            "            * *Example Values*: -\n",
            "        \n",
            "        \n",
            "          + *Name*: MaxEntScan\n",
            "        \n",
            "            * *Type*: Boolean\n",
            "            * *Description*: \n",
            "            * *Default*: 0\n",
            "            * *Example Values*: -\n",
            "        \n",
            "        \n",
            "          + *Name*: Phenotypes\n",
            "        \n",
            "            * *Type*: Boolean\n",
            "            * *Description*: \n",
            "            * *Default*: 0\n",
            "            * *Example Values*: -\n",
            "        \n",
            "        \n",
            "          + *Name*: appris\n",
            "        \n",
            "            * *Type*: Boolean\n",
            "            * *Description*: Include APPRIS isoform annotation\n",
            "            * *Default*: 0\n",
            "            * *Example Values*: -\n",
            "        \n",
            "        \n",
            "          + *Name*: callback\n",
            "        \n",
            "            * *Type*: String\n",
            "            * *Description*: Name of the callback subroutine to be returned by the requested JSONP response. Required ONLY when using JSONP as the serialisation method. Please see , .\n",
            "            * *Default*: -\n",
            "            * *Example Values*: randomlygeneratedname\n",
            "        \n",
            "        \n",
            "          + *Name*: canonical\n",
            "        \n",
            "            * *Type*: Boolean\n",
            "            * *Description*: Include a flag indicating the canonical transcript for a gene\n",
            "            * *Default*: 0\n",
            "            * *Example Values*: -\n",
            "        \n",
            "        \n",
            "          + *Name*: ccds\n",
            "        \n",
            "            * *Type*: Boolean\n",
            "            * *Description*: Include CCDS transcript identifiers\n",
            "            * *Default*: 0\n",
            "            * *Example Values*: -\n",
            "        \n",
            "        \n",
            "          + *Name*: dbNSFP\n",
            "        \n",
            "            * *Type*: String\n",
            "            * *Description*: \n",
            "            * *Default*: Not used\n",
            "            * *Example Values*: LRT_pred,MutationTaster_pred\n",
            "        \n",
            "        \n",
            "          + *Name*: dbscSNV\n",
            "        \n",
            "            * *Type*: Boolean\n",
            "            * *Description*: \n",
            "            * *Default*: 0\n",
            "            * *Example Values*: -\n",
            "        \n",
            "        \n",
            "          + *Name*: distance\n",
            "        \n",
            "            * *Type*: Integer\n",
            "            * *Description*: Change the distance to transcript for which VEP assigns upstream and downstream consequences\n",
            "            * *Default*: 5000\n",
            "            * *Example Values*: -\n",
            "        \n",
            "        \n",
            "          + *Name*: domains\n",
            "        \n",
            "            * *Type*: Boolean\n",
            "            * *Description*: Include names of overlapping protein domains\n",
            "            * *Default*: 0\n",
            "            * *Example Values*: -\n",
            "        \n",
            "        \n",
            "          + *Name*: failed\n",
            "        \n",
            "            * *Type*: Boolean\n",
            "            * *Description*: \n",
            "            * *Default*: 0\n",
            "            * *Example Values*: -\n",
            "        \n",
            "        \n",
            "          + *Name*: hgvs\n",
            "        \n",
            "            * *Type*: Boolean\n",
            "            * *Description*: Include HGVS nomenclature based on Ensembl stable identifiers\n",
            "            * *Default*: 0\n",
            "            * *Example Values*: -\n",
            "        \n",
            "        \n",
            "          + *Name*: merged\n",
            "        \n",
            "            * *Type*: Boolean\n",
            "            * *Description*: Use merged Ensembl and RefSeq transcript set to report consequences (human only)\n",
            "            * *Default*: 0\n",
            "            * *Example Values*: -\n",
            "        \n",
            "        \n",
            "          + *Name*: miRNA\n",
            "        \n",
            "            * *Type*: Boolean\n",
            "            * *Description*: \n",
            "            * *Default*: 0\n",
            "            * *Example Values*: -\n",
            "        \n",
            "        \n",
            "          + *Name*: minimal\n",
            "        \n",
            "            * *Type*: Boolean\n",
            "            * *Description*: Convert alleles to their most minimal representation before consequence calculation i.e. sequence that is identical between each pair of reference and alternate alleles is trimmed off from both ends, with coordinates adjusted accordingly. Note this may lead to discrepancies between input coordinates and coordinates reported by VEP relative to transcript sequences\n",
            "            * *Default*: 0\n",
            "            * *Example Values*: -\n",
            "        \n",
            "        \n",
            "          + *Name*: numbers\n",
            "        \n",
            "            * *Type*: Boolean\n",
            "            * *Description*: Include affected exon and intron positions within the transcript\n",
            "            * *Default*: 0\n",
            "            * *Example Values*: -\n",
            "        \n",
            "        \n",
            "          + *Name*: protein\n",
            "        \n",
            "            * *Type*: Boolean\n",
            "            * *Description*: Include Ensembl protein identifiers\n",
            "            * *Default*: 0\n",
            "            * *Example Values*: -\n",
            "        \n",
            "        \n",
            "          + *Name*: refseq\n",
            "        \n",
            "            * *Type*: Boolean\n",
            "            * *Description*: Use RefSeq transcript set to report consequences (human only)\n",
            "            * *Default*: 0\n",
            "            * *Example Values*: -\n",
            "        \n",
            "        \n",
            "          + *Name*: transcript_id\n",
            "        \n",
            "            * *Type*: String\n",
            "            * *Description*: Filter results by Transcript ID\n",
            "            * *Default*: Not Used\n",
            "            * *Example Values*: -\n",
            "        \n",
            "        \n",
            "          + *Name*: tsl\n",
            "        \n",
            "            * *Type*: Boolean\n",
            "            * *Description*: Include transcript support level (TSL) annotation\n",
            "            * *Default*: 0\n",
            "            * *Example Values*: -\n",
            "        \n",
            "        \n",
            "          + *Name*: uniprot\n",
            "        \n",
            "            * *Type*: Boolean\n",
            "            * *Description*: Include best match accessions for translated protein products from three UniProt-related databases (SWISSPROT, TREMBL and UniParc)\n",
            "            * *Default*: 0\n",
            "            * *Example Values*: -\n",
            "        \n",
            "        \n",
            "          + *Name*: \n",
            "        \n",
            "            * *Type*: Boolean\n",
            "            * *Description*: \n",
            "            * *Default*: 0\n",
            "            * *Example Values*: -\n",
            "        \n",
            "        \n",
            "          + *Name*: xref_refseq\n",
            "        \n",
            "            * *Type*: Boolean\n",
            "            * *Description*: Include aligned RefSeq mRNA identifiers for transcript. NB: theRefSeq and Ensembl transcripts aligned in this way MAY NOT, AND FREQUENTLY WILL NOT, match exactly in sequence, exon structure and protein product\n",
            "            * *Default*: 0\n",
            "            * *Example Values*: -\n",
            "        \n",
            "        \n",
            "        **Resource info**\n",
            "        \n",
            "        - **Methods**:  GET\n",
            "        - **Response formats**: json, xml, jsonp\n",
            "        \n",
            "        \n",
            "        **More info**\n",
            "        \n",
            "        https://rest.ensembl.org/documentation/info/vep_id_get\n",
            "    \n",
            "    vep_id_post(*args, **kwargs)\n",
            "            VEP ``POST vep/:species/id``\n",
            "        \n",
            "        Fetch variant consequences for multiple ids\n",
            "        \n",
            "        \n",
            "        **Parameters**\n",
            "        \n",
            "        - Required:\n",
            "        \n",
            "          + *Name*: species\n",
            "        \n",
            "            * *Type*: String\n",
            "            * *Description*: Species name/alias\n",
            "            * *Default*: -\n",
            "            * *Example Values*: homo_sapiens, human\n",
            "        \n",
            "        \n",
            "        - Optional:\n",
            "        \n",
            "          + *Name*: Blosum62\n",
            "        \n",
            "            * *Type*: Boolean\n",
            "            * *Description*: \n",
            "            * *Default*: 0\n",
            "            * *Example Values*: -\n",
            "        \n",
            "        \n",
            "          + *Name*: GeneSplicer\n",
            "        \n",
            "            * *Type*: Boolean\n",
            "            * *Description*: \n",
            "            * *Default*: 0\n",
            "            * *Example Values*: -\n",
            "        \n",
            "        \n",
            "          + *Name*: MaxEntScan\n",
            "        \n",
            "            * *Type*: Boolean\n",
            "            * *Description*: \n",
            "            * *Default*: 0\n",
            "            * *Example Values*: -\n",
            "        \n",
            "        \n",
            "          + *Name*: Phenotypes\n",
            "        \n",
            "            * *Type*: Boolean\n",
            "            * *Description*: \n",
            "            * *Default*: 0\n",
            "            * *Example Values*: -\n",
            "        \n",
            "        \n",
            "          + *Name*: appris\n",
            "        \n",
            "            * *Type*: Boolean\n",
            "            * *Description*: Include APPRIS isoform annotation\n",
            "            * *Default*: 0\n",
            "            * *Example Values*: -\n",
            "        \n",
            "        \n",
            "          + *Name*: callback\n",
            "        \n",
            "            * *Type*: String\n",
            "            * *Description*: Name of the callback subroutine to be returned by the requested JSONP response. Required ONLY when using JSONP as the serialisation method. Please see , .\n",
            "            * *Default*: -\n",
            "            * *Example Values*: randomlygeneratedname\n",
            "        \n",
            "        \n",
            "          + *Name*: canonical\n",
            "        \n",
            "            * *Type*: Boolean\n",
            "            * *Description*: Include a flag indicating the canonical transcript for a gene\n",
            "            * *Default*: 0\n",
            "            * *Example Values*: -\n",
            "        \n",
            "        \n",
            "          + *Name*: ccds\n",
            "        \n",
            "            * *Type*: Boolean\n",
            "            * *Description*: Include CCDS transcript identifiers\n",
            "            * *Default*: 0\n",
            "            * *Example Values*: -\n",
            "        \n",
            "        \n",
            "          + *Name*: dbNSFP\n",
            "        \n",
            "            * *Type*: String\n",
            "            * *Description*: \n",
            "            * *Default*: Not used\n",
            "            * *Example Values*: LRT_pred,MutationTaster_pred\n",
            "        \n",
            "        \n",
            "          + *Name*: dbscSNV\n",
            "        \n",
            "            * *Type*: Boolean\n",
            "            * *Description*: \n",
            "            * *Default*: 0\n",
            "            * *Example Values*: -\n",
            "        \n",
            "        \n",
            "          + *Name*: distance\n",
            "        \n",
            "            * *Type*: Integer\n",
            "            * *Description*: Change the distance to transcript for which VEP assigns upstream and downstream consequences\n",
            "            * *Default*: 5000\n",
            "            * *Example Values*: -\n",
            "        \n",
            "        \n",
            "          + *Name*: domains\n",
            "        \n",
            "            * *Type*: Boolean\n",
            "            * *Description*: Include names of overlapping protein domains\n",
            "            * *Default*: 0\n",
            "            * *Example Values*: -\n",
            "        \n",
            "        \n",
            "          + *Name*: failed\n",
            "        \n",
            "            * *Type*: Boolean\n",
            "            * *Description*: \n",
            "            * *Default*: 0\n",
            "            * *Example Values*: -\n",
            "        \n",
            "        \n",
            "          + *Name*: hgvs\n",
            "        \n",
            "            * *Type*: Boolean\n",
            "            * *Description*: Include HGVS nomenclature based on Ensembl stable identifiers\n",
            "            * *Default*: 0\n",
            "            * *Example Values*: -\n",
            "        \n",
            "        \n",
            "          + *Name*: merged\n",
            "        \n",
            "            * *Type*: Boolean\n",
            "            * *Description*: Use merged Ensembl and RefSeq transcript set to report consequences (human only)\n",
            "            * *Default*: 0\n",
            "            * *Example Values*: -\n",
            "        \n",
            "        \n",
            "          + *Name*: miRNA\n",
            "        \n",
            "            * *Type*: Boolean\n",
            "            * *Description*: \n",
            "            * *Default*: 0\n",
            "            * *Example Values*: -\n",
            "        \n",
            "        \n",
            "          + *Name*: minimal\n",
            "        \n",
            "            * *Type*: Boolean\n",
            "            * *Description*: Convert alleles to their most minimal representation before consequence calculation i.e. sequence that is identical between each pair of reference and alternate alleles is trimmed off from both ends, with coordinates adjusted accordingly. Note this may lead to discrepancies between input coordinates and coordinates reported by VEP relative to transcript sequences\n",
            "            * *Default*: 0\n",
            "            * *Example Values*: -\n",
            "        \n",
            "        \n",
            "          + *Name*: numbers\n",
            "        \n",
            "            * *Type*: Boolean\n",
            "            * *Description*: Include affected exon and intron positions within the transcript\n",
            "            * *Default*: 0\n",
            "            * *Example Values*: -\n",
            "        \n",
            "        \n",
            "          + *Name*: protein\n",
            "        \n",
            "            * *Type*: Boolean\n",
            "            * *Description*: Include Ensembl protein identifiers\n",
            "            * *Default*: 0\n",
            "            * *Example Values*: -\n",
            "        \n",
            "        \n",
            "          + *Name*: refseq\n",
            "        \n",
            "            * *Type*: Boolean\n",
            "            * *Description*: Use RefSeq transcript set to report consequences (human only)\n",
            "            * *Default*: 0\n",
            "            * *Example Values*: -\n",
            "        \n",
            "        \n",
            "          + *Name*: transcript_id\n",
            "        \n",
            "            * *Type*: String\n",
            "            * *Description*: Filter results by Transcript ID\n",
            "            * *Default*: Not Used\n",
            "            * *Example Values*: -\n",
            "        \n",
            "        \n",
            "          + *Name*: tsl\n",
            "        \n",
            "            * *Type*: Boolean\n",
            "            * *Description*: Include transcript support level (TSL) annotation\n",
            "            * *Default*: 0\n",
            "            * *Example Values*: -\n",
            "        \n",
            "        \n",
            "          + *Name*: uniprot\n",
            "        \n",
            "            * *Type*: Boolean\n",
            "            * *Description*: Include best match accessions for translated protein products from three UniProt-related databases (SWISSPROT, TREMBL and UniParc)\n",
            "            * *Default*: 0\n",
            "            * *Example Values*: -\n",
            "        \n",
            "        \n",
            "          + *Name*: \n",
            "        \n",
            "            * *Type*: Boolean\n",
            "            * *Description*: \n",
            "            * *Default*: 0\n",
            "            * *Example Values*: -\n",
            "        \n",
            "        \n",
            "          + *Name*: xref_refseq\n",
            "        \n",
            "            * *Type*: Boolean\n",
            "            * *Description*: Include aligned RefSeq mRNA identifiers for transcript. NB: theRefSeq and Ensembl transcripts aligned in this way MAY NOT, AND FREQUENTLY WILL NOT, match exactly in sequence, exon structure and protein product\n",
            "            * *Default*: 0\n",
            "            * *Example Values*: -\n",
            "        \n",
            "        \n",
            "        **Resource info**\n",
            "        \n",
            "        - **Methods**:  POST\n",
            "        - **Response formats**: json, xml, jsonp\n",
            "        - **Maximum POST size**:        200\n",
            "        \n",
            "        \n",
            "        **More info**\n",
            "        \n",
            "        https://rest.ensembl.org/documentation/info/vep_id_post\n",
            "    \n",
            "    vep_region_get(*args, **kwargs)\n",
            "            VEP ``GET vep/:species/region/:region/:allele/``\n",
            "        \n",
            "        Fetch variant consequences\n",
            "        \n",
            "        \n",
            "        **Parameters**\n",
            "        \n",
            "        - Required:\n",
            "        \n",
            "          + *Name*: allele\n",
            "        \n",
            "            * *Type*: String\n",
            "            * *Description*: Variation allele\n",
            "            * *Default*: -\n",
            "            * *Example Values*: C, DUP\n",
            "        \n",
            "        \n",
            "          + *Name*: region\n",
            "        \n",
            "            * *Type*: String\n",
            "            * *Description*: Query region. We only support the current assembly\n",
            "            * *Default*: -\n",
            "            * *Example Values*: 9:22125503-22125502:1, 7:100318423-100321323:1\n",
            "        \n",
            "        \n",
            "          + *Name*: species\n",
            "        \n",
            "            * *Type*: String\n",
            "            * *Description*: Species name/alias\n",
            "            * *Default*: -\n",
            "            * *Example Values*: homo_sapiens, human\n",
            "        \n",
            "        \n",
            "        - Optional:\n",
            "        \n",
            "          + *Name*: Blosum62\n",
            "        \n",
            "            * *Type*: Boolean\n",
            "            * *Description*: \n",
            "            * *Default*: 0\n",
            "            * *Example Values*: -\n",
            "        \n",
            "        \n",
            "          + *Name*: Conservation\n",
            "        \n",
            "            * *Type*: Boolean\n",
            "            * *Description*: \n",
            "            * *Default*: 0\n",
            "            * *Example Values*: -\n",
            "        \n",
            "        \n",
            "          + *Name*: GeneSplicer\n",
            "        \n",
            "            * *Type*: Boolean\n",
            "            * *Description*: \n",
            "            * *Default*: 0\n",
            "            * *Example Values*: -\n",
            "        \n",
            "        \n",
            "          + *Name*: MaxEntScan\n",
            "        \n",
            "            * *Type*: Boolean\n",
            "            * *Description*: \n",
            "            * *Default*: 0\n",
            "            * *Example Values*: -\n",
            "        \n",
            "        \n",
            "          + *Name*: Phenotypes\n",
            "        \n",
            "            * *Type*: Boolean\n",
            "            * *Description*: \n",
            "            * *Default*: 0\n",
            "            * *Example Values*: -\n",
            "        \n",
            "        \n",
            "          + *Name*: appris\n",
            "        \n",
            "            * *Type*: Boolean\n",
            "            * *Description*: Include APPRIS isoform annotation\n",
            "            * *Default*: 0\n",
            "            * *Example Values*: -\n",
            "        \n",
            "        \n",
            "          + *Name*: callback\n",
            "        \n",
            "            * *Type*: String\n",
            "            * *Description*: Name of the callback subroutine to be returned by the requested JSONP response. Required ONLY when using JSONP as the serialisation method. Please see , .\n",
            "            * *Default*: -\n",
            "            * *Example Values*: randomlygeneratedname\n",
            "        \n",
            "        \n",
            "          + *Name*: canonical\n",
            "        \n",
            "            * *Type*: Boolean\n",
            "            * *Description*: Include a flag indicating the canonical transcript for a gene\n",
            "            * *Default*: 0\n",
            "            * *Example Values*: -\n",
            "        \n",
            "        \n",
            "          + *Name*: ccds\n",
            "        \n",
            "            * *Type*: Boolean\n",
            "            * *Description*: Include CCDS transcript identifiers\n",
            "            * *Default*: 0\n",
            "            * *Example Values*: -\n",
            "        \n",
            "        \n",
            "          + *Name*: dbNSFP\n",
            "        \n",
            "            * *Type*: String\n",
            "            * *Description*: \n",
            "            * *Default*: Not used\n",
            "            * *Example Values*: LRT_pred,MutationTaster_pred\n",
            "        \n",
            "        \n",
            "          + *Name*: dbscSNV\n",
            "        \n",
            "            * *Type*: Boolean\n",
            "            * *Description*: \n",
            "            * *Default*: 0\n",
            "            * *Example Values*: -\n",
            "        \n",
            "        \n",
            "          + *Name*: distance\n",
            "        \n",
            "            * *Type*: Integer\n",
            "            * *Description*: Change the distance to transcript for which VEP assigns upstream and downstream consequences\n",
            "            * *Default*: 5000\n",
            "            * *Example Values*: -\n",
            "        \n",
            "        \n",
            "          + *Name*: domains\n",
            "        \n",
            "            * *Type*: Boolean\n",
            "            * *Description*: Include names of overlapping protein domains\n",
            "            * *Default*: 0\n",
            "            * *Example Values*: -\n",
            "        \n",
            "        \n",
            "          + *Name*: failed\n",
            "        \n",
            "            * *Type*: Boolean\n",
            "            * *Description*: \n",
            "            * *Default*: 0\n",
            "            * *Example Values*: -\n",
            "        \n",
            "        \n",
            "          + *Name*: hgvs\n",
            "        \n",
            "            * *Type*: Boolean\n",
            "            * *Description*: Include HGVS nomenclature based on Ensembl stable identifiers\n",
            "            * *Default*: 0\n",
            "            * *Example Values*: -\n",
            "        \n",
            "        \n",
            "          + *Name*: merged\n",
            "        \n",
            "            * *Type*: Boolean\n",
            "            * *Description*: Use merged Ensembl and RefSeq transcript set to report consequences (human only)\n",
            "            * *Default*: 0\n",
            "            * *Example Values*: -\n",
            "        \n",
            "        \n",
            "          + *Name*: miRNA\n",
            "        \n",
            "            * *Type*: Boolean\n",
            "            * *Description*: \n",
            "            * *Default*: 0\n",
            "            * *Example Values*: -\n",
            "        \n",
            "        \n",
            "          + *Name*: minimal\n",
            "        \n",
            "            * *Type*: Boolean\n",
            "            * *Description*: Convert alleles to their most minimal representation before consequence calculation i.e. sequence that is identical between each pair of reference and alternate alleles is trimmed off from both ends, with coordinates adjusted accordingly. Note this may lead to discrepancies between input coordinates and coordinates reported by VEP relative to transcript sequences\n",
            "            * *Default*: 0\n",
            "            * *Example Values*: -\n",
            "        \n",
            "        \n",
            "          + *Name*: numbers\n",
            "        \n",
            "            * *Type*: Boolean\n",
            "            * *Description*: Include affected exon and intron positions within the transcript\n",
            "            * *Default*: 0\n",
            "            * *Example Values*: -\n",
            "        \n",
            "        \n",
            "          + *Name*: protein\n",
            "        \n",
            "            * *Type*: Boolean\n",
            "            * *Description*: Include Ensembl protein identifiers\n",
            "            * *Default*: 0\n",
            "            * *Example Values*: -\n",
            "        \n",
            "        \n",
            "          + *Name*: refseq\n",
            "        \n",
            "            * *Type*: Boolean\n",
            "            * *Description*: Use RefSeq transcript set to report consequences (human only)\n",
            "            * *Default*: 0\n",
            "            * *Example Values*: -\n",
            "        \n",
            "        \n",
            "          + *Name*: transcript_id\n",
            "        \n",
            "            * *Type*: String\n",
            "            * *Description*: Filter results by Transcript ID\n",
            "            * *Default*: Not Used\n",
            "            * *Example Values*: -\n",
            "        \n",
            "        \n",
            "          + *Name*: tsl\n",
            "        \n",
            "            * *Type*: Boolean\n",
            "            * *Description*: Include transcript support level (TSL) annotation\n",
            "            * *Default*: 0\n",
            "            * *Example Values*: -\n",
            "        \n",
            "        \n",
            "          + *Name*: uniprot\n",
            "        \n",
            "            * *Type*: Boolean\n",
            "            * *Description*: Include best match accessions for translated protein products from three UniProt-related databases (SWISSPROT, TREMBL and UniParc)\n",
            "            * *Default*: 0\n",
            "            * *Example Values*: -\n",
            "        \n",
            "        \n",
            "          + *Name*: \n",
            "        \n",
            "            * *Type*: Boolean\n",
            "            * *Description*: \n",
            "            * *Default*: 0\n",
            "            * *Example Values*: -\n",
            "        \n",
            "        \n",
            "          + *Name*: xref_refseq\n",
            "        \n",
            "            * *Type*: Boolean\n",
            "            * *Description*: Include aligned RefSeq mRNA identifiers for transcript. NB: theRefSeq and Ensembl transcripts aligned in this way MAY NOT, AND FREQUENTLY WILL NOT, match exactly in sequence, exon structure and protein product\n",
            "            * *Default*: 0\n",
            "            * *Example Values*: -\n",
            "        \n",
            "        \n",
            "        **Resource info**\n",
            "        \n",
            "        - **Methods**:  GET\n",
            "        - **Response formats**: json, xml, jsonp\n",
            "        \n",
            "        \n",
            "        **More info**\n",
            "        \n",
            "        https://rest.ensembl.org/documentation/info/vep_region_get\n",
            "    \n",
            "    vep_region_post(*args, **kwargs)\n",
            "            VEP ``POST vep/:species/region``\n",
            "        \n",
            "        Fetch variant consequences for multiple regions\n",
            "        \n",
            "        \n",
            "        **Parameters**\n",
            "        \n",
            "        - Required:\n",
            "        \n",
            "          + *Name*: species\n",
            "        \n",
            "            * *Type*: String\n",
            "            * *Description*: Species name/alias\n",
            "            * *Default*: -\n",
            "            * *Example Values*: homo_sapiens, human\n",
            "        \n",
            "        \n",
            "        - Optional:\n",
            "        \n",
            "          + *Name*: Blosum62\n",
            "        \n",
            "            * *Type*: Boolean\n",
            "            * *Description*: \n",
            "            * *Default*: 0\n",
            "            * *Example Values*: -\n",
            "        \n",
            "        \n",
            "          + *Name*: GeneSplicer\n",
            "        \n",
            "            * *Type*: Boolean\n",
            "            * *Description*: \n",
            "            * *Default*: 0\n",
            "            * *Example Values*: -\n",
            "        \n",
            "        \n",
            "          + *Name*: MaxEntScan\n",
            "        \n",
            "            * *Type*: Boolean\n",
            "            * *Description*: \n",
            "            * *Default*: 0\n",
            "            * *Example Values*: -\n",
            "        \n",
            "        \n",
            "          + *Name*: Phenotypes\n",
            "        \n",
            "            * *Type*: Boolean\n",
            "            * *Description*: \n",
            "            * *Default*: 0\n",
            "            * *Example Values*: -\n",
            "        \n",
            "        \n",
            "          + *Name*: appris\n",
            "        \n",
            "            * *Type*: Boolean\n",
            "            * *Description*: Include APPRIS isoform annotation\n",
            "            * *Default*: 0\n",
            "            * *Example Values*: -\n",
            "        \n",
            "        \n",
            "          + *Name*: callback\n",
            "        \n",
            "            * *Type*: String\n",
            "            * *Description*: Name of the callback subroutine to be returned by the requested JSONP response. Required ONLY when using JSONP as the serialisation method. Please see , .\n",
            "            * *Default*: -\n",
            "            * *Example Values*: randomlygeneratedname\n",
            "        \n",
            "        \n",
            "          + *Name*: canonical\n",
            "        \n",
            "            * *Type*: Boolean\n",
            "            * *Description*: Include a flag indicating the canonical transcript for a gene\n",
            "            * *Default*: 0\n",
            "            * *Example Values*: -\n",
            "        \n",
            "        \n",
            "          + *Name*: ccds\n",
            "        \n",
            "            * *Type*: Boolean\n",
            "            * *Description*: Include CCDS transcript identifiers\n",
            "            * *Default*: 0\n",
            "            * *Example Values*: -\n",
            "        \n",
            "        \n",
            "          + *Name*: dbNSFP\n",
            "        \n",
            "            * *Type*: String\n",
            "            * *Description*: \n",
            "            * *Default*: Not used\n",
            "            * *Example Values*: LRT_pred,MutationTaster_pred\n",
            "        \n",
            "        \n",
            "          + *Name*: dbscSNV\n",
            "        \n",
            "            * *Type*: Boolean\n",
            "            * *Description*: \n",
            "            * *Default*: 0\n",
            "            * *Example Values*: -\n",
            "        \n",
            "        \n",
            "          + *Name*: distance\n",
            "        \n",
            "            * *Type*: Integer\n",
            "            * *Description*: Change the distance to transcript for which VEP assigns upstream and downstream consequences\n",
            "            * *Default*: 5000\n",
            "            * *Example Values*: -\n",
            "        \n",
            "        \n",
            "          + *Name*: domains\n",
            "        \n",
            "            * *Type*: Boolean\n",
            "            * *Description*: Include names of overlapping protein domains\n",
            "            * *Default*: 0\n",
            "            * *Example Values*: -\n",
            "        \n",
            "        \n",
            "          + *Name*: failed\n",
            "        \n",
            "            * *Type*: Boolean\n",
            "            * *Description*: \n",
            "            * *Default*: 0\n",
            "            * *Example Values*: -\n",
            "        \n",
            "        \n",
            "          + *Name*: hgvs\n",
            "        \n",
            "            * *Type*: Boolean\n",
            "            * *Description*: Include HGVS nomenclature based on Ensembl stable identifiers\n",
            "            * *Default*: 0\n",
            "            * *Example Values*: -\n",
            "        \n",
            "        \n",
            "          + *Name*: merged\n",
            "        \n",
            "            * *Type*: Boolean\n",
            "            * *Description*: Use merged Ensembl and RefSeq transcript set to report consequences (human only)\n",
            "            * *Default*: 0\n",
            "            * *Example Values*: -\n",
            "        \n",
            "        \n",
            "          + *Name*: miRNA\n",
            "        \n",
            "            * *Type*: Boolean\n",
            "            * *Description*: \n",
            "            * *Default*: 0\n",
            "            * *Example Values*: -\n",
            "        \n",
            "        \n",
            "          + *Name*: minimal\n",
            "        \n",
            "            * *Type*: Boolean\n",
            "            * *Description*: Convert alleles to their most minimal representation before consequence calculation i.e. sequence that is identical between each pair of reference and alternate alleles is trimmed off from both ends, with coordinates adjusted accordingly. Note this may lead to discrepancies between input coordinates and coordinates reported by VEP relative to transcript sequences\n",
            "            * *Default*: 0\n",
            "            * *Example Values*: -\n",
            "        \n",
            "        \n",
            "          + *Name*: numbers\n",
            "        \n",
            "            * *Type*: Boolean\n",
            "            * *Description*: Include affected exon and intron positions within the transcript\n",
            "            * *Default*: 0\n",
            "            * *Example Values*: -\n",
            "        \n",
            "        \n",
            "          + *Name*: protein\n",
            "        \n",
            "            * *Type*: Boolean\n",
            "            * *Description*: Include Ensembl protein identifiers\n",
            "            * *Default*: 0\n",
            "            * *Example Values*: -\n",
            "        \n",
            "        \n",
            "          + *Name*: refseq\n",
            "        \n",
            "            * *Type*: Boolean\n",
            "            * *Description*: Use RefSeq transcript set to report consequences (human only)\n",
            "            * *Default*: 0\n",
            "            * *Example Values*: -\n",
            "        \n",
            "        \n",
            "          + *Name*: transcript_id\n",
            "        \n",
            "            * *Type*: String\n",
            "            * *Description*: Filter results by Transcript ID\n",
            "            * *Default*: Not Used\n",
            "            * *Example Values*: -\n",
            "        \n",
            "        \n",
            "          + *Name*: tsl\n",
            "        \n",
            "            * *Type*: Boolean\n",
            "            * *Description*: Include transcript support level (TSL) annotation\n",
            "            * *Default*: 0\n",
            "            * *Example Values*: -\n",
            "        \n",
            "        \n",
            "          + *Name*: uniprot\n",
            "        \n",
            "            * *Type*: Boolean\n",
            "            * *Description*: Include best match accessions for translated protein products from three UniProt-related databases (SWISSPROT, TREMBL and UniParc)\n",
            "            * *Default*: 0\n",
            "            * *Example Values*: -\n",
            "        \n",
            "        \n",
            "          + *Name*: \n",
            "        \n",
            "            * *Type*: Boolean\n",
            "            * *Description*: \n",
            "            * *Default*: 0\n",
            "            * *Example Values*: -\n",
            "        \n",
            "        \n",
            "          + *Name*: xref_refseq\n",
            "        \n",
            "            * *Type*: Boolean\n",
            "            * *Description*: Include aligned RefSeq mRNA identifiers for transcript. NB: theRefSeq and Ensembl transcripts aligned in this way MAY NOT, AND FREQUENTLY WILL NOT, match exactly in sequence, exon structure and protein product\n",
            "            * *Default*: 0\n",
            "            * *Example Values*: -\n",
            "        \n",
            "        \n",
            "        **Resource info**\n",
            "        \n",
            "        - **Methods**:  POST\n",
            "        - **Response formats**: json, xml, jsonp\n",
            "        - **Maximum POST size**:        200\n",
            "        \n",
            "        \n",
            "        **More info**\n",
            "        \n",
            "        https://rest.ensembl.org/documentation/info/vep_region_post\n",
            "    \n",
            "    xref_external(*args, **kwargs)\n",
            "            Cross References ``GET xrefs/symbol/:species/:symbol``\n",
            "        \n",
            "        Looks up an external symbol and returns all Ensembl objects linked to it. This can be a display name for a gene/transcript/translation, a synonym or an externally linked reference. If a gene's transcript is linked to the supplied symbol the service will return both gene and transcript (it supports transient links).\n",
            "        \n",
            "        \n",
            "        **Parameters**\n",
            "        \n",
            "        - Required:\n",
            "        \n",
            "          + *Name*: species\n",
            "        \n",
            "            * *Type*: String\n",
            "            * *Description*: Species name/alias\n",
            "            * *Default*: -\n",
            "            * *Example Values*: homo_sapiens, human\n",
            "        \n",
            "        \n",
            "          + *Name*: symbol\n",
            "        \n",
            "            * *Type*: String\n",
            "            * *Description*: Symbol or display name of a gene\n",
            "            * *Default*: -\n",
            "            * *Example Values*: BRCA2\n",
            "        \n",
            "        \n",
            "        - Optional:\n",
            "        \n",
            "          + *Name*: callback\n",
            "        \n",
            "            * *Type*: String\n",
            "            * *Description*: Name of the callback subroutine to be returned by the requested JSONP response. Required ONLY when using JSONP as the serialisation method. Please see , .\n",
            "            * *Default*: -\n",
            "            * *Example Values*: randomlygeneratedname\n",
            "        \n",
            "        \n",
            "          + *Name*: db_type\n",
            "        \n",
            "            * *Type*: String\n",
            "            * *Description*: Restrict the search to a database other than the default. Useful if you need to use a DB other than core\n",
            "            * *Default*: core\n",
            "            * *Example Values*: core, otherfeatures\n",
            "        \n",
            "        \n",
            "          + *Name*: external_db\n",
            "        \n",
            "            * *Type*: String\n",
            "            * *Description*: Filter by external database\n",
            "            * *Default*: -\n",
            "            * *Example Values*: HGNC\n",
            "        \n",
            "        \n",
            "          + *Name*: object_type\n",
            "        \n",
            "            * *Type*: String\n",
            "            * *Description*: Filter by feature type\n",
            "            * *Default*: -\n",
            "            * *Example Values*: gene, transcript\n",
            "        \n",
            "        \n",
            "        **Resource info**\n",
            "        \n",
            "        - **Methods**:  GET\n",
            "        - **Response formats**: json, xml, jsonp\n",
            "        \n",
            "        \n",
            "        **More info**\n",
            "        \n",
            "        https://rest.ensembl.org/documentation/info/xref_external\n",
            "    \n",
            "    xref_id(*args, **kwargs)\n",
            "            Cross References ``GET xrefs/id/:id``\n",
            "        \n",
            "        Perform lookups of Ensembl Identifiers and retrieve their external references in other databases\n",
            "        \n",
            "        \n",
            "        **Parameters**\n",
            "        \n",
            "        - Required:\n",
            "        \n",
            "          + *Name*: id\n",
            "        \n",
            "            * *Type*: String\n",
            "            * *Description*: An Ensembl Stable ID\n",
            "            * *Default*: -\n",
            "            * *Example Values*: ENSG00000157764\n",
            "        \n",
            "        \n",
            "        - Optional:\n",
            "        \n",
            "          + *Name*: all_levels\n",
            "        \n",
            "            * *Type*: Boolean\n",
            "            * *Description*: Set to find all genetic features linked to the stable ID, and fetch all external references for them. Specifying this on a gene will also return values from its transcripts and translations\n",
            "            * *Default*: 0\n",
            "            * *Example Values*: 1\n",
            "        \n",
            "        \n",
            "          + *Name*: callback\n",
            "        \n",
            "            * *Type*: String\n",
            "            * *Description*: Name of the callback subroutine to be returned by the requested JSONP response. Required ONLY when using JSONP as the serialisation method. Please see , .\n",
            "            * *Default*: -\n",
            "            * *Example Values*: randomlygeneratedname\n",
            "        \n",
            "        \n",
            "          + *Name*: db_type\n",
            "        \n",
            "            * *Type*: String\n",
            "            * *Description*: Restrict the search to a database other than the default. Useful if you need to use a DB other than core\n",
            "            * *Default*: core\n",
            "            * *Example Values*: core, otherfeatures\n",
            "        \n",
            "        \n",
            "          + *Name*: external_db\n",
            "        \n",
            "            * *Type*: String\n",
            "            * *Description*: Filter by external database\n",
            "            * *Default*: -\n",
            "            * *Example Values*: HGNC\n",
            "        \n",
            "        \n",
            "          + *Name*: object_type\n",
            "        \n",
            "            * *Type*: String\n",
            "            * *Description*: Filter by feature type\n",
            "            * *Default*: -\n",
            "            * *Example Values*: gene, transcript\n",
            "        \n",
            "        \n",
            "          + *Name*: species\n",
            "        \n",
            "            * *Type*: String\n",
            "            * *Description*: Species name/alias\n",
            "            * *Default*: -\n",
            "            * *Example Values*: homo_sapiens, human\n",
            "        \n",
            "        \n",
            "        **Resource info**\n",
            "        \n",
            "        - **Methods**:  GET\n",
            "        - **Response formats**: json, xml, jsonp\n",
            "        \n",
            "        \n",
            "        **More info**\n",
            "        \n",
            "        https://rest.ensembl.org/documentation/info/xref_id\n",
            "    \n",
            "    xref_name(*args, **kwargs)\n",
            "            Cross References ``GET xrefs/name/:species/:name``\n",
            "        \n",
            "        Performs a lookup based upon the primary accession or display label of an external reference and returning the information we hold about the entry\n",
            "        \n",
            "        \n",
            "        **Parameters**\n",
            "        \n",
            "        - Required:\n",
            "        \n",
            "          + *Name*: name\n",
            "        \n",
            "            * *Type*: String\n",
            "            * *Description*: Symbol or display name of a gene\n",
            "            * *Default*: -\n",
            "            * *Example Values*: BRCA2\n",
            "        \n",
            "        \n",
            "          + *Name*: species\n",
            "        \n",
            "            * *Type*: String\n",
            "            * *Description*: Species name/alias\n",
            "            * *Default*: -\n",
            "            * *Example Values*: homo_sapiens, human\n",
            "        \n",
            "        \n",
            "        - Optional:\n",
            "        \n",
            "          + *Name*: callback\n",
            "        \n",
            "            * *Type*: String\n",
            "            * *Description*: Name of the callback subroutine to be returned by the requested JSONP response. Required ONLY when using JSONP as the serialisation method. Please see , .\n",
            "            * *Default*: -\n",
            "            * *Example Values*: randomlygeneratedname\n",
            "        \n",
            "        \n",
            "          + *Name*: db_type\n",
            "        \n",
            "            * *Type*: String\n",
            "            * *Description*: Restrict the search to a database other than the default. Useful if you need to use a DB other than core\n",
            "            * *Default*: core\n",
            "            * *Example Values*: core, otherfeatures\n",
            "        \n",
            "        \n",
            "          + *Name*: external_db\n",
            "        \n",
            "            * *Type*: String\n",
            "            * *Description*: Filter by external database\n",
            "            * *Default*: -\n",
            "            * *Example Values*: HGNC\n",
            "        \n",
            "        \n",
            "        **Resource info**\n",
            "        \n",
            "        - **Methods**:  GET\n",
            "        - **Response formats**: json, xml, jsonp\n",
            "        \n",
            "        \n",
            "        **More info**\n",
            "        \n",
            "        https://rest.ensembl.org/documentation/info/xref_name\n",
            "\n",
            "DATA\n",
            "    docstring = 'Archive ``POST archive/id``\\n\\nRetrieve the latest....ens...\n",
            "    endpoint = {'category': 'Archive', 'description': 'Retrieve the latest...\n",
            "\n",
            "FILE\n",
            "    c:\\users\\aayus\\appdata\\local\\packages\\pythonsoftwarefoundation.python.3.10_qbz5n2kfra8p0\\localcache\\local-packages\\python310\\site-packages\\ensembl_rest\\__init__.py\n",
            "\n",
            "\n"
          ]
        }
      ],
      "source": [
        "help(ensembl_rest)"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "yUFf7YexYeFG"
      },
      "source": [
        "## Getting data"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 11,
      "metadata": {},
      "outputs": [],
      "source": [
        "def read_gene_ids_from_file(file_path):\n",
        "    try:\n",
        "        with open(file_path, 'r') as file:\n",
        "            # Skip the first line\n",
        "            next(file)\n",
        "            \n",
        "            # Read gene IDs from the remaining lines\n",
        "            gene_ids = [line.strip() for line in file]\n",
        "\n",
        "        return gene_ids\n",
        "    except FileNotFoundError:\n",
        "        print(f\"Error: File '{file_path}' not found.\")\n",
        "        return []\n",
        "    except Exception as e:\n",
        "        print(f\"Error: {str(e)}\")\n",
        "        return []\n"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 15,
      "metadata": {},
      "outputs": [],
      "source": [
        "# Function to get cds sequence\n",
        "def get_cds(transcript_id):\n",
        "    address = f\"https://rest.ensembl.org/sequence/id/{transcript_id}?multiple_sequences=1;type=cds\"\n",
        "    r = requests.get(address, headers={ \"Content-Type\" : \"text/x-fasta\"})\n",
        "    # Ensure that there are no issues with the sequence request (e.g. multiple sequences available but not adequately requested)\n",
        "    if not r.ok:\n",
        "        r.raise_for_status()\n",
        "        sys.exit()\n",
        "    raw_output = r.text\n",
        "    # Extract only the nucleotide sequence and format into a single string\n",
        "    pattern = re.compile('(?:^|\\n)[ATGC]+')\n",
        "    matches = pattern.findall(raw_output)\n",
        "\n",
        "    return ''.join(matches).replace('\\n', '')\n",
        "\n",
        "def get_promoter_terminator(transcript_id, promoter_length=1000, terminator_length=500):\n",
        "\n",
        "    sequence = ensembl_rest.sequence_id(id=gene_id, type=\"cdna\", expand_5prime=promoter_length, expand_3prime=terminator_length)[\"seq\"]\n",
        "    \n",
        "    # Use expand_5prime and expand_3prime to get promoter and terminator sequences\n",
        "    promoter_sequence = sequence[:promoter_length]\n",
        "    terminator_sequence = sequence[-terminator_length:]\n",
        "    \n",
        "    return promoter_sequence, terminator_sequence\n",
        "\n",
        "def extract_utr_information(data):\n",
        "    utr_data = data.get('UTR', [])\n",
        "    \n",
        "    utr5_list = []\n",
        "    utr3_list = []\n",
        "\n",
        "    for utr_entry in utr_data:\n",
        "        utr_type = utr_entry.get('type', '')\n",
        "        utr_start = utr_entry.get('start', None)\n",
        "        utr_end = utr_entry.get('end', None)\n",
        "\n",
        "        if utr_type and utr_start is not None and utr_end is not None:\n",
        "            if utr_type == 'five_prime_utr':\n",
        "                utr5_list.append((utr_start, utr_end))\n",
        "            elif utr_type == 'three_prime_utr':\n",
        "                utr3_list.append((utr_start, utr_end))\n",
        "\n",
        "    chromosome = utr_entry['seq_region_name'] if utr_data else None\n",
        "    strand = data.get('strand', None)\n",
        "\n",
        "    return utr5_list, utr3_list, chromosome, strand\n",
        "\n",
        "def get_utr_sequence(chromosome, strand, start, end, species=\"human\"):\n",
        "    \n",
        "    return ensembl_rest.sequence_region(region=f\"{chromosome}:{start}..{end}:{strand}\", species=species)[\"seq\"]\n",
        "\n",
        "def get_full_utr_sequence(list_utr_coordinates, chromosome, strand):\n",
        "    \n",
        "    concatenated_sequence = \"\"\n",
        "\n",
        "    for start, end in list_utr_coordinates:\n",
        "        sequence = get_utr_sequence(chromosome, strand, start, end)\n",
        "        concatenated_sequence += sequence\n",
        "\n",
        "    return concatenated_sequence"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 16,
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "ohzGKTWKYg4R",
        "outputId": "dfcbaf2b-901b-48a6-d328-e7d9b62e20fa"
      },
      "outputs": [
        {
          "name": "stdout",
          "output_type": "stream",
          "text": [
            "Extracting data for gene ID : ENSG00000000003\n",
            "Extracting data for gene ID : ENSG00000000005\n",
            "Extracting data for gene ID : ENSG00000000419\n",
            "Extracting data for gene ID : ENSG00000000457\n",
            "Extracting data for gene ID : ENSG00000000460\n",
            "Data extraction complete.\n"
          ]
        }
      ],
      "source": [
        "# Define the gene IDs you are interested in\n",
        "# gene_ids = [\"ENSG00000000003\", \"ENSG00000000005\", \"ENSG00000000419\"]  # Replace with your gene IDs\n",
        "file_path = \"homo_sapiens_genes_small.txt\"\n",
        "gene_ids = read_gene_ids_from_file(file_path)\n",
        "\n",
        "# Initialize a CSV writer\n",
        "csv_file = open(\"ensembl_data_homo_sapiens.csv\", \"w\", newline=\"\")\n",
        "csv_writer = csv.writer(csv_file)\n",
        "\n",
        "# Write the header to the CSV file\n",
        "csv_writer.writerow([\"ensembl_gene_id\", \"transcript_id\", \"promoter\", \"utr5\", \"cds\", \"utr3\", \"terminator\"])\n",
        "\n",
        "\n",
        "# Loop through each gene ID and retrieve the data\n",
        "for gene_id in gene_ids:\n",
        "    print(f\"Extracting data for gene ID : {gene_id}\")\n",
        "    gene_data = ensembl_rest.lookup(species='homo sapiens', id=gene_id)\n",
        "    \n",
        "    # Get transcript ID\n",
        "    transcript_id = gene_data[\"canonical_transcript\"].split(\".\")[0]\n",
        "\n",
        "    # Retrieve promoter, CDS, and terminator sequences\n",
        "    cds_sequence = get_cds(transcript_id)\n",
        "    promoter_sequence, terminator_sequence = get_promoter_terminator(transcript_id)\n",
        "    \n",
        "    # Retrieve UTR sequences\n",
        "    transcript_data = ensembl_rest.lookup(id=transcript_id, params={'expand':True,'utr':True})\n",
        "    utr5_coord_list, utr3_coord_list, chromosome, strand = extract_utr_information(transcript_data)\n",
        "    utr5_sequence = get_full_utr_sequence(utr5_coord_list, chromosome, strand)\n",
        "    utr3_sequence = get_full_utr_sequence(utr3_coord_list, chromosome, strand)\n",
        "\n",
        "    # Write the row to the CSV file\n",
        "    csv_writer.writerow([gene_id, transcript_id, promoter_sequence, utr5_sequence, cds_sequence, utr3_sequence, terminator_sequence])\n",
        "\n",
        "# Close the CSV file\n",
        "csv_file.close()\n",
        "\n",
        "print(\"Data extraction complete.\")\n"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 14,
      "metadata": {},
      "outputs": [
        {
          "data": {
            "text/html": [
              "<div>\n",
              "<style scoped>\n",
              "    .dataframe tbody tr th:only-of-type {\n",
              "        vertical-align: middle;\n",
              "    }\n",
              "\n",
              "    .dataframe tbody tr th {\n",
              "        vertical-align: top;\n",
              "    }\n",
              "\n",
              "    .dataframe thead th {\n",
              "        text-align: right;\n",
              "    }\n",
              "</style>\n",
              "<table border=\"1\" class=\"dataframe\">\n",
              "  <thead>\n",
              "    <tr style=\"text-align: right;\">\n",
              "      <th></th>\n",
              "      <th>ensembl_gene_id</th>\n",
              "      <th>transcript_id</th>\n",
              "      <th>promoter</th>\n",
              "      <th>utr5</th>\n",
              "      <th>cds</th>\n",
              "      <th>utr3</th>\n",
              "      <th>terminator</th>\n",
              "    </tr>\n",
              "  </thead>\n",
              "  <tbody>\n",
              "    <tr>\n",
              "      <th>0</th>\n",
              "      <td>ENSG00000000003</td>\n",
              "      <td>ENST00000373020</td>\n",
              "      <td>AGCTCTTCAGTAGTTTCTGAACATCTAGACGGTAGGATGTAGAATA...</td>\n",
              "      <td>AGTTGTGGACGCTCGTAAGTTTTCGGCAGTTTCCGGGGAGACTCGG...</td>\n",
              "      <td>ATGGCGTCCCCGTCTCGGAGACTGCAGACTAAACCAGTCATTACTT...</td>\n",
              "      <td>CCCAATGTATCTGTGGGCCTATTCCTCTCTACCTTTAAGGACATTT...</td>\n",
              "      <td>AGGAAACATTTCACCTAATGATAACTGGTACAAAGGAAAGTTCACG...</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>1</th>\n",
              "      <td>ENSG00000000005</td>\n",
              "      <td>ENST00000373031</td>\n",
              "      <td>AGCCGACTCACTTGCAACTCCACCTCAGCAGTGGTCTCTCAGTCCT...</td>\n",
              "      <td>AGCCGACTCACTTGCAACTCCACCTCAGCAGTGGTCTCTCAGTCCT...</td>\n",
              "      <td>ATGGCAAAGAATCCTCCAGAGAATTGTGAAGACTGTCACATTCTAA...</td>\n",
              "      <td>TAGGAGGTTTGAGCTCAAATGCTTAAACTGCTGGCAACATATAATA...</td>\n",
              "      <td>AGGAGCTTAATTTCTCACAAGCCCCCAACTCATACAGTCTTCTAGT...</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>2</th>\n",
              "      <td>ENSG00000000419</td>\n",
              "      <td>ENST00000371588</td>\n",
              "      <td>ACTGCGCTAGTGGACAGCCGAGCCCACCGCAGCCCACGATTAGCAC...</td>\n",
              "      <td>AGTTCCGCC</td>\n",
              "      <td>ATGGCCTCCTTGGAAGTCAGTCGTAGTCCTCGCAGGTCTCGGCGGG...</td>\n",
              "      <td>AAGAAAGATACTCATTTATAGTTACGTTCATTTCAGGTTAAACATG...</td>\n",
              "      <td>ATTTTATACTTACTTTAAGGACCTCTGTTGTAGGACTAATCCTATT...</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>3</th>\n",
              "      <td>ENSG00000000457</td>\n",
              "      <td>ENST00000367771</td>\n",
              "      <td>GCACCTCTACTGTTTGCTACAAGTGGCCAGCAGCCATTTTGGATTT...</td>\n",
              "      <td>GTAGTGGCCACAGCCTTACAGGCAGGCAGGGGTGGTTGGTGTCAAC...</td>\n",
              "      <td>ATGGGATCAGAGAACAGTGCTTTAAAGAGCTATACACTGAGAGAAC...</td>\n",
              "      <td>CAATAGATGTGAGTTAAACTTTAGGAAAAAGGATTCCCTTTTTTTA...</td>\n",
              "      <td>AACTCATGATTCCTTCAGGGTCCCTAAGGATAATACAAAATTAACT...</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>4</th>\n",
              "      <td>ENSG00000000460</td>\n",
              "      <td>ENST00000359326</td>\n",
              "      <td>AACCCGCTCGGGTCCCCTTCCACACTGTGGAAGCTTTGTTCTTTCG...</td>\n",
              "      <td>ACTGCGAGTTTCCGGTCTGGGCTTTGGCGGGTCTGGTTTGAAGCTC...</td>\n",
              "      <td>ATGTTTTTACCTCATATGAACCACCTGACATTGGAACAGACTTTCT...</td>\n",
              "      <td>AACTTATCACTAGGCAGAACTGGGTTTGATGCTTTGTCAACTGAAA...</td>\n",
              "      <td>TGGATAATGAGCTCCTGGGATGGGAAAAGCAGGCCACTTTGGGGTT...</td>\n",
              "    </tr>\n",
              "  </tbody>\n",
              "</table>\n",
              "</div>"
            ],
            "text/plain": [
              "   ensembl_gene_id    transcript_id  \\\n",
              "0  ENSG00000000003  ENST00000373020   \n",
              "1  ENSG00000000005  ENST00000373031   \n",
              "2  ENSG00000000419  ENST00000371588   \n",
              "3  ENSG00000000457  ENST00000367771   \n",
              "4  ENSG00000000460  ENST00000359326   \n",
              "\n",
              "                                            promoter  \\\n",
              "0  AGCTCTTCAGTAGTTTCTGAACATCTAGACGGTAGGATGTAGAATA...   \n",
              "1  AGCCGACTCACTTGCAACTCCACCTCAGCAGTGGTCTCTCAGTCCT...   \n",
              "2  ACTGCGCTAGTGGACAGCCGAGCCCACCGCAGCCCACGATTAGCAC...   \n",
              "3  GCACCTCTACTGTTTGCTACAAGTGGCCAGCAGCCATTTTGGATTT...   \n",
              "4  AACCCGCTCGGGTCCCCTTCCACACTGTGGAAGCTTTGTTCTTTCG...   \n",
              "\n",
              "                                                utr5  \\\n",
              "0  AGTTGTGGACGCTCGTAAGTTTTCGGCAGTTTCCGGGGAGACTCGG...   \n",
              "1  AGCCGACTCACTTGCAACTCCACCTCAGCAGTGGTCTCTCAGTCCT...   \n",
              "2                                          AGTTCCGCC   \n",
              "3  GTAGTGGCCACAGCCTTACAGGCAGGCAGGGGTGGTTGGTGTCAAC...   \n",
              "4  ACTGCGAGTTTCCGGTCTGGGCTTTGGCGGGTCTGGTTTGAAGCTC...   \n",
              "\n",
              "                                                 cds  \\\n",
              "0  ATGGCGTCCCCGTCTCGGAGACTGCAGACTAAACCAGTCATTACTT...   \n",
              "1  ATGGCAAAGAATCCTCCAGAGAATTGTGAAGACTGTCACATTCTAA...   \n",
              "2  ATGGCCTCCTTGGAAGTCAGTCGTAGTCCTCGCAGGTCTCGGCGGG...   \n",
              "3  ATGGGATCAGAGAACAGTGCTTTAAAGAGCTATACACTGAGAGAAC...   \n",
              "4  ATGTTTTTACCTCATATGAACCACCTGACATTGGAACAGACTTTCT...   \n",
              "\n",
              "                                                utr3  \\\n",
              "0  CCCAATGTATCTGTGGGCCTATTCCTCTCTACCTTTAAGGACATTT...   \n",
              "1  TAGGAGGTTTGAGCTCAAATGCTTAAACTGCTGGCAACATATAATA...   \n",
              "2  AAGAAAGATACTCATTTATAGTTACGTTCATTTCAGGTTAAACATG...   \n",
              "3  CAATAGATGTGAGTTAAACTTTAGGAAAAAGGATTCCCTTTTTTTA...   \n",
              "4  AACTTATCACTAGGCAGAACTGGGTTTGATGCTTTGTCAACTGAAA...   \n",
              "\n",
              "                                          terminator  \n",
              "0  AGGAAACATTTCACCTAATGATAACTGGTACAAAGGAAAGTTCACG...  \n",
              "1  AGGAGCTTAATTTCTCACAAGCCCCCAACTCATACAGTCTTCTAGT...  \n",
              "2  ATTTTATACTTACTTTAAGGACCTCTGTTGTAGGACTAATCCTATT...  \n",
              "3  AACTCATGATTCCTTCAGGGTCCCTAAGGATAATACAAAATTAACT...  \n",
              "4  TGGATAATGAGCTCCTGGGATGGGAAAAGCAGGCCACTTTGGGGTT...  "
            ]
          },
          "execution_count": 14,
          "metadata": {},
          "output_type": "execute_result"
        }
      ],
      "source": [
        "genetic_data = pd.read_csv(\"ensembl_data_homo_sapiens.csv\")\n",
        "genetic_data"
      ]
    }
  ],
  "metadata": {
    "colab": {
      "provenance": []
    },
    "kernelspec": {
      "display_name": "Python 3",
      "name": "python3"
    },
    "language_info": {
      "codemirror_mode": {
        "name": "ipython",
        "version": 3
      },
      "file_extension": ".py",
      "mimetype": "text/x-python",
      "name": "python",
      "nbconvert_exporter": "python",
      "pygments_lexer": "ipython3",
      "version": "3.10.11"
    }
  },
  "nbformat": 4,
  "nbformat_minor": 0
}
