<!DOCTYPE html>
<html class="writer-html5" lang="en" >
<head>
    <meta charset="utf-8" />
    <meta http-equiv="X-UA-Compatible" content="IE=edge" />
    <meta name="viewport" content="width=device-width, initial-scale=1.0" />
      <link rel="shortcut icon" href="../../../img/favicon.ico" />
    <title>dna_feature_extraction - Genomic Data Extraction</title>
    <link rel="stylesheet" href="../../../css/theme.css" />
    <link rel="stylesheet" href="../../../css/theme_extra.css" />
        <link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/highlight.js/10.5.0/styles/github.min.css" />
        <link href="../../../assets/_mkdocstrings.css" rel="stylesheet" />
    
      <script>
        // Current page data
        var mkdocs_page_name = "dna_feature_extraction";
        var mkdocs_page_input_path = "genomic_data_extraction/dna/dna_feature_extraction.md";
        var mkdocs_page_url = null;
      </script>
    
    <script src="../../../js/jquery-3.6.0.min.js" defer></script>
    <!--[if lt IE 9]>
      <script src="../../../js/html5shiv.min.js"></script>
    <![endif]-->
      <script src="https://cdnjs.cloudflare.com/ajax/libs/highlight.js/10.5.0/highlight.min.js"></script>
      <script>hljs.initHighlightingOnLoad();</script> 
</head>

<body class="wy-body-for-nav" role="document">

  <div class="wy-grid-for-nav">
    <nav data-toggle="wy-nav-shift" class="wy-nav-side stickynav">
    <div class="wy-side-scroll">
      <div class="wy-side-nav-search">
          <a href="../../.." class="icon icon-home"> Genomic Data Extraction
        </a>
      </div>

      <div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="Navigation menu">
              <ul>
                <li class="toctree-l1"><a class="reference internal" href="../../..">Home</a>
                </li>
              </ul>
              <p class="caption"><span class="caption-text">genomic_data_extraction</span></p>
              <ul class="current">
                  <li class="toctree-l1 current"><a class="reference internal current" href="#">dna</a>
    <ul class="current">
                <li class="toctree-l2"><a class="reference internal" href="../dna_extraction/">dna_extraction</a>
                </li>
                <li class="toctree-l2 current"><a class="reference internal current" href="./">dna_feature_extraction</a>
    <ul class="current">
    </ul>
                </li>
                <li class="toctree-l2"><a class="reference internal" href="../ensembl_api/">ensembl_api</a>
                </li>
    </ul>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="#">rna</a>
    <ul>
                <li class="toctree-l2"><a class="reference internal" href="#">data_conversion_helper_functions</a>
    <ul>
                <li class="toctree-l3"><a class="reference internal" href="../../rna/data_conversion_helper_functions/convert_quantsf_to_csv/">convert_quantsf_to_csv</a>
                </li>
                <li class="toctree-l3"><a class="reference internal" href="../../rna/data_conversion_helper_functions/create_expression_matrix/">create_expression_matrix</a>
                </li>
                <li class="toctree-l3"><a class="reference internal" href="../../rna/data_conversion_helper_functions/create_samplesheet_csv/">create_samplesheet_csv</a>
                </li>
                <li class="toctree-l3"><a class="reference internal" href="../../rna/data_conversion_helper_functions/process_expression_matrix/">process_expression_matrix</a>
                </li>
    </ul>
                </li>
                <li class="toctree-l2"><a class="reference internal" href="#">rna_download_logic</a>
    <ul>
                <li class="toctree-l3"><a class="reference internal" href="../../rna/rna_download_logic/mRNA_fastq_download/">mRNA_fastq_download</a>
                </li>
                <li class="toctree-l3"><a class="reference internal" href="../../rna/rna_download_logic/query_and_csv_production/">query_and_csv_production</a>
                </li>
    </ul>
                </li>
                <li class="toctree-l2"><a class="reference internal" href="../../rna/rna_extraction/">rna_extraction</a>
                </li>
    </ul>
                  </li>
              </ul>
      </div>
    </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap">
      <nav class="wy-nav-top" role="navigation" aria-label="Mobile navigation menu">
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="../../..">Genomic Data Extraction</a>
        
      </nav>
      <div class="wy-nav-content">
        <div class="rst-content"><div role="navigation" aria-label="breadcrumbs navigation">
  <ul class="wy-breadcrumbs">
    <li><a href="../../.." class="icon icon-home" alt="Docs"></a> &raquo;</li>
          <li>genomic_data_extraction &raquo;</li>
          <li>dna &raquo;</li>
      <li>dna_feature_extraction</li>
    <li class="wy-breadcrumbs-aside">
          <a href="https://github.com/TransferLearningForDNA/genomic_data_extraction/edit/master/docs/genomic_data_extraction/dna/dna_feature_extraction.md" class="icon icon-github"> Edit on GitHub</a>
    </li>
  </ul>
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
            <div class="section" itemprop="articleBody">
              
                <div class="doc doc-object doc-module">



<a id="dna.dna_feature_extraction"></a>
  <div class="doc doc-contents first">

  

  <div class="doc doc-children">










<div class="doc doc-object doc-function">



<h2 id="dna.dna_feature_extraction.compute_cds_codon_frequencies" class="doc doc-heading">
          <code class="highlight language-python">compute_cds_codon_frequencies(cds, codons)</code>

</h2>


  <div class="doc doc-contents ">
  
      <p>Compute the frequency of every possible codon in the cds.</p>



<table class="field-list">
  <colgroup>
    <col class="field-name" />
    <col class="field-body" />
  </colgroup>
  <tbody valign="top">
    <tr class="field">
      <th class="field-name">Parameters:</th>
      <td class="field-body">
        <ul class="first simple">
            <li>
              <b><code>cds</code></b>
                  (<code>str</code>)
              –
              <div class="doc-md-description">
                <p>cds sequence</p>
              </div>
            </li>
            <li>
              <b><code>codons</code></b>
                  (<code><span title="typing.List">List</span>[str]</code>)
              –
              <div class="doc-md-description">
                <p>list of strings representing all 64 possible codons</p>
              </div>
            </li>
        </ul>
      </td>
    </tr>
  </tbody>
</table>


<table class="field-list">
  <colgroup>
    <col class="field-name" />
    <col class="field-body" />
  </colgroup>
  <tbody valign="top">
    <tr class="field">
    <th class="field-name">Returns:</th>
    <td class="field-body">
      <ul class="first simple">
          <li>
                <code><span title="typing.Dict">Dict</span>[str, float]</code>
            –
            <div class="doc-md-description">
              <p>Dict[str, float]: dictionary of codon frequencies for all 64 possible codons,</p>
            </div>
          </li>
          <li>
                <code><span title="typing.Dict">Dict</span>[str, float]</code>
            –
            <div class="doc-md-description">
              <p>with codons as keys and their frequencies as values.</p>
            </div>
          </li>
      </ul>
    </td>
    </tr>
  </tbody>
</table>
          <details class="quote">
            <summary>Source code in <code>dna/dna_feature_extraction.py</code></summary>
            <pre class="highlight"><code class="language-python">def compute_cds_codon_frequencies(cds: str, codons: List[str]) -&gt; Dict[str, float]:
    """Compute the frequency of every possible codon in the cds.

    Args:
        cds (str): cds sequence

        codons (List[str]): list of strings representing all 64 possible codons
        e.g. "AGA"

    Returns:
        Dict[str, float]: dictionary of codon frequencies for all 64 possible codons,
        with codons as keys and their frequencies as values.
    """
    cds_length = len(cds)
    codon_count = cds_length // 3
    # Placeholder for codon frequencies
    codon_frequencies = {codon: 0 for codon in codons}

    # Address situation in which the cds provided is too short
    if codon_count == 0:
        return codon_frequencies

    # Compute codon frequencies based on the sequence in the row
    for i in range(0, cds_length - 2, 3):
        codon = cds[i : i + 3]
        if len(codon) == 3:
            codon_frequencies[codon] += 1

    for codon in codon_frequencies:
        codon_frequencies[codon] /= codon_count

    return codon_frequencies</code></pre>
          </details>
  </div>

</div>


<div class="doc doc-object doc-function">



<h2 id="dna.dna_feature_extraction.compute_gc_content_sequence_components" class="doc doc-heading">
          <code class="highlight language-python">compute_gc_content_sequence_components(utr5, cds, utr3)</code>

</h2>


  <div class="doc doc-contents ">
  
      <p>Compute the GC content of the cds, utr3 and utr5.</p>
<p>GC content is defined as (G + C)/(A + T + G + C), where each letter
represents the number of times that nucleotide appears in the DNA sequence.</p>



<table class="field-list">
  <colgroup>
    <col class="field-name" />
    <col class="field-body" />
  </colgroup>
  <tbody valign="top">
    <tr class="field">
      <th class="field-name">Parameters:</th>
      <td class="field-body">
        <ul class="first simple">
            <li>
              <b><code>utr5</code></b>
                  (<code>str</code>)
              –
              <div class="doc-md-description">
                <p>5' UTR DNA sequence.</p>
              </div>
            </li>
            <li>
              <b><code>cds</code></b>
                  (<code>str</code>)
              –
              <div class="doc-md-description">
                <p>CDS DNA sequence.</p>
              </div>
            </li>
            <li>
              <b><code>utr3</code></b>
                  (<code>str</code>)
              –
              <div class="doc-md-description">
                <p>3'UTR DNA sequence.</p>
              </div>
            </li>
        </ul>
      </td>
    </tr>
  </tbody>
</table>


<table class="field-list">
  <colgroup>
    <col class="field-name" />
    <col class="field-body" />
  </colgroup>
  <tbody valign="top">
    <tr class="field">
    <th class="field-name">Returns:</th>
    <td class="field-body">
      <ul class="first simple">
          <li>
                <code><span title="typing.Dict">Dict</span>[str, float]</code>
            –
            <div class="doc-md-description">
              <p>Dict[str, float]: A dictionary with keys 'cds_gc', 'utr5_gc', and 'utr3_gc',
              and float values representing the GC content of these sequences.</p>
            </div>
          </li>
      </ul>
    </td>
    </tr>
  </tbody>
</table>
          <details class="quote">
            <summary>Source code in <code>dna/dna_feature_extraction.py</code></summary>
            <pre class="highlight"><code class="language-python">def compute_gc_content_sequence_components(
    utr5: str, cds: str, utr3: str
) -&gt; Dict[str, float]:
    """Compute the GC content of the cds, utr3 and utr5.

    GC content is defined as (G + C)/(A + T + G + C), where each letter
    represents the number of times that nucleotide appears in the DNA sequence.

    Args:
        utr5 (str): 5' UTR DNA sequence.
        cds (str): CDS DNA sequence.
        utr3 (str): 3'UTR DNA sequence.

    Returns:
        Dict[str, float]: A dictionary with keys 'cds_gc', 'utr5_gc', and 'utr3_gc',
                          and float values representing the GC content of these sequences.
    """
    utr5_gc_count = count_gc_nucleotides(utr5)
    cds_gc_count = count_gc_nucleotides(cds)
    utr3_gc_count = count_gc_nucleotides(utr3)

    utr5_length = len(utr5)
    cds_length = len(cds)
    utr3_length = len(utr3)

    utr5_gc = utr5_gc_count / utr5_length if utr5_length != 0 else ""
    cds_gc = cds_gc_count / cds_length if cds_length != 0 else ""
    utr3_gc = utr3_gc_count / utr3_length if utr3_length != 0 else ""

    gc_content = {"utr5_gc": utr5_gc, "cds_gc": cds_gc, "utr3_gc": utr3_gc}

    return gc_content</code></pre>
          </details>
  </div>

</div>


<div class="doc doc-object doc-function">



<h2 id="dna.dna_feature_extraction.compute_gc_content_wobble_positions" class="doc doc-heading">
          <code class="highlight language-python">compute_gc_content_wobble_positions(cds)</code>

</h2>


  <div class="doc doc-contents ">
  
      <p>Compute the GC content of wobble positions 2 and 3 in the cds.</p>
<p>GC content is defined as (G + C)/(A + T + G + C), where each letter
represents the number of times that nucleotide appears in a given
wobble position</p>



<table class="field-list">
  <colgroup>
    <col class="field-name" />
    <col class="field-body" />
  </colgroup>
  <tbody valign="top">
    <tr class="field">
      <th class="field-name">Parameters:</th>
      <td class="field-body">
        <ul class="first simple">
            <li>
              <b><code>cds</code></b>
                  (<code>str</code>)
              –
              <div class="doc-md-description">
                <p>CDS DNA sequence.</p>
              </div>
            </li>
        </ul>
      </td>
    </tr>
  </tbody>
</table>


<table class="field-list">
  <colgroup>
    <col class="field-name" />
    <col class="field-body" />
  </colgroup>
  <tbody valign="top">
    <tr class="field">
    <th class="field-name">Returns:</th>
    <td class="field-body">
      <ul class="first simple">
          <li>
                <code><span title="typing.Dict">Dict</span>[str, float]</code>
            –
            <div class="doc-md-description">
              <p>Dict[str, float]: A dictionary with keys 'wobble2_gc' and 'wobble3_gc',
              and float values representing the GC content at wobble positions 2 and 3
              in the cds, respectively.</p>
            </div>
          </li>
      </ul>
    </td>
    </tr>
  </tbody>
</table>
          <details class="quote">
            <summary>Source code in <code>dna/dna_feature_extraction.py</code></summary>
            <pre class="highlight"><code class="language-python">def compute_gc_content_wobble_positions(cds: str) -&gt; Dict[str, float]:
    """Compute the GC content of wobble positions 2 and 3 in the cds.

    GC content is defined as (G + C)/(A + T + G + C), where each letter
    represents the number of times that nucleotide appears in a given
    wobble position

    Args:
        cds (str): CDS DNA sequence.

    Returns:
        Dict[str, float]: A dictionary with keys 'wobble2_gc' and 'wobble3_gc',
                          and float values representing the GC content at wobble positions 2 and 3
                          in the cds, respectively.
    """
    # Extract the nucleotides in positions 2 and 3 into separate lists
    wobble2_nucleotides = cds[1::3]
    wobble3_nucleotides = cds[2::3]

    wobble2_gc_count = wobble2_nucleotides.count("G") + wobble2_nucleotides.count("C")
    wobble3_gc_count = wobble3_nucleotides.count("G") + wobble3_nucleotides.count("C")

    # Compute the GC content
    wobble2_gc = (
        wobble2_gc_count / len(wobble2_nucleotides)
        if len(wobble2_nucleotides) != 0
        else ""
    )

    wobble3_gc = (
        wobble3_gc_count / len(wobble3_nucleotides)
        if len(wobble3_nucleotides) != 0
        else ""
    )

    gc_content = {"cds_wobble2_gc": wobble2_gc, "cds_wobble3_gc": wobble3_gc}

    return gc_content</code></pre>
          </details>
  </div>

</div>


<div class="doc doc-object doc-function">



<h2 id="dna.dna_feature_extraction.compute_lengths" class="doc doc-heading">
          <code class="highlight language-python">compute_lengths(cds, utr5, utr3)</code>

</h2>


  <div class="doc doc-contents ">
  
      <p>Compute the length of the cds, utr3 and utr5 DNA sequences.</p>



<table class="field-list">
  <colgroup>
    <col class="field-name" />
    <col class="field-body" />
  </colgroup>
  <tbody valign="top">
    <tr class="field">
      <th class="field-name">Parameters:</th>
      <td class="field-body">
        <ul class="first simple">
            <li>
              <b><code>cds</code></b>
                  (<code>str</code>)
              –
              <div class="doc-md-description">
                <p>CDS DNA sequence.</p>
              </div>
            </li>
            <li>
              <b><code>utr5</code></b>
                  (<code>str</code>)
              –
              <div class="doc-md-description">
                <p>5' UTR DNA sequence.</p>
              </div>
            </li>
            <li>
              <b><code>utr3</code></b>
                  (<code>str</code>)
              –
              <div class="doc-md-description">
                <p>3' UTR DNA sequence.</p>
              </div>
            </li>
        </ul>
      </td>
    </tr>
  </tbody>
</table>


<table class="field-list">
  <colgroup>
    <col class="field-name" />
    <col class="field-body" />
  </colgroup>
  <tbody valign="top">
    <tr class="field">
    <th class="field-name">Returns:</th>
    <td class="field-body">
      <ul class="first simple">
          <li>
                <code><span title="typing.Dict">Dict</span>[str, int]</code>
            –
            <div class="doc-md-description">
              <p>Dict[str, int]: A dictionary with keys 'cds', 'utr5', and 'utr3', and integer values
            representing the lengths of these sequences.</p>
            </div>
          </li>
      </ul>
    </td>
    </tr>
  </tbody>
</table>
          <details class="quote">
            <summary>Source code in <code>dna/dna_feature_extraction.py</code></summary>
            <pre class="highlight"><code class="language-python">def compute_lengths(cds: str, utr5: str, utr3: str) -&gt; Dict[str, int]:
    """Compute the length of the cds, utr3 and utr5 DNA sequences.

    Args:
        cds (str): CDS DNA sequence.
        utr5 (str): 5' UTR DNA sequence.
        utr3 (str): 3' UTR DNA sequence.

    Returns:
        Dict[str, int]: A dictionary with keys 'cds', 'utr5', and 'utr3', and integer values
                        representing the lengths of these sequences.
    """
    lengths = {
        "cds_length": len(cds) if len(cds) &gt; 0 else "",
        "utr5_length": len(utr5) if len(utr5) &gt; 0 else "",
        "utr3_length": len(utr3) if len(utr3) &gt; 0 else "",
    }

    return lengths</code></pre>
          </details>
  </div>

</div>


<div class="doc doc-object doc-function">



<h2 id="dna.dna_feature_extraction.count_gc_nucleotides" class="doc doc-heading">
          <code class="highlight language-python">count_gc_nucleotides(sequence)</code>

</h2>


  <div class="doc doc-contents ">
  
      <p>Count the number of times G and C nucleotides appear in a DNA sequence.</p>



<table class="field-list">
  <colgroup>
    <col class="field-name" />
    <col class="field-body" />
  </colgroup>
  <tbody valign="top">
    <tr class="field">
      <th class="field-name">Parameters:</th>
      <td class="field-body">
        <ul class="first simple">
            <li>
              <b><code>sequence</code></b>
                  (<code>str</code>)
              –
              <div class="doc-md-description">
                <p>DNA sequence</p>
              </div>
            </li>
        </ul>
      </td>
    </tr>
  </tbody>
</table>


<table class="field-list">
  <colgroup>
    <col class="field-name" />
    <col class="field-body" />
  </colgroup>
  <tbody valign="top">
    <tr class="field">
    <th class="field-name">Returns:</th>
    <td class="field-body">
      <ul class="first simple">
          <li>
<b><code>int</code></b>(                <code>int</code>
)            –
            <div class="doc-md-description">
              <p>Sum of G and C counts in the sequence.</p>
            </div>
          </li>
      </ul>
    </td>
    </tr>
  </tbody>
</table>
          <details class="quote">
            <summary>Source code in <code>dna/dna_feature_extraction.py</code></summary>
            <pre class="highlight"><code class="language-python">def count_gc_nucleotides(sequence: str) -&gt; int:
    """Count the number of times G and C nucleotides appear in a DNA sequence.

    Args:
        sequence (str): DNA sequence

    Returns:
        int: Sum of G and C counts in the sequence.
    """
    g_content = sequence.count("G")
    c_content = sequence.count("C")
    return g_content + c_content</code></pre>
          </details>
  </div>

</div>


<div class="doc doc-object doc-function">



<h2 id="dna.dna_feature_extraction.extract_dna_features" class="doc doc-heading">
          <code class="highlight language-python">extract_dna_features(folder_path)</code>

</h2>


  <div class="doc doc-contents ">
  
      <p>Extract and compute DNA features for each CSV file containing genomic sequences.</p>
<p>Add new columns to the existing CSV files with computed codon frequencies, lengths of
different DNA segments (utr5, cds, utr3), GC content in DNA segments, and GC content at
wobble positions.</p>



<table class="field-list">
  <colgroup>
    <col class="field-name" />
    <col class="field-body" />
  </colgroup>
  <tbody valign="top">
    <tr class="field">
      <th class="field-name">Parameters:</th>
      <td class="field-body">
        <ul class="first simple">
            <li>
              <b><code>folder_path</code></b>
                  (<code>str</code>)
              –
              <div class="doc-md-description">
                <p>The path to the folder containing genomic data CSV files.</p>
              </div>
            </li>
        </ul>
      </td>
    </tr>
  </tbody>
</table>


<table class="field-list">
  <colgroup>
    <col class="field-name" />
    <col class="field-body" />
  </colgroup>
  <tbody valign="top">
    <tr class="field">
    <th class="field-name">Returns:</th>
    <td class="field-body">
      <ul class="first simple">
          <li>
<b><code>None</code></b>(                <code>None</code>
)            –
            <div class="doc-md-description">
              <p>This function does not return a value but modifies the files in the specified directory.</p>
            </div>
          </li>
      </ul>
    </td>
    </tr>
  </tbody>
</table>
          <details class="quote">
            <summary>Source code in <code>dna/dna_feature_extraction.py</code></summary>
            <pre class="highlight"><code class="language-python">def extract_dna_features(folder_path: str) -&gt; None:
    """Extract and compute DNA features for each CSV file containing genomic sequences.

    Add new columns to the existing CSV files with computed codon frequencies, lengths of
    different DNA segments (utr5, cds, utr3), GC content in DNA segments, and GC content at
    wobble positions.

    Args:
        folder_path (str): The path to the folder containing genomic data CSV files.

    Returns:
        None: This function does not return a value but modifies the files in the specified directory.
    """
    # Iterate over files in the directory
    for filename in os.listdir(folder_path):
        if filename.endswith(".csv"):
            file_path = os.path.join(folder_path, filename)

            # Create a temporary file to write the modified data
            temp_file = tempfile.NamedTemporaryFile(
                mode="w", delete=False, newline="", encoding="utf-8"
            )

            # Open the CSV file for reading
            with open(
                file_path, "r", newline="", encoding="utf-8"
            ) as infile, temp_file:
                reader = csv.DictReader(infile)

                # Define the fieldnames for the output CSV
                header = reader.fieldnames

                if header != [
                    "ensembl_gene_id",
                    "transcript_id",
                    "promoter",
                    "utr5",
                    "cds",
                    "utr3",
                    "terminator",
                ]:
                    continue
                print("Extracting DNA features from:", filename)

                new_columns = []
                codons = [
                    "".join(combination) for combination in product("ACGT", repeat=3)
                ]
                new_columns.extend(codons)
                new_columns.extend(
                    [
                        "cds_length",
                        "utr5_length",
                        "utr3_length",
                        "utr5_gc",
                        "cds_gc",
                        "utr3_gc",
                        "cds_wobble2_gc",
                        "cds_wobble3_gc",
                    ]
                )
                header.extend(new_columns)

                # Open the CSV file for writing
                writer = csv.DictWriter(temp_file, fieldnames=header)
                writer.writeheader()

                # Compute features for each gene
                for row in reader:
                    # Add data for new columns (assuming new_columns is a list of values)
                    utr5 = row.get("utr5")
                    cds = row.get("cds")
                    utr3 = row.get("utr3")

                    row.update(compute_cds_codon_frequencies(cds=cds, codons=codons))
                    row.update(compute_lengths(cds=cds, utr5=utr5, utr3=utr3))
                    row.update(
                        compute_gc_content_sequence_components(
                            utr5=utr5, cds=cds, utr3=utr3
                        )
                    )
                    row.update(compute_gc_content_wobble_positions(cds))

                    # Write the modified row to the temporary file
                    writer.writerow(row)

            # Replace the original file with the temporary file
            shutil.move(temp_file.name, file_path)</code></pre>
          </details>
  </div>

</div>



  </div>

  </div>

</div>
              
            </div>
          </div><footer>
    <div class="rst-footer-buttons" role="navigation" aria-label="Footer Navigation">
        <a href="../dna_extraction/" class="btn btn-neutral float-left" title="dna_extraction"><span class="icon icon-circle-arrow-left"></span> Previous</a>
        <a href="../ensembl_api/" class="btn btn-neutral float-right" title="ensembl_api">Next <span class="icon icon-circle-arrow-right"></span></a>
    </div>

  <hr/>

  <div role="contentinfo">
    <!-- Copyright etc -->
  </div>

  Built with <a href="https://www.mkdocs.org/">MkDocs</a> using a <a href="https://github.com/readthedocs/sphinx_rtd_theme">theme</a> provided by <a href="https://readthedocs.org">Read the Docs</a>.
</footer>
          
        </div>
      </div>

    </section>

  </div>

  <div class="rst-versions" role="note" aria-label="Versions">
  <span class="rst-current-version" data-toggle="rst-current-version">
    
        <span>
          <a href="https://github.com/TransferLearningForDNA/genomic_data_extraction" class="fa fa-github" style="color: #fcfcfc"> GitHub</a>
        </span>
    
    
      <span><a href="../dna_extraction/" style="color: #fcfcfc">&laquo; Previous</a></span>
    
    
      <span><a href="../ensembl_api/" style="color: #fcfcfc">Next &raquo;</a></span>
    
  </span>
</div>
    <script>var base_url = '../../..';</script>
    <script src="../../../js/theme_extra.js" defer></script>
    <script src="../../../js/theme.js" defer></script>
    <script defer>
        window.onload = function () {
            SphinxRtdTheme.Navigation.enable(true);
        };
    </script>

</body>
</html>
