<!DOCTYPE html>
<html class="writer-html5" lang="en" >
<head>
    <meta charset="utf-8" />
    <meta http-equiv="X-UA-Compatible" content="IE=edge" />
    <meta name="viewport" content="width=device-width, initial-scale=1.0" />
      <link rel="shortcut icon" href="../../../../img/favicon.ico" />
    <title>convert_quantsf_to_csv - Genomic Data Extraction</title>
    <link rel="stylesheet" href="../../../../css/theme.css" />
    <link rel="stylesheet" href="../../../../css/theme_extra.css" />
        <link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/highlight.js/10.5.0/styles/github.min.css" />
        <link href="../../../../assets/_mkdocstrings.css" rel="stylesheet" />
    
      <script>
        // Current page data
        var mkdocs_page_name = "convert_quantsf_to_csv";
        var mkdocs_page_input_path = "genomic_data_extraction/rna/data_conversion_helper_functions/convert_quantsf_to_csv.md";
        var mkdocs_page_url = null;
      </script>
    
    <script src="../../../../js/jquery-3.6.0.min.js" defer></script>
    <!--[if lt IE 9]>
      <script src="../../../../js/html5shiv.min.js"></script>
    <![endif]-->
      <script src="https://cdnjs.cloudflare.com/ajax/libs/highlight.js/10.5.0/highlight.min.js"></script>
      <script>hljs.initHighlightingOnLoad();</script> 
</head>

<body class="wy-body-for-nav" role="document">

  <div class="wy-grid-for-nav">
    <nav data-toggle="wy-nav-shift" class="wy-nav-side stickynav">
    <div class="wy-side-scroll">
      <div class="wy-side-nav-search">
          <a href="../../../.." class="icon icon-home"> Genomic Data Extraction
        </a>
      </div>

      <div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="Navigation menu">
              <ul>
                <li class="toctree-l1"><a class="reference internal" href="../../../..">Home</a>
                </li>
              </ul>
              <p class="caption"><span class="caption-text">genomic_data_extraction</span></p>
              <ul class="current">
                  <li class="toctree-l1"><a class="reference internal" href="#">dna</a>
    <ul>
                <li class="toctree-l2"><a class="reference internal" href="../../../dna/dna_extraction/">dna_extraction</a>
                </li>
                <li class="toctree-l2"><a class="reference internal" href="../../../dna/dna_feature_extraction/">dna_feature_extraction</a>
                </li>
                <li class="toctree-l2"><a class="reference internal" href="../../../dna/ensembl_api/">ensembl_api</a>
                </li>
    </ul>
                  </li>
                  <li class="toctree-l1 current"><a class="reference internal current" href="#">rna</a>
    <ul class="current">
                <li class="toctree-l2"><a class="reference internal" href="../../rna_extraction/">rna_extraction</a>
                </li>
                <li class="toctree-l2 current"><a class="reference internal current" href="#">data_conversion_helper_functions</a>
    <ul class="current">
                <li class="toctree-l3 current"><a class="reference internal current" href="./">convert_quantsf_to_csv</a>
    <ul class="current">
    </ul>
                </li>
                <li class="toctree-l3"><a class="reference internal" href="../create_expression_matrix/">create_expression_matrix</a>
                </li>
                <li class="toctree-l3"><a class="reference internal" href="../process_expression_matrix/">process_expression_matrix</a>
                </li>
                <li class="toctree-l3"><a class="reference internal" href="../create_samplesheet_csv/">create_samplesheet_csv</a>
                </li>
    </ul>
                </li>
                <li class="toctree-l2"><a class="reference internal" href="#">rna_download_logic</a>
    <ul>
                <li class="toctree-l3"><a class="reference internal" href="../../rna_download_logic/mRNA_fastq_download/">mRNA_fastq_download</a>
                </li>
                <li class="toctree-l3"><a class="reference internal" href="../../rna_download_logic/query_and_csv_production/">query_and_csv_production</a>
                </li>
    </ul>
                </li>
    </ul>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../../../dataset_integration/">dataset_integration</a>
                  </li>
              </ul>
      </div>
    </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap">
      <nav class="wy-nav-top" role="navigation" aria-label="Mobile navigation menu">
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="../../../..">Genomic Data Extraction</a>
        
      </nav>
      <div class="wy-nav-content">
        <div class="rst-content"><div role="navigation" aria-label="breadcrumbs navigation">
  <ul class="wy-breadcrumbs">
    <li><a href="../../../.." class="icon icon-home" alt="Docs"></a> &raquo;</li>
          <li>genomic_data_extraction &raquo;</li>
          <li>rna &raquo;</li>
          <li>data_conversion_helper_functions &raquo;</li>
      <li>convert_quantsf_to_csv</li>
    <li class="wy-breadcrumbs-aside">
          <a href="https://github.com/TransferLearningForDNA/genomic_data_extraction/edit/master/docs/genomic_data_extraction/rna/data_conversion_helper_functions/convert_quantsf_to_csv.md" class="icon icon-github"> Edit on GitHub</a>
    </li>
  </ul>
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
            <div class="section" itemprop="articleBody">
              
                <div class="doc doc-object doc-module">



<a id="rna.data_conversion_helper_functions.convert_quantsf_to_csv"></a>
  <div class="doc doc-contents first">

  

  <div class="doc doc-children">










<div class="doc doc-object doc-function">



<h2 id="rna.data_conversion_helper_functions.convert_quantsf_to_csv.convert_all_species_files" class="doc doc-heading">
          <code class="highlight language-python">convert_all_species_files(folder_path)</code>

</h2>


  <div class="doc doc-contents ">
  
      <p>Convert quantification files from sf to csv for each species.</p>



<table class="field-list">
  <colgroup>
    <col class="field-name" />
    <col class="field-body" />
  </colgroup>
  <tbody valign="top">
    <tr class="field">
      <th class="field-name">Parameters:</th>
      <td class="field-body">
        <ul class="first simple">
            <li>
              <b><code>folder_path</code></b>
                  (<code>str) </code>)
              –
              <div class="doc-md-description">
                <p>Path to raw quant files folder.</p>
              </div>
            </li>
        </ul>
      </td>
    </tr>
  </tbody>
</table>


<table class="field-list">
  <colgroup>
    <col class="field-name" />
    <col class="field-body" />
  </colgroup>
  <tbody valign="top">
    <tr class="field">
    <th class="field-name">Returns:</th>
    <td class="field-body">
      <ul class="first simple">
          <li>
<b><code>None</code></b>(                <code>None</code>
)            –
            <div class="doc-md-description">
              <p>This function does not return a value but outputs files to the specified directory.</p>
            </div>
          </li>
      </ul>
    </td>
    </tr>
  </tbody>
</table>
          <details class="quote">
            <summary>Source code in <code>rna/data_conversion_helper_functions/convert_quantsf_to_csv.py</code></summary>
            <pre class="highlight"><code class="language-python">def convert_all_species_files(folder_path: str) -&gt; None:
    """Convert quantification files from sf to csv for each species.

    Args:
        folder_path (str) : Path to raw quant files folder.

    Returns:
        None: This function does not return a value but outputs files to the specified directory.
    """
    # Iterate over items in the directory
    for item in os.listdir(folder_path):
        if item == ".gitignore":
            continue
        item_path = os.path.join(folder_path, item)  # Get the full path of the item
        if os.path.isdir(item_path):  # Check if the item is a directory
            print(f"\nConverting quant files for species: {item}")

            # Check if the directory containing sf files exists
            sf_files_path = os.path.join(item_path, "sf_files")
            if not os.path.isdir(sf_files_path):
                print(f"The directory {sf_files_path} does not exist.")
            elif not os.listdir(sf_files_path):
                print(f"The directory {sf_files_path} is empty.")
            else:
                csv_files_path = os.path.join(
                    item_path, "csv_files"
                )  # path to store csv files
                # Convert all quant sf files to csv
                convert_quant_output_to_csv(sf_files_path, csv_files_path)</code></pre>
          </details>
  </div>

</div>


<div class="doc doc-object doc-function">



<h2 id="rna.data_conversion_helper_functions.convert_quantsf_to_csv.convert_quant_output_to_csv" class="doc doc-heading">
          <code class="highlight language-python">convert_quant_output_to_csv(input_path, output_path)</code>

</h2>


  <div class="doc doc-contents ">
  
      <p>Convert quantification files, Salmon (nf-core rna-seq) output, to csv files.</p>



<table class="field-list">
  <colgroup>
    <col class="field-name" />
    <col class="field-body" />
  </colgroup>
  <tbody valign="top">
    <tr class="field">
      <th class="field-name">Parameters:</th>
      <td class="field-body">
        <ul class="first simple">
            <li>
              <b><code>input_path</code></b>
                  (<code>str</code>)
              –
              <div class="doc-md-description">
                <p>quant_DRR513083.sf files folder path (tsv: tab-separated values)</p>
              </div>
            </li>
            <li>
              <b><code>output_path</code></b>
                  (<code>str</code>)
              –
              <div class="doc-md-description">
                <p>folder path to store the csv version of the quant files</p>
              </div>
            </li>
        </ul>
      </td>
    </tr>
  </tbody>
</table>


<table class="field-list">
  <colgroup>
    <col class="field-name" />
    <col class="field-body" />
  </colgroup>
  <tbody valign="top">
    <tr class="field">
    <th class="field-name">Returns:</th>
    <td class="field-body">
      <ul class="first simple">
          <li>
<b><code>None</code></b>(                <code>None</code>
)            –
            <div class="doc-md-description">
              <p>This function does not return a value but outputs files to the specified directory.</p>
            </div>
          </li>
      </ul>
    </td>
    </tr>
  </tbody>
</table>
          <details class="quote">
            <summary>Source code in <code>rna/data_conversion_helper_functions/convert_quantsf_to_csv.py</code></summary>
            <pre class="highlight"><code class="language-python">def convert_quant_output_to_csv(input_path: str, output_path: str) -&gt; None:
    """Convert quantification files, Salmon (nf-core rna-seq) output, to csv files.

    Args:
        input_path (str): quant_DRR513083.sf files folder path (tsv: tab-separated values)
        output_path (str): folder path to store the csv version of the quant files

    Returns:
        None: This function does not return a value but outputs files to the specified directory.
    """
    # Iterate over files in the directory
    for filename in os.listdir(input_path):
        if filename.endswith(".sf"):
            input_file_path = os.path.join(input_path, filename)
            output_file_path = os.path.join(
                output_path, filename.replace(".sf", ".csv")
            )

            with open(input_file_path, "r", encoding="utf-8") as input_file, open(
                output_file_path, "w", newline="", encoding="utf-8"
            ) as output_file:
                # Create a csv reader object to read from the tsv file (tab-delimited)
                tsv_reader = csv.reader(input_file, delimiter="\t")

                # Create a csv writer object to write to the csv file (comma-delimited)
                csv_writer = csv.writer(output_file)

                # Loop through each row in the input file and write it to the output file
                for row in tsv_reader:
                    csv_writer.writerow(row)</code></pre>
          </details>
  </div>

</div>



  </div>

  </div>

</div>
              
            </div>
          </div><footer>
    <div class="rst-footer-buttons" role="navigation" aria-label="Footer Navigation">
        <a href="../../rna_extraction/" class="btn btn-neutral float-left" title="rna_extraction"><span class="icon icon-circle-arrow-left"></span> Previous</a>
        <a href="../create_expression_matrix/" class="btn btn-neutral float-right" title="create_expression_matrix">Next <span class="icon icon-circle-arrow-right"></span></a>
    </div>

  <hr/>

  <div role="contentinfo">
    <!-- Copyright etc -->
  </div>

  Built with <a href="https://www.mkdocs.org/">MkDocs</a> using a <a href="https://github.com/readthedocs/sphinx_rtd_theme">theme</a> provided by <a href="https://readthedocs.org">Read the Docs</a>.
</footer>
          
        </div>
      </div>

    </section>

  </div>

  <div class="rst-versions" role="note" aria-label="Versions">
  <span class="rst-current-version" data-toggle="rst-current-version">
    
        <span>
          <a href="https://github.com/TransferLearningForDNA/genomic_data_extraction" class="fa fa-github" style="color: #fcfcfc"> GitHub</a>
        </span>
    
    
      <span><a href="../../rna_extraction/" style="color: #fcfcfc">&laquo; Previous</a></span>
    
    
      <span><a href="../create_expression_matrix/" style="color: #fcfcfc">Next &raquo;</a></span>
    
  </span>
</div>
    <script>var base_url = '../../../..';</script>
    <script src="../../../../js/theme_extra.js" defer></script>
    <script src="../../../../js/theme.js" defer></script>
    <script defer>
        window.onload = function () {
            SphinxRtdTheme.Navigation.enable(true);
        };
    </script>

</body>
</html>
