<!DOCTYPE html>
<html class="writer-html5" lang="en" >
<head>
    <meta charset="utf-8" />
    <meta http-equiv="X-UA-Compatible" content="IE=edge" />
    <meta name="viewport" content="width=device-width, initial-scale=1.0" />
      <link rel="shortcut icon" href="../../../img/favicon.ico" />
    <title>ensembl_api - Genomic Data Extraction</title>
    <link rel="stylesheet" href="../../../css/theme.css" />
    <link rel="stylesheet" href="../../../css/theme_extra.css" />
        <link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/highlight.js/10.5.0/styles/github.min.css" />
        <link href="../../../assets/_mkdocstrings.css" rel="stylesheet" />
    
      <script>
        // Current page data
        var mkdocs_page_name = "ensembl_api";
        var mkdocs_page_input_path = "genomic_data_extraction/dna/ensembl_api.md";
        var mkdocs_page_url = null;
      </script>
    
    <script src="../../../js/jquery-3.6.0.min.js" defer></script>
    <!--[if lt IE 9]>
      <script src="../../../js/html5shiv.min.js"></script>
    <![endif]-->
      <script src="https://cdnjs.cloudflare.com/ajax/libs/highlight.js/10.5.0/highlight.min.js"></script>
      <script>hljs.initHighlightingOnLoad();</script> 
</head>

<body class="wy-body-for-nav" role="document">

  <div class="wy-grid-for-nav">
    <nav data-toggle="wy-nav-shift" class="wy-nav-side stickynav">
    <div class="wy-side-scroll">
      <div class="wy-side-nav-search">
          <a href="../../.." class="icon icon-home"> Genomic Data Extraction
        </a>
      </div>

      <div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="Navigation menu">
              <ul>
                <li class="toctree-l1"><a class="reference internal" href="../../..">Home</a>
                </li>
              </ul>
              <p class="caption"><span class="caption-text">genomic_data_extraction</span></p>
              <ul class="current">
                  <li class="toctree-l1 current"><a class="reference internal current" href="#">dna</a>
    <ul class="current">
                <li class="toctree-l2"><a class="reference internal" href="../dna_extraction/">dna_extraction</a>
                </li>
                <li class="toctree-l2"><a class="reference internal" href="../dna_feature_extraction/">dna_feature_extraction</a>
                </li>
                <li class="toctree-l2 current"><a class="reference internal current" href="./">ensembl_api</a>
    <ul class="current">
    </ul>
                </li>
    </ul>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="#">rna</a>
    <ul>
                <li class="toctree-l2"><a class="reference internal" href="../../rna/rna_extraction/">rna_extraction</a>
                </li>
                <li class="toctree-l2"><a class="reference internal" href="#">data_conversion_helper_functions</a>
    <ul>
                <li class="toctree-l3"><a class="reference internal" href="../../rna/data_conversion_helper_functions/convert_quantsf_to_csv/">convert_quantsf_to_csv</a>
                </li>
                <li class="toctree-l3"><a class="reference internal" href="../../rna/data_conversion_helper_functions/create_expression_matrix/">create_expression_matrix</a>
                </li>
                <li class="toctree-l3"><a class="reference internal" href="../../rna/data_conversion_helper_functions/process_expression_matrix/">process_expression_matrix</a>
                </li>
                <li class="toctree-l3"><a class="reference internal" href="../../rna/data_conversion_helper_functions/create_samplesheet_csv/">create_samplesheet_csv</a>
                </li>
    </ul>
                </li>
                <li class="toctree-l2"><a class="reference internal" href="#">rna_download_logic</a>
    <ul>
                <li class="toctree-l3"><a class="reference internal" href="../../rna/rna_download_logic/mRNA_fastq_download/">mRNA_fastq_download</a>
                </li>
                <li class="toctree-l3"><a class="reference internal" href="../../rna/rna_download_logic/query_and_csv_production/">query_and_csv_production</a>
                </li>
    </ul>
                </li>
    </ul>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../../dataset_integration/">dataset_integration</a>
                  </li>
              </ul>
      </div>
    </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap">
      <nav class="wy-nav-top" role="navigation" aria-label="Mobile navigation menu">
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="../../..">Genomic Data Extraction</a>
        
      </nav>
      <div class="wy-nav-content">
        <div class="rst-content"><div role="navigation" aria-label="breadcrumbs navigation">
  <ul class="wy-breadcrumbs">
    <li><a href="../../.." class="icon icon-home" alt="Docs"></a> &raquo;</li>
          <li>genomic_data_extraction &raquo;</li>
          <li>dna &raquo;</li>
      <li>ensembl_api</li>
    <li class="wy-breadcrumbs-aside">
          <a href="https://github.com/TransferLearningForDNA/genomic_data_extraction/edit/master/docs/genomic_data_extraction/dna/ensembl_api.md" class="icon icon-github"> Edit on GitHub</a>
    </li>
  </ul>
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
            <div class="section" itemprop="articleBody">
              
                <div class="doc doc-object doc-module">



<a id="dna.ensembl_api"></a>
  <div class="doc doc-contents first">

  

  <div class="doc doc-children">










<div class="doc doc-object doc-function">



<h2 id="dna.ensembl_api.extract_utr_information" class="doc doc-heading">
          <code class="highlight language-python">extract_utr_information(data)</code>

</h2>


  <div class="doc doc-contents ">
  
      <p>Extracts information about 5' and 3' UTRs (Untranslated Regions) from the provided data.</p>



<table class="field-list">
  <colgroup>
    <col class="field-name" />
    <col class="field-body" />
  </colgroup>
  <tbody valign="top">
    <tr class="field">
      <th class="field-name">Parameters:</th>
      <td class="field-body">
        <ul class="first simple">
            <li>
              <b><code>data</code></b>
                  (<code>dict</code>)
              –
              <div class="doc-md-description">
                <p>A dictionary containing information about UTRs.</p>
              </div>
            </li>
        </ul>
      </td>
    </tr>
  </tbody>
</table>


<table class="field-list">
  <colgroup>
    <col class="field-name" />
    <col class="field-body" />
  </colgroup>
  <tbody valign="top">
    <tr class="field">
    <th class="field-name">Returns:</th>
    <td class="field-body">
      <ul class="first simple">
          <li>
<b><code>tuple</code></b>(                <code><span title="typing.Tuple">Tuple</span>[<span title="typing.List">List</span>[<span title="typing.Tuple">Tuple</span>[int, int]], <span title="typing.List">List</span>[<span title="typing.Tuple">Tuple</span>[int, int]], <span title="typing.Optional">Optional</span>[str], <span title="typing.Optional">Optional</span>[int]]</code>
)            –
            <div class="doc-md-description">
              <p>A tuple containing lists of 5' UTRs, 3' UTRs, chromosome, and strand information.</p>
            </div>
          </li>
      </ul>
    </td>
    </tr>
  </tbody>
</table>
          <details class="quote">
            <summary>Source code in <code>dna/ensembl_api.py</code></summary>
            <pre class="highlight"><code class="language-python">def extract_utr_information(
    data: dict,
) -&gt; Tuple[List[Tuple[int, int]], List[Tuple[int, int]], Optional[str], Optional[int]]:
    """Extracts information about 5' and 3' UTRs (Untranslated Regions) from the provided data.

    Args:
        data (dict): A dictionary containing information about UTRs.

    Returns:
        tuple: A tuple containing lists of 5' UTRs, 3' UTRs, chromosome, and strand information.
    """
    # Retrieve UTR data from the input dictionary
    utr_data = data.get("UTR", [])

    # Initialize lists to store 5' UTR and 3' UTR information
    utr5_list = []
    utr3_list = []

    # Iterate through UTR entries in the data
    for utr_entry in utr_data:
        utr_type = utr_entry.get("type", "")
        utr_start = utr_entry.get("start", None)
        utr_end = utr_entry.get("end", None)

        # Check if the UTR entry has valid type, start, and end information
        if utr_type and utr_start is not None and utr_end is not None:
            # Categorize UTRs into 5' and 3' UTR lists
            if utr_type == "five_prime_utr":
                utr5_list.append((utr_start, utr_end))
            elif utr_type == "three_prime_utr":
                utr3_list.append((utr_start, utr_end))

    # Extract chromosome and strand information
    chromosome = utr_entry["seq_region_name"] if utr_data else None
    strand = data.get("strand", None)

    return utr5_list, utr3_list, chromosome, strand</code></pre>
          </details>
  </div>

</div>


<div class="doc doc-object doc-function">



<h2 id="dna.ensembl_api.get_cds" class="doc doc-heading">
          <code class="highlight language-python">get_cds(transcript_id)</code>

</h2>


  <div class="doc doc-contents ">
  
      <p>Retrieves the coding sequence (CDS) for a given Ensembl transcript ID.</p>



<table class="field-list">
  <colgroup>
    <col class="field-name" />
    <col class="field-body" />
  </colgroup>
  <tbody valign="top">
    <tr class="field">
      <th class="field-name">Parameters:</th>
      <td class="field-body">
        <ul class="first simple">
            <li>
              <b><code>transcript_id</code></b>
                  (<code>str</code>)
              –
              <div class="doc-md-description">
                <p>Ensembl transcript ID for the target gene.</p>
              </div>
            </li>
        </ul>
      </td>
    </tr>
  </tbody>
</table>


<table class="field-list">
  <colgroup>
    <col class="field-name" />
    <col class="field-body" />
  </colgroup>
  <tbody valign="top">
    <tr class="field">
    <th class="field-name">Returns:</th>
    <td class="field-body">
      <ul class="first simple">
          <li>
<b><code>str</code></b>(                <code>str</code>
)            –
            <div class="doc-md-description">
              <p>The nucleotide sequence of the coding sequence (CDS).
 Returns an empty string in case of an error.</p>
            </div>
          </li>
      </ul>
    </td>
    </tr>
  </tbody>
</table>
          <details class="quote">
            <summary>Source code in <code>dna/ensembl_api.py</code></summary>
            <pre class="highlight"><code class="language-python">def get_cds(transcript_id: str) -&gt; str:
    """Retrieves the coding sequence (CDS) for a given Ensembl transcript ID.

    Args:
        transcript_id (str): Ensembl transcript ID for the target gene.

    Returns:
        str: The nucleotide sequence of the coding sequence (CDS).
             Returns an empty string in case of an error.
    """
    # Construct the REST API URL for retrieving CDS
    address = f"https://rest.ensembl.org/sequence/id/{transcript_id}?multiple_sequences=1;type=cds"

    try:
        # Make a GET request to the Ensembl REST API
        r = requests.get(address, headers={"Content-Type": "text/x-fasta"}, timeout=30)

        # Ensure that there are no issues with the sequence request
        r.raise_for_status()

        # Extract only the nucleotide sequence and format into a single string
        raw_output = r.text
        pattern = re.compile("(?:^|\n)[ATGC]+")
        matches = pattern.findall(raw_output)
        cds_sequence = "".join(matches).replace("\n", "")

        return cds_sequence

    except requests.exceptions.RequestException as e:
        # If there's an error with the request, print the error and return an empty string
        print(f"Error with the request for {transcript_id}: {e}")
        return ""</code></pre>
          </details>
  </div>

</div>


<div class="doc doc-object doc-function">



<h2 id="dna.ensembl_api.get_data_as_csv" class="doc doc-heading">
          <code class="highlight language-python">get_data_as_csv(file_paths, output_directory)</code>

</h2>


  <div class="doc doc-contents ">
  
      <p>Retrieves data for gene IDs from Ensembl, processes it, and saves it as CSV files.</p>



<table class="field-list">
  <colgroup>
    <col class="field-name" />
    <col class="field-body" />
  </colgroup>
  <tbody valign="top">
    <tr class="field">
      <th class="field-name">Parameters:</th>
      <td class="field-body">
        <ul class="first simple">
            <li>
              <b><code>file_paths</code></b>
                  (<code><span title="typing.List">List</span>[str]</code>)
              –
              <div class="doc-md-description">
                <p>List of file paths containing gene IDs.</p>
              </div>
            </li>
            <li>
              <b><code>output_directory</code></b>
                  (<code>str</code>)
              –
              <div class="doc-md-description">
                <p>Directory where CSV files will be saved.</p>
              </div>
            </li>
        </ul>
      </td>
    </tr>
  </tbody>
</table>


<table class="field-list">
  <colgroup>
    <col class="field-name" />
    <col class="field-body" />
  </colgroup>
  <tbody valign="top">
    <tr class="field">
    <th class="field-name">Returns:</th>
    <td class="field-body">
      <ul class="first simple">
          <li>
<b><code>None</code></b>(                <code>None</code>
)            –
            <div class="doc-md-description">
              <p>This function does not return a value but outputs files to the specified directory.</p>
            </div>
          </li>
      </ul>
    </td>
    </tr>
  </tbody>
</table>
          <details class="quote">
            <summary>Source code in <code>dna/ensembl_api.py</code></summary>
            <pre class="highlight"><code class="language-python">def get_data_as_csv(file_paths: List[str], output_directory: str) -&gt; None:
    """Retrieves data for gene IDs from Ensembl, processes it, and saves it as CSV files.

    Args:
        file_paths (List[str]): List of file paths containing gene IDs.
        output_directory (str): Directory where CSV files will be saved.

    Returns:
        None: This function does not return a value but outputs files to the specified directory.
    """
    # Create the directory if it doesn't exist
    os.makedirs(output_directory, exist_ok=True)

    for file_path in file_paths:
        # Read gene IDs from the file
        gene_ids = read_gene_ids_from_file(file_path)

        # Generate output filename and species name
        filename = "ensembl_data_" + get_species_name(file_path) + ".csv"
        species = " ".join(get_species_name(file_path).split("_"))

        print(f"Starting data extraction for {species}.")

        # Initialize a CSV writer
        filename = os.path.join(output_directory, filename)
        csv_file = open(filename, "w", newline="", encoding="utf-8")
        csv_writer = csv.writer(csv_file)

        # Write the header to the CSV file
        csv_writer.writerow(
            [
                "ensembl_gene_id",
                "transcript_id",
                "promoter",
                "utr5",
                "cds",
                "utr3",
                "terminator",
            ]
        )

        # Loop through each gene ID and retrieve the data
        for gene_id in gene_ids:
            time.sleep(2)
            print(f"Extracting data for gene ID : {gene_id}")

            try:
                gene_data = ensembl_rest.lookup(species=species, id=gene_id)

            except ensembl_rest.core.restclient.HTTPError as e:
                if e.response.status_code == 429:  # Check for rate limit exceeded error
                    print("Rate limit exceeded. Waiting before retrying...")
                    time.sleep(1)
                    gene_data = ensembl_rest.lookup(species=species, id=gene_id)

            # Get transcript ID
            transcript_id = gene_data["canonical_transcript"].split(".")[0]

            # Retrieve promoter, CDS, and terminator sequences
            cds_sequence = get_cds(transcript_id)
            if cds_sequence == "":
                continue
            promoter_sequence, terminator_sequence = get_promoter_terminator(
                transcript_id
            )

            # Retrieve UTR sequences
            transcript_data = request_with_retry(transcript_id)
            if transcript_data == {}:
                continue

            utr5_coord_list, utr3_coord_list, chromosome, strand = (
                extract_utr_information(transcript_data)
            )
            utr5_sequence = get_full_utr_sequence(
                utr5_coord_list, chromosome, strand, species=get_species_name(file_path)
            )
            utr3_sequence = get_full_utr_sequence(
                utr3_coord_list, chromosome, strand, species=get_species_name(file_path)
            )

            # Write the row to the CSV file
            csv_writer.writerow(
                [
                    gene_id,
                    transcript_id,
                    promoter_sequence,
                    utr5_sequence,
                    cds_sequence,
                    utr3_sequence,
                    terminator_sequence,
                ]
            )

        # Close the CSV file
        csv_file.close()

        print(f"Data extraction for {species} is now complete.")</code></pre>
          </details>
  </div>

</div>


<div class="doc doc-object doc-function">



<h2 id="dna.ensembl_api.get_full_utr_sequence" class="doc doc-heading">
          <code class="highlight language-python">get_full_utr_sequence(list_utr_coordinates, chromosome, strand, species)</code>

</h2>


  <div class="doc doc-contents ">
  
      <p>Retrieves the concatenated nucleotide sequence of multiple UTRs from the Ensembl database.</p>



<table class="field-list">
  <colgroup>
    <col class="field-name" />
    <col class="field-body" />
  </colgroup>
  <tbody valign="top">
    <tr class="field">
      <th class="field-name">Parameters:</th>
      <td class="field-body">
        <ul class="first simple">
            <li>
              <b><code>list_utr_coordinates</code></b>
                  (<code><span title="typing.List">List</span>[<span title="typing.Tuple">Tuple</span>[int, int]]</code>)
              –
              <div class="doc-md-description">
                <p>A list of tuples representing UTR start and end coordinates.</p>
              </div>
            </li>
            <li>
              <b><code>chromosome</code></b>
                  (<code>str</code>)
              –
              <div class="doc-md-description">
                <p>Chromosome name or identifier.</p>
              </div>
            </li>
            <li>
              <b><code>strand</code></b>
                  (<code>int</code>)
              –
              <div class="doc-md-description">
                <p>Strand information (1 for forward strand, -1 for reverse strand).</p>
              </div>
            </li>
        </ul>
      </td>
    </tr>
  </tbody>
</table>


<table class="field-list">
  <colgroup>
    <col class="field-name" />
    <col class="field-body" />
  </colgroup>
  <tbody valign="top">
    <tr class="field">
    <th class="field-name">Returns:</th>
    <td class="field-body">
      <ul class="first simple">
          <li>
<b><code>str</code></b>(                <code>str</code>
)            –
            <div class="doc-md-description">
              <p>The concatenated nucleotide sequence of the specified UTRs.</p>
            </div>
          </li>
      </ul>
    </td>
    </tr>
  </tbody>
</table>
          <details class="quote">
            <summary>Source code in <code>dna/ensembl_api.py</code></summary>
            <pre class="highlight"><code class="language-python">def get_full_utr_sequence(
    list_utr_coordinates: List[Tuple[int, int]],
    chromosome: str,
    strand: int,
    species: str,
) -&gt; str:
    """Retrieves the concatenated nucleotide sequence of multiple UTRs from the Ensembl database.

    Args:
        list_utr_coordinates (List[Tuple[int, int]]): A list of tuples representing UTR start and end coordinates.
        chromosome (str): Chromosome name or identifier.
        strand (int): Strand information (1 for forward strand, -1 for reverse strand).

    Returns:
        str: The concatenated nucleotide sequence of the specified UTRs.
    """
    # Initialize an empty string to store the concatenated UTR sequence
    concatenated_sequence = ""

    # Iterate through UTR coordinates and retrieve individual UTR sequences
    for start, end in list_utr_coordinates:
        sequence = get_utr_sequence(chromosome, strand, start, end, species)
        concatenated_sequence += sequence

    return concatenated_sequence</code></pre>
          </details>
  </div>

</div>


<div class="doc doc-object doc-function">



<h2 id="dna.ensembl_api.get_promoter_terminator" class="doc doc-heading">
          <code class="highlight language-python">get_promoter_terminator(transcript_id, promoter_length=1000, terminator_length=500)</code>

</h2>


  <div class="doc doc-contents ">
  
      <p>Retrieves the promoter and terminator sequences for a given Ensembl transcript ID.</p>



<table class="field-list">
  <colgroup>
    <col class="field-name" />
    <col class="field-body" />
  </colgroup>
  <tbody valign="top">
    <tr class="field">
      <th class="field-name">Parameters:</th>
      <td class="field-body">
        <ul class="first simple">
            <li>
              <b><code>transcript_id</code></b>
                  (<code>str</code>)
              –
              <div class="doc-md-description">
                <p>Ensembl transcript ID for the target gene.</p>
              </div>
            </li>
            <li>
              <b><code>promoter_length</code></b>
                  (<code><span title="typing.Optional">Optional</span>[int]</code>, default:
                      <code>1000</code>
)
              –
              <div class="doc-md-description">
                <p>Length of the promoter sequence (default is 1000).</p>
              </div>
            </li>
            <li>
              <b><code>terminator_length</code></b>
                  (<code><span title="typing.Optional">Optional</span>[int]</code>, default:
                      <code>500</code>
)
              –
              <div class="doc-md-description">
                <p>Length of the terminator sequence (default is 500).</p>
              </div>
            </li>
        </ul>
      </td>
    </tr>
  </tbody>
</table>


<table class="field-list">
  <colgroup>
    <col class="field-name" />
    <col class="field-body" />
  </colgroup>
  <tbody valign="top">
    <tr class="field">
    <th class="field-name">Returns:</th>
    <td class="field-body">
      <ul class="first simple">
          <li>
                <code><span title="typing.Tuple">Tuple</span>[str, str]</code>
            –
            <div class="doc-md-description">
              <p>Tuple[str, str]: A tuple containing the promoter and terminator sequences as strings.</p>
            </div>
          </li>
      </ul>
    </td>
    </tr>
  </tbody>
</table>
          <details class="quote">
            <summary>Source code in <code>dna/ensembl_api.py</code></summary>
            <pre class="highlight"><code class="language-python">def get_promoter_terminator(
    transcript_id: str,
    promoter_length: Optional[int] = 1000,
    terminator_length: Optional[int] = 500,
) -&gt; Tuple[str, str]:
    """Retrieves the promoter and terminator sequences for a given Ensembl transcript ID.

    Args:
        transcript_id (str): Ensembl transcript ID for the target gene.
        promoter_length (Optional[int]): Length of the promoter sequence (default is 1000).
        terminator_length (Optional[int]): Length of the terminator sequence (default is 500).

    Returns:
        Tuple[str, str]: A tuple containing the promoter and terminator sequences as strings.
    """
    # Construct the REST API URL for retrieving genomic sequence with specified 5' and 3' expansions
    address = f"https://rest.ensembl.org/sequence/id/{transcript_id}?type=genomic;expand_5prime=1000;expand_3prime=500"

    try:
        # Make a GET request to the Ensembl REST API
        r = requests.get(address, headers={"Content-Type": "text/x-fasta"}, timeout=30)

        # Ensure that there are no issues with the sequence request
        r.raise_for_status()

        # Remove unwanted characters to produce only the nucleotide sequence and format into a single string
        raw_output = r.text
        pattern = re.compile("(?:^|\n)[ATGC]+")
        matches = pattern.findall(raw_output)
        sequence = "".join(matches).replace("\n", "")

        # Extract the promoter and terminator sequence from the entire sequence
        promoter_sequence = sequence[:promoter_length]
        terminator_sequence = sequence[-terminator_length:]

        return promoter_sequence, terminator_sequence

    except requests.exceptions.RequestException as e:
        # If there's an error with the request, print the error and return an empty string
        print(f"Error with the promoter-terminator request for {transcript_id}: {e}")
        return "", ""</code></pre>
          </details>
  </div>

</div>


<div class="doc doc-object doc-function">



<h2 id="dna.ensembl_api.get_species_name" class="doc doc-heading">
          <code class="highlight language-python">get_species_name(file_path)</code>

</h2>


  <div class="doc doc-contents ">
  
      <p>Extracts the species name from a file path.</p>



<table class="field-list">
  <colgroup>
    <col class="field-name" />
    <col class="field-body" />
  </colgroup>
  <tbody valign="top">
    <tr class="field">
      <th class="field-name">Parameters:</th>
      <td class="field-body">
        <ul class="first simple">
            <li>
              <b><code>file_path</code></b>
                  (<code>str</code>)
              –
              <div class="doc-md-description">
                <p>The path to the file containing the species information.</p>
              </div>
            </li>
        </ul>
      </td>
    </tr>
  </tbody>
</table>


<table class="field-list">
  <colgroup>
    <col class="field-name" />
    <col class="field-body" />
  </colgroup>
  <tbody valign="top">
    <tr class="field">
    <th class="field-name">Returns:</th>
    <td class="field-body">
      <ul class="first simple">
          <li>
<b><code>str</code></b>(                <code>str</code>
)            –
            <div class="doc-md-description">
              <p>The extracted species name.</p>
            </div>
          </li>
      </ul>
    </td>
    </tr>
  </tbody>
</table>
          <details class="quote">
            <summary>Source code in <code>dna/ensembl_api.py</code></summary>
            <pre class="highlight"><code class="language-python">def get_species_name(file_path: str) -&gt; str:
    """Extracts the species name from a file path.

    Args:
        file_path (str): The path to the file containing the species information.

    Returns:
        str: The extracted species name.
    """
    # Extract the filename from the path
    filename = os.path.basename(file_path)

    # Remove the extension
    filename_without_extension = filename.rsplit(".", 1)[0]

    # Split the filename by underscore and get the last two parts
    species = "_".join(filename_without_extension.split("_")[:2])

    return species</code></pre>
          </details>
  </div>

</div>


<div class="doc doc-object doc-function">



<h2 id="dna.ensembl_api.get_utr_sequence" class="doc doc-heading">
          <code class="highlight language-python">get_utr_sequence(chromosome, strand, start, end, species)</code>

</h2>


  <div class="doc doc-contents ">
  
      <p>Retrieves the nucleotide sequence of a UTR (Untranslated Region) from the Ensembl database.</p>



<table class="field-list">
  <colgroup>
    <col class="field-name" />
    <col class="field-body" />
  </colgroup>
  <tbody valign="top">
    <tr class="field">
      <th class="field-name">Parameters:</th>
      <td class="field-body">
        <ul class="first simple">
            <li>
              <b><code>chromosome</code></b>
                  (<code>str</code>)
              –
              <div class="doc-md-description">
                <p>Chromosome name or identifier.</p>
              </div>
            </li>
            <li>
              <b><code>strand</code></b>
                  (<code>int</code>)
              –
              <div class="doc-md-description">
                <p>Strand information (1 for forward strand, -1 for reverse strand).</p>
              </div>
            </li>
            <li>
              <b><code>start</code></b>
                  (<code>int</code>)
              –
              <div class="doc-md-description">
                <p>Start position of the UTR on the chromosome.</p>
              </div>
            </li>
            <li>
              <b><code>end</code></b>
                  (<code>int</code>)
              –
              <div class="doc-md-description">
                <p>End position of the UTR on the chromosome.</p>
              </div>
            </li>
            <li>
              <b><code>species</code></b>
                  (<code>str</code>)
              –
              <div class="doc-md-description">
                <p>Species for which the UTR sequence is requested.</p>
              </div>
            </li>
        </ul>
      </td>
    </tr>
  </tbody>
</table>


<table class="field-list">
  <colgroup>
    <col class="field-name" />
    <col class="field-body" />
  </colgroup>
  <tbody valign="top">
    <tr class="field">
    <th class="field-name">Returns:</th>
    <td class="field-body">
      <ul class="first simple">
          <li>
<b><code>str</code></b>(                <code>str</code>
)            –
            <div class="doc-md-description">
              <p>The nucleotide sequence of the specified UTR.</p>
            </div>
          </li>
      </ul>
    </td>
    </tr>
  </tbody>
</table>
          <details class="quote">
            <summary>Source code in <code>dna/ensembl_api.py</code></summary>
            <pre class="highlight"><code class="language-python">def get_utr_sequence(
    chromosome: str, strand: int, start: int, end: int, species: str
) -&gt; str:
    """Retrieves the nucleotide sequence of a UTR (Untranslated Region) from the Ensembl database.

    Args:
        chromosome (str): Chromosome name or identifier.
        strand (int): Strand information (1 for forward strand, -1 for reverse strand).
        start (int): Start position of the UTR on the chromosome.
        end (int): End position of the UTR on the chromosome.
        species (str): Species for which the UTR sequence is requested.

    Returns:
        str: The nucleotide sequence of the specified UTR.
    """
    # Use Ensembl REST API to retrieve UTR sequence for the specified region
    region = f"{chromosome}:{start}..{end}:{strand}"
    try:
        utr_sequence = ensembl_rest.sequence_region(region=region, species=species)[
            "seq"
        ]
    except ensembl_rest.core.restclient.HTTPError as e:
        if e.response.status_code == 429:  # Check for rate limit exceeded error
            print("Rate limit exceeded. Waiting before retrying...")
            time.sleep(1)
            utr_sequence = get_utr_sequence(chromosome, strand, start, end, species)

    return utr_sequence</code></pre>
          </details>
  </div>

</div>


<div class="doc doc-object doc-function">



<h2 id="dna.ensembl_api.read_gene_ids_from_file" class="doc doc-heading">
          <code class="highlight language-python">read_gene_ids_from_file(file_path)</code>

</h2>


  <div class="doc doc-contents ">
  
      <p>Reads gene IDs from a file, skipping the first line.</p>



<table class="field-list">
  <colgroup>
    <col class="field-name" />
    <col class="field-body" />
  </colgroup>
  <tbody valign="top">
    <tr class="field">
      <th class="field-name">Parameters:</th>
      <td class="field-body">
        <ul class="first simple">
            <li>
              <b><code>file_path</code></b>
                  (<code>str</code>)
              –
              <div class="doc-md-description">
                <p>The path to the file containing gene IDs.</p>
              </div>
            </li>
        </ul>
      </td>
    </tr>
  </tbody>
</table>


<table class="field-list">
  <colgroup>
    <col class="field-name" />
    <col class="field-body" />
  </colgroup>
  <tbody valign="top">
    <tr class="field">
    <th class="field-name">Returns:</th>
    <td class="field-body">
      <ul class="first simple">
          <li>
                <code><span title="typing.List">List</span>[str]</code>
            –
            <div class="doc-md-description">
              <p>List[str]: A list of gene IDs read from the file.</p>
            </div>
          </li>
      </ul>
    </td>
    </tr>
  </tbody>
</table>
          <details class="quote">
            <summary>Source code in <code>dna/ensembl_api.py</code></summary>
            <pre class="highlight"><code class="language-python">def read_gene_ids_from_file(file_path: str) -&gt; List[str]:
    """Reads gene IDs from a file, skipping the first line.

    Args:
        file_path (str): The path to the file containing gene IDs.

    Returns:
        List[str]: A list of gene IDs read from the file.
    """
    try:
        with open(file_path, "r", encoding="utf-8") as file:
            # Skip the first line
            next(file)

            # Read gene IDs from the remaining lines
            gene_ids = [line.strip() for line in file]

        return gene_ids
    except FileNotFoundError:
        # Handle file not found error
        print(f"Error: File '{file_path}' not found.")
        return []
    except Exception as e:
        # Handle other exceptions
        print(f"Error: {str(e)}")
        return []</code></pre>
          </details>
  </div>

</div>


<div class="doc doc-object doc-function">



<h2 id="dna.ensembl_api.request_with_retry" class="doc doc-heading">
          <code class="highlight language-python">request_with_retry(transcript_id)</code>

</h2>


  <div class="doc doc-contents ">
  
      <p>Retries the request for a transcript ID</p>



<table class="field-list">
  <colgroup>
    <col class="field-name" />
    <col class="field-body" />
  </colgroup>
  <tbody valign="top">
    <tr class="field">
      <th class="field-name">Parameters:</th>
      <td class="field-body">
        <ul class="first simple">
            <li>
              <b><code>transcript_id</code></b>
                  (<code>str</code>)
              –
              <div class="doc-md-description">
                <p>Ensembl transcript ID for the target gene.</p>
              </div>
            </li>
        </ul>
      </td>
    </tr>
  </tbody>
</table>


<table class="field-list">
  <colgroup>
    <col class="field-name" />
    <col class="field-body" />
  </colgroup>
  <tbody valign="top">
    <tr class="field">
    <th class="field-name">Returns:</th>
    <td class="field-body">
      <ul class="first simple">
          <li>
<b><code>dict</code></b>(                <code><span title="typing.Dict">Dict</span></code>
)            –
            <div class="doc-md-description">
              <p>A dictionary containing information about the transcript.</p>
            </div>
          </li>
      </ul>
    </td>
    </tr>
  </tbody>
</table>
          <details class="quote">
            <summary>Source code in <code>dna/ensembl_api.py</code></summary>
            <pre class="highlight"><code class="language-python">def request_with_retry(transcript_id: str) -&gt; Dict:
    """Retries the request for a transcript ID

    Args:
        transcript_id (str): Ensembl transcript ID for the target gene.

    Returns:
        dict: A dictionary containing information about the transcript.
    """
    while True:
        try:
            transcript_data = ensembl_rest.lookup(
                id=transcript_id, params={"expand": True, "utr": True}
            )
            return transcript_data
        except ensembl_rest.core.restclient.HTTPError as e:
            if e.response.status_code == 429:  # Check for rate limit exceeded error
                print("Rate limit exceeded. Waiting before retrying...")
                time.sleep(1)
                continue

            print(f"Error with the request for {transcript_id}: {e}")
            return {}</code></pre>
          </details>
  </div>

</div>



  </div>

  </div>

</div>
              
            </div>
          </div><footer>
    <div class="rst-footer-buttons" role="navigation" aria-label="Footer Navigation">
        <a href="../dna_feature_extraction/" class="btn btn-neutral float-left" title="dna_feature_extraction"><span class="icon icon-circle-arrow-left"></span> Previous</a>
        <a href="../../rna/rna_extraction/" class="btn btn-neutral float-right" title="rna_extraction">Next <span class="icon icon-circle-arrow-right"></span></a>
    </div>

  <hr/>

  <div role="contentinfo">
    <!-- Copyright etc -->
  </div>

  Built with <a href="https://www.mkdocs.org/">MkDocs</a> using a <a href="https://github.com/readthedocs/sphinx_rtd_theme">theme</a> provided by <a href="https://readthedocs.org">Read the Docs</a>.
</footer>
          
        </div>
      </div>

    </section>

  </div>

  <div class="rst-versions" role="note" aria-label="Versions">
  <span class="rst-current-version" data-toggle="rst-current-version">
    
        <span>
          <a href="https://github.com/TransferLearningForDNA/genomic_data_extraction" class="fa fa-github" style="color: #fcfcfc"> GitHub</a>
        </span>
    
    
      <span><a href="../dna_feature_extraction/" style="color: #fcfcfc">&laquo; Previous</a></span>
    
    
      <span><a href="../../rna/rna_extraction/" style="color: #fcfcfc">Next &raquo;</a></span>
    
  </span>
</div>
    <script>var base_url = '../../..';</script>
    <script src="../../../js/theme_extra.js" defer></script>
    <script src="../../../js/theme.js" defer></script>
    <script defer>
        window.onload = function () {
            SphinxRtdTheme.Navigation.enable(true);
        };
    </script>

</body>
</html>
