<!DOCTYPE html>
<html class="writer-html5" lang="en" >
<head>
    <meta charset="utf-8" />
    <meta http-equiv="X-UA-Compatible" content="IE=edge" />
    <meta name="viewport" content="width=device-width, initial-scale=1.0" />
      <link rel="shortcut icon" href="../../../img/favicon.ico" />
    <title>ensembl_api - Genomic Data Extraction</title>
    <link rel="stylesheet" href="../../../css/theme.css" />
    <link rel="stylesheet" href="../../../css/theme_extra.css" />
        <link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/highlight.js/10.5.0/styles/github.min.css" />
        <link href="../../../assets/_mkdocstrings.css" rel="stylesheet" />
    
      <script>
        // Current page data
        var mkdocs_page_name = "ensembl_api";
        var mkdocs_page_input_path = "genomic_data_extraction\\dna\\ensembl_api.md";
        var mkdocs_page_url = null;
      </script>
    
    <script src="../../../js/jquery-3.6.0.min.js" defer></script>
    <!--[if lt IE 9]>
      <script src="../../../js/html5shiv.min.js"></script>
    <![endif]-->
      <script src="https://cdnjs.cloudflare.com/ajax/libs/highlight.js/10.5.0/highlight.min.js"></script>
      <script>hljs.initHighlightingOnLoad();</script> 
</head>

<body class="wy-body-for-nav" role="document">

  <div class="wy-grid-for-nav">
    <nav data-toggle="wy-nav-shift" class="wy-nav-side stickynav">
    <div class="wy-side-scroll">
      <div class="wy-side-nav-search">
          <a href="../../.." class="icon icon-home"> Genomic Data Extraction
        </a>
      </div>

      <div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="Navigation menu">
              <ul>
                <li class="toctree-l1"><a class="reference internal" href="../../..">Home</a>
                </li>
              </ul>
              <p class="caption"><span class="caption-text">genomic_data_extraction</span></p>
              <ul class="current">
                  <li class="toctree-l1 current"><a class="reference internal current" href="#">dna</a>
    <ul class="current">
                <li class="toctree-l2"><a class="reference internal" href="../dna_extraction/">dna_extraction</a>
                </li>
                <li class="toctree-l2"><a class="reference internal" href="../dna_feature_extraction/">dna_feature_extraction</a>
                </li>
                <li class="toctree-l2 current"><a class="reference internal current" href="./">ensembl_api</a>
    <ul class="current">
    </ul>
                </li>
    </ul>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="#">rna</a>
    <ul>
                <li class="toctree-l2"><a class="reference internal" href="../../rna/rna_extraction/">rna_extraction</a>
                </li>
                <li class="toctree-l2"><a class="reference internal" href="#">data_conversion_helper_functions</a>
    <ul>
                <li class="toctree-l3"><a class="reference internal" href="../../rna/data_conversion_helper_functions/convert_quantsf_to_csv/">convert_quantsf_to_csv</a>
                </li>
                <li class="toctree-l3"><a class="reference internal" href="../../rna/data_conversion_helper_functions/create_expression_matrix/">create_expression_matrix</a>
                </li>
                <li class="toctree-l3"><a class="reference internal" href="../../rna/data_conversion_helper_functions/process_expression_matrix/">process_expression_matrix</a>
                </li>
                <li class="toctree-l3"><a class="reference internal" href="../../rna/data_conversion_helper_functions/create_samplesheet_csv/">create_samplesheet_csv</a>
                </li>
    </ul>
                </li>
                <li class="toctree-l2"><a class="reference internal" href="#">rna_download_logic</a>
    <ul>
                <li class="toctree-l3"><a class="reference internal" href="../../rna/rna_download_logic/mRNA_fastq_download/">mRNA_fastq_download</a>
                </li>
                <li class="toctree-l3"><a class="reference internal" href="../../rna/rna_download_logic/query_and_csv_production/">query_and_csv_production</a>
                </li>
    </ul>
                </li>
    </ul>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../../dataset_integration/">dataset_integration</a>
                  </li>
              </ul>
      </div>
    </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap">
      <nav class="wy-nav-top" role="navigation" aria-label="Mobile navigation menu">
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="../../..">Genomic Data Extraction</a>
        
      </nav>
      <div class="wy-nav-content">
        <div class="rst-content"><div role="navigation" aria-label="breadcrumbs navigation">
  <ul class="wy-breadcrumbs">
    <li><a href="../../.." class="icon icon-home" alt="Docs"></a> &raquo;</li>
          <li>genomic_data_extraction &raquo;</li>
          <li>dna &raquo;</li>
      <li>ensembl_api</li>
    <li class="wy-breadcrumbs-aside">
          <a href="https://github.com/TransferLearningForDNA/genomic_data_extraction/edit/master/docs/genomic_data_extraction/dna/ensembl_api.md" class="icon icon-github"> Edit on GitHub</a>
    </li>
  </ul>
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
            <div class="section" itemprop="articleBody">
              
                <div class="doc doc-object doc-module">



<a id="dna.ensembl_api"></a>
  <div class="doc doc-contents first">

  

  <div class="doc doc-children">










<div class="doc doc-object doc-function">



<h2 id="dna.ensembl_api.extract_utr_information" class="doc doc-heading">
          <code class="highlight language-python"><span class="n">extract_utr_information</span><span class="p">(</span><span class="n">data</span><span class="p">)</span></code>

</h2>


  <div class="doc doc-contents ">
  
      <p>Extracts information about 5' and 3' UTRs (Untranslated Regions) from the provided data.</p>



<table class="field-list">
  <colgroup>
    <col class="field-name" />
    <col class="field-body" />
  </colgroup>
  <tbody valign="top">
    <tr class="field">
      <th class="field-name">Parameters:</th>
      <td class="field-body">
        <ul class="first simple">
            <li>
              <b><code>data</code></b>
                  (<code>dict</code>)
              –
              <div class="doc-md-description">
                <p>A dictionary containing information about UTRs.</p>
              </div>
            </li>
        </ul>
      </td>
    </tr>
  </tbody>
</table>


<table class="field-list">
  <colgroup>
    <col class="field-name" />
    <col class="field-body" />
  </colgroup>
  <tbody valign="top">
    <tr class="field">
    <th class="field-name">Returns:</th>
    <td class="field-body">
      <ul class="first simple">
          <li>
<b><code>tuple</code></b>(                <code><span title="typing.Tuple">Tuple</span>[<span title="typing.List">List</span>[<span title="typing.Tuple">Tuple</span>[int, int]], <span title="typing.List">List</span>[<span title="typing.Tuple">Tuple</span>[int, int]], <span title="typing.Optional">Optional</span>[str], <span title="typing.Optional">Optional</span>[int]]</code>
)            –
            <div class="doc-md-description">
              <p>A tuple containing lists of 5' UTRs, 3' UTRs, chromosome, and strand information.</p>
            </div>
          </li>
      </ul>
    </td>
    </tr>
  </tbody>
</table>
          <details class="quote">
            <summary>Source code in <code>dna\ensembl_api.py</code></summary>
            <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">115</span>
<span class="normal">116</span>
<span class="normal">117</span>
<span class="normal">118</span>
<span class="normal">119</span>
<span class="normal">120</span>
<span class="normal">121</span>
<span class="normal">122</span>
<span class="normal">123</span>
<span class="normal">124</span>
<span class="normal">125</span>
<span class="normal">126</span>
<span class="normal">127</span>
<span class="normal">128</span>
<span class="normal">129</span>
<span class="normal">130</span>
<span class="normal">131</span>
<span class="normal">132</span>
<span class="normal">133</span>
<span class="normal">134</span>
<span class="normal">135</span>
<span class="normal">136</span>
<span class="normal">137</span>
<span class="normal">138</span>
<span class="normal">139</span>
<span class="normal">140</span>
<span class="normal">141</span>
<span class="normal">142</span>
<span class="normal">143</span>
<span class="normal">144</span>
<span class="normal">145</span>
<span class="normal">146</span>
<span class="normal">147</span>
<span class="normal">148</span>
<span class="normal">149</span>
<span class="normal">150</span>
<span class="normal">151</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span> <span class="nf">extract_utr_information</span><span class="p">(</span>
    <span class="n">data</span><span class="p">:</span> <span class="nb">dict</span><span class="p">,</span>
<span class="p">)</span> <span class="o">-&gt;</span> <span class="n">Tuple</span><span class="p">[</span><span class="n">List</span><span class="p">[</span><span class="n">Tuple</span><span class="p">[</span><span class="nb">int</span><span class="p">,</span> <span class="nb">int</span><span class="p">]],</span> <span class="n">List</span><span class="p">[</span><span class="n">Tuple</span><span class="p">[</span><span class="nb">int</span><span class="p">,</span> <span class="nb">int</span><span class="p">]],</span> <span class="n">Optional</span><span class="p">[</span><span class="nb">str</span><span class="p">],</span> <span class="n">Optional</span><span class="p">[</span><span class="nb">int</span><span class="p">]]:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Extracts information about 5&#39; and 3&#39; UTRs (Untranslated Regions) from the provided data.</span>

<span class="sd">    Args:</span>
<span class="sd">        data (dict): A dictionary containing information about UTRs.</span>

<span class="sd">    Returns:</span>
<span class="sd">        tuple: A tuple containing lists of 5&#39; UTRs, 3&#39; UTRs, chromosome, and strand information.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="c1"># Retrieve UTR data from the input dictionary</span>
    <span class="n">utr_data</span> <span class="o">=</span> <span class="n">data</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="s2">&quot;UTR&quot;</span><span class="p">,</span> <span class="p">[])</span>

    <span class="c1"># Initialize lists to store 5&#39; UTR and 3&#39; UTR information</span>
    <span class="n">utr5_list</span> <span class="o">=</span> <span class="p">[]</span>
    <span class="n">utr3_list</span> <span class="o">=</span> <span class="p">[]</span>

    <span class="c1"># Iterate through UTR entries in the data</span>
    <span class="k">for</span> <span class="n">utr_entry</span> <span class="ow">in</span> <span class="n">utr_data</span><span class="p">:</span>
        <span class="n">utr_type</span> <span class="o">=</span> <span class="n">utr_entry</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="s2">&quot;type&quot;</span><span class="p">,</span> <span class="s2">&quot;&quot;</span><span class="p">)</span>
        <span class="n">utr_start</span> <span class="o">=</span> <span class="n">utr_entry</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="s2">&quot;start&quot;</span><span class="p">,</span> <span class="kc">None</span><span class="p">)</span>
        <span class="n">utr_end</span> <span class="o">=</span> <span class="n">utr_entry</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="s2">&quot;end&quot;</span><span class="p">,</span> <span class="kc">None</span><span class="p">)</span>

        <span class="c1"># Check if the UTR entry has valid type, start, and end information</span>
        <span class="k">if</span> <span class="n">utr_type</span> <span class="ow">and</span> <span class="n">utr_start</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span> <span class="ow">and</span> <span class="n">utr_end</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>
            <span class="c1"># Categorize UTRs into 5&#39; and 3&#39; UTR lists</span>
            <span class="k">if</span> <span class="n">utr_type</span> <span class="o">==</span> <span class="s2">&quot;five_prime_utr&quot;</span><span class="p">:</span>
                <span class="n">utr5_list</span><span class="o">.</span><span class="n">append</span><span class="p">((</span><span class="n">utr_start</span><span class="p">,</span> <span class="n">utr_end</span><span class="p">))</span>
            <span class="k">elif</span> <span class="n">utr_type</span> <span class="o">==</span> <span class="s2">&quot;three_prime_utr&quot;</span><span class="p">:</span>
                <span class="n">utr3_list</span><span class="o">.</span><span class="n">append</span><span class="p">((</span><span class="n">utr_start</span><span class="p">,</span> <span class="n">utr_end</span><span class="p">))</span>

    <span class="c1"># Extract chromosome and strand information</span>
    <span class="n">chromosome</span> <span class="o">=</span> <span class="n">utr_entry</span><span class="p">[</span><span class="s2">&quot;seq_region_name&quot;</span><span class="p">]</span> <span class="k">if</span> <span class="n">utr_data</span> <span class="k">else</span> <span class="kc">None</span>
    <span class="n">strand</span> <span class="o">=</span> <span class="n">data</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="s2">&quot;strand&quot;</span><span class="p">,</span> <span class="kc">None</span><span class="p">)</span>

    <span class="k">return</span> <span class="n">utr5_list</span><span class="p">,</span> <span class="n">utr3_list</span><span class="p">,</span> <span class="n">chromosome</span><span class="p">,</span> <span class="n">strand</span>
</code></pre></div></td></tr></table></div>
          </details>
  </div>

</div>


<div class="doc doc-object doc-function">



<h2 id="dna.ensembl_api.get_cds" class="doc doc-heading">
          <code class="highlight language-python"><span class="n">get_cds</span><span class="p">(</span><span class="n">transcript_id</span><span class="p">)</span></code>

</h2>


  <div class="doc doc-contents ">
  
      <p>Retrieves the coding sequence (CDS) for a given Ensembl transcript ID.</p>



<table class="field-list">
  <colgroup>
    <col class="field-name" />
    <col class="field-body" />
  </colgroup>
  <tbody valign="top">
    <tr class="field">
      <th class="field-name">Parameters:</th>
      <td class="field-body">
        <ul class="first simple">
            <li>
              <b><code>transcript_id</code></b>
                  (<code>str</code>)
              –
              <div class="doc-md-description">
                <p>Ensembl transcript ID for the target gene.</p>
              </div>
            </li>
        </ul>
      </td>
    </tr>
  </tbody>
</table>


<table class="field-list">
  <colgroup>
    <col class="field-name" />
    <col class="field-body" />
  </colgroup>
  <tbody valign="top">
    <tr class="field">
    <th class="field-name">Returns:</th>
    <td class="field-body">
      <ul class="first simple">
          <li>
<b><code>str</code></b>(                <code>str</code>
)            –
            <div class="doc-md-description">
              <p>The nucleotide sequence of the coding sequence (CDS).
 Returns an empty string in case of an error.</p>
            </div>
          </li>
      </ul>
    </td>
    </tr>
  </tbody>
</table>
          <details class="quote">
            <summary>Source code in <code>dna\ensembl_api.py</code></summary>
            <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">38</span>
<span class="normal">39</span>
<span class="normal">40</span>
<span class="normal">41</span>
<span class="normal">42</span>
<span class="normal">43</span>
<span class="normal">44</span>
<span class="normal">45</span>
<span class="normal">46</span>
<span class="normal">47</span>
<span class="normal">48</span>
<span class="normal">49</span>
<span class="normal">50</span>
<span class="normal">51</span>
<span class="normal">52</span>
<span class="normal">53</span>
<span class="normal">54</span>
<span class="normal">55</span>
<span class="normal">56</span>
<span class="normal">57</span>
<span class="normal">58</span>
<span class="normal">59</span>
<span class="normal">60</span>
<span class="normal">61</span>
<span class="normal">62</span>
<span class="normal">63</span>
<span class="normal">64</span>
<span class="normal">65</span>
<span class="normal">66</span>
<span class="normal">67</span>
<span class="normal">68</span>
<span class="normal">69</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span> <span class="nf">get_cds</span><span class="p">(</span><span class="n">transcript_id</span><span class="p">:</span> <span class="nb">str</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="nb">str</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Retrieves the coding sequence (CDS) for a given Ensembl transcript ID.</span>

<span class="sd">    Args:</span>
<span class="sd">        transcript_id (str): Ensembl transcript ID for the target gene.</span>

<span class="sd">    Returns:</span>
<span class="sd">        str: The nucleotide sequence of the coding sequence (CDS).</span>
<span class="sd">             Returns an empty string in case of an error.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="c1"># Construct the REST API URL for retrieving CDS</span>
    <span class="n">address</span> <span class="o">=</span> <span class="sa">f</span><span class="s2">&quot;https://rest.ensembl.org/sequence/id/</span><span class="si">{</span><span class="n">transcript_id</span><span class="si">}</span><span class="s2">?multiple_sequences=1;type=cds&quot;</span>

    <span class="k">try</span><span class="p">:</span>
        <span class="c1"># Make a GET request to the Ensembl REST API</span>
        <span class="n">r</span> <span class="o">=</span> <span class="n">requests</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="n">address</span><span class="p">,</span> <span class="n">headers</span><span class="o">=</span><span class="p">{</span><span class="s2">&quot;Content-Type&quot;</span><span class="p">:</span> <span class="s2">&quot;text/x-fasta&quot;</span><span class="p">},</span> <span class="n">timeout</span><span class="o">=</span><span class="mi">30</span><span class="p">)</span>

        <span class="c1"># Ensure that there are no issues with the sequence request</span>
        <span class="n">r</span><span class="o">.</span><span class="n">raise_for_status</span><span class="p">()</span>

        <span class="c1"># Extract only the nucleotide sequence and format into a single string</span>
        <span class="n">raw_output</span> <span class="o">=</span> <span class="n">r</span><span class="o">.</span><span class="n">text</span>
        <span class="n">pattern</span> <span class="o">=</span> <span class="n">re</span><span class="o">.</span><span class="n">compile</span><span class="p">(</span><span class="s2">&quot;(?:^|</span><span class="se">\n</span><span class="s2">)[ATGC]+&quot;</span><span class="p">)</span>
        <span class="n">matches</span> <span class="o">=</span> <span class="n">pattern</span><span class="o">.</span><span class="n">findall</span><span class="p">(</span><span class="n">raw_output</span><span class="p">)</span>
        <span class="n">cds_sequence</span> <span class="o">=</span> <span class="s2">&quot;&quot;</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">matches</span><span class="p">)</span><span class="o">.</span><span class="n">replace</span><span class="p">(</span><span class="s2">&quot;</span><span class="se">\n</span><span class="s2">&quot;</span><span class="p">,</span> <span class="s2">&quot;&quot;</span><span class="p">)</span>

        <span class="k">return</span> <span class="n">cds_sequence</span>

    <span class="k">except</span> <span class="n">requests</span><span class="o">.</span><span class="n">exceptions</span><span class="o">.</span><span class="n">RequestException</span> <span class="k">as</span> <span class="n">e</span><span class="p">:</span>
        <span class="c1"># If there&#39;s an error with the request, print the error and return an empty string</span>
        <span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Error with the request for </span><span class="si">{</span><span class="n">transcript_id</span><span class="si">}</span><span class="s2">: </span><span class="si">{</span><span class="n">e</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
        <span class="k">return</span> <span class="s2">&quot;&quot;</span>
</code></pre></div></td></tr></table></div>
          </details>
  </div>

</div>


<div class="doc doc-object doc-function">



<h2 id="dna.ensembl_api.get_data_as_csv" class="doc doc-heading">
          <code class="highlight language-python"><span class="n">get_data_as_csv</span><span class="p">(</span><span class="n">file_paths</span><span class="p">,</span> <span class="n">output_directory</span><span class="p">)</span></code>

</h2>


  <div class="doc doc-contents ">
  
      <p>Retrieves data for gene IDs from Ensembl, processes it, and saves it as CSV files.</p>



<table class="field-list">
  <colgroup>
    <col class="field-name" />
    <col class="field-body" />
  </colgroup>
  <tbody valign="top">
    <tr class="field">
      <th class="field-name">Parameters:</th>
      <td class="field-body">
        <ul class="first simple">
            <li>
              <b><code>file_paths</code></b>
                  (<code><span title="typing.List">List</span>[str]</code>)
              –
              <div class="doc-md-description">
                <p>List of file paths containing gene IDs.</p>
              </div>
            </li>
            <li>
              <b><code>output_directory</code></b>
                  (<code>str</code>)
              –
              <div class="doc-md-description">
                <p>Directory where CSV files will be saved.</p>
              </div>
            </li>
        </ul>
      </td>
    </tr>
  </tbody>
</table>


<table class="field-list">
  <colgroup>
    <col class="field-name" />
    <col class="field-body" />
  </colgroup>
  <tbody valign="top">
    <tr class="field">
    <th class="field-name">Returns:</th>
    <td class="field-body">
      <ul class="first simple">
          <li>
<b><code>None</code></b>(                <code>None</code>
)            –
            <div class="doc-md-description">
              <p>This function does not return a value but outputs files to the specified directory.</p>
            </div>
          </li>
      </ul>
    </td>
    </tr>
  </tbody>
</table>
          <details class="quote">
            <summary>Source code in <code>dna\ensembl_api.py</code></summary>
            <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">257</span>
<span class="normal">258</span>
<span class="normal">259</span>
<span class="normal">260</span>
<span class="normal">261</span>
<span class="normal">262</span>
<span class="normal">263</span>
<span class="normal">264</span>
<span class="normal">265</span>
<span class="normal">266</span>
<span class="normal">267</span>
<span class="normal">268</span>
<span class="normal">269</span>
<span class="normal">270</span>
<span class="normal">271</span>
<span class="normal">272</span>
<span class="normal">273</span>
<span class="normal">274</span>
<span class="normal">275</span>
<span class="normal">276</span>
<span class="normal">277</span>
<span class="normal">278</span>
<span class="normal">279</span>
<span class="normal">280</span>
<span class="normal">281</span>
<span class="normal">282</span>
<span class="normal">283</span>
<span class="normal">284</span>
<span class="normal">285</span>
<span class="normal">286</span>
<span class="normal">287</span>
<span class="normal">288</span>
<span class="normal">289</span>
<span class="normal">290</span>
<span class="normal">291</span>
<span class="normal">292</span>
<span class="normal">293</span>
<span class="normal">294</span>
<span class="normal">295</span>
<span class="normal">296</span>
<span class="normal">297</span>
<span class="normal">298</span>
<span class="normal">299</span>
<span class="normal">300</span>
<span class="normal">301</span>
<span class="normal">302</span>
<span class="normal">303</span>
<span class="normal">304</span>
<span class="normal">305</span>
<span class="normal">306</span>
<span class="normal">307</span>
<span class="normal">308</span>
<span class="normal">309</span>
<span class="normal">310</span>
<span class="normal">311</span>
<span class="normal">312</span>
<span class="normal">313</span>
<span class="normal">314</span>
<span class="normal">315</span>
<span class="normal">316</span>
<span class="normal">317</span>
<span class="normal">318</span>
<span class="normal">319</span>
<span class="normal">320</span>
<span class="normal">321</span>
<span class="normal">322</span>
<span class="normal">323</span>
<span class="normal">324</span>
<span class="normal">325</span>
<span class="normal">326</span>
<span class="normal">327</span>
<span class="normal">328</span>
<span class="normal">329</span>
<span class="normal">330</span>
<span class="normal">331</span>
<span class="normal">332</span>
<span class="normal">333</span>
<span class="normal">334</span>
<span class="normal">335</span>
<span class="normal">336</span>
<span class="normal">337</span>
<span class="normal">338</span>
<span class="normal">339</span>
<span class="normal">340</span>
<span class="normal">341</span>
<span class="normal">342</span>
<span class="normal">343</span>
<span class="normal">344</span>
<span class="normal">345</span>
<span class="normal">346</span>
<span class="normal">347</span>
<span class="normal">348</span>
<span class="normal">349</span>
<span class="normal">350</span>
<span class="normal">351</span>
<span class="normal">352</span>
<span class="normal">353</span>
<span class="normal">354</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span> <span class="nf">get_data_as_csv</span><span class="p">(</span><span class="n">file_paths</span><span class="p">:</span> <span class="n">List</span><span class="p">[</span><span class="nb">str</span><span class="p">],</span> <span class="n">output_directory</span><span class="p">:</span> <span class="nb">str</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="kc">None</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Retrieves data for gene IDs from Ensembl, processes it, and saves it as CSV files.</span>

<span class="sd">    Args:</span>
<span class="sd">        file_paths (List[str]): List of file paths containing gene IDs.</span>
<span class="sd">        output_directory (str): Directory where CSV files will be saved.</span>

<span class="sd">    Returns:</span>
<span class="sd">        None: This function does not return a value but outputs files to the specified directory.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="c1"># Create the directory if it doesn&#39;t exist</span>
    <span class="n">os</span><span class="o">.</span><span class="n">makedirs</span><span class="p">(</span><span class="n">output_directory</span><span class="p">,</span> <span class="n">exist_ok</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>

    <span class="k">for</span> <span class="n">file_path</span> <span class="ow">in</span> <span class="n">file_paths</span><span class="p">:</span>
        <span class="c1"># Read gene IDs from the file</span>
        <span class="n">gene_ids</span> <span class="o">=</span> <span class="n">read_gene_ids_from_file</span><span class="p">(</span><span class="n">file_path</span><span class="p">)</span>

        <span class="c1"># Generate output filename and species name</span>
        <span class="n">filename</span> <span class="o">=</span> <span class="s2">&quot;ensembl_data_&quot;</span> <span class="o">+</span> <span class="n">get_species_name</span><span class="p">(</span><span class="n">file_path</span><span class="p">)</span> <span class="o">+</span> <span class="s2">&quot;.csv&quot;</span>
        <span class="n">species</span> <span class="o">=</span> <span class="s2">&quot; &quot;</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">get_species_name</span><span class="p">(</span><span class="n">file_path</span><span class="p">)</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s2">&quot;_&quot;</span><span class="p">))</span>

        <span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Starting data extraction for </span><span class="si">{</span><span class="n">species</span><span class="si">}</span><span class="s2">.&quot;</span><span class="p">)</span>

        <span class="c1"># Initialize a CSV writer</span>
        <span class="n">filename</span> <span class="o">=</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">output_directory</span><span class="p">,</span> <span class="n">filename</span><span class="p">)</span>
        <span class="n">csv_file</span> <span class="o">=</span> <span class="nb">open</span><span class="p">(</span><span class="n">filename</span><span class="p">,</span> <span class="s2">&quot;w&quot;</span><span class="p">,</span> <span class="n">newline</span><span class="o">=</span><span class="s2">&quot;&quot;</span><span class="p">,</span> <span class="n">encoding</span><span class="o">=</span><span class="s2">&quot;utf-8&quot;</span><span class="p">)</span>
        <span class="n">csv_writer</span> <span class="o">=</span> <span class="n">csv</span><span class="o">.</span><span class="n">writer</span><span class="p">(</span><span class="n">csv_file</span><span class="p">)</span>

        <span class="c1"># Write the header to the CSV file</span>
        <span class="n">csv_writer</span><span class="o">.</span><span class="n">writerow</span><span class="p">(</span>
            <span class="p">[</span>
                <span class="s2">&quot;ensembl_gene_id&quot;</span><span class="p">,</span>
                <span class="s2">&quot;transcript_id&quot;</span><span class="p">,</span>
                <span class="s2">&quot;promoter&quot;</span><span class="p">,</span>
                <span class="s2">&quot;utr5&quot;</span><span class="p">,</span>
                <span class="s2">&quot;cds&quot;</span><span class="p">,</span>
                <span class="s2">&quot;utr3&quot;</span><span class="p">,</span>
                <span class="s2">&quot;terminator&quot;</span><span class="p">,</span>
            <span class="p">]</span>
        <span class="p">)</span>

        <span class="c1"># Loop through each gene ID and retrieve the data</span>
        <span class="k">for</span> <span class="n">gene_id</span> <span class="ow">in</span> <span class="n">gene_ids</span><span class="p">:</span>
            <span class="n">time</span><span class="o">.</span><span class="n">sleep</span><span class="p">(</span><span class="mi">2</span><span class="p">)</span>
            <span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Extracting data for gene ID : </span><span class="si">{</span><span class="n">gene_id</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>

            <span class="k">try</span><span class="p">:</span>
                <span class="n">gene_data</span> <span class="o">=</span> <span class="n">ensembl_rest</span><span class="o">.</span><span class="n">lookup</span><span class="p">(</span><span class="n">species</span><span class="o">=</span><span class="n">species</span><span class="p">,</span> <span class="nb">id</span><span class="o">=</span><span class="n">gene_id</span><span class="p">)</span>

            <span class="k">except</span> <span class="n">ensembl_rest</span><span class="o">.</span><span class="n">core</span><span class="o">.</span><span class="n">restclient</span><span class="o">.</span><span class="n">HTTPError</span> <span class="k">as</span> <span class="n">e</span><span class="p">:</span>
                <span class="k">if</span> <span class="n">e</span><span class="o">.</span><span class="n">response</span><span class="o">.</span><span class="n">status_code</span> <span class="o">==</span> <span class="mi">429</span><span class="p">:</span>  <span class="c1"># Check for rate limit exceeded error</span>
                    <span class="nb">print</span><span class="p">(</span><span class="s2">&quot;Rate limit exceeded. Waiting before retrying...&quot;</span><span class="p">)</span>
                    <span class="n">time</span><span class="o">.</span><span class="n">sleep</span><span class="p">(</span><span class="mi">1</span><span class="p">)</span>
                    <span class="n">gene_data</span> <span class="o">=</span> <span class="n">ensembl_rest</span><span class="o">.</span><span class="n">lookup</span><span class="p">(</span><span class="n">species</span><span class="o">=</span><span class="n">species</span><span class="p">,</span> <span class="nb">id</span><span class="o">=</span><span class="n">gene_id</span><span class="p">)</span>

            <span class="c1"># Get transcript ID</span>
            <span class="n">transcript_id</span> <span class="o">=</span> <span class="n">gene_data</span><span class="p">[</span><span class="s2">&quot;canonical_transcript&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s2">&quot;.&quot;</span><span class="p">)[</span><span class="mi">0</span><span class="p">]</span>

            <span class="c1"># Retrieve promoter, CDS, and terminator sequences</span>
            <span class="n">cds_sequence</span> <span class="o">=</span> <span class="n">get_cds</span><span class="p">(</span><span class="n">transcript_id</span><span class="p">)</span>
            <span class="k">if</span> <span class="n">cds_sequence</span> <span class="o">==</span> <span class="s2">&quot;&quot;</span><span class="p">:</span>
                <span class="k">continue</span>
            <span class="n">promoter_sequence</span><span class="p">,</span> <span class="n">terminator_sequence</span> <span class="o">=</span> <span class="n">get_promoter_terminator</span><span class="p">(</span>
                <span class="n">transcript_id</span>
            <span class="p">)</span>

            <span class="c1"># Retrieve UTR sequences</span>
            <span class="n">transcript_data</span> <span class="o">=</span> <span class="n">request_with_retry</span><span class="p">(</span><span class="n">transcript_id</span><span class="p">)</span>
            <span class="k">if</span> <span class="n">transcript_data</span> <span class="o">==</span> <span class="p">{}:</span>
                <span class="k">continue</span>

            <span class="n">utr5_coord_list</span><span class="p">,</span> <span class="n">utr3_coord_list</span><span class="p">,</span> <span class="n">chromosome</span><span class="p">,</span> <span class="n">strand</span> <span class="o">=</span> <span class="p">(</span>
                <span class="n">extract_utr_information</span><span class="p">(</span><span class="n">transcript_data</span><span class="p">)</span>
            <span class="p">)</span>
            <span class="n">utr5_sequence</span> <span class="o">=</span> <span class="n">get_full_utr_sequence</span><span class="p">(</span>
                <span class="n">utr5_coord_list</span><span class="p">,</span> <span class="n">chromosome</span><span class="p">,</span> <span class="n">strand</span><span class="p">,</span> <span class="n">species</span><span class="o">=</span><span class="n">get_species_name</span><span class="p">(</span><span class="n">file_path</span><span class="p">)</span>
            <span class="p">)</span>
            <span class="n">utr3_sequence</span> <span class="o">=</span> <span class="n">get_full_utr_sequence</span><span class="p">(</span>
                <span class="n">utr3_coord_list</span><span class="p">,</span> <span class="n">chromosome</span><span class="p">,</span> <span class="n">strand</span><span class="p">,</span> <span class="n">species</span><span class="o">=</span><span class="n">get_species_name</span><span class="p">(</span><span class="n">file_path</span><span class="p">)</span>
            <span class="p">)</span>

            <span class="c1"># Write the row to the CSV file</span>
            <span class="n">csv_writer</span><span class="o">.</span><span class="n">writerow</span><span class="p">(</span>
                <span class="p">[</span>
                    <span class="n">gene_id</span><span class="p">,</span>
                    <span class="n">transcript_id</span><span class="p">,</span>
                    <span class="n">promoter_sequence</span><span class="p">,</span>
                    <span class="n">utr5_sequence</span><span class="p">,</span>
                    <span class="n">cds_sequence</span><span class="p">,</span>
                    <span class="n">utr3_sequence</span><span class="p">,</span>
                    <span class="n">terminator_sequence</span><span class="p">,</span>
                <span class="p">]</span>
            <span class="p">)</span>

        <span class="c1"># Close the CSV file</span>
        <span class="n">csv_file</span><span class="o">.</span><span class="n">close</span><span class="p">()</span>

        <span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Data extraction for </span><span class="si">{</span><span class="n">species</span><span class="si">}</span><span class="s2"> is now complete.&quot;</span><span class="p">)</span>
</code></pre></div></td></tr></table></div>
          </details>
  </div>

</div>


<div class="doc doc-object doc-function">



<h2 id="dna.ensembl_api.get_full_utr_sequence" class="doc doc-heading">
          <code class="highlight language-python"><span class="n">get_full_utr_sequence</span><span class="p">(</span><span class="n">list_utr_coordinates</span><span class="p">,</span> <span class="n">chromosome</span><span class="p">,</span> <span class="n">strand</span><span class="p">,</span> <span class="n">species</span><span class="p">)</span></code>

</h2>


  <div class="doc doc-contents ">
  
      <p>Retrieves the concatenated nucleotide sequence of multiple UTRs from the Ensembl database.</p>



<table class="field-list">
  <colgroup>
    <col class="field-name" />
    <col class="field-body" />
  </colgroup>
  <tbody valign="top">
    <tr class="field">
      <th class="field-name">Parameters:</th>
      <td class="field-body">
        <ul class="first simple">
            <li>
              <b><code>list_utr_coordinates</code></b>
                  (<code><span title="typing.List">List</span>[<span title="typing.Tuple">Tuple</span>[int, int]]</code>)
              –
              <div class="doc-md-description">
                <p>A list of tuples representing UTR start and end coordinates.</p>
              </div>
            </li>
            <li>
              <b><code>chromosome</code></b>
                  (<code>str</code>)
              –
              <div class="doc-md-description">
                <p>Chromosome name or identifier.</p>
              </div>
            </li>
            <li>
              <b><code>strand</code></b>
                  (<code>int</code>)
              –
              <div class="doc-md-description">
                <p>Strand information (1 for forward strand, -1 for reverse strand).</p>
              </div>
            </li>
        </ul>
      </td>
    </tr>
  </tbody>
</table>


<table class="field-list">
  <colgroup>
    <col class="field-name" />
    <col class="field-body" />
  </colgroup>
  <tbody valign="top">
    <tr class="field">
    <th class="field-name">Returns:</th>
    <td class="field-body">
      <ul class="first simple">
          <li>
<b><code>str</code></b>(                <code>str</code>
)            –
            <div class="doc-md-description">
              <p>The concatenated nucleotide sequence of the specified UTRs.</p>
            </div>
          </li>
      </ul>
    </td>
    </tr>
  </tbody>
</table>
          <details class="quote">
            <summary>Source code in <code>dna\ensembl_api.py</code></summary>
            <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">184</span>
<span class="normal">185</span>
<span class="normal">186</span>
<span class="normal">187</span>
<span class="normal">188</span>
<span class="normal">189</span>
<span class="normal">190</span>
<span class="normal">191</span>
<span class="normal">192</span>
<span class="normal">193</span>
<span class="normal">194</span>
<span class="normal">195</span>
<span class="normal">196</span>
<span class="normal">197</span>
<span class="normal">198</span>
<span class="normal">199</span>
<span class="normal">200</span>
<span class="normal">201</span>
<span class="normal">202</span>
<span class="normal">203</span>
<span class="normal">204</span>
<span class="normal">205</span>
<span class="normal">206</span>
<span class="normal">207</span>
<span class="normal">208</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span> <span class="nf">get_full_utr_sequence</span><span class="p">(</span>
    <span class="n">list_utr_coordinates</span><span class="p">:</span> <span class="n">List</span><span class="p">[</span><span class="n">Tuple</span><span class="p">[</span><span class="nb">int</span><span class="p">,</span> <span class="nb">int</span><span class="p">]],</span>
    <span class="n">chromosome</span><span class="p">:</span> <span class="nb">str</span><span class="p">,</span>
    <span class="n">strand</span><span class="p">:</span> <span class="nb">int</span><span class="p">,</span>
    <span class="n">species</span><span class="p">:</span> <span class="nb">str</span><span class="p">,</span>
<span class="p">)</span> <span class="o">-&gt;</span> <span class="nb">str</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Retrieves the concatenated nucleotide sequence of multiple UTRs from the Ensembl database.</span>

<span class="sd">    Args:</span>
<span class="sd">        list_utr_coordinates (List[Tuple[int, int]]): A list of tuples representing UTR start and end coordinates.</span>
<span class="sd">        chromosome (str): Chromosome name or identifier.</span>
<span class="sd">        strand (int): Strand information (1 for forward strand, -1 for reverse strand).</span>

<span class="sd">    Returns:</span>
<span class="sd">        str: The concatenated nucleotide sequence of the specified UTRs.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="c1"># Initialize an empty string to store the concatenated UTR sequence</span>
    <span class="n">concatenated_sequence</span> <span class="o">=</span> <span class="s2">&quot;&quot;</span>

    <span class="c1"># Iterate through UTR coordinates and retrieve individual UTR sequences</span>
    <span class="k">for</span> <span class="n">start</span><span class="p">,</span> <span class="n">end</span> <span class="ow">in</span> <span class="n">list_utr_coordinates</span><span class="p">:</span>
        <span class="n">sequence</span> <span class="o">=</span> <span class="n">get_utr_sequence</span><span class="p">(</span><span class="n">chromosome</span><span class="p">,</span> <span class="n">strand</span><span class="p">,</span> <span class="n">start</span><span class="p">,</span> <span class="n">end</span><span class="p">,</span> <span class="n">species</span><span class="p">)</span>
        <span class="n">concatenated_sequence</span> <span class="o">+=</span> <span class="n">sequence</span>

    <span class="k">return</span> <span class="n">concatenated_sequence</span>
</code></pre></div></td></tr></table></div>
          </details>
  </div>

</div>


<div class="doc doc-object doc-function">



<h2 id="dna.ensembl_api.get_promoter_terminator" class="doc doc-heading">
          <code class="highlight language-python"><span class="n">get_promoter_terminator</span><span class="p">(</span><span class="n">transcript_id</span><span class="p">,</span> <span class="n">promoter_length</span><span class="o">=</span><span class="mi">1000</span><span class="p">,</span> <span class="n">terminator_length</span><span class="o">=</span><span class="mi">500</span><span class="p">)</span></code>

</h2>


  <div class="doc doc-contents ">
  
      <p>Retrieves the promoter and terminator sequences for a given Ensembl transcript ID.</p>



<table class="field-list">
  <colgroup>
    <col class="field-name" />
    <col class="field-body" />
  </colgroup>
  <tbody valign="top">
    <tr class="field">
      <th class="field-name">Parameters:</th>
      <td class="field-body">
        <ul class="first simple">
            <li>
              <b><code>transcript_id</code></b>
                  (<code>str</code>)
              –
              <div class="doc-md-description">
                <p>Ensembl transcript ID for the target gene.</p>
              </div>
            </li>
            <li>
              <b><code>promoter_length</code></b>
                  (<code><span title="typing.Optional">Optional</span>[int]</code>, default:
                      <code>1000</code>
)
              –
              <div class="doc-md-description">
                <p>Length of the promoter sequence (default is 1000).</p>
              </div>
            </li>
            <li>
              <b><code>terminator_length</code></b>
                  (<code><span title="typing.Optional">Optional</span>[int]</code>, default:
                      <code>500</code>
)
              –
              <div class="doc-md-description">
                <p>Length of the terminator sequence (default is 500).</p>
              </div>
            </li>
        </ul>
      </td>
    </tr>
  </tbody>
</table>


<table class="field-list">
  <colgroup>
    <col class="field-name" />
    <col class="field-body" />
  </colgroup>
  <tbody valign="top">
    <tr class="field">
    <th class="field-name">Returns:</th>
    <td class="field-body">
      <ul class="first simple">
          <li>
                <code><span title="typing.Tuple">Tuple</span>[str, str]</code>
            –
            <div class="doc-md-description">
              <p>Tuple[str, str]: A tuple containing the promoter and terminator sequences as strings.</p>
            </div>
          </li>
      </ul>
    </td>
    </tr>
  </tbody>
</table>
          <details class="quote">
            <summary>Source code in <code>dna\ensembl_api.py</code></summary>
            <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal"> 72</span>
<span class="normal"> 73</span>
<span class="normal"> 74</span>
<span class="normal"> 75</span>
<span class="normal"> 76</span>
<span class="normal"> 77</span>
<span class="normal"> 78</span>
<span class="normal"> 79</span>
<span class="normal"> 80</span>
<span class="normal"> 81</span>
<span class="normal"> 82</span>
<span class="normal"> 83</span>
<span class="normal"> 84</span>
<span class="normal"> 85</span>
<span class="normal"> 86</span>
<span class="normal"> 87</span>
<span class="normal"> 88</span>
<span class="normal"> 89</span>
<span class="normal"> 90</span>
<span class="normal"> 91</span>
<span class="normal"> 92</span>
<span class="normal"> 93</span>
<span class="normal"> 94</span>
<span class="normal"> 95</span>
<span class="normal"> 96</span>
<span class="normal"> 97</span>
<span class="normal"> 98</span>
<span class="normal"> 99</span>
<span class="normal">100</span>
<span class="normal">101</span>
<span class="normal">102</span>
<span class="normal">103</span>
<span class="normal">104</span>
<span class="normal">105</span>
<span class="normal">106</span>
<span class="normal">107</span>
<span class="normal">108</span>
<span class="normal">109</span>
<span class="normal">110</span>
<span class="normal">111</span>
<span class="normal">112</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span> <span class="nf">get_promoter_terminator</span><span class="p">(</span>
    <span class="n">transcript_id</span><span class="p">:</span> <span class="nb">str</span><span class="p">,</span>
    <span class="n">promoter_length</span><span class="p">:</span> <span class="n">Optional</span><span class="p">[</span><span class="nb">int</span><span class="p">]</span> <span class="o">=</span> <span class="mi">1000</span><span class="p">,</span>
    <span class="n">terminator_length</span><span class="p">:</span> <span class="n">Optional</span><span class="p">[</span><span class="nb">int</span><span class="p">]</span> <span class="o">=</span> <span class="mi">500</span><span class="p">,</span>
<span class="p">)</span> <span class="o">-&gt;</span> <span class="n">Tuple</span><span class="p">[</span><span class="nb">str</span><span class="p">,</span> <span class="nb">str</span><span class="p">]:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Retrieves the promoter and terminator sequences for a given Ensembl transcript ID.</span>

<span class="sd">    Args:</span>
<span class="sd">        transcript_id (str): Ensembl transcript ID for the target gene.</span>
<span class="sd">        promoter_length (Optional[int]): Length of the promoter sequence (default is 1000).</span>
<span class="sd">        terminator_length (Optional[int]): Length of the terminator sequence (default is 500).</span>

<span class="sd">    Returns:</span>
<span class="sd">        Tuple[str, str]: A tuple containing the promoter and terminator sequences as strings.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="c1"># Construct the REST API URL for retrieving genomic sequence with specified 5&#39; and 3&#39; expansions</span>
    <span class="n">address</span> <span class="o">=</span> <span class="sa">f</span><span class="s2">&quot;https://rest.ensembl.org/sequence/id/</span><span class="si">{</span><span class="n">transcript_id</span><span class="si">}</span><span class="s2">?type=genomic;expand_5prime=1000;expand_3prime=500&quot;</span>

    <span class="k">try</span><span class="p">:</span>
        <span class="c1"># Make a GET request to the Ensembl REST API</span>
        <span class="n">r</span> <span class="o">=</span> <span class="n">requests</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="n">address</span><span class="p">,</span> <span class="n">headers</span><span class="o">=</span><span class="p">{</span><span class="s2">&quot;Content-Type&quot;</span><span class="p">:</span> <span class="s2">&quot;text/x-fasta&quot;</span><span class="p">},</span> <span class="n">timeout</span><span class="o">=</span><span class="mi">30</span><span class="p">)</span>

        <span class="c1"># Ensure that there are no issues with the sequence request</span>
        <span class="n">r</span><span class="o">.</span><span class="n">raise_for_status</span><span class="p">()</span>

        <span class="c1"># Remove unwanted characters to produce only the nucleotide sequence and format into a single string</span>
        <span class="n">raw_output</span> <span class="o">=</span> <span class="n">r</span><span class="o">.</span><span class="n">text</span>
        <span class="n">pattern</span> <span class="o">=</span> <span class="n">re</span><span class="o">.</span><span class="n">compile</span><span class="p">(</span><span class="s2">&quot;(?:^|</span><span class="se">\n</span><span class="s2">)[ATGC]+&quot;</span><span class="p">)</span>
        <span class="n">matches</span> <span class="o">=</span> <span class="n">pattern</span><span class="o">.</span><span class="n">findall</span><span class="p">(</span><span class="n">raw_output</span><span class="p">)</span>
        <span class="n">sequence</span> <span class="o">=</span> <span class="s2">&quot;&quot;</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">matches</span><span class="p">)</span><span class="o">.</span><span class="n">replace</span><span class="p">(</span><span class="s2">&quot;</span><span class="se">\n</span><span class="s2">&quot;</span><span class="p">,</span> <span class="s2">&quot;&quot;</span><span class="p">)</span>

        <span class="c1"># Extract the promoter and terminator sequence from the entire sequence</span>
        <span class="n">promoter_sequence</span> <span class="o">=</span> <span class="n">sequence</span><span class="p">[:</span><span class="n">promoter_length</span><span class="p">]</span>
        <span class="n">terminator_sequence</span> <span class="o">=</span> <span class="n">sequence</span><span class="p">[</span><span class="o">-</span><span class="n">terminator_length</span><span class="p">:]</span>

        <span class="k">return</span> <span class="n">promoter_sequence</span><span class="p">,</span> <span class="n">terminator_sequence</span>

    <span class="k">except</span> <span class="n">requests</span><span class="o">.</span><span class="n">exceptions</span><span class="o">.</span><span class="n">RequestException</span> <span class="k">as</span> <span class="n">e</span><span class="p">:</span>
        <span class="c1"># If there&#39;s an error with the request, print the error and return an empty string</span>
        <span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Error with the promoter-terminator request for </span><span class="si">{</span><span class="n">transcript_id</span><span class="si">}</span><span class="s2">: </span><span class="si">{</span><span class="n">e</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
        <span class="k">return</span> <span class="s2">&quot;&quot;</span><span class="p">,</span> <span class="s2">&quot;&quot;</span>
</code></pre></div></td></tr></table></div>
          </details>
  </div>

</div>


<div class="doc doc-object doc-function">



<h2 id="dna.ensembl_api.get_species_name" class="doc doc-heading">
          <code class="highlight language-python"><span class="n">get_species_name</span><span class="p">(</span><span class="n">file_path</span><span class="p">)</span></code>

</h2>


  <div class="doc doc-contents ">
  
      <p>Extracts the species name from a file path.</p>



<table class="field-list">
  <colgroup>
    <col class="field-name" />
    <col class="field-body" />
  </colgroup>
  <tbody valign="top">
    <tr class="field">
      <th class="field-name">Parameters:</th>
      <td class="field-body">
        <ul class="first simple">
            <li>
              <b><code>file_path</code></b>
                  (<code>str</code>)
              –
              <div class="doc-md-description">
                <p>The path to the file containing the species information.</p>
              </div>
            </li>
        </ul>
      </td>
    </tr>
  </tbody>
</table>


<table class="field-list">
  <colgroup>
    <col class="field-name" />
    <col class="field-body" />
  </colgroup>
  <tbody valign="top">
    <tr class="field">
    <th class="field-name">Returns:</th>
    <td class="field-body">
      <ul class="first simple">
          <li>
<b><code>str</code></b>(                <code>str</code>
)            –
            <div class="doc-md-description">
              <p>The extracted species name.</p>
            </div>
          </li>
      </ul>
    </td>
    </tr>
  </tbody>
</table>
          <details class="quote">
            <summary>Source code in <code>dna\ensembl_api.py</code></summary>
            <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">211</span>
<span class="normal">212</span>
<span class="normal">213</span>
<span class="normal">214</span>
<span class="normal">215</span>
<span class="normal">216</span>
<span class="normal">217</span>
<span class="normal">218</span>
<span class="normal">219</span>
<span class="normal">220</span>
<span class="normal">221</span>
<span class="normal">222</span>
<span class="normal">223</span>
<span class="normal">224</span>
<span class="normal">225</span>
<span class="normal">226</span>
<span class="normal">227</span>
<span class="normal">228</span>
<span class="normal">229</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span> <span class="nf">get_species_name</span><span class="p">(</span><span class="n">file_path</span><span class="p">:</span> <span class="nb">str</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="nb">str</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Extracts the species name from a file path.</span>

<span class="sd">    Args:</span>
<span class="sd">        file_path (str): The path to the file containing the species information.</span>

<span class="sd">    Returns:</span>
<span class="sd">        str: The extracted species name.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="c1"># Extract the filename from the path</span>
    <span class="n">filename</span> <span class="o">=</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">basename</span><span class="p">(</span><span class="n">file_path</span><span class="p">)</span>

    <span class="c1"># Remove the extension</span>
    <span class="n">filename_without_extension</span> <span class="o">=</span> <span class="n">filename</span><span class="o">.</span><span class="n">rsplit</span><span class="p">(</span><span class="s2">&quot;.&quot;</span><span class="p">,</span> <span class="mi">1</span><span class="p">)[</span><span class="mi">0</span><span class="p">]</span>

    <span class="c1"># Split the filename by underscore and get the last two parts</span>
    <span class="n">species</span> <span class="o">=</span> <span class="s2">&quot;_&quot;</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">filename_without_extension</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s2">&quot;_&quot;</span><span class="p">)[:</span><span class="mi">2</span><span class="p">])</span>

    <span class="k">return</span> <span class="n">species</span>
</code></pre></div></td></tr></table></div>
          </details>
  </div>

</div>


<div class="doc doc-object doc-function">



<h2 id="dna.ensembl_api.get_utr_sequence" class="doc doc-heading">
          <code class="highlight language-python"><span class="n">get_utr_sequence</span><span class="p">(</span><span class="n">chromosome</span><span class="p">,</span> <span class="n">strand</span><span class="p">,</span> <span class="n">start</span><span class="p">,</span> <span class="n">end</span><span class="p">,</span> <span class="n">species</span><span class="p">)</span></code>

</h2>


  <div class="doc doc-contents ">
  
      <p>Retrieves the nucleotide sequence of a UTR (Untranslated Region) from the Ensembl database.</p>



<table class="field-list">
  <colgroup>
    <col class="field-name" />
    <col class="field-body" />
  </colgroup>
  <tbody valign="top">
    <tr class="field">
      <th class="field-name">Parameters:</th>
      <td class="field-body">
        <ul class="first simple">
            <li>
              <b><code>chromosome</code></b>
                  (<code>str</code>)
              –
              <div class="doc-md-description">
                <p>Chromosome name or identifier.</p>
              </div>
            </li>
            <li>
              <b><code>strand</code></b>
                  (<code>int</code>)
              –
              <div class="doc-md-description">
                <p>Strand information (1 for forward strand, -1 for reverse strand).</p>
              </div>
            </li>
            <li>
              <b><code>start</code></b>
                  (<code>int</code>)
              –
              <div class="doc-md-description">
                <p>Start position of the UTR on the chromosome.</p>
              </div>
            </li>
            <li>
              <b><code>end</code></b>
                  (<code>int</code>)
              –
              <div class="doc-md-description">
                <p>End position of the UTR on the chromosome.</p>
              </div>
            </li>
            <li>
              <b><code>species</code></b>
                  (<code>str</code>)
              –
              <div class="doc-md-description">
                <p>Species for which the UTR sequence is requested.</p>
              </div>
            </li>
        </ul>
      </td>
    </tr>
  </tbody>
</table>


<table class="field-list">
  <colgroup>
    <col class="field-name" />
    <col class="field-body" />
  </colgroup>
  <tbody valign="top">
    <tr class="field">
    <th class="field-name">Returns:</th>
    <td class="field-body">
      <ul class="first simple">
          <li>
<b><code>str</code></b>(                <code>str</code>
)            –
            <div class="doc-md-description">
              <p>The nucleotide sequence of the specified UTR.</p>
            </div>
          </li>
      </ul>
    </td>
    </tr>
  </tbody>
</table>
          <details class="quote">
            <summary>Source code in <code>dna\ensembl_api.py</code></summary>
            <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">154</span>
<span class="normal">155</span>
<span class="normal">156</span>
<span class="normal">157</span>
<span class="normal">158</span>
<span class="normal">159</span>
<span class="normal">160</span>
<span class="normal">161</span>
<span class="normal">162</span>
<span class="normal">163</span>
<span class="normal">164</span>
<span class="normal">165</span>
<span class="normal">166</span>
<span class="normal">167</span>
<span class="normal">168</span>
<span class="normal">169</span>
<span class="normal">170</span>
<span class="normal">171</span>
<span class="normal">172</span>
<span class="normal">173</span>
<span class="normal">174</span>
<span class="normal">175</span>
<span class="normal">176</span>
<span class="normal">177</span>
<span class="normal">178</span>
<span class="normal">179</span>
<span class="normal">180</span>
<span class="normal">181</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span> <span class="nf">get_utr_sequence</span><span class="p">(</span>
    <span class="n">chromosome</span><span class="p">:</span> <span class="nb">str</span><span class="p">,</span> <span class="n">strand</span><span class="p">:</span> <span class="nb">int</span><span class="p">,</span> <span class="n">start</span><span class="p">:</span> <span class="nb">int</span><span class="p">,</span> <span class="n">end</span><span class="p">:</span> <span class="nb">int</span><span class="p">,</span> <span class="n">species</span><span class="p">:</span> <span class="nb">str</span>
<span class="p">)</span> <span class="o">-&gt;</span> <span class="nb">str</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Retrieves the nucleotide sequence of a UTR (Untranslated Region) from the Ensembl database.</span>

<span class="sd">    Args:</span>
<span class="sd">        chromosome (str): Chromosome name or identifier.</span>
<span class="sd">        strand (int): Strand information (1 for forward strand, -1 for reverse strand).</span>
<span class="sd">        start (int): Start position of the UTR on the chromosome.</span>
<span class="sd">        end (int): End position of the UTR on the chromosome.</span>
<span class="sd">        species (str): Species for which the UTR sequence is requested.</span>

<span class="sd">    Returns:</span>
<span class="sd">        str: The nucleotide sequence of the specified UTR.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="c1"># Use Ensembl REST API to retrieve UTR sequence for the specified region</span>
    <span class="n">region</span> <span class="o">=</span> <span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">chromosome</span><span class="si">}</span><span class="s2">:</span><span class="si">{</span><span class="n">start</span><span class="si">}</span><span class="s2">..</span><span class="si">{</span><span class="n">end</span><span class="si">}</span><span class="s2">:</span><span class="si">{</span><span class="n">strand</span><span class="si">}</span><span class="s2">&quot;</span>
    <span class="k">try</span><span class="p">:</span>
        <span class="n">utr_sequence</span> <span class="o">=</span> <span class="n">ensembl_rest</span><span class="o">.</span><span class="n">sequence_region</span><span class="p">(</span><span class="n">region</span><span class="o">=</span><span class="n">region</span><span class="p">,</span> <span class="n">species</span><span class="o">=</span><span class="n">species</span><span class="p">)[</span>
            <span class="s2">&quot;seq&quot;</span>
        <span class="p">]</span>
    <span class="k">except</span> <span class="n">ensembl_rest</span><span class="o">.</span><span class="n">core</span><span class="o">.</span><span class="n">restclient</span><span class="o">.</span><span class="n">HTTPError</span> <span class="k">as</span> <span class="n">e</span><span class="p">:</span>
        <span class="k">if</span> <span class="n">e</span><span class="o">.</span><span class="n">response</span><span class="o">.</span><span class="n">status_code</span> <span class="o">==</span> <span class="mi">429</span><span class="p">:</span>  <span class="c1"># Check for rate limit exceeded error</span>
            <span class="nb">print</span><span class="p">(</span><span class="s2">&quot;Rate limit exceeded. Waiting before retrying...&quot;</span><span class="p">)</span>
            <span class="n">time</span><span class="o">.</span><span class="n">sleep</span><span class="p">(</span><span class="mi">1</span><span class="p">)</span>
            <span class="n">utr_sequence</span> <span class="o">=</span> <span class="n">get_utr_sequence</span><span class="p">(</span><span class="n">chromosome</span><span class="p">,</span> <span class="n">strand</span><span class="p">,</span> <span class="n">start</span><span class="p">,</span> <span class="n">end</span><span class="p">,</span> <span class="n">species</span><span class="p">)</span>

    <span class="k">return</span> <span class="n">utr_sequence</span>
</code></pre></div></td></tr></table></div>
          </details>
  </div>

</div>


<div class="doc doc-object doc-function">



<h2 id="dna.ensembl_api.read_gene_ids_from_file" class="doc doc-heading">
          <code class="highlight language-python"><span class="n">read_gene_ids_from_file</span><span class="p">(</span><span class="n">file_path</span><span class="p">)</span></code>

</h2>


  <div class="doc doc-contents ">
  
      <p>Reads gene IDs from a file, skipping the first line.</p>



<table class="field-list">
  <colgroup>
    <col class="field-name" />
    <col class="field-body" />
  </colgroup>
  <tbody valign="top">
    <tr class="field">
      <th class="field-name">Parameters:</th>
      <td class="field-body">
        <ul class="first simple">
            <li>
              <b><code>file_path</code></b>
                  (<code>str</code>)
              –
              <div class="doc-md-description">
                <p>The path to the file containing gene IDs.</p>
              </div>
            </li>
        </ul>
      </td>
    </tr>
  </tbody>
</table>


<table class="field-list">
  <colgroup>
    <col class="field-name" />
    <col class="field-body" />
  </colgroup>
  <tbody valign="top">
    <tr class="field">
    <th class="field-name">Returns:</th>
    <td class="field-body">
      <ul class="first simple">
          <li>
                <code><span title="typing.List">List</span>[str]</code>
            –
            <div class="doc-md-description">
              <p>List[str]: A list of gene IDs read from the file.</p>
            </div>
          </li>
      </ul>
    </td>
    </tr>
  </tbody>
</table>
          <details class="quote">
            <summary>Source code in <code>dna\ensembl_api.py</code></summary>
            <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">10</span>
<span class="normal">11</span>
<span class="normal">12</span>
<span class="normal">13</span>
<span class="normal">14</span>
<span class="normal">15</span>
<span class="normal">16</span>
<span class="normal">17</span>
<span class="normal">18</span>
<span class="normal">19</span>
<span class="normal">20</span>
<span class="normal">21</span>
<span class="normal">22</span>
<span class="normal">23</span>
<span class="normal">24</span>
<span class="normal">25</span>
<span class="normal">26</span>
<span class="normal">27</span>
<span class="normal">28</span>
<span class="normal">29</span>
<span class="normal">30</span>
<span class="normal">31</span>
<span class="normal">32</span>
<span class="normal">33</span>
<span class="normal">34</span>
<span class="normal">35</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span> <span class="nf">read_gene_ids_from_file</span><span class="p">(</span><span class="n">file_path</span><span class="p">:</span> <span class="nb">str</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="n">List</span><span class="p">[</span><span class="nb">str</span><span class="p">]:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Reads gene IDs from a file, skipping the first line.</span>

<span class="sd">    Args:</span>
<span class="sd">        file_path (str): The path to the file containing gene IDs.</span>

<span class="sd">    Returns:</span>
<span class="sd">        List[str]: A list of gene IDs read from the file.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="k">try</span><span class="p">:</span>
        <span class="k">with</span> <span class="nb">open</span><span class="p">(</span><span class="n">file_path</span><span class="p">,</span> <span class="s2">&quot;r&quot;</span><span class="p">,</span> <span class="n">encoding</span><span class="o">=</span><span class="s2">&quot;utf-8&quot;</span><span class="p">)</span> <span class="k">as</span> <span class="n">file</span><span class="p">:</span>
            <span class="c1"># Skip the first line</span>
            <span class="nb">next</span><span class="p">(</span><span class="n">file</span><span class="p">)</span>

            <span class="c1"># Read gene IDs from the remaining lines</span>
            <span class="n">gene_ids</span> <span class="o">=</span> <span class="p">[</span><span class="n">line</span><span class="o">.</span><span class="n">strip</span><span class="p">()</span> <span class="k">for</span> <span class="n">line</span> <span class="ow">in</span> <span class="n">file</span><span class="p">]</span>

        <span class="k">return</span> <span class="n">gene_ids</span>
    <span class="k">except</span> <span class="ne">FileNotFoundError</span><span class="p">:</span>
        <span class="c1"># Handle file not found error</span>
        <span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Error: File &#39;</span><span class="si">{</span><span class="n">file_path</span><span class="si">}</span><span class="s2">&#39; not found.&quot;</span><span class="p">)</span>
        <span class="k">return</span> <span class="p">[]</span>
    <span class="k">except</span> <span class="ne">Exception</span> <span class="k">as</span> <span class="n">e</span><span class="p">:</span>
        <span class="c1"># Handle other exceptions</span>
        <span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Error: </span><span class="si">{</span><span class="nb">str</span><span class="p">(</span><span class="n">e</span><span class="p">)</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
        <span class="k">return</span> <span class="p">[]</span>
</code></pre></div></td></tr></table></div>
          </details>
  </div>

</div>


<div class="doc doc-object doc-function">



<h2 id="dna.ensembl_api.request_with_retry" class="doc doc-heading">
          <code class="highlight language-python"><span class="n">request_with_retry</span><span class="p">(</span><span class="n">transcript_id</span><span class="p">)</span></code>

</h2>


  <div class="doc doc-contents ">
  
      <p>Retries the request for a transcript ID</p>



<table class="field-list">
  <colgroup>
    <col class="field-name" />
    <col class="field-body" />
  </colgroup>
  <tbody valign="top">
    <tr class="field">
      <th class="field-name">Parameters:</th>
      <td class="field-body">
        <ul class="first simple">
            <li>
              <b><code>transcript_id</code></b>
                  (<code>str</code>)
              –
              <div class="doc-md-description">
                <p>Ensembl transcript ID for the target gene.</p>
              </div>
            </li>
        </ul>
      </td>
    </tr>
  </tbody>
</table>


<table class="field-list">
  <colgroup>
    <col class="field-name" />
    <col class="field-body" />
  </colgroup>
  <tbody valign="top">
    <tr class="field">
    <th class="field-name">Returns:</th>
    <td class="field-body">
      <ul class="first simple">
          <li>
<b><code>dict</code></b>(                <code><span title="typing.Dict">Dict</span></code>
)            –
            <div class="doc-md-description">
              <p>A dictionary containing information about the transcript.</p>
            </div>
          </li>
      </ul>
    </td>
    </tr>
  </tbody>
</table>
          <details class="quote">
            <summary>Source code in <code>dna\ensembl_api.py</code></summary>
            <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">232</span>
<span class="normal">233</span>
<span class="normal">234</span>
<span class="normal">235</span>
<span class="normal">236</span>
<span class="normal">237</span>
<span class="normal">238</span>
<span class="normal">239</span>
<span class="normal">240</span>
<span class="normal">241</span>
<span class="normal">242</span>
<span class="normal">243</span>
<span class="normal">244</span>
<span class="normal">245</span>
<span class="normal">246</span>
<span class="normal">247</span>
<span class="normal">248</span>
<span class="normal">249</span>
<span class="normal">250</span>
<span class="normal">251</span>
<span class="normal">252</span>
<span class="normal">253</span>
<span class="normal">254</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span> <span class="nf">request_with_retry</span><span class="p">(</span><span class="n">transcript_id</span><span class="p">:</span> <span class="nb">str</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="n">Dict</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Retries the request for a transcript ID</span>

<span class="sd">    Args:</span>
<span class="sd">        transcript_id (str): Ensembl transcript ID for the target gene.</span>

<span class="sd">    Returns:</span>
<span class="sd">        dict: A dictionary containing information about the transcript.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="k">while</span> <span class="kc">True</span><span class="p">:</span>
        <span class="k">try</span><span class="p">:</span>
            <span class="n">transcript_data</span> <span class="o">=</span> <span class="n">ensembl_rest</span><span class="o">.</span><span class="n">lookup</span><span class="p">(</span>
                <span class="nb">id</span><span class="o">=</span><span class="n">transcript_id</span><span class="p">,</span> <span class="n">params</span><span class="o">=</span><span class="p">{</span><span class="s2">&quot;expand&quot;</span><span class="p">:</span> <span class="kc">True</span><span class="p">,</span> <span class="s2">&quot;utr&quot;</span><span class="p">:</span> <span class="kc">True</span><span class="p">}</span>
            <span class="p">)</span>
            <span class="k">return</span> <span class="n">transcript_data</span>
        <span class="k">except</span> <span class="n">ensembl_rest</span><span class="o">.</span><span class="n">core</span><span class="o">.</span><span class="n">restclient</span><span class="o">.</span><span class="n">HTTPError</span> <span class="k">as</span> <span class="n">e</span><span class="p">:</span>
            <span class="k">if</span> <span class="n">e</span><span class="o">.</span><span class="n">response</span><span class="o">.</span><span class="n">status_code</span> <span class="o">==</span> <span class="mi">429</span><span class="p">:</span>  <span class="c1"># Check for rate limit exceeded error</span>
                <span class="nb">print</span><span class="p">(</span><span class="s2">&quot;Rate limit exceeded. Waiting before retrying...&quot;</span><span class="p">)</span>
                <span class="n">time</span><span class="o">.</span><span class="n">sleep</span><span class="p">(</span><span class="mi">1</span><span class="p">)</span>
                <span class="k">continue</span>

            <span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Error with the request for </span><span class="si">{</span><span class="n">transcript_id</span><span class="si">}</span><span class="s2">: </span><span class="si">{</span><span class="n">e</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
            <span class="k">return</span> <span class="p">{}</span>
</code></pre></div></td></tr></table></div>
          </details>
  </div>

</div>



  </div>

  </div>

</div>
              
            </div>
          </div><footer>
    <div class="rst-footer-buttons" role="navigation" aria-label="Footer Navigation">
        <a href="../dna_feature_extraction/" class="btn btn-neutral float-left" title="dna_feature_extraction"><span class="icon icon-circle-arrow-left"></span> Previous</a>
        <a href="../../rna/rna_extraction/" class="btn btn-neutral float-right" title="rna_extraction">Next <span class="icon icon-circle-arrow-right"></span></a>
    </div>

  <hr/>

  <div role="contentinfo">
    <!-- Copyright etc -->
  </div>

  Built with <a href="https://www.mkdocs.org/">MkDocs</a> using a <a href="https://github.com/readthedocs/sphinx_rtd_theme">theme</a> provided by <a href="https://readthedocs.org">Read the Docs</a>.
</footer>
          
        </div>
      </div>

    </section>

  </div>

  <div class="rst-versions" role="note" aria-label="Versions">
  <span class="rst-current-version" data-toggle="rst-current-version">
    
        <span>
          <a href="https://github.com/TransferLearningForDNA/genomic_data_extraction" class="fa fa-github" style="color: #fcfcfc"> GitHub</a>
        </span>
    
    
      <span><a href="../dna_feature_extraction/" style="color: #fcfcfc">&laquo; Previous</a></span>
    
    
      <span><a href="../../rna/rna_extraction/" style="color: #fcfcfc">Next &raquo;</a></span>
    
  </span>
</div>
    <script>var base_url = '../../..';</script>
    <script src="../../../js/theme_extra.js" defer></script>
    <script src="../../../js/theme.js" defer></script>
    <script defer>
        window.onload = function () {
            SphinxRtdTheme.Navigation.enable(true);
        };
    </script>

</body>
</html>
