<!DOCTYPE html>
<html class="writer-html5" lang="en" >
<head>
    <meta charset="utf-8" />
    <meta http-equiv="X-UA-Compatible" content="IE=edge" />
    <meta name="viewport" content="width=device-width, initial-scale=1.0" />
      <link rel="shortcut icon" href="../../../../img/favicon.ico" />
    <title>create_expression_matrix - Genomic Data Extraction</title>
    <link rel="stylesheet" href="../../../../css/theme.css" />
    <link rel="stylesheet" href="../../../../css/theme_extra.css" />
        <link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/highlight.js/10.5.0/styles/github.min.css" />
        <link href="../../../../assets/_mkdocstrings.css" rel="stylesheet" />
    
      <script>
        // Current page data
        var mkdocs_page_name = "create_expression_matrix";
        var mkdocs_page_input_path = "genomic_data_extraction\\rna\\data_conversion_helper_functions\\create_expression_matrix.md";
        var mkdocs_page_url = null;
      </script>
    
    <script src="../../../../js/jquery-3.6.0.min.js" defer></script>
    <!--[if lt IE 9]>
      <script src="../../../../js/html5shiv.min.js"></script>
    <![endif]-->
      <script src="https://cdnjs.cloudflare.com/ajax/libs/highlight.js/10.5.0/highlight.min.js"></script>
      <script>hljs.initHighlightingOnLoad();</script> 
</head>

<body class="wy-body-for-nav" role="document">

  <div class="wy-grid-for-nav">
    <nav data-toggle="wy-nav-shift" class="wy-nav-side stickynav">
    <div class="wy-side-scroll">
      <div class="wy-side-nav-search">
          <a href="../../../.." class="icon icon-home"> Genomic Data Extraction
        </a>
      </div>

      <div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="Navigation menu">
              <ul>
                <li class="toctree-l1"><a class="reference internal" href="../../../..">Home</a>
                </li>
              </ul>
              <p class="caption"><span class="caption-text">genomic_data_extraction</span></p>
              <ul class="current">
                  <li class="toctree-l1"><a class="reference internal" href="#">dna</a>
    <ul>
                <li class="toctree-l2"><a class="reference internal" href="../../../dna/dna_extraction/">dna_extraction</a>
                </li>
                <li class="toctree-l2"><a class="reference internal" href="../../../dna/dna_feature_extraction/">dna_feature_extraction</a>
                </li>
                <li class="toctree-l2"><a class="reference internal" href="../../../dna/ensembl_api/">ensembl_api</a>
                </li>
    </ul>
                  </li>
                  <li class="toctree-l1 current"><a class="reference internal current" href="#">rna</a>
    <ul class="current">
                <li class="toctree-l2 current"><a class="reference internal current" href="#">data_conversion_helper_functions</a>
    <ul class="current">
                <li class="toctree-l3"><a class="reference internal" href="../convert_quantsf_to_csv/">convert_quantsf_to_csv</a>
                </li>
                <li class="toctree-l3 current"><a class="reference internal current" href="./">create_expression_matrix</a>
    <ul class="current">
    </ul>
                </li>
                <li class="toctree-l3"><a class="reference internal" href="../create_samplesheet_csv/">create_samplesheet_csv</a>
                </li>
                <li class="toctree-l3"><a class="reference internal" href="../process_expression_matrix/">process_expression_matrix</a>
                </li>
    </ul>
                </li>
                <li class="toctree-l2"><a class="reference internal" href="#">rna_download_logic</a>
    <ul>
                <li class="toctree-l3"><a class="reference internal" href="../../rna_download_logic/mRNA_fastq_download/">mRNA_fastq_download</a>
                </li>
                <li class="toctree-l3"><a class="reference internal" href="../../rna_download_logic/query_and_csv_production/">query_and_csv_production</a>
                </li>
    </ul>
                </li>
                <li class="toctree-l2"><a class="reference internal" href="../../rna_extraction/">rna_extraction</a>
                </li>
    </ul>
                  </li>
              </ul>
      </div>
    </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap">
      <nav class="wy-nav-top" role="navigation" aria-label="Mobile navigation menu">
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="../../../..">Genomic Data Extraction</a>
        
      </nav>
      <div class="wy-nav-content">
        <div class="rst-content"><div role="navigation" aria-label="breadcrumbs navigation">
  <ul class="wy-breadcrumbs">
    <li><a href="../../../.." class="icon icon-home" alt="Docs"></a> &raquo;</li>
          <li>genomic_data_extraction &raquo;</li>
          <li>rna &raquo;</li>
          <li>data_conversion_helper_functions &raquo;</li>
      <li>create_expression_matrix</li>
    <li class="wy-breadcrumbs-aside">
          <a href="https://github.com/TransferLearningForDNA/genomic_data_extraction/edit/master/docs/genomic_data_extraction/rna/data_conversion_helper_functions/create_expression_matrix.md" class="icon icon-github"> Edit on GitHub</a>
    </li>
  </ul>
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
            <div class="section" itemprop="articleBody">
              
                <div class="doc doc-object doc-module">



<a id="rna.data_conversion_helper_functions.create_expression_matrix"></a>
  <div class="doc doc-contents first">

  

  <div class="doc doc-children">










<div class="doc doc-object doc-function">



<h2 id="rna.data_conversion_helper_functions.create_expression_matrix.create_expression_matrix" class="doc doc-heading">
          <code class="highlight language-python"><span class="n">create_expression_matrix</span><span class="p">(</span><span class="n">raw_data_path</span><span class="p">,</span> <span class="n">processed_data_path</span><span class="p">)</span></code>

</h2>


  <div class="doc doc-contents ">
  
      <p>Create the expression matrices for all species.</p>



<table class="field-list">
  <colgroup>
    <col class="field-name" />
    <col class="field-body" />
  </colgroup>
  <tbody valign="top">
    <tr class="field">
      <th class="field-name">Parameters:</th>
      <td class="field-body">
        <ul class="first simple">
            <li>
              <b><code>raw_data_path</code></b>
                  (<code>str</code>)
              –
              <div class="doc-md-description">
                <p>Path to the folder containing raw quant files.</p>
              </div>
            </li>
            <li>
              <b><code>processed_data_path</code></b>
                  (<code>str</code>)
              –
              <div class="doc-md-description">
                <p>Path to store the processed expression matrix csv files.</p>
              </div>
            </li>
        </ul>
      </td>
    </tr>
  </tbody>
</table>


<table class="field-list">
  <colgroup>
    <col class="field-name" />
    <col class="field-body" />
  </colgroup>
  <tbody valign="top">
    <tr class="field">
    <th class="field-name">Returns:</th>
    <td class="field-body">
      <ul class="first simple">
          <li>
<b><code>None</code></b>(                <code>None</code>
)            –
            <div class="doc-md-description">
              <p>This function does not return a value but outputs files to the specified directory.</p>
            </div>
          </li>
      </ul>
    </td>
    </tr>
  </tbody>
</table>
          <details class="quote">
            <summary>Source code in <code>rna\data_conversion_helper_functions\create_expression_matrix.py</code></summary>
            <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal"> 40</span>
<span class="normal"> 41</span>
<span class="normal"> 42</span>
<span class="normal"> 43</span>
<span class="normal"> 44</span>
<span class="normal"> 45</span>
<span class="normal"> 46</span>
<span class="normal"> 47</span>
<span class="normal"> 48</span>
<span class="normal"> 49</span>
<span class="normal"> 50</span>
<span class="normal"> 51</span>
<span class="normal"> 52</span>
<span class="normal"> 53</span>
<span class="normal"> 54</span>
<span class="normal"> 55</span>
<span class="normal"> 56</span>
<span class="normal"> 57</span>
<span class="normal"> 58</span>
<span class="normal"> 59</span>
<span class="normal"> 60</span>
<span class="normal"> 61</span>
<span class="normal"> 62</span>
<span class="normal"> 63</span>
<span class="normal"> 64</span>
<span class="normal"> 65</span>
<span class="normal"> 66</span>
<span class="normal"> 67</span>
<span class="normal"> 68</span>
<span class="normal"> 69</span>
<span class="normal"> 70</span>
<span class="normal"> 71</span>
<span class="normal"> 72</span>
<span class="normal"> 73</span>
<span class="normal"> 74</span>
<span class="normal"> 75</span>
<span class="normal"> 76</span>
<span class="normal"> 77</span>
<span class="normal"> 78</span>
<span class="normal"> 79</span>
<span class="normal"> 80</span>
<span class="normal"> 81</span>
<span class="normal"> 82</span>
<span class="normal"> 83</span>
<span class="normal"> 84</span>
<span class="normal"> 85</span>
<span class="normal"> 86</span>
<span class="normal"> 87</span>
<span class="normal"> 88</span>
<span class="normal"> 89</span>
<span class="normal"> 90</span>
<span class="normal"> 91</span>
<span class="normal"> 92</span>
<span class="normal"> 93</span>
<span class="normal"> 94</span>
<span class="normal"> 95</span>
<span class="normal"> 96</span>
<span class="normal"> 97</span>
<span class="normal"> 98</span>
<span class="normal"> 99</span>
<span class="normal">100</span>
<span class="normal">101</span>
<span class="normal">102</span>
<span class="normal">103</span>
<span class="normal">104</span>
<span class="normal">105</span>
<span class="normal">106</span>
<span class="normal">107</span>
<span class="normal">108</span>
<span class="normal">109</span>
<span class="normal">110</span>
<span class="normal">111</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span> <span class="nf">create_expression_matrix</span><span class="p">(</span><span class="n">raw_data_path</span><span class="p">:</span> <span class="nb">str</span><span class="p">,</span> <span class="n">processed_data_path</span><span class="p">:</span> <span class="nb">str</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="kc">None</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Create the expression matrices for all species.</span>

<span class="sd">    Args:</span>
<span class="sd">        raw_data_path (str): Path to the folder containing raw quant files.</span>
<span class="sd">        processed_data_path (str): Path to store the processed expression matrix csv files.</span>

<span class="sd">    Returns:</span>
<span class="sd">        None: This function does not return a value but outputs files to the specified directory.</span>
<span class="sd">    &quot;&quot;&quot;</span>

    <span class="c1"># Iterate over species</span>
    <span class="k">for</span> <span class="n">species</span> <span class="ow">in</span> <span class="n">os</span><span class="o">.</span><span class="n">listdir</span><span class="p">(</span><span class="n">raw_data_path</span><span class="p">):</span>
        <span class="n">raw_csv_data_path</span> <span class="o">=</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">raw_data_path</span><span class="p">,</span> <span class="n">species</span><span class="p">,</span> <span class="s2">&quot;csv_files&quot;</span><span class="p">)</span>

        <span class="c1"># Skip if it&#39;s not a directory (e.g., .DS_Store)</span>
        <span class="k">if</span> <span class="ow">not</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">isdir</span><span class="p">(</span><span class="n">raw_csv_data_path</span><span class="p">):</span>
            <span class="k">continue</span>
        <span class="k">if</span> <span class="ow">not</span> <span class="n">os</span><span class="o">.</span><span class="n">listdir</span><span class="p">(</span><span class="n">raw_csv_data_path</span><span class="p">):</span>
            <span class="k">continue</span>

        <span class="c1"># Initialise empty DataFrames for the abundance, length and counts matrices</span>
        <span class="n">abundance_mat</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">()</span>  <span class="c1"># TPM</span>
        <span class="n">length_mat</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">()</span>  <span class="c1"># EffectiveLength</span>
        <span class="n">counts_mat</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">()</span>  <span class="c1"># NumReads</span>

        <span class="c1"># Iterate through each quant csv file</span>
        <span class="k">for</span> <span class="n">quant_file</span> <span class="ow">in</span> <span class="n">os</span><span class="o">.</span><span class="n">listdir</span><span class="p">(</span><span class="n">raw_csv_data_path</span><span class="p">):</span>
            <span class="n">file_path</span> <span class="o">=</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">raw_csv_data_path</span><span class="p">,</span> <span class="n">quant_file</span><span class="p">)</span>
            <span class="c1"># Read the csv file into a DataFrame</span>
            <span class="c1"># Abundance TPM matrix</span>
            <span class="n">abundance_df</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">read_csv</span><span class="p">(</span><span class="n">file_path</span><span class="p">,</span> <span class="n">usecols</span><span class="o">=</span><span class="p">[</span><span class="s2">&quot;Name&quot;</span><span class="p">,</span> <span class="s2">&quot;TPM&quot;</span><span class="p">])</span>
            <span class="c1"># Effective Length matrix</span>
            <span class="n">length_df</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">read_csv</span><span class="p">(</span><span class="n">file_path</span><span class="p">,</span> <span class="n">usecols</span><span class="o">=</span><span class="p">[</span><span class="s2">&quot;Name&quot;</span><span class="p">,</span> <span class="s2">&quot;EffectiveLength&quot;</span><span class="p">])</span>
            <span class="c1"># Counts NumReads matrix</span>
            <span class="n">counts_df</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">read_csv</span><span class="p">(</span><span class="n">file_path</span><span class="p">,</span> <span class="n">usecols</span><span class="o">=</span><span class="p">[</span><span class="s2">&quot;Name&quot;</span><span class="p">,</span> <span class="s2">&quot;NumReads&quot;</span><span class="p">])</span>

            <span class="c1"># Rename &#39;TPM&#39; column to the run ID (file name e.g. &#39;quant_DRR513083.csv&#39;)</span>
            <span class="n">abundance_df</span> <span class="o">=</span> <span class="n">abundance_df</span><span class="o">.</span><span class="n">rename</span><span class="p">(</span>
                <span class="n">columns</span><span class="o">=</span><span class="p">{</span><span class="s2">&quot;TPM&quot;</span><span class="p">:</span> <span class="n">quant_file</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s2">&quot;_&quot;</span><span class="p">)[</span><span class="mi">1</span><span class="p">][:</span><span class="o">-</span><span class="mi">4</span><span class="p">]}</span>
            <span class="p">)</span>
            <span class="n">length_df</span> <span class="o">=</span> <span class="n">length_df</span><span class="o">.</span><span class="n">rename</span><span class="p">(</span>
                <span class="n">columns</span><span class="o">=</span><span class="p">{</span><span class="s2">&quot;EffectiveLength&quot;</span><span class="p">:</span> <span class="n">quant_file</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s2">&quot;_&quot;</span><span class="p">)[</span><span class="mi">1</span><span class="p">][:</span><span class="o">-</span><span class="mi">4</span><span class="p">]}</span>
            <span class="p">)</span>
            <span class="n">counts_df</span> <span class="o">=</span> <span class="n">counts_df</span><span class="o">.</span><span class="n">rename</span><span class="p">(</span>
                <span class="n">columns</span><span class="o">=</span><span class="p">{</span><span class="s2">&quot;NumReads&quot;</span><span class="p">:</span> <span class="n">quant_file</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s2">&quot;_&quot;</span><span class="p">)[</span><span class="mi">1</span><span class="p">][:</span><span class="o">-</span><span class="mi">4</span><span class="p">]}</span>
            <span class="p">)</span>
            <span class="c1"># Set &#39;Name&#39; as index</span>
            <span class="n">abundance_df</span> <span class="o">=</span> <span class="n">abundance_df</span><span class="o">.</span><span class="n">set_index</span><span class="p">(</span><span class="s2">&quot;Name&quot;</span><span class="p">)</span>
            <span class="n">length_df</span> <span class="o">=</span> <span class="n">length_df</span><span class="o">.</span><span class="n">set_index</span><span class="p">(</span><span class="s2">&quot;Name&quot;</span><span class="p">)</span>
            <span class="n">counts_df</span> <span class="o">=</span> <span class="n">counts_df</span><span class="o">.</span><span class="n">set_index</span><span class="p">(</span><span class="s2">&quot;Name&quot;</span><span class="p">)</span>
            <span class="c1"># Remove duplicates</span>
            <span class="n">abundance_df</span> <span class="o">=</span> <span class="n">abundance_df</span><span class="p">[</span><span class="o">~</span><span class="n">abundance_df</span><span class="o">.</span><span class="n">index</span><span class="o">.</span><span class="n">duplicated</span><span class="p">(</span><span class="n">keep</span><span class="o">=</span><span class="s2">&quot;first&quot;</span><span class="p">)]</span>
            <span class="n">length_df</span> <span class="o">=</span> <span class="n">length_df</span><span class="p">[</span><span class="o">~</span><span class="n">length_df</span><span class="o">.</span><span class="n">index</span><span class="o">.</span><span class="n">duplicated</span><span class="p">(</span><span class="n">keep</span><span class="o">=</span><span class="s2">&quot;first&quot;</span><span class="p">)]</span>
            <span class="n">counts_df</span> <span class="o">=</span> <span class="n">counts_df</span><span class="p">[</span><span class="o">~</span><span class="n">counts_df</span><span class="o">.</span><span class="n">index</span><span class="o">.</span><span class="n">duplicated</span><span class="p">(</span><span class="n">keep</span><span class="o">=</span><span class="s2">&quot;first&quot;</span><span class="p">)]</span>

            <span class="c1"># Concatenate with the main expression matrix</span>
            <span class="c1"># If the index (Name) does not exist in expression_matrix, NaN values will be added for other columns</span>
            <span class="n">abundance_mat</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">concat</span><span class="p">([</span><span class="n">abundance_mat</span><span class="p">,</span> <span class="n">abundance_df</span><span class="p">],</span> <span class="n">axis</span><span class="o">=</span><span class="mi">1</span><span class="p">,</span> <span class="n">sort</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>
            <span class="n">length_mat</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">concat</span><span class="p">([</span><span class="n">length_mat</span><span class="p">,</span> <span class="n">length_df</span><span class="p">],</span> <span class="n">axis</span><span class="o">=</span><span class="mi">1</span><span class="p">,</span> <span class="n">sort</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>
            <span class="n">counts_mat</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">concat</span><span class="p">([</span><span class="n">counts_mat</span><span class="p">,</span> <span class="n">counts_df</span><span class="p">],</span> <span class="n">axis</span><span class="o">=</span><span class="mi">1</span><span class="p">,</span> <span class="n">sort</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>

        <span class="c1"># Calculate the length scaled TPM values</span>
        <span class="n">length_scaled_tpm_mat</span> <span class="o">=</span> <span class="n">get_length_scaled_tpm_matrix</span><span class="p">(</span>
            <span class="n">counts_mat</span><span class="p">,</span> <span class="n">abundance_mat</span><span class="p">,</span> <span class="n">length_mat</span>
        <span class="p">)</span>

        <span class="c1"># Save the processed expression matrix to a CSV file</span>
        <span class="n">expression_matrix_path</span> <span class="o">=</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">processed_data_path</span><span class="p">,</span> <span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">species</span><span class="si">}</span><span class="s2">.csv&quot;</span><span class="p">)</span>
        <span class="n">length_scaled_tpm_mat</span><span class="o">.</span><span class="n">to_csv</span><span class="p">(</span><span class="n">expression_matrix_path</span><span class="p">)</span>

        <span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;</span><span class="se">\n</span><span class="s2">Expression matrix for </span><span class="si">{</span><span class="n">species</span><span class="si">}</span><span class="s2"> created successfully.&quot;</span><span class="p">)</span>
</code></pre></div></td></tr></table></div>
          </details>
  </div>

</div>


<div class="doc doc-object doc-function">



<h2 id="rna.data_conversion_helper_functions.create_expression_matrix.get_length_scaled_tpm_matrix" class="doc doc-heading">
          <code class="highlight language-python"><span class="n">get_length_scaled_tpm_matrix</span><span class="p">(</span><span class="n">counts_mat</span><span class="p">,</span> <span class="n">abundance_mat</span><span class="p">,</span> <span class="n">length_mat</span><span class="p">)</span></code>

</h2>


  <div class="doc doc-contents ">
  
      <p>Generate length scaled TPM matrix.</p>



<table class="field-list">
  <colgroup>
    <col class="field-name" />
    <col class="field-body" />
  </colgroup>
  <tbody valign="top">
    <tr class="field">
      <th class="field-name">Parameters:</th>
      <td class="field-body">
        <ul class="first simple">
            <li>
              <b><code>counts_mat</code></b>
                  (<code>DataFrame</code>)
              –
              <div class="doc-md-description">
                <p>A matrix of original counts (NumReads).</p>
              </div>
            </li>
            <li>
              <b><code>abundance_mat</code></b>
                  (<code>DataFrame</code>)
              –
              <div class="doc-md-description">
                <p>A matrix of abundances (TPM).</p>
              </div>
            </li>
            <li>
              <b><code>length_mat</code></b>
                  (<code>DataFrame</code>)
              –
              <div class="doc-md-description">
                <p>A matrix of effective lengths.</p>
              </div>
            </li>
        </ul>
      </td>
    </tr>
  </tbody>
</table>


<table class="field-list">
  <colgroup>
    <col class="field-name" />
    <col class="field-body" />
  </colgroup>
  <tbody valign="top">
    <tr class="field">
    <th class="field-name">Returns:</th>
    <td class="field-body">
      <ul class="first simple">
          <li>
<b><code>DataFrame</code></b>(                <code><span title="pandas.DataFrame">DataFrame</span></code>
)            –
            <div class="doc-md-description">
              <p>A matrix of length scaled TPM values indexed by transcript ID.</p>
            </div>
          </li>
      </ul>
    </td>
    </tr>
  </tbody>
</table>
          <details class="quote">
            <summary>Source code in <code>rna\data_conversion_helper_functions\create_expression_matrix.py</code></summary>
            <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal"> 5</span>
<span class="normal"> 6</span>
<span class="normal"> 7</span>
<span class="normal"> 8</span>
<span class="normal"> 9</span>
<span class="normal">10</span>
<span class="normal">11</span>
<span class="normal">12</span>
<span class="normal">13</span>
<span class="normal">14</span>
<span class="normal">15</span>
<span class="normal">16</span>
<span class="normal">17</span>
<span class="normal">18</span>
<span class="normal">19</span>
<span class="normal">20</span>
<span class="normal">21</span>
<span class="normal">22</span>
<span class="normal">23</span>
<span class="normal">24</span>
<span class="normal">25</span>
<span class="normal">26</span>
<span class="normal">27</span>
<span class="normal">28</span>
<span class="normal">29</span>
<span class="normal">30</span>
<span class="normal">31</span>
<span class="normal">32</span>
<span class="normal">33</span>
<span class="normal">34</span>
<span class="normal">35</span>
<span class="normal">36</span>
<span class="normal">37</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span> <span class="nf">get_length_scaled_tpm_matrix</span><span class="p">(</span>
    <span class="n">counts_mat</span><span class="p">:</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">,</span> <span class="n">abundance_mat</span><span class="p">:</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">,</span> <span class="n">length_mat</span><span class="p">:</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span>
<span class="p">)</span> <span class="o">-&gt;</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Generate length scaled TPM matrix.</span>

<span class="sd">    Args:</span>
<span class="sd">        counts_mat (DataFrame): A matrix of original counts (NumReads).</span>
<span class="sd">        abundance_mat (DataFrame): A matrix of abundances (TPM).</span>
<span class="sd">        length_mat (DataFrame): A matrix of effective lengths.</span>

<span class="sd">    Returns:</span>
<span class="sd">        DataFrame: A matrix of length scaled TPM values indexed by transcript ID.</span>
<span class="sd">    &quot;&quot;&quot;</span>

    <span class="c1"># Calculate the sum of counts across all transcripts for each sample: per-sample library size</span>
    <span class="n">counts_sum</span> <span class="o">=</span> <span class="n">counts_mat</span><span class="o">.</span><span class="n">sum</span><span class="p">(</span><span class="n">axis</span><span class="o">=</span><span class="mi">0</span><span class="p">)</span>

    <span class="c1"># Calculate the average transcript lengths (averaged over samples)</span>
    <span class="n">length_means</span> <span class="o">=</span> <span class="n">length_mat</span><span class="o">.</span><span class="n">mean</span><span class="p">(</span><span class="n">axis</span><span class="o">=</span><span class="mi">1</span><span class="p">)</span>

    <span class="c1"># Multiply abundance matrix (TPM) by average transcript lengths</span>
    <span class="n">length_tpm_mat</span> <span class="o">=</span> <span class="n">abundance_mat</span><span class="o">.</span><span class="n">multiply</span><span class="p">(</span><span class="n">length_means</span><span class="p">,</span> <span class="n">axis</span><span class="o">=</span><span class="mi">0</span><span class="p">)</span>

    <span class="c1"># Calculate the sum of new counts across transcripts for each sample</span>
    <span class="n">new_sum</span> <span class="o">=</span> <span class="n">length_tpm_mat</span><span class="o">.</span><span class="n">sum</span><span class="p">(</span><span class="n">axis</span><span class="o">=</span><span class="mi">0</span><span class="p">)</span>

    <span class="c1"># Determine the scaling factor for each sample</span>
    <span class="n">scaling_factor</span> <span class="o">=</span> <span class="n">counts_sum</span> <span class="o">/</span> <span class="n">new_sum</span>

    <span class="c1"># Apply the scaling factor to scale new counts to match the original total counts</span>
    <span class="n">length_scaled_tpm_mat</span> <span class="o">=</span> <span class="n">length_tpm_mat</span><span class="o">.</span><span class="n">multiply</span><span class="p">(</span><span class="n">scaling_factor</span><span class="p">,</span> <span class="n">axis</span><span class="o">=</span><span class="mi">1</span><span class="p">)</span>

    <span class="k">return</span> <span class="n">length_scaled_tpm_mat</span>
</code></pre></div></td></tr></table></div>
          </details>
  </div>

</div>



  </div>

  </div>

</div>
              
            </div>
          </div><footer>
    <div class="rst-footer-buttons" role="navigation" aria-label="Footer Navigation">
        <a href="../convert_quantsf_to_csv/" class="btn btn-neutral float-left" title="convert_quantsf_to_csv"><span class="icon icon-circle-arrow-left"></span> Previous</a>
        <a href="../create_samplesheet_csv/" class="btn btn-neutral float-right" title="create_samplesheet_csv">Next <span class="icon icon-circle-arrow-right"></span></a>
    </div>

  <hr/>

  <div role="contentinfo">
    <!-- Copyright etc -->
  </div>

  Built with <a href="https://www.mkdocs.org/">MkDocs</a> using a <a href="https://github.com/readthedocs/sphinx_rtd_theme">theme</a> provided by <a href="https://readthedocs.org">Read the Docs</a>.
</footer>
          
        </div>
      </div>

    </section>

  </div>

  <div class="rst-versions" role="note" aria-label="Versions">
  <span class="rst-current-version" data-toggle="rst-current-version">
    
        <span>
          <a href="https://github.com/TransferLearningForDNA/genomic_data_extraction" class="fa fa-github" style="color: #fcfcfc"> GitHub</a>
        </span>
    
    
      <span><a href="../convert_quantsf_to_csv/" style="color: #fcfcfc">&laquo; Previous</a></span>
    
    
      <span><a href="../create_samplesheet_csv/" style="color: #fcfcfc">Next &raquo;</a></span>
    
  </span>
</div>
    <script>var base_url = '../../../..';</script>
    <script src="../../../../js/theme_extra.js" defer></script>
    <script src="../../../../js/theme.js" defer></script>
    <script defer>
        window.onload = function () {
            SphinxRtdTheme.Navigation.enable(true);
        };
    </script>

</body>
</html>
