<!DOCTYPE html>
<html class="writer-html5" lang="en" >
<head>
    <meta charset="utf-8" />
    <meta http-equiv="X-UA-Compatible" content="IE=edge" />
    <meta name="viewport" content="width=device-width, initial-scale=1.0" />
      <link rel="shortcut icon" href="../../../../img/favicon.ico" />
    <title>create_expression_matrix - Genomic Data Extraction</title>
    <link rel="stylesheet" href="../../../../css/theme.css" />
    <link rel="stylesheet" href="../../../../css/theme_extra.css" />
        <link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/highlight.js/10.5.0/styles/github.min.css" />
        <link href="../../../../assets/_mkdocstrings.css" rel="stylesheet" />
    
      <script>
        // Current page data
        var mkdocs_page_name = "create_expression_matrix";
        var mkdocs_page_input_path = "genomic_data_extraction/rna/data_conversion_helper_functions/create_expression_matrix.md";
        var mkdocs_page_url = null;
      </script>
    
    <script src="../../../../js/jquery-3.6.0.min.js" defer></script>
    <!--[if lt IE 9]>
      <script src="../../../../js/html5shiv.min.js"></script>
    <![endif]-->
      <script src="https://cdnjs.cloudflare.com/ajax/libs/highlight.js/10.5.0/highlight.min.js"></script>
      <script>hljs.initHighlightingOnLoad();</script> 
</head>

<body class="wy-body-for-nav" role="document">

  <div class="wy-grid-for-nav">
    <nav data-toggle="wy-nav-shift" class="wy-nav-side stickynav">
    <div class="wy-side-scroll">
      <div class="wy-side-nav-search">
          <a href="../../../.." class="icon icon-home"> Genomic Data Extraction
        </a>
      </div>

      <div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="Navigation menu">
              <ul>
                <li class="toctree-l1"><a class="reference internal" href="../../../..">Home</a>
                </li>
              </ul>
              <p class="caption"><span class="caption-text">genomic_data_extraction</span></p>
              <ul class="current">
                  <li class="toctree-l1"><a class="reference internal" href="#">dna</a>
    <ul>
                <li class="toctree-l2"><a class="reference internal" href="../../../dna/dna_extraction/">dna_extraction</a>
                </li>
                <li class="toctree-l2"><a class="reference internal" href="../../../dna/dna_feature_extraction/">dna_feature_extraction</a>
                </li>
                <li class="toctree-l2"><a class="reference internal" href="../../../dna/ensembl_api/">ensembl_api</a>
                </li>
    </ul>
                  </li>
                  <li class="toctree-l1 current"><a class="reference internal current" href="#">rna</a>
    <ul class="current">
                <li class="toctree-l2 current"><a class="reference internal current" href="#">data_conversion_helper_functions</a>
    <ul class="current">
                <li class="toctree-l3"><a class="reference internal" href="../convert_quantsf_to_csv/">convert_quantsf_to_csv</a>
                </li>
                <li class="toctree-l3 current"><a class="reference internal current" href="./">create_expression_matrix</a>
    <ul class="current">
    </ul>
                </li>
                <li class="toctree-l3"><a class="reference internal" href="../create_samplesheet_csv/">create_samplesheet_csv</a>
                </li>
                <li class="toctree-l3"><a class="reference internal" href="../process_expression_matrix/">process_expression_matrix</a>
                </li>
    </ul>
                </li>
                <li class="toctree-l2"><a class="reference internal" href="#">rna_download_logic</a>
    <ul>
                <li class="toctree-l3"><a class="reference internal" href="../../rna_download_logic/mRNA_fastq_download/">mRNA_fastq_download</a>
                </li>
                <li class="toctree-l3"><a class="reference internal" href="../../rna_download_logic/query_and_csv_production/">query_and_csv_production</a>
                </li>
    </ul>
                </li>
                <li class="toctree-l2"><a class="reference internal" href="../../rna_extraction/">rna_extraction</a>
                </li>
    </ul>
                  </li>
              </ul>
      </div>
    </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap">
      <nav class="wy-nav-top" role="navigation" aria-label="Mobile navigation menu">
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="../../../..">Genomic Data Extraction</a>
        
      </nav>
      <div class="wy-nav-content">
        <div class="rst-content"><div role="navigation" aria-label="breadcrumbs navigation">
  <ul class="wy-breadcrumbs">
    <li><a href="../../../.." class="icon icon-home" alt="Docs"></a> &raquo;</li>
          <li>genomic_data_extraction &raquo;</li>
          <li>rna &raquo;</li>
          <li>data_conversion_helper_functions &raquo;</li>
      <li>create_expression_matrix</li>
    <li class="wy-breadcrumbs-aside">
          <a href="https://github.com/TransferLearningForDNA/genomic_data_extraction/edit/master/docs/genomic_data_extraction/rna/data_conversion_helper_functions/create_expression_matrix.md" class="icon icon-github"> Edit on GitHub</a>
    </li>
  </ul>
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
            <div class="section" itemprop="articleBody">
              
                <div class="doc doc-object doc-module">



<a id="rna.data_conversion_helper_functions.create_expression_matrix"></a>
  <div class="doc doc-contents first">

  

  <div class="doc doc-children">










<div class="doc doc-object doc-function">



<h2 id="rna.data_conversion_helper_functions.create_expression_matrix.create_expression_matrix" class="doc doc-heading">
          <code class="highlight language-python">create_expression_matrix(raw_data_path, processed_data_path)</code>

</h2>


  <div class="doc doc-contents ">
  
      <p>Create the expression matrices for all species.</p>



<table class="field-list">
  <colgroup>
    <col class="field-name" />
    <col class="field-body" />
  </colgroup>
  <tbody valign="top">
    <tr class="field">
      <th class="field-name">Parameters:</th>
      <td class="field-body">
        <ul class="first simple">
            <li>
              <b><code>raw_data_path</code></b>
                  (<code>str</code>)
              –
              <div class="doc-md-description">
                <p>Path to the folder containing raw quant files.</p>
              </div>
            </li>
            <li>
              <b><code>processed_data_path</code></b>
                  (<code>str</code>)
              –
              <div class="doc-md-description">
                <p>Path to store the processed expression matrix csv files.</p>
              </div>
            </li>
        </ul>
      </td>
    </tr>
  </tbody>
</table>


<table class="field-list">
  <colgroup>
    <col class="field-name" />
    <col class="field-body" />
  </colgroup>
  <tbody valign="top">
    <tr class="field">
    <th class="field-name">Returns:</th>
    <td class="field-body">
      <ul class="first simple">
          <li>
<b><code>None</code></b>(                <code>None</code>
)            –
            <div class="doc-md-description">
              <p>This function does not return a value but outputs files to the specified directory.</p>
            </div>
          </li>
      </ul>
    </td>
    </tr>
  </tbody>
</table>
          <details class="quote">
            <summary>Source code in <code>rna/data_conversion_helper_functions/create_expression_matrix.py</code></summary>
            <pre class="highlight"><code class="language-python">def create_expression_matrix(raw_data_path: str, processed_data_path: str) -&gt; None:
    """Create the expression matrices for all species.

    Args:
        raw_data_path (str): Path to the folder containing raw quant files.
        processed_data_path (str): Path to store the processed expression matrix csv files.

    Returns:
        None: This function does not return a value but outputs files to the specified directory.
    """

    # Iterate over species
    for species in os.listdir(raw_data_path):
        raw_csv_data_path = os.path.join(raw_data_path, species, "csv_files")

        # Skip if it's not a directory (e.g., .DS_Store)
        if not os.path.isdir(raw_csv_data_path):
            continue
        if not os.listdir(raw_csv_data_path):
            continue

        # Initialise empty DataFrames for the abundance, length and counts matrices
        abundance_mat = pd.DataFrame()  # TPM
        length_mat = pd.DataFrame()  # EffectiveLength
        counts_mat = pd.DataFrame()  # NumReads

        # Iterate through each quant csv file
        for quant_file in os.listdir(raw_csv_data_path):
            file_path = os.path.join(raw_csv_data_path, quant_file)
            # Read the csv file into a DataFrame
            # Abundance TPM matrix
            abundance_df = pd.read_csv(file_path, usecols=["Name", "TPM"])
            # Effective Length matrix
            length_df = pd.read_csv(file_path, usecols=["Name", "EffectiveLength"])
            # Counts NumReads matrix
            counts_df = pd.read_csv(file_path, usecols=["Name", "NumReads"])

            # Rename 'TPM' column to the run ID (file name e.g. 'quant_DRR513083.csv')
            abundance_df = abundance_df.rename(
                columns={"TPM": quant_file.split("_")[1][:-4]}
            )
            length_df = length_df.rename(
                columns={"EffectiveLength": quant_file.split("_")[1][:-4]}
            )
            counts_df = counts_df.rename(
                columns={"NumReads": quant_file.split("_")[1][:-4]}
            )
            # Set 'Name' as index
            abundance_df = abundance_df.set_index("Name")
            length_df = length_df.set_index("Name")
            counts_df = counts_df.set_index("Name")
            # Remove duplicates
            abundance_df = abundance_df[~abundance_df.index.duplicated(keep="first")]
            length_df = length_df[~length_df.index.duplicated(keep="first")]
            counts_df = counts_df[~counts_df.index.duplicated(keep="first")]

            # Concatenate with the main expression matrix
            # If the index (Name) does not exist in expression_matrix, NaN values will be added for other columns
            abundance_mat = pd.concat([abundance_mat, abundance_df], axis=1, sort=False)
            length_mat = pd.concat([length_mat, length_df], axis=1, sort=False)
            counts_mat = pd.concat([counts_mat, counts_df], axis=1, sort=False)

        # Calculate the length scaled TPM values
        length_scaled_tpm_mat = get_length_scaled_tpm_matrix(
            counts_mat, abundance_mat, length_mat
        )

        # Save the processed expression matrix to a CSV file
        expression_matrix_path = os.path.join(processed_data_path, f"{species}.csv")
        length_scaled_tpm_mat.to_csv(expression_matrix_path)

        print(f"\nExpression matrix for {species} created successfully.")</code></pre>
          </details>
  </div>

</div>


<div class="doc doc-object doc-function">



<h2 id="rna.data_conversion_helper_functions.create_expression_matrix.get_length_scaled_tpm_matrix" class="doc doc-heading">
          <code class="highlight language-python">get_length_scaled_tpm_matrix(counts_mat, abundance_mat, length_mat)</code>

</h2>


  <div class="doc doc-contents ">
  
      <p>Generate length scaled TPM matrix.</p>



<table class="field-list">
  <colgroup>
    <col class="field-name" />
    <col class="field-body" />
  </colgroup>
  <tbody valign="top">
    <tr class="field">
      <th class="field-name">Parameters:</th>
      <td class="field-body">
        <ul class="first simple">
            <li>
              <b><code>counts_mat</code></b>
                  (<code>DataFrame</code>)
              –
              <div class="doc-md-description">
                <p>A matrix of original counts (NumReads).</p>
              </div>
            </li>
            <li>
              <b><code>abundance_mat</code></b>
                  (<code>DataFrame</code>)
              –
              <div class="doc-md-description">
                <p>A matrix of abundances (TPM).</p>
              </div>
            </li>
            <li>
              <b><code>length_mat</code></b>
                  (<code>DataFrame</code>)
              –
              <div class="doc-md-description">
                <p>A matrix of effective lengths.</p>
              </div>
            </li>
        </ul>
      </td>
    </tr>
  </tbody>
</table>


<table class="field-list">
  <colgroup>
    <col class="field-name" />
    <col class="field-body" />
  </colgroup>
  <tbody valign="top">
    <tr class="field">
    <th class="field-name">Returns:</th>
    <td class="field-body">
      <ul class="first simple">
          <li>
<b><code>DataFrame</code></b>(                <code><span title="pandas.DataFrame">DataFrame</span></code>
)            –
            <div class="doc-md-description">
              <p>A matrix of length scaled TPM values indexed by transcript ID.</p>
            </div>
          </li>
      </ul>
    </td>
    </tr>
  </tbody>
</table>
          <details class="quote">
            <summary>Source code in <code>rna/data_conversion_helper_functions/create_expression_matrix.py</code></summary>
            <pre class="highlight"><code class="language-python">def get_length_scaled_tpm_matrix(
    counts_mat: pd.DataFrame, abundance_mat: pd.DataFrame, length_mat: pd.DataFrame
) -&gt; pd.DataFrame:
    """Generate length scaled TPM matrix.

    Args:
        counts_mat (DataFrame): A matrix of original counts (NumReads).
        abundance_mat (DataFrame): A matrix of abundances (TPM).
        length_mat (DataFrame): A matrix of effective lengths.

    Returns:
        DataFrame: A matrix of length scaled TPM values indexed by transcript ID.
    """

    # Calculate the sum of counts across all transcripts for each sample: per-sample library size
    counts_sum = counts_mat.sum(axis=0)

    # Calculate the average transcript lengths (averaged over samples)
    length_means = length_mat.mean(axis=1)

    # Multiply abundance matrix (TPM) by average transcript lengths
    length_tpm_mat = abundance_mat.multiply(length_means, axis=0)

    # Calculate the sum of new counts across transcripts for each sample
    new_sum = length_tpm_mat.sum(axis=0)

    # Determine the scaling factor for each sample
    scaling_factor = counts_sum / new_sum

    # Apply the scaling factor to scale new counts to match the original total counts
    length_scaled_tpm_mat = length_tpm_mat.multiply(scaling_factor, axis=1)

    return length_scaled_tpm_mat</code></pre>
          </details>
  </div>

</div>



  </div>

  </div>

</div>
              
            </div>
          </div><footer>
    <div class="rst-footer-buttons" role="navigation" aria-label="Footer Navigation">
        <a href="../convert_quantsf_to_csv/" class="btn btn-neutral float-left" title="convert_quantsf_to_csv"><span class="icon icon-circle-arrow-left"></span> Previous</a>
        <a href="../create_samplesheet_csv/" class="btn btn-neutral float-right" title="create_samplesheet_csv">Next <span class="icon icon-circle-arrow-right"></span></a>
    </div>

  <hr/>

  <div role="contentinfo">
    <!-- Copyright etc -->
  </div>

  Built with <a href="https://www.mkdocs.org/">MkDocs</a> using a <a href="https://github.com/readthedocs/sphinx_rtd_theme">theme</a> provided by <a href="https://readthedocs.org">Read the Docs</a>.
</footer>
          
        </div>
      </div>

    </section>

  </div>

  <div class="rst-versions" role="note" aria-label="Versions">
  <span class="rst-current-version" data-toggle="rst-current-version">
    
        <span>
          <a href="https://github.com/TransferLearningForDNA/genomic_data_extraction" class="fa fa-github" style="color: #fcfcfc"> GitHub</a>
        </span>
    
    
      <span><a href="../convert_quantsf_to_csv/" style="color: #fcfcfc">&laquo; Previous</a></span>
    
    
      <span><a href="../create_samplesheet_csv/" style="color: #fcfcfc">Next &raquo;</a></span>
    
  </span>
</div>
    <script>var base_url = '../../../..';</script>
    <script src="../../../../js/theme_extra.js" defer></script>
    <script src="../../../../js/theme.js" defer></script>
    <script defer>
        window.onload = function () {
            SphinxRtdTheme.Navigation.enable(true);
        };
    </script>

</body>
</html>
